{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "887ed803-c6dd-4da6-9845-3afb0c5f8392",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'scripts'",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mModuleNotFoundError\u001b[39m                       Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[2]\u001b[39m\u001b[32m, line 3\u001b[39m\n\u001b[32m      1\u001b[39m \u001b[38;5;28;01mimport\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01msys\u001b[39;00m,\u001b[38;5;250m \u001b[39m\u001b[34;01mos\u001b[39;00m,\u001b[38;5;250m \u001b[39m\u001b[34;01mjson\u001b[39;00m\n\u001b[32m      2\u001b[39m sys.path.append(\u001b[33m'\u001b[39m\u001b[33m../\u001b[39m\u001b[33m'\u001b[39m)\n\u001b[32m----> \u001b[39m\u001b[32m3\u001b[39m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01mscripts\u001b[39;00m\u001b[34;01m.\u001b[39;00m\u001b[34;01muniprot\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m UniprotInterface\n\u001b[32m      4\u001b[39m \u001b[38;5;28;01mimport\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01mpandas\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mas\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01mpd\u001b[39;00m\n",
      "\u001b[31mModuleNotFoundError\u001b[39m: No module named 'scripts'"
     ]
    }
   ],
   "source": [
    "from src.uniprot import UniprotInterface\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5bc2dc57",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e4834c7",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "def replace_char_at_index(s, i, new_char):\n",
    "    if i < 0 or i >= len(s):\n",
    "        raise IndexError(\"Index out of range.\")\n",
    "    return s[:i] + new_char + s[i+1:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b318b89f",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "ids = [\"Q75UA4\"]\n",
    "from_db = 'UniProtKB_AC-ID'\n",
    "to_db = 'UniProtKB'\n",
    "disease = \"CRC\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0a647744",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "downloader = UniprotInterface()\n",
    "\n",
    "job_id = downloader.submit_id_mapping(from_db=from_db, to_db=to_db, ids=ids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eef323a8",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetched: 1 / 1\n"
     ]
    }
   ],
   "source": [
    "if downloader.check_id_mapping_results_ready(job_id):\n",
    "    link = downloader.get_id_mapping_results_link(job_id)\n",
    "    results = downloader.get_id_mapping_results_search(link)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b9396a36",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "with open(\"results.json\", \"w\") as f:\n",
    "    json.dump(results, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "243e35c2",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'MKFGKFVLLAASTALAVVGLGGPAAADSTPQAQPSIIGGSNATSGPWAARLFVNGRQNCTATIIAPQYILTAKHCVSSSGTYTFRIGSLDQTSGGTMATGSTITRYPGSADLAIVRLTTSVNATYSPLGSVGDVSVGQNVSVYGWGATSQCGSEINCQSRYLKVATVRVNSISCSDYTGGVAVCANRVNGITAGGDSGGPMFASGRQVGVASTSDRVNNTAYTNITRYRSWISQVAGV'"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results['results'][0]['to']['sequence']['value']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d06bc8c",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Q75UA4\n"
     ]
    }
   ],
   "source": [
    "for result in results['results']:\n",
    "    print(result['from'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f536ba57",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "export_data = []\n",
    "sequence = results['results'][0]['to']['sequence']['value']\n",
    "for feature in results['results'][0]['to']['features']:\n",
    "    row = []\n",
    "    if feature['type'] == 'Natural variant' and disease in feature['description']:     \n",
    "        row.append(feature['featureId'])\n",
    "        location_start = feature['location']['start']['value']\n",
    "        location_end = feature['location']['end']['value']\n",
    "        if location_start == location_end:\n",
    "            row.append(location_start)\n",
    "            original_sequence = feature['alternativeSequence']['originalSequence']\n",
    "            new_sequence = feature['alternativeSequence']['alternativeSequences'][0]\n",
    "            row.append(f\"{original_sequence}->{new_sequence}\")\n",
    "            row.append(replace_char_at_index(sequence, int(location_start)-1, new_sequence))\n",
    "        else:\n",
    "            row.append(f\"{location_start}-{location_end}\")\n",
    "            row.append(\"missing\")\n",
    "            row.append(sequence[:int(location_start)-1] + sequence[int(location_end)-1:])\n",
    "        export_data.append(row)\n",
    "export_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c43c304e",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "df = pd.DataFrame(export_data, columns=[\"variant id\", \"position\", \"change\", \"sequence\"])\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0b712ab6",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "df.to_csv(\"results.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41389a75",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[{'database': 'PubMed', 'id': '11133465'}, {'database': 'DOI', 'id': '10.1128/AEM.67.1.345-353.2001'}]\n",
      "[{'database': 'PubMed', 'id': '16237016'}, {'database': 'DOI', 'id': '10.1128/JB.187.21.7333-7340.2005'}]\n"
     ]
    }
   ],
   "source": [
    "result = results['results'][0]\n",
    "for reference in result['to']['references']:\n",
    "    print(reference['citation']['citationCrossReferences'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e9e1647-5992-4b1e-a28f-c1d73fd8af0f",
   "metadata": {},
   "outputs": [],
   "source": [
    "references_list = []\n",
    "result = results['results'][0]                 \n",
    "\n",
    "try:\n",
    "    for r in result['to']['references']:\n",
    "        tmp = {}\n",
    "        tmp[\"citacionCrossReferences\"] = r['citation']['citationCrossReferences']\n",
    "        tmp.update({\"title\": r['citation']['title']})\n",
    "        references_list.append(tmp)\n",
    "except KeyError:\n",
    "    pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "28767add-3dbd-45fd-9c7e-8a519ff59da0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'citacionCrossReferences': [{'database': 'PubMed', 'id': '11133465'},\n",
       "   {'database': 'DOI', 'id': '10.1128/AEM.67.1.345-353.2001'}],\n",
       "  'title': 'Purification and characterization of an extracellular poly(L-lactic acid) depolymerase from a soil isolate, Amycolatopsis sp. strain K104-1.'},\n",
       " {'citacionCrossReferences': [{'database': 'PubMed', 'id': '16237016'},\n",
       "   {'database': 'DOI', 'id': '10.1128/JB.187.21.7333-7340.2005'}],\n",
       "  'title': 'Gene cloning and molecular characterization of an extracellular poly(L-lactic acid) depolymerase from Amycolatopsis sp. strain K104-1.'}]"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "references_list"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1aa7cd1e",
   "metadata": {},
   "source": [
    "## Blast"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1102bd78",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, argparse\n",
    "import shutil\n",
    "import subprocess\n",
    "import tarfile\n",
    "from pathlib import Path\n",
    "from urllib.request import urlopen\n",
    "import re\n",
    "from typing import List\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "DB_DIR = os.path.join(\"scripts\", \"db\")\n",
    "BLAST_BASE_URL = \"https://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/LATEST/\"\n",
    "UNIPROT_BASE_URL = \"https://ftp.uniprot.org/pub/databases/uniprot/current_release\"\n",
    "BLAST_DIR = Path(\"blast_bin\")\n",
    "#https://ftp.uniprot.org/pub/databases/uniprot/uniref/uniref100/uniref100.fasta.gz\n",
    "#https://ftp.uniprot.org/pub/databases/uniprot/current_release/uniref/uniref100/uniref.xsd\n",
    "#https://ftp.uniprot.org/pub/databases/uniprot/uniref/uniref100/uniref100.xml.gz\n",
    "#https://ftp.uniprot.org/pub/databases/uniprot/uniref/uniref90/uniref90.xml.gz\n",
    "#https://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/complete/uniprot_sprot.xml.gz\n",
    "#https://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/complete/uniprot_trembl.fasta.gz\n",
    "\n",
    "databases = {\n",
    "    \"uniprotkb_reviewed\": \"knowledgebase/complete/uniprot_sprot\",\n",
    "    \"uniprotkb_unreviewed\": \"knowledgebase/complete/uniprot_trembl\",\n",
    "    \"uniref100\": \"uniref/niref100/uniref100\",\n",
    "    \"uniref90\": \"uniref/uniref90/uniref90\",\n",
    "    \"uniref50\": \"uniref/uniref50/uniref50\",\n",
    "}\n",
    "\n",
    "def download_uniprot_database(db_name: str, extension: str = \"xml\"):\n",
    "    \"\"\" Download a Uniprot database from the Uniprot FTP server.\n",
    "    Args:\n",
    "        db_name (str): Name of the database to download.\n",
    "        extension (str): File extension of the database. Default is \"xml\".\n",
    "    \"\"\"\n",
    "\n",
    "    if db_name not in databases:\n",
    "        raise ValueError(f\"Database {db_name} is not supported. Supported databases are: {', '.join(databases.keys())}.\")\n",
    "    \n",
    "    db_path = os.path.join(DB_DIR, f\"{db_name}.{extension}\")\n",
    "    \n",
    "    if not os.path.exists(db_path):\n",
    "        os.makedirs(DB_DIR, exist_ok=True)\n",
    "        url = f\"{UNIPROT_BASE_URL}/{databases[db_name]}.{extension}.gz\"\n",
    "        os.system(f\"wget {url} -O {db_path}.gz\")\n",
    "        print(f\"Unzipping {db_path}...\")\n",
    "        subprocess.run([\"gunzip\", db_path], check=True)\n",
    "    else:\n",
    "        print(f\"Database {db_name} already exists at {db_path}.\")\n",
    "\n",
    "def get_latest_version_url():\n",
    "    \"\"\"Retrieve the latest BLAST+ tarball URL from the NCBI FTP site.\"\"\"\n",
    "    with urlopen(BLAST_BASE_URL) as response:\n",
    "        html = response.read().decode(\"utf-8\")\n",
    "    # Look for something like: ncbi-blast-2.16.0+-x64-linux.tar.gz\n",
    "    match = re.search(r'ncbi-blast-(\\d+\\.\\d+\\.\\d+\\+)-x64-linux\\.tar\\.gz', html)\n",
    "    if match:\n",
    "        version = match.group(1)\n",
    "        tar_name = f\"ncbi-blast-{version}-x64-linux.tar.gz\"\n",
    "        return version, BLAST_BASE_URL + tar_name\n",
    "    else:\n",
    "        raise RuntimeError(\"Could not find the latest BLAST version from NCBI.\")\n",
    "\n",
    "def is_blast_installed():\n",
    "    \"\"\"Check if 'blastp' is available in the system PATH.\"\"\"\n",
    "    try:\n",
    "        subprocess.run([\"blastp\", \"-version\"], check=True, stdout=subprocess.DEVNULL)\n",
    "        return True\n",
    "    except (subprocess.CalledProcessError, FileNotFoundError):\n",
    "        return False\n",
    "\n",
    "\n",
    "def download_and_extract_blast(version: str, url: str):\n",
    "    \"\"\"Download and extract the BLAST+ tarball.\"\"\"\n",
    "    tarball_name = url.split(\"/\")[-1]\n",
    "    if not Path(tarball_name).exists():\n",
    "        print(f\"Downloading BLAST+ {version}...\")\n",
    "        subprocess.run([\"wget\", url], check=True)\n",
    "\n",
    "    print(\"Extracting BLAST+...\")\n",
    "    with tarfile.open(tarball_name, \"r:gz\") as tar:\n",
    "        tar.extractall(BLAST_DIR)\n",
    "    print(f\"BLAST extracted to: {BLAST_DIR.resolve()}\")\n",
    "\n",
    "\n",
    "def get_local_blastp_path(version: str):\n",
    "    \"\"\"Return the path to local blastp binary.\"\"\"\n",
    "    return BLAST_DIR / f\"ncbi-blast-{version}\" / \"bin\" / \"blastp\"\n",
    "\n",
    "\n",
    "def check_blast():\n",
    "    \"\"\"Ensure BLAST is installed. Return path to `blastp` binary.\"\"\"\n",
    "    if is_blast_installed():\n",
    "        print(\"System-wide BLAST is installed.\")\n",
    "        return shutil.which(\"blastp\")\n",
    "    else:\n",
    "        version, url = get_latest_version_url()\n",
    "        local_blastp = get_local_blastp_path(version)\n",
    "        if not local_blastp.exists():\n",
    "            print(f\"BLAST {version} not found locally. Installing...\")\n",
    "            BLAST_DIR.mkdir(exist_ok=True)\n",
    "            download_and_extract_blast(version, url)\n",
    "        else:\n",
    "            print(f\"Using already downloaded BLAST {version}.\")\n",
    "        return str(local_blastp)\n",
    "\n",
    "def make_blast_database(db_name: str, db_type: str = \"prot\", extension: str = \"xml\"):\n",
    "    \"\"\"Create a BLAST database from the Uniprot database.\"\"\"\n",
    "    db_path = os.path.join(DB_DIR, f\"{db_name}.{extension}\")\n",
    "    if not os.path.exists(db_path):\n",
    "        raise FileNotFoundError(f\"Database {db_name} not found at {db_path}. Please download it first.\")\n",
    "    \n",
    "    # Check if the database is already created\n",
    "    blast_db_path = os.path.join(DB_DIR, db_name)\n",
    "    extensions = [\".pdb\", \".phr\", \".pin\", \".psq\", \".pot\", \".psq\", \".ptf\", \".pto\"]\n",
    "    makedb = False\n",
    "    # For all extensions check if exists if there is one failing makedb again\n",
    "    for ext in extensions:\n",
    "        if not os.path.exists(blast_db_path + \"/db\" + ext):\n",
    "            makedb = True\n",
    "            break\n",
    "    if makedb:\n",
    "        print(f\"Creating BLAST database for {db_name}...\")\n",
    "        blast_db_cmd = [\n",
    "            \"makeblastdb\",\n",
    "            \"-in\", db_path,\n",
    "            \"-dbtype\", db_type,\n",
    "            \"-out\", os.path.join(DB_DIR, db_name) + \"/db\",\n",
    "        ]\n",
    "    \n",
    "        subprocess.run(blast_db_cmd, check=True)\n",
    "        print(f\"BLAST database created at: {os.path.join(DB_DIR, databases[db_name])}\")\n",
    "    else:\n",
    "        print(f\"BLAST database already exists at {blast_db_path}. No need to create it again.\")\n",
    "\n",
    "def run_blast(sequences: List[str], db_name: str, blast_type: str = \"blastp\", evalue: float = 0.001):\n",
    "    \"\"\"Run BLAST search.\"\"\"\n",
    "    blast_db_path = os.path.join(DB_DIR, db_name)\n",
    "    if not os.path.exists(blast_db_path):\n",
    "        raise FileNotFoundError(f\"Database {db_name} not found at {blast_db_path}. Please download it first.\")\n",
    "\n",
    "    # Make tmp directory if it does not exist\n",
    "    os.makedirs(\"tmp\", exist_ok=True)\n",
    "\n",
    "    # Write sequences to a temporary file\n",
    "    with open(\"tmp/sequences.fasta\", \"w\") as f:\n",
    "        for i, seq in enumerate(sequences):\n",
    "            f.write(f\">{i}\\n{seq}\\n\")\n",
    "    \n",
    "    blast_cmd = [\n",
    "        blast_type,\n",
    "        \"-query\", \"tmp/sequences.fasta\",\n",
    "        \"-db\", blast_db_path + \"/db\",\n",
    "        \"-outfmt\", \"6\",\n",
    "        \"-evalue\", str(evalue),\n",
    "    ]\n",
    "    \n",
    "    print(f\"Running BLAST search...\")\n",
    "    with open(\"tmp/blast_results.txt\", \"w\") as f:\n",
    "        subprocess.run(blast_cmd, stdout=f, check=True)\n",
    "    print(f\"BLAST results saved to tmp/blast_results.txt\")\n",
    "    # Clean up temporary file\n",
    "    os.remove(\"tmp/sequences.fasta\")\n",
    "\n",
    "def parse_blast_results(file_path: str, identity_threshold: float = 90.0):\n",
    "    \"\"\"Parse BLAST results from a file.\"\"\"\n",
    "    with open(file_path, \"r\") as f:\n",
    "        results = f.readlines()\n",
    "    \n",
    "    parsed_results = []\n",
    "    for line in results:\n",
    "        fields = line.strip().split(\"\\t\")\n",
    "        identity = float(fields[2])\n",
    "        if identity >= identity_threshold:\n",
    "            parsed_results.append({\n",
    "                \"query\": fields[0],\n",
    "                \"subject\": fields[1],\n",
    "                \"identity\": fields[2],\n",
    "                \"alignment_length\": fields[3],\n",
    "                \"evalue\": fields[4],\n",
    "                \"bit_score\": fields[5],\n",
    "            })\n",
    "    \n",
    "    return parsed_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "de905467",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Database uniprotkb_reviewed already exists at scripts/db/uniprotkb_reviewed.fasta.\n",
      "System-wide BLAST is installed.\n",
      "Using blastp at: /home/diego/micromamba/envs/bioseqdownloader/bin/blastp\n",
      "BLAST database already exists at scripts/db/uniprotkb_reviewed. No need to create it again.\n",
      "Running BLAST search...\n",
      "BLAST results saved to tmp/blast_results.txt\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequences</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MARPLLGKTSSVRRRLESLSACSIFFFLRKFCQKMASLVFLNSPVY...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MSFKVYDPIAELIATQFPTSNPDLQIINNDVLVVSPHKITLPMGPQ...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NYTETAQAIARSWRAGSHDRLKARGEAVAVTVHRLVAVPRGRDTPR...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           sequences  id\n",
       "0  MARPLLGKTSSVRRRLESLSACSIFFFLRKFCQKMASLVFLNSPVY...   0\n",
       "1  MSFKVYDPIAELIATQFPTSNPDLQIINNDVLVVSPHKITLPMGPQ...   1\n",
       "2  NYTETAQAIARSWRAGSHDRLKARGEAVAVTVHRLVAVPRGRDTPR...   2"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv(\"data/test.csv\")\n",
    "sequences = df[\"sequences\"].dropna().unique().tolist()\n",
    "    \n",
    "download_uniprot_database(\"uniprotkb_reviewed\", \"fasta\")\n",
    "    \n",
    "blastp_path = check_blast()\n",
    "print(f\"Using blastp at: {blastp_path}\")\n",
    "\n",
    "make_blast_database(\"uniprotkb_reviewed\", extension=\"fasta\")\n",
    "\n",
    "run_blast(sequences, \"uniprotkb_reviewed\", blast_type=\"blastp\", evalue=0.0001)\n",
    "\n",
    "results = parse_blast_results(\"tmp/blast_results.txt\")\n",
    "\n",
    "# Convert to DataFrame\n",
    "sequences_df = pd.DataFrame(sequences, columns=[\"sequences\"])\n",
    "sequences_df[\"id\"] = sequences_df.index\n",
    "\n",
    "sequences_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4009c945",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_blast = pd.DataFrame(results)\n",
    "\n",
    "df_blast = df_blast.rename(columns={\"query\": \"id\", \"subject\": \"subject_id\"})\n",
    "df_blast[\"id\"] = df_blast[\"id\"].astype(int)\n",
    "df_blast = df_blast.merge(sequences_df, on=\"id\", how=\"left\")\n",
    "df_blast = df_blast.drop(columns=[\"id\"])\n",
    "df_blast = df_blast.rename(columns={\"sequences\": \"sequence\"})\n",
    "\n",
    "# Separate subject into source, accession, entry_name\n",
    "df_blast[\"source\"] = df_blast[\"subject_id\"].apply(lambda x: x.split(\"|\")[0])\n",
    "df_blast[\"accession\"] = df_blast[\"subject_id\"].apply(lambda x: x.split(\"|\")[1])\n",
    "df_blast[\"entry_name\"] = df_blast[\"subject_id\"].apply(lambda x: x.split(\"|\")[2])\n",
    "df_blast = df_blast.drop(columns=[\"subject_id\"])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a32f8e17",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identity</th>\n",
       "      <th>alignment_length</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bit_score</th>\n",
       "      <th>sequence</th>\n",
       "      <th>source</th>\n",
       "      <th>accession</th>\n",
       "      <th>entry_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>100.000</td>\n",
       "      <td>438</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>MARPLLGKTSSVRRRLESLSACSIFFFLRKFCQKMASLVFLNSPVY...</td>\n",
       "      <td>sp</td>\n",
       "      <td>Q6GZX2</td>\n",
       "      <td>003R_FRG3G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>100.000</td>\n",
       "      <td>180</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>MSFKVYDPIAELIATQFPTSNPDLQIINNDVLVVSPHKITLPMGPQ...</td>\n",
       "      <td>sp</td>\n",
       "      <td>Q197F2</td>\n",
       "      <td>008L_IIV3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>100.000</td>\n",
       "      <td>50</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>MSFKVYDPIAELIATQFPTSNPDLQIINNDVLVVSPHKITLPMGPQ...</td>\n",
       "      <td>sp</td>\n",
       "      <td>Q6GZW6</td>\n",
       "      <td>009L_FRG3G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>100.000</td>\n",
       "      <td>345</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NYTETAQAIARSWRAGSHDRLKARGEAVAVTVHRLVAVPRGRDTPR...</td>\n",
       "      <td>sp</td>\n",
       "      <td>Q6GZW6</td>\n",
       "      <td>009L_FRG3G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  identity alignment_length evalue bit_score  \\\n",
       "0  100.000              438      0         0   \n",
       "1  100.000              180      0         0   \n",
       "2  100.000               50      0         0   \n",
       "3  100.000              345      0         0   \n",
       "\n",
       "                                            sequence source accession  \\\n",
       "0  MARPLLGKTSSVRRRLESLSACSIFFFLRKFCQKMASLVFLNSPVY...     sp    Q6GZX2   \n",
       "1  MSFKVYDPIAELIATQFPTSNPDLQIINNDVLVVSPHKITLPMGPQ...     sp    Q197F2   \n",
       "2  MSFKVYDPIAELIATQFPTSNPDLQIINNDVLVVSPHKITLPMGPQ...     sp    Q6GZW6   \n",
       "3  NYTETAQAIARSWRAGSHDRLKARGEAVAVTVHRLVAVPRGRDTPR...     sp    Q6GZW6   \n",
       "\n",
       "   entry_name  \n",
       "0  003R_FRG3G  \n",
       "1   008L_IIV3  \n",
       "2  009L_FRG3G  \n",
       "3  009L_FRG3G  "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_blast"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d944e22",
   "metadata": {},
   "source": [
    "## GO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5899c06f",
   "metadata": {},
   "outputs": [],
   "source": [
    "from src.description_go import *\n",
    "import os, ast\n",
    "import pandas as pd\n",
    "from tqdm import tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6b45cd2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[DESCRIPTION_GO] Getting Gen Ontology\n",
      "Docker version 28.0.0, build f9ced58158\n",
      "[DESCRIPTION_GO] Metastudent found.\n",
      "[DESCRIPTION_GO] Go terms found in input data.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>entry_type</th>\n",
       "      <th>protein_name</th>\n",
       "      <th>ec_numbers</th>\n",
       "      <th>organism</th>\n",
       "      <th>taxon_id</th>\n",
       "      <th>sequence</th>\n",
       "      <th>length</th>\n",
       "      <th>go_terms</th>\n",
       "      <th>pfam_ids</th>\n",
       "      <th>references</th>\n",
       "      <th>features</th>\n",
       "      <th>keywords</th>\n",
       "      <th>source_db</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [uniprot_id, entry_type, protein_name, ec_numbers, organism, taxon_id, sequence, length, go_terms, pfam_ids, references, features, keywords, source_db]\n",
       "Index: []"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DOCKER_IMAGE_NAME = \"metastudent\"\n",
    "DOCKER_CONTAINER_NAME = \"metastudent_container\"\n",
    "HOST_INPUT_FILE = os.path.abspath(\"tmp/sequences.fasta\")\n",
    "HOST_OUTPUT_DIR = os.path.abspath(\"tmp/\")\n",
    "CONTAINER_INPUT_FILE = \"/app/input.fasta\"\n",
    "CONTAINER_OUTPUT_FILE = \"/app/output.result\"\n",
    "\n",
    "\n",
    "print(\"[DESCRIPTION_GO] Getting Gen Ontology\")\n",
    "tqdm.pandas()\n",
    "\n",
    "if not check_dependencies(DOCKER_IMAGE_NAME):\n",
    "    print(\"[DESCRIPTION_GO] Metastudent not found. Installing...\")\n",
    "    install_dependencies(DOCKER_IMAGE_NAME)\n",
    "else:\n",
    "    print(\"[DESCRIPTION_GO] Metastudent found.\")\n",
    "\n",
    "input_df = pd.read_csv(\"results/umami_uniprot.csv\")\n",
    "obsolete_df = pd.read_csv(\"scripts/resources/amiGO_data.csv\", sep=\"\\t\", names=[\"id_go\", \"description\", \"is_obsolete\"])\n",
    "input_df['go_terms'] = input_df['go_terms'].apply(lambda x: ast.literal_eval(x) if isinstance(x, str) else x)\n",
    "\n",
    "parsed_df = pd.DataFrame()\n",
    "if os.path.isfile(f\"{HOST_OUTPUT_DIR}/output.BPO.txt\") and \\\n",
    "        os.path.isfile(f\"{HOST_OUTPUT_DIR}/output.CCO.txt\") and \\\n",
    "        os.path.isfile(f\"{HOST_OUTPUT_DIR}/output.MFO.txt\"):\n",
    "    print(\"[DESCRIPTION_GO] Metastudent results found.\")\n",
    "    parsed_df = parse_outputs(\"uniprot_id\")\n",
    "\n",
    "# Filter input_df with go_terms ~= null\n",
    "input_df_with_go_terms = input_df[input_df[\"go_terms\"].apply(lambda x: isinstance(x, list) and len(x) > 0)]\n",
    "input_df = input_df[input_df[\"go_terms\"].apply(lambda x: isinstance(x, list) and len(x) == 0)]\n",
    "\n",
    "if not input_df_with_go_terms.empty:\n",
    "    print(\"[DESCRIPTION_GO] Go terms found in input data.\")\n",
    "    input_df_with_go_terms = input_df_with_go_terms[[\"uniprot_id\", \"go_terms\"]]\n",
    "    input_df_with_go_terms = input_df_with_go_terms.explode(\"go_terms\")\n",
    "    parsed_df = pd.concat(\n",
    "        [\n",
    "            parsed_df,\n",
    "            pd.merge(\n",
    "                input_df_with_go_terms, \n",
    "                obsolete_df, \n",
    "                left_on=\"go_terms\", \n",
    "                right_on=\"id_go\", \n",
    "                how=\"left\"\n",
    "            )\n",
    "            .drop(columns=[\"go_terms\"])\n",
    "            .rename(columns={\"id_go\": \"go\"})  \n",
    "        ]\n",
    "    )\n",
    "\n",
    "input_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6a2a9d8e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[DESCRIPTION_GO] 0 sequences have not been processed.\n"
     ]
    }
   ],
   "source": [
    "if not parsed_df.empty:\n",
    "    # Check if all sequences have been processed\n",
    "    parsed_ids = parsed_df[\"uniprot_id\"].unique()\n",
    "    input_ids = input_df[\"uniprot_id\"].unique()\n",
    "    if len(parsed_ids) == len(input_ids):\n",
    "        print(\"[DESCRIPTION_GO] All sequences have been processed.\")\n",
    "        input_df = pd.DataFrame()\n",
    "    else:\n",
    "        input_df = input_df[~input_df[\"uniprot_id\"].isin(parsed_ids)]\n",
    "        print(f\"[DESCRIPTION_GO] {len(input_df)} sequences have not been processed.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d815e6cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.makedirs(HOST_OUTPUT_DIR, exist_ok=True)\n",
    "if not input_df.empty:\n",
    "    print(\"[DESCRIPTION_GO] Running in batches of 50...\")\n",
    "    for i in tqdm(range(0, len(input_df), 50)):\n",
    "        run_in_batches(input_df[i:i+50], HOST_OUTPUT_DIR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7cffac13",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File '/home/diego/Documents/PythonProjects/BioSeqDownloader/tmp/output.BPO.txt' not found.\n"
     ]
    },
    {
     "ename": "NameError",
     "evalue": "name 'exit' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mFileNotFoundError\u001b[39m                         Traceback (most recent call last)",
      "\u001b[36mFile \u001b[39m\u001b[32m~/Documents/PythonProjects/BioSeqDownloader/scripts/description_go.py:27\u001b[39m, in \u001b[36mparse_outputs\u001b[39m\u001b[34m(id_column)\u001b[39m\n\u001b[32m     26\u001b[39m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[32m---> \u001b[39m\u001b[32m27\u001b[39m     df_go = \u001b[43mpd\u001b[49m\u001b[43m.\u001b[49m\u001b[43mread_csv\u001b[49m\u001b[43m(\u001b[49m\n\u001b[32m     28\u001b[39m \u001b[43m            \u001b[49m\u001b[43mfile\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\n\u001b[32m     29\u001b[39m \u001b[43m            \u001b[49m\u001b[43mheader\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\n\u001b[32m     30\u001b[39m \u001b[43m            \u001b[49m\u001b[43msep\u001b[49m\u001b[43m=\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[38;5;130;43;01m\\t\u001b[39;49;00m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\n\u001b[32m     31\u001b[39m \u001b[43m            \u001b[49m\u001b[43mnames\u001b[49m\u001b[43m=\u001b[49m\u001b[43m[\u001b[49m\u001b[43mid_column\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mgo\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mprobability\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mterm\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m     32\u001b[39m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m     33\u001b[39m     df_go[\u001b[33m'\u001b[39m\u001b[33mgo_type\u001b[39m\u001b[33m'\u001b[39m] = file_type\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/micromamba/envs/bioseqdownloader/lib/python3.13/site-packages/pandas/io/parsers/readers.py:1026\u001b[39m, in \u001b[36mread_csv\u001b[39m\u001b[34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, date_format, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, encoding_errors, dialect, on_bad_lines, delim_whitespace, low_memory, memory_map, float_precision, storage_options, dtype_backend)\u001b[39m\n\u001b[32m   1024\u001b[39m kwds.update(kwds_defaults)\n\u001b[32m-> \u001b[39m\u001b[32m1026\u001b[39m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_read\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilepath_or_buffer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/micromamba/envs/bioseqdownloader/lib/python3.13/site-packages/pandas/io/parsers/readers.py:620\u001b[39m, in \u001b[36m_read\u001b[39m\u001b[34m(filepath_or_buffer, kwds)\u001b[39m\n\u001b[32m    619\u001b[39m \u001b[38;5;66;03m# Create the parser.\u001b[39;00m\n\u001b[32m--> \u001b[39m\u001b[32m620\u001b[39m parser = \u001b[43mTextFileReader\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilepath_or_buffer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m*\u001b[49m\u001b[43m*\u001b[49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    622\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m chunksize \u001b[38;5;129;01mor\u001b[39;00m iterator:\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/micromamba/envs/bioseqdownloader/lib/python3.13/site-packages/pandas/io/parsers/readers.py:1620\u001b[39m, in \u001b[36mTextFileReader.__init__\u001b[39m\u001b[34m(self, f, engine, **kwds)\u001b[39m\n\u001b[32m   1619\u001b[39m \u001b[38;5;28mself\u001b[39m.handles: IOHandles | \u001b[38;5;28;01mNone\u001b[39;00m = \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[32m-> \u001b[39m\u001b[32m1620\u001b[39m \u001b[38;5;28mself\u001b[39m._engine = \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43m_make_engine\u001b[49m\u001b[43m(\u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43mengine\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/micromamba/envs/bioseqdownloader/lib/python3.13/site-packages/pandas/io/parsers/readers.py:1880\u001b[39m, in \u001b[36mTextFileReader._make_engine\u001b[39m\u001b[34m(self, f, engine)\u001b[39m\n\u001b[32m   1879\u001b[39m         mode += \u001b[33m\"\u001b[39m\u001b[33mb\u001b[39m\u001b[33m\"\u001b[39m\n\u001b[32m-> \u001b[39m\u001b[32m1880\u001b[39m \u001b[38;5;28mself\u001b[39m.handles = \u001b[43mget_handle\u001b[49m\u001b[43m(\u001b[49m\n\u001b[32m   1881\u001b[39m \u001b[43m    \u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1882\u001b[39m \u001b[43m    \u001b[49m\u001b[43mmode\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1883\u001b[39m \u001b[43m    \u001b[49m\u001b[43mencoding\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43moptions\u001b[49m\u001b[43m.\u001b[49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mencoding\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1884\u001b[39m \u001b[43m    \u001b[49m\u001b[43mcompression\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43moptions\u001b[49m\u001b[43m.\u001b[49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mcompression\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1885\u001b[39m \u001b[43m    \u001b[49m\u001b[43mmemory_map\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43moptions\u001b[49m\u001b[43m.\u001b[49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mmemory_map\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1886\u001b[39m \u001b[43m    \u001b[49m\u001b[43mis_text\u001b[49m\u001b[43m=\u001b[49m\u001b[43mis_text\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1887\u001b[39m \u001b[43m    \u001b[49m\u001b[43merrors\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43moptions\u001b[49m\u001b[43m.\u001b[49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mencoding_errors\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mstrict\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1888\u001b[39m \u001b[43m    \u001b[49m\u001b[43mstorage_options\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43moptions\u001b[49m\u001b[43m.\u001b[49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mstorage_options\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1889\u001b[39m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m   1890\u001b[39m \u001b[38;5;28;01massert\u001b[39;00m \u001b[38;5;28mself\u001b[39m.handles \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/micromamba/envs/bioseqdownloader/lib/python3.13/site-packages/pandas/io/common.py:873\u001b[39m, in \u001b[36mget_handle\u001b[39m\u001b[34m(path_or_buf, mode, encoding, compression, memory_map, is_text, errors, storage_options)\u001b[39m\n\u001b[32m    871\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m ioargs.encoding \u001b[38;5;129;01mand\u001b[39;00m \u001b[33m\"\u001b[39m\u001b[33mb\u001b[39m\u001b[33m\"\u001b[39m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m ioargs.mode:\n\u001b[32m    872\u001b[39m     \u001b[38;5;66;03m# Encoding\u001b[39;00m\n\u001b[32m--> \u001b[39m\u001b[32m873\u001b[39m     handle = \u001b[38;5;28;43mopen\u001b[39;49m\u001b[43m(\u001b[49m\n\u001b[32m    874\u001b[39m \u001b[43m        \u001b[49m\u001b[43mhandle\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    875\u001b[39m \u001b[43m        \u001b[49m\u001b[43mioargs\u001b[49m\u001b[43m.\u001b[49m\u001b[43mmode\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    876\u001b[39m \u001b[43m        \u001b[49m\u001b[43mencoding\u001b[49m\u001b[43m=\u001b[49m\u001b[43mioargs\u001b[49m\u001b[43m.\u001b[49m\u001b[43mencoding\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    877\u001b[39m \u001b[43m        \u001b[49m\u001b[43merrors\u001b[49m\u001b[43m=\u001b[49m\u001b[43merrors\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    878\u001b[39m \u001b[43m        \u001b[49m\u001b[43mnewline\u001b[49m\u001b[43m=\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[32m    879\u001b[39m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    880\u001b[39m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[32m    881\u001b[39m     \u001b[38;5;66;03m# Binary mode\u001b[39;00m\n",
      "\u001b[31mFileNotFoundError\u001b[39m: [Errno 2] No such file or directory: '/home/diego/Documents/PythonProjects/BioSeqDownloader/tmp/output.BPO.txt'",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[31mNameError\u001b[39m                                 Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[23]\u001b[39m\u001b[32m, line 4\u001b[39m\n\u001b[32m      1\u001b[39m test = pd.concat(\n\u001b[32m      2\u001b[39m     [\n\u001b[32m      3\u001b[39m         parsed_df,\n\u001b[32m----> \u001b[39m\u001b[32m4\u001b[39m         \u001b[43mparse_outputs\u001b[49m\u001b[43m(\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43muniprot_id\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[32m      5\u001b[39m     ]\n\u001b[32m      6\u001b[39m )\n\u001b[32m      8\u001b[39m test = test.sort_values(by=\u001b[33m\"\u001b[39m\u001b[33muniprot_id\u001b[39m\u001b[33m\"\u001b[39m)\n\u001b[32m      9\u001b[39m test = test.merge(obsolete_df, left_on=\u001b[33m\"\u001b[39m\u001b[33mgo\u001b[39m\u001b[33m\"\u001b[39m, right_on=\u001b[33m\"\u001b[39m\u001b[33mid_go\u001b[39m\u001b[33m\"\u001b[39m, how=\u001b[33m\"\u001b[39m\u001b[33mleft\u001b[39m\u001b[33m\"\u001b[39m)\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/Documents/PythonProjects/BioSeqDownloader/scripts/description_go.py:40\u001b[39m, in \u001b[36mparse_outputs\u001b[39m\u001b[34m(id_column)\u001b[39m\n\u001b[32m     38\u001b[39m     \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mFileNotFoundError\u001b[39;00m:\n\u001b[32m     39\u001b[39m         \u001b[38;5;28mprint\u001b[39m(\u001b[33mf\u001b[39m\u001b[33m\"\u001b[39m\u001b[33mFile \u001b[39m\u001b[33m'\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mfile\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m'\u001b[39m\u001b[33m not found.\u001b[39m\u001b[33m\"\u001b[39m)\n\u001b[32m---> \u001b[39m\u001b[32m40\u001b[39m         \u001b[43mexit\u001b[49m(\u001b[32m1\u001b[39m)\n\u001b[32m     42\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m results:\n\u001b[32m     43\u001b[39m     \u001b[38;5;66;03m# df = pd.DataFrame({\u001b[39;00m\n\u001b[32m     44\u001b[39m     \u001b[38;5;66;03m#     \"id_seq\": [\"None\"],\u001b[39;00m\n\u001b[32m   (...)\u001b[39m\u001b[32m     48\u001b[39m     \u001b[38;5;66;03m#     \"go\": [\"No results found\"]\u001b[39;00m\n\u001b[32m     49\u001b[39m     \u001b[38;5;66;03m# })\u001b[39;00m\n\u001b[32m     50\u001b[39m     df = pd.DataFrame(columns=[id_column, \u001b[33m\"\u001b[39m\u001b[33mgo\u001b[39m\u001b[33m\"\u001b[39m, \u001b[33m\"\u001b[39m\u001b[33mprobability\u001b[39m\u001b[33m\"\u001b[39m, \u001b[33m\"\u001b[39m\u001b[33mterm\u001b[39m\u001b[33m\"\u001b[39m, \u001b[33m\"\u001b[39m\u001b[33mgo_type\u001b[39m\u001b[33m\"\u001b[39m])\n",
      "\u001b[31mNameError\u001b[39m: name 'exit' is not defined"
     ]
    }
   ],
   "source": [
    "test = pd.concat(\n",
    "    [\n",
    "        parsed_df,\n",
    "        parse_outputs(\"uniprot_id\")\n",
    "    ]\n",
    ")\n",
    "    \n",
    "test = test.sort_values(by=\"uniprot_id\")\n",
    "test = test.merge(obsolete_df, left_on=\"go\", right_on=\"id_go\", how=\"left\")\n",
    "test = test.drop(columns=[\"id_go\"])\n",
    "\n",
    "test"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "504746e9",
   "metadata": {},
   "source": [
    "## Uniprot query"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "4a4b02aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "os.chdir(\"src\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ec84c845",
   "metadata": {},
   "outputs": [],
   "source": [
    "query=\"organism_name:homo sapiens (human) AND length:[15 TO 30] AND reviewed:true\"\n",
    "fields=\"accession,protein_name,sequence,ec,lineage,organism_name,xref_pfam,xref_alphafolddb,xref_pdb,go_id\"\n",
    "sort=\"accession asc\"\n",
    "download=True\n",
    "format=\"json\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "6f64e04b",
   "metadata": {},
   "outputs": [],
   "source": [
    "from uniprot import UniprotInterface\n",
    "\n",
    "uniprot = UniprotInterface()\n",
    "response = uniprot.submit_stream(\n",
    "    query=query,\n",
    "    fields=fields,\n",
    "    sort=sort,\n",
    "    include_isoform=True,\n",
    "    download=download,\n",
    "    format=format\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "02c0f301",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query</th>\n",
       "      <th>accession</th>\n",
       "      <th>protein_name</th>\n",
       "      <th>organism_name</th>\n",
       "      <th>taxon_id</th>\n",
       "      <th>ineage</th>\n",
       "      <th>sequence</th>\n",
       "      <th>length</th>\n",
       "      <th>go_terms</th>\n",
       "      <th>pfam_ids</th>\n",
       "      <th>alphafold_ids</th>\n",
       "      <th>pdb_ids</th>\n",
       "      <th>kegg_ids</th>\n",
       "      <th>brenda_ids</th>\n",
       "      <th>reactome_ids</th>\n",
       "      <th>refseq_ids</th>\n",
       "      <th>interpro_ids</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>A0A075B6S0</td>\n",
       "      <td>T cell receptor gamma joining 1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>NYYKKLFGSGTTLVVT</td>\n",
       "      <td>16</td>\n",
       "      <td>[GO:0042101, GO:0002250]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[A0A075B6S0]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>A0A075B6Y3</td>\n",
       "      <td>T cell receptor alpha joining 3</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>GYSSASKIIFGSGTRLSIRP</td>\n",
       "      <td>20</td>\n",
       "      <td>[GO:0042101, GO:0002250]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[A0A075B6Y3]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>A0A075B6Y9</td>\n",
       "      <td>T cell receptor alpha joining 42</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>YGGSQGNLIFGKGTKLSVKP</td>\n",
       "      <td>20</td>\n",
       "      <td>[GO:0005886, GO:0002250]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[A0A075B6Y9]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>A0A075B700</td>\n",
       "      <td>T cell receptor alpha joining 31</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>NNNARLMFGDGTQLVVKP</td>\n",
       "      <td>18</td>\n",
       "      <td>[GO:0005886, GO:0002250]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[A0A075B700]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>A0A075B706</td>\n",
       "      <td>T cell receptor delta joining 1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>TDKLIFGKGTRVTVEP</td>\n",
       "      <td>16</td>\n",
       "      <td>[GO:0042101, GO:0002250]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[A0A075B706]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>A0A0A0MT70</td>\n",
       "      <td>T cell receptor beta joining 2-6</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>SGANVLTFGAGSRLTVL</td>\n",
       "      <td>17</td>\n",
       "      <td>[GO:0042101, GO:0002250]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[A0A0A0MT70]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>A0A0A0MT78</td>\n",
       "      <td>T cell receptor beta joining 2-7</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>SYEQYFGPGTRLTVT</td>\n",
       "      <td>15</td>\n",
       "      <td>[GO:0042101, GO:0002250]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>A0A0A0MT87</td>\n",
       "      <td>T cell receptor beta joining 2-4</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>AKNIQYFGAGTRLSVL</td>\n",
       "      <td>16</td>\n",
       "      <td>[GO:0042101, GO:0002250]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[A0A0A0MT87]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>A0A0A0MT94</td>\n",
       "      <td>T cell receptor beta joining 2-2</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>NTGELFFGEGSRLTVL</td>\n",
       "      <td>16</td>\n",
       "      <td>[GO:0042101, GO:0002250]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[A0A0A0MT94]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>A0A0A0MTA4</td>\n",
       "      <td>T cell receptor beta joining 2-5</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>QETQYFGPGTRLLVL</td>\n",
       "      <td>15</td>\n",
       "      <td>[GO:0042101, GO:0002250]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>A0A0A0MTA7</td>\n",
       "      <td>T cell receptor beta joining 2-1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>SYNEQFFGPGTRLTVL</td>\n",
       "      <td>16</td>\n",
       "      <td>[GO:0042101, GO:0002250]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[A0A0A0MTA7]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>A0A0B4J200</td>\n",
       "      <td>T cell receptor beta joining 2-3</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>STDTQYFGPGTRLTVL</td>\n",
       "      <td>16</td>\n",
       "      <td>[GO:0042101, GO:0002250]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[A0A0B4J200]</td>\n",
       "      <td>[7NDT]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>A0A0C4DH62</td>\n",
       "      <td>Immunoglobulin heavy joining 1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>AEYFQHWGQGTLVTVSS</td>\n",
       "      <td>17</td>\n",
       "      <td>[GO:0005576, GO:0019814, GO:0005886, GO:0002250]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[A0A0C4DH62]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>A0A0C5B5G6</td>\n",
       "      <td>Mitochondrial-derived peptide MOTS-c</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>MRWQEMGYIFYPRKLR</td>\n",
       "      <td>16</td>\n",
       "      <td>[GO:0005615, GO:0005739, GO:0005634, GO:000367...</td>\n",
       "      <td>[PF21945]</td>\n",
       "      <td>[A0A0C5B5G6]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>A0A0J9YWP8</td>\n",
       "      <td>T cell receptor beta joining 1-3</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>SGNTIYFGEGSWLTVV</td>\n",
       "      <td>16</td>\n",
       "      <td>[GO:0042101, GO:0002250]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[A0A0J9YWP8]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>A0A0J9YWX3</td>\n",
       "      <td>T cell receptor beta joining 1-6</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>SYNSPLHFGNGTRLTVT</td>\n",
       "      <td>17</td>\n",
       "      <td>[GO:0042101, GO:0002250]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[A0A0J9YWX3]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>A0A0J9YX06</td>\n",
       "      <td>T cell receptor beta joining 1-2</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>NYGYTFGSGTRLTVV</td>\n",
       "      <td>15</td>\n",
       "      <td>[GO:0042101, GO:0002250]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[7NDQ, 7NDU]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>A0A0J9YXA8</td>\n",
       "      <td>T cell receptor beta joining 1-1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>NTEAFFGQGTRLTVV</td>\n",
       "      <td>15</td>\n",
       "      <td>[GO:0042101, GO:0002250]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>A0A0J9YXG5</td>\n",
       "      <td>T cell receptor beta joining 1-4</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>TNEKLFFGSGTQLSVL</td>\n",
       "      <td>16</td>\n",
       "      <td>[GO:0042101, GO:0002250]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[A0A0J9YXG5]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>A0A0J9YXM7</td>\n",
       "      <td>T cell receptor beta joining 1-5</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>SNQPQHFGDGTRLSIL</td>\n",
       "      <td>16</td>\n",
       "      <td>[GO:0042101, GO:0002250]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[A0A0J9YXM7]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>A0A1B0GWH6</td>\n",
       "      <td>LASP1 neighbor protein</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>MLDIFILMFFAIIGLVILSYIIYLL</td>\n",
       "      <td>25</td>\n",
       "      <td>[GO:0016020]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[A0A1B0GWH6]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>C0HM02</td>\n",
       "      <td>PRKCH upstream open reading frame 2</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>MASRGALRRCLSPGLPRLLHLSRGLA</td>\n",
       "      <td>26</td>\n",
       "      <td>[GO:0120283, GO:0030291, GO:0045182, GO:003419...</td>\n",
       "      <td>[PF21952]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>C0HMA1</td>\n",
       "      <td>HSPA8-interacting micropeptide miPEP155</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>MEMALMVAQTRKGKSVV</td>\n",
       "      <td>17</td>\n",
       "      <td>[GO:0005737, GO:0005634, GO:0030544, GO:000237...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>D0EPY3</td>\n",
       "      <td>Small integral membrane protein 47</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>MNFQENVTLAMALFTILTSIYFFNKAQQ</td>\n",
       "      <td>28</td>\n",
       "      <td>[GO:0016020]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[D0EPY3]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>P0CJ68</td>\n",
       "      <td>Humanin-like 1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>MAPRGFSCLLLSTSEIDLPVKRRT</td>\n",
       "      <td>24</td>\n",
       "      <td>[GO:0005737, GO:0005576, GO:0048019, GO:1900118]</td>\n",
       "      <td>[PF15040]</td>\n",
       "      <td>[P0CJ68]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>P0CJ69</td>\n",
       "      <td>Humanin-like 2</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>MAPRGFSCLLLSTSEIDLPVKRLLSSVF</td>\n",
       "      <td>28</td>\n",
       "      <td>[GO:0005737, GO:0005576, GO:0048019, GO:1900118]</td>\n",
       "      <td>[PF15040]</td>\n",
       "      <td>[P0CJ69]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>P0CJ70</td>\n",
       "      <td>Humanin-like 3</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>MATRRFSCLLLSTSEIDLSVKRRI</td>\n",
       "      <td>24</td>\n",
       "      <td>[GO:0005737, GO:0005576, GO:0048019, GO:1900118]</td>\n",
       "      <td>[PF15040]</td>\n",
       "      <td>[P0CJ70]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>P0CJ71</td>\n",
       "      <td>Humanin-like 4</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>MATQGFSCLLLSVSEIDLSMKRQYKQIR</td>\n",
       "      <td>28</td>\n",
       "      <td>[GO:0005737, GO:0005576, GO:0048019, GO:1900118]</td>\n",
       "      <td>[PF15040]</td>\n",
       "      <td>[P0CJ71]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>P0CJ72</td>\n",
       "      <td>Humanin-like 5</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>MATPGFSCLLLSTSEIDLPMKRRV</td>\n",
       "      <td>24</td>\n",
       "      <td>[GO:0005576, GO:0005739, GO:0048019, GO:007273...</td>\n",
       "      <td>[PF15040]</td>\n",
       "      <td>[P0CJ72]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>P0CJ73</td>\n",
       "      <td>Humanin-like 6</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>MTPRGFSCLLLPTSETDLPVKRRT</td>\n",
       "      <td>24</td>\n",
       "      <td>[GO:0005737, GO:0005576, GO:0048019, GO:1900118]</td>\n",
       "      <td>[PF15040]</td>\n",
       "      <td>[P0CJ73]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>P0CJ74</td>\n",
       "      <td>Humanin-like 7</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>MATGGFGCLLLLIREIDLSVKRQI</td>\n",
       "      <td>24</td>\n",
       "      <td>[GO:0005737, GO:0005576, GO:0048019, GO:1900118]</td>\n",
       "      <td>[PF15040]</td>\n",
       "      <td>[P0CJ74]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>P0CJ75</td>\n",
       "      <td>Humanin-like 8</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>MAPRGFSCLLLSTSEIDLPVKRRA</td>\n",
       "      <td>24</td>\n",
       "      <td>[GO:0005737, GO:0005576, GO:0048019, GO:1900118]</td>\n",
       "      <td>[PF15040]</td>\n",
       "      <td>[P0CJ75]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>P0CJ76</td>\n",
       "      <td>Humanin-like 9</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>MARRGFSCLLLSTTATDLPVKRRT</td>\n",
       "      <td>24</td>\n",
       "      <td>[GO:0005737, GO:0005576, GO:0048019, GO:1900118]</td>\n",
       "      <td>[PF15040]</td>\n",
       "      <td>[P0CJ76]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>P0CJ77</td>\n",
       "      <td>Humanin-like 10</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>MTTRGFSCLLLLIREIDLSAKRRI</td>\n",
       "      <td>24</td>\n",
       "      <td>[GO:0005737, GO:0005576, GO:0048019, GO:1900118]</td>\n",
       "      <td>[PF15040]</td>\n",
       "      <td>[P0CJ77]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>P0DMP1</td>\n",
       "      <td>Humanin-like 12</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>MAPRGFSCLLLSTSEIDLPVKRRA</td>\n",
       "      <td>24</td>\n",
       "      <td>[GO:0005737, GO:0005576]</td>\n",
       "      <td>[PF15040]</td>\n",
       "      <td>[P0DMP1]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>P16070-2</td>\n",
       "      <td>Isoform 2 of CD44 antigen</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>MDKFWWHAAWGLCLVPLSLAQIGVGRRKS</td>\n",
       "      <td>29</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>P59103-2</td>\n",
       "      <td>Isoform 2 of D-amino acid oxidase regulator</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>MLEKLMGADSLQLFRV</td>\n",
       "      <td>16</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>P62945</td>\n",
       "      <td>Small ribosomal subunit protein eS32</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>MRAKWRKKRMRRLKRKRRKMRQRSK</td>\n",
       "      <td>25</td>\n",
       "      <td>[GO:0005737, GO:0005829, GO:0022625, GO:000578...</td>\n",
       "      <td>[PF05162]</td>\n",
       "      <td>[P62945]</td>\n",
       "      <td>[4UG0, 4V6X, 5A2Q, 5AJ0, 5LKS, 5OA3, 5VYC, 6G5...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>P84351</td>\n",
       "      <td>Osteocalcin</td>\n",
       "      <td>Homo sapiens neanderthalensis</td>\n",
       "      <td>63221</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>REVCELNPDCDELADHIGFQEAYR</td>\n",
       "      <td>24</td>\n",
       "      <td>[GO:0005576, GO:0005509, GO:0005179, GO:004684...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[P84351]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>Q5TGZ0-2</td>\n",
       "      <td>Isoform 2 of MICOS complex subunit MIC10</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>MSESELGRKWDRCLADAVVKIESW</td>\n",
       "      <td>24</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>Q68G75-2</td>\n",
       "      <td>Isoform 2 of LEM domain-containing protein 1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>MVDVKCLSDCKLQNQLEKLGFSPGPILLA</td>\n",
       "      <td>29</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>Q6Q0C1-2</td>\n",
       "      <td>Isoform 2 of Solute carrier family 25 member 47</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>MDFVAGAIGGVCGVAVGYPLDTVKGLLALP</td>\n",
       "      <td>30</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>Q8IVG9</td>\n",
       "      <td>Humanin</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>MAPRGFSCLLLLTSEIDLPVKRRA</td>\n",
       "      <td>24</td>\n",
       "      <td>[GO:0005737, GO:0005576, GO:0005615, GO:000573...</td>\n",
       "      <td>[PF15040]</td>\n",
       "      <td>[Q8IVG9]</td>\n",
       "      <td>[1Y32, 2GD3, 5GIW, 7WVX]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>Q8IZ96-15</td>\n",
       "      <td>Isoform 16 of CKLF-like MARVEL transmembrane d...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>MDPEHAKPESSEAPSGNLKQPETAAALR</td>\n",
       "      <td>28</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>Q8N688-2</td>\n",
       "      <td>Isoform 2 of Beta-defensin 123</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>MKLLLLTLTVLLLLSQLTPDLSNPESTC</td>\n",
       "      <td>28</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>Q8N769-2</td>\n",
       "      <td>Isoform 2 of Putative uncharacterized protein ...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>MGREMKKTGTPRPFRIEDPNQQPT</td>\n",
       "      <td>24</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>Q9NRI7</td>\n",
       "      <td>Putative pancreatic polypeptide 2</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>MAAACRCLSLLLLSTCVALLL</td>\n",
       "      <td>21</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[Q9NRI7]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>Q9NZM6-4</td>\n",
       "      <td>Isoform 4 of Polycystin-2-like protein 2</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>MAEASRWHRGGYVISIFGHFCAW</td>\n",
       "      <td>23</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>S4R3P1</td>\n",
       "      <td>Humanin-like 13</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>MDTQGFSCLLLLISEIDLSVKRRI</td>\n",
       "      <td>24</td>\n",
       "      <td>[GO:0005737, GO:0005576, GO:0048019, GO:1900118]</td>\n",
       "      <td>[PF15040]</td>\n",
       "      <td>[S4R3P1]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>S4R3Y5</td>\n",
       "      <td>Humanin-like 11</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>MATRGFSCLLLVISEIDLSVKRWV</td>\n",
       "      <td>24</td>\n",
       "      <td>[GO:0005737, GO:0005576, GO:0048019, GO:1900118]</td>\n",
       "      <td>[PF15040]</td>\n",
       "      <td>[S4R3Y5]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                query   accession  \\\n",
       "0   organism_name:homo sapiens (human) AND length:...  A0A075B6S0   \n",
       "1   organism_name:homo sapiens (human) AND length:...  A0A075B6Y3   \n",
       "2   organism_name:homo sapiens (human) AND length:...  A0A075B6Y9   \n",
       "3   organism_name:homo sapiens (human) AND length:...  A0A075B700   \n",
       "4   organism_name:homo sapiens (human) AND length:...  A0A075B706   \n",
       "5   organism_name:homo sapiens (human) AND length:...  A0A0A0MT70   \n",
       "6   organism_name:homo sapiens (human) AND length:...  A0A0A0MT78   \n",
       "7   organism_name:homo sapiens (human) AND length:...  A0A0A0MT87   \n",
       "8   organism_name:homo sapiens (human) AND length:...  A0A0A0MT94   \n",
       "9   organism_name:homo sapiens (human) AND length:...  A0A0A0MTA4   \n",
       "10  organism_name:homo sapiens (human) AND length:...  A0A0A0MTA7   \n",
       "11  organism_name:homo sapiens (human) AND length:...  A0A0B4J200   \n",
       "12  organism_name:homo sapiens (human) AND length:...  A0A0C4DH62   \n",
       "13  organism_name:homo sapiens (human) AND length:...  A0A0C5B5G6   \n",
       "14  organism_name:homo sapiens (human) AND length:...  A0A0J9YWP8   \n",
       "15  organism_name:homo sapiens (human) AND length:...  A0A0J9YWX3   \n",
       "16  organism_name:homo sapiens (human) AND length:...  A0A0J9YX06   \n",
       "17  organism_name:homo sapiens (human) AND length:...  A0A0J9YXA8   \n",
       "18  organism_name:homo sapiens (human) AND length:...  A0A0J9YXG5   \n",
       "19  organism_name:homo sapiens (human) AND length:...  A0A0J9YXM7   \n",
       "20  organism_name:homo sapiens (human) AND length:...  A0A1B0GWH6   \n",
       "21  organism_name:homo sapiens (human) AND length:...      C0HM02   \n",
       "22  organism_name:homo sapiens (human) AND length:...      C0HMA1   \n",
       "23  organism_name:homo sapiens (human) AND length:...      D0EPY3   \n",
       "24  organism_name:homo sapiens (human) AND length:...      P0CJ68   \n",
       "25  organism_name:homo sapiens (human) AND length:...      P0CJ69   \n",
       "26  organism_name:homo sapiens (human) AND length:...      P0CJ70   \n",
       "27  organism_name:homo sapiens (human) AND length:...      P0CJ71   \n",
       "28  organism_name:homo sapiens (human) AND length:...      P0CJ72   \n",
       "29  organism_name:homo sapiens (human) AND length:...      P0CJ73   \n",
       "30  organism_name:homo sapiens (human) AND length:...      P0CJ74   \n",
       "31  organism_name:homo sapiens (human) AND length:...      P0CJ75   \n",
       "32  organism_name:homo sapiens (human) AND length:...      P0CJ76   \n",
       "33  organism_name:homo sapiens (human) AND length:...      P0CJ77   \n",
       "34  organism_name:homo sapiens (human) AND length:...      P0DMP1   \n",
       "35  organism_name:homo sapiens (human) AND length:...    P16070-2   \n",
       "36  organism_name:homo sapiens (human) AND length:...    P59103-2   \n",
       "37  organism_name:homo sapiens (human) AND length:...      P62945   \n",
       "38  organism_name:homo sapiens (human) AND length:...      P84351   \n",
       "39  organism_name:homo sapiens (human) AND length:...    Q5TGZ0-2   \n",
       "40  organism_name:homo sapiens (human) AND length:...    Q68G75-2   \n",
       "41  organism_name:homo sapiens (human) AND length:...    Q6Q0C1-2   \n",
       "42  organism_name:homo sapiens (human) AND length:...      Q8IVG9   \n",
       "43  organism_name:homo sapiens (human) AND length:...   Q8IZ96-15   \n",
       "44  organism_name:homo sapiens (human) AND length:...    Q8N688-2   \n",
       "45  organism_name:homo sapiens (human) AND length:...    Q8N769-2   \n",
       "46  organism_name:homo sapiens (human) AND length:...      Q9NRI7   \n",
       "47  organism_name:homo sapiens (human) AND length:...    Q9NZM6-4   \n",
       "48  organism_name:homo sapiens (human) AND length:...      S4R3P1   \n",
       "49  organism_name:homo sapiens (human) AND length:...      S4R3Y5   \n",
       "\n",
       "                                         protein_name  \\\n",
       "0                     T cell receptor gamma joining 1   \n",
       "1                     T cell receptor alpha joining 3   \n",
       "2                    T cell receptor alpha joining 42   \n",
       "3                    T cell receptor alpha joining 31   \n",
       "4                     T cell receptor delta joining 1   \n",
       "5                    T cell receptor beta joining 2-6   \n",
       "6                    T cell receptor beta joining 2-7   \n",
       "7                    T cell receptor beta joining 2-4   \n",
       "8                    T cell receptor beta joining 2-2   \n",
       "9                    T cell receptor beta joining 2-5   \n",
       "10                   T cell receptor beta joining 2-1   \n",
       "11                   T cell receptor beta joining 2-3   \n",
       "12                     Immunoglobulin heavy joining 1   \n",
       "13               Mitochondrial-derived peptide MOTS-c   \n",
       "14                   T cell receptor beta joining 1-3   \n",
       "15                   T cell receptor beta joining 1-6   \n",
       "16                   T cell receptor beta joining 1-2   \n",
       "17                   T cell receptor beta joining 1-1   \n",
       "18                   T cell receptor beta joining 1-4   \n",
       "19                   T cell receptor beta joining 1-5   \n",
       "20                             LASP1 neighbor protein   \n",
       "21                PRKCH upstream open reading frame 2   \n",
       "22            HSPA8-interacting micropeptide miPEP155   \n",
       "23                 Small integral membrane protein 47   \n",
       "24                                     Humanin-like 1   \n",
       "25                                     Humanin-like 2   \n",
       "26                                     Humanin-like 3   \n",
       "27                                     Humanin-like 4   \n",
       "28                                     Humanin-like 5   \n",
       "29                                     Humanin-like 6   \n",
       "30                                     Humanin-like 7   \n",
       "31                                     Humanin-like 8   \n",
       "32                                     Humanin-like 9   \n",
       "33                                    Humanin-like 10   \n",
       "34                                    Humanin-like 12   \n",
       "35                          Isoform 2 of CD44 antigen   \n",
       "36        Isoform 2 of D-amino acid oxidase regulator   \n",
       "37               Small ribosomal subunit protein eS32   \n",
       "38                                        Osteocalcin   \n",
       "39           Isoform 2 of MICOS complex subunit MIC10   \n",
       "40       Isoform 2 of LEM domain-containing protein 1   \n",
       "41    Isoform 2 of Solute carrier family 25 member 47   \n",
       "42                                            Humanin   \n",
       "43  Isoform 16 of CKLF-like MARVEL transmembrane d...   \n",
       "44                     Isoform 2 of Beta-defensin 123   \n",
       "45  Isoform 2 of Putative uncharacterized protein ...   \n",
       "46                  Putative pancreatic polypeptide 2   \n",
       "47           Isoform 4 of Polycystin-2-like protein 2   \n",
       "48                                    Humanin-like 13   \n",
       "49                                    Humanin-like 11   \n",
       "\n",
       "                    organism_name  taxon_id  \\\n",
       "0                    Homo sapiens      9606   \n",
       "1                    Homo sapiens      9606   \n",
       "2                    Homo sapiens      9606   \n",
       "3                    Homo sapiens      9606   \n",
       "4                    Homo sapiens      9606   \n",
       "5                    Homo sapiens      9606   \n",
       "6                    Homo sapiens      9606   \n",
       "7                    Homo sapiens      9606   \n",
       "8                    Homo sapiens      9606   \n",
       "9                    Homo sapiens      9606   \n",
       "10                   Homo sapiens      9606   \n",
       "11                   Homo sapiens      9606   \n",
       "12                   Homo sapiens      9606   \n",
       "13                   Homo sapiens      9606   \n",
       "14                   Homo sapiens      9606   \n",
       "15                   Homo sapiens      9606   \n",
       "16                   Homo sapiens      9606   \n",
       "17                   Homo sapiens      9606   \n",
       "18                   Homo sapiens      9606   \n",
       "19                   Homo sapiens      9606   \n",
       "20                   Homo sapiens      9606   \n",
       "21                   Homo sapiens      9606   \n",
       "22                   Homo sapiens      9606   \n",
       "23                   Homo sapiens      9606   \n",
       "24                   Homo sapiens      9606   \n",
       "25                   Homo sapiens      9606   \n",
       "26                   Homo sapiens      9606   \n",
       "27                   Homo sapiens      9606   \n",
       "28                   Homo sapiens      9606   \n",
       "29                   Homo sapiens      9606   \n",
       "30                   Homo sapiens      9606   \n",
       "31                   Homo sapiens      9606   \n",
       "32                   Homo sapiens      9606   \n",
       "33                   Homo sapiens      9606   \n",
       "34                   Homo sapiens      9606   \n",
       "35                   Homo sapiens      9606   \n",
       "36                   Homo sapiens      9606   \n",
       "37                   Homo sapiens      9606   \n",
       "38  Homo sapiens neanderthalensis     63221   \n",
       "39                   Homo sapiens      9606   \n",
       "40                   Homo sapiens      9606   \n",
       "41                   Homo sapiens      9606   \n",
       "42                   Homo sapiens      9606   \n",
       "43                   Homo sapiens      9606   \n",
       "44                   Homo sapiens      9606   \n",
       "45                   Homo sapiens      9606   \n",
       "46                   Homo sapiens      9606   \n",
       "47                   Homo sapiens      9606   \n",
       "48                   Homo sapiens      9606   \n",
       "49                   Homo sapiens      9606   \n",
       "\n",
       "                                               ineage  \\\n",
       "0   [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "1   [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "2   [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "3   [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "4   [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "5   [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "6   [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "7   [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "8   [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "9   [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "10  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "11  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "12  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "13  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "14  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "15  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "16  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "17  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "18  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "19  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "20  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "21  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "22  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "23  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "24  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "25  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "26  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "27  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "28  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "29  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "30  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "31  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "32  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "33  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "34  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "35  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "36  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "37  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "38  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "39  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "40  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "41  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "42  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "43  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "44  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "45  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "46  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "47  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "48  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "49  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "\n",
       "                          sequence  length  \\\n",
       "0                 NYYKKLFGSGTTLVVT      16   \n",
       "1             GYSSASKIIFGSGTRLSIRP      20   \n",
       "2             YGGSQGNLIFGKGTKLSVKP      20   \n",
       "3               NNNARLMFGDGTQLVVKP      18   \n",
       "4                 TDKLIFGKGTRVTVEP      16   \n",
       "5                SGANVLTFGAGSRLTVL      17   \n",
       "6                  SYEQYFGPGTRLTVT      15   \n",
       "7                 AKNIQYFGAGTRLSVL      16   \n",
       "8                 NTGELFFGEGSRLTVL      16   \n",
       "9                  QETQYFGPGTRLLVL      15   \n",
       "10                SYNEQFFGPGTRLTVL      16   \n",
       "11                STDTQYFGPGTRLTVL      16   \n",
       "12               AEYFQHWGQGTLVTVSS      17   \n",
       "13                MRWQEMGYIFYPRKLR      16   \n",
       "14                SGNTIYFGEGSWLTVV      16   \n",
       "15               SYNSPLHFGNGTRLTVT      17   \n",
       "16                 NYGYTFGSGTRLTVV      15   \n",
       "17                 NTEAFFGQGTRLTVV      15   \n",
       "18                TNEKLFFGSGTQLSVL      16   \n",
       "19                SNQPQHFGDGTRLSIL      16   \n",
       "20       MLDIFILMFFAIIGLVILSYIIYLL      25   \n",
       "21      MASRGALRRCLSPGLPRLLHLSRGLA      26   \n",
       "22               MEMALMVAQTRKGKSVV      17   \n",
       "23    MNFQENVTLAMALFTILTSIYFFNKAQQ      28   \n",
       "24        MAPRGFSCLLLSTSEIDLPVKRRT      24   \n",
       "25    MAPRGFSCLLLSTSEIDLPVKRLLSSVF      28   \n",
       "26        MATRRFSCLLLSTSEIDLSVKRRI      24   \n",
       "27    MATQGFSCLLLSVSEIDLSMKRQYKQIR      28   \n",
       "28        MATPGFSCLLLSTSEIDLPMKRRV      24   \n",
       "29        MTPRGFSCLLLPTSETDLPVKRRT      24   \n",
       "30        MATGGFGCLLLLIREIDLSVKRQI      24   \n",
       "31        MAPRGFSCLLLSTSEIDLPVKRRA      24   \n",
       "32        MARRGFSCLLLSTTATDLPVKRRT      24   \n",
       "33        MTTRGFSCLLLLIREIDLSAKRRI      24   \n",
       "34        MAPRGFSCLLLSTSEIDLPVKRRA      24   \n",
       "35   MDKFWWHAAWGLCLVPLSLAQIGVGRRKS      29   \n",
       "36                MLEKLMGADSLQLFRV      16   \n",
       "37       MRAKWRKKRMRRLKRKRRKMRQRSK      25   \n",
       "38        REVCELNPDCDELADHIGFQEAYR      24   \n",
       "39        MSESELGRKWDRCLADAVVKIESW      24   \n",
       "40   MVDVKCLSDCKLQNQLEKLGFSPGPILLA      29   \n",
       "41  MDFVAGAIGGVCGVAVGYPLDTVKGLLALP      30   \n",
       "42        MAPRGFSCLLLLTSEIDLPVKRRA      24   \n",
       "43    MDPEHAKPESSEAPSGNLKQPETAAALR      28   \n",
       "44    MKLLLLTLTVLLLLSQLTPDLSNPESTC      28   \n",
       "45        MGREMKKTGTPRPFRIEDPNQQPT      24   \n",
       "46           MAAACRCLSLLLLSTCVALLL      21   \n",
       "47         MAEASRWHRGGYVISIFGHFCAW      23   \n",
       "48        MDTQGFSCLLLLISEIDLSVKRRI      24   \n",
       "49        MATRGFSCLLLVISEIDLSVKRWV      24   \n",
       "\n",
       "                                             go_terms   pfam_ids  \\\n",
       "0                            [GO:0042101, GO:0002250]         []   \n",
       "1                            [GO:0042101, GO:0002250]         []   \n",
       "2                            [GO:0005886, GO:0002250]         []   \n",
       "3                            [GO:0005886, GO:0002250]         []   \n",
       "4                            [GO:0042101, GO:0002250]         []   \n",
       "5                            [GO:0042101, GO:0002250]         []   \n",
       "6                            [GO:0042101, GO:0002250]         []   \n",
       "7                            [GO:0042101, GO:0002250]         []   \n",
       "8                            [GO:0042101, GO:0002250]         []   \n",
       "9                            [GO:0042101, GO:0002250]         []   \n",
       "10                           [GO:0042101, GO:0002250]         []   \n",
       "11                           [GO:0042101, GO:0002250]         []   \n",
       "12   [GO:0005576, GO:0019814, GO:0005886, GO:0002250]         []   \n",
       "13  [GO:0005615, GO:0005739, GO:0005634, GO:000367...  [PF21945]   \n",
       "14                           [GO:0042101, GO:0002250]         []   \n",
       "15                           [GO:0042101, GO:0002250]         []   \n",
       "16                           [GO:0042101, GO:0002250]         []   \n",
       "17                           [GO:0042101, GO:0002250]         []   \n",
       "18                           [GO:0042101, GO:0002250]         []   \n",
       "19                           [GO:0042101, GO:0002250]         []   \n",
       "20                                       [GO:0016020]         []   \n",
       "21  [GO:0120283, GO:0030291, GO:0045182, GO:003419...  [PF21952]   \n",
       "22  [GO:0005737, GO:0005634, GO:0030544, GO:000237...         []   \n",
       "23                                       [GO:0016020]         []   \n",
       "24   [GO:0005737, GO:0005576, GO:0048019, GO:1900118]  [PF15040]   \n",
       "25   [GO:0005737, GO:0005576, GO:0048019, GO:1900118]  [PF15040]   \n",
       "26   [GO:0005737, GO:0005576, GO:0048019, GO:1900118]  [PF15040]   \n",
       "27   [GO:0005737, GO:0005576, GO:0048019, GO:1900118]  [PF15040]   \n",
       "28  [GO:0005576, GO:0005739, GO:0048019, GO:007273...  [PF15040]   \n",
       "29   [GO:0005737, GO:0005576, GO:0048019, GO:1900118]  [PF15040]   \n",
       "30   [GO:0005737, GO:0005576, GO:0048019, GO:1900118]  [PF15040]   \n",
       "31   [GO:0005737, GO:0005576, GO:0048019, GO:1900118]  [PF15040]   \n",
       "32   [GO:0005737, GO:0005576, GO:0048019, GO:1900118]  [PF15040]   \n",
       "33   [GO:0005737, GO:0005576, GO:0048019, GO:1900118]  [PF15040]   \n",
       "34                           [GO:0005737, GO:0005576]  [PF15040]   \n",
       "35                                               None       None   \n",
       "36                                               None       None   \n",
       "37  [GO:0005737, GO:0005829, GO:0022625, GO:000578...  [PF05162]   \n",
       "38  [GO:0005576, GO:0005509, GO:0005179, GO:004684...         []   \n",
       "39                                               None       None   \n",
       "40                                               None       None   \n",
       "41                                               None       None   \n",
       "42  [GO:0005737, GO:0005576, GO:0005615, GO:000573...  [PF15040]   \n",
       "43                                               None       None   \n",
       "44                                               None       None   \n",
       "45                                               None       None   \n",
       "46                                                 []         []   \n",
       "47                                               None       None   \n",
       "48   [GO:0005737, GO:0005576, GO:0048019, GO:1900118]  [PF15040]   \n",
       "49   [GO:0005737, GO:0005576, GO:0048019, GO:1900118]  [PF15040]   \n",
       "\n",
       "   alphafold_ids                                            pdb_ids kegg_ids  \\\n",
       "0   [A0A075B6S0]                                                 []       []   \n",
       "1   [A0A075B6Y3]                                                 []       []   \n",
       "2   [A0A075B6Y9]                                                 []       []   \n",
       "3   [A0A075B700]                                                 []       []   \n",
       "4   [A0A075B706]                                                 []       []   \n",
       "5   [A0A0A0MT70]                                                 []       []   \n",
       "6             []                                                 []       []   \n",
       "7   [A0A0A0MT87]                                                 []       []   \n",
       "8   [A0A0A0MT94]                                                 []       []   \n",
       "9             []                                                 []       []   \n",
       "10  [A0A0A0MTA7]                                                 []       []   \n",
       "11  [A0A0B4J200]                                             [7NDT]       []   \n",
       "12  [A0A0C4DH62]                                                 []       []   \n",
       "13  [A0A0C5B5G6]                                                 []       []   \n",
       "14  [A0A0J9YWP8]                                                 []       []   \n",
       "15  [A0A0J9YWX3]                                                 []       []   \n",
       "16            []                                       [7NDQ, 7NDU]       []   \n",
       "17            []                                                 []       []   \n",
       "18  [A0A0J9YXG5]                                                 []       []   \n",
       "19  [A0A0J9YXM7]                                                 []       []   \n",
       "20  [A0A1B0GWH6]                                                 []       []   \n",
       "21            []                                                 []       []   \n",
       "22            []                                                 []       []   \n",
       "23      [D0EPY3]                                                 []       []   \n",
       "24      [P0CJ68]                                                 []       []   \n",
       "25      [P0CJ69]                                                 []       []   \n",
       "26      [P0CJ70]                                                 []       []   \n",
       "27      [P0CJ71]                                                 []       []   \n",
       "28      [P0CJ72]                                                 []       []   \n",
       "29      [P0CJ73]                                                 []       []   \n",
       "30      [P0CJ74]                                                 []       []   \n",
       "31      [P0CJ75]                                                 []       []   \n",
       "32      [P0CJ76]                                                 []       []   \n",
       "33      [P0CJ77]                                                 []       []   \n",
       "34      [P0DMP1]                                                 []       []   \n",
       "35          None                                               None     None   \n",
       "36          None                                               None     None   \n",
       "37      [P62945]  [4UG0, 4V6X, 5A2Q, 5AJ0, 5LKS, 5OA3, 5VYC, 6G5...       []   \n",
       "38      [P84351]                                                 []       []   \n",
       "39          None                                               None     None   \n",
       "40          None                                               None     None   \n",
       "41          None                                               None     None   \n",
       "42      [Q8IVG9]                           [1Y32, 2GD3, 5GIW, 7WVX]       []   \n",
       "43          None                                               None     None   \n",
       "44          None                                               None     None   \n",
       "45          None                                               None     None   \n",
       "46      [Q9NRI7]                                                 []       []   \n",
       "47          None                                               None     None   \n",
       "48      [S4R3P1]                                                 []       []   \n",
       "49      [S4R3Y5]                                                 []       []   \n",
       "\n",
       "   brenda_ids reactome_ids refseq_ids interpro_ids  \n",
       "0          []           []         []           []  \n",
       "1          []           []         []           []  \n",
       "2          []           []         []           []  \n",
       "3          []           []         []           []  \n",
       "4          []           []         []           []  \n",
       "5          []           []         []           []  \n",
       "6          []           []         []           []  \n",
       "7          []           []         []           []  \n",
       "8          []           []         []           []  \n",
       "9          []           []         []           []  \n",
       "10         []           []         []           []  \n",
       "11         []           []         []           []  \n",
       "12         []           []         []           []  \n",
       "13         []           []         []           []  \n",
       "14         []           []         []           []  \n",
       "15         []           []         []           []  \n",
       "16         []           []         []           []  \n",
       "17         []           []         []           []  \n",
       "18         []           []         []           []  \n",
       "19         []           []         []           []  \n",
       "20         []           []         []           []  \n",
       "21         []           []         []           []  \n",
       "22         []           []         []           []  \n",
       "23         []           []         []           []  \n",
       "24         []           []         []           []  \n",
       "25         []           []         []           []  \n",
       "26         []           []         []           []  \n",
       "27         []           []         []           []  \n",
       "28         []           []         []           []  \n",
       "29         []           []         []           []  \n",
       "30         []           []         []           []  \n",
       "31         []           []         []           []  \n",
       "32         []           []         []           []  \n",
       "33         []           []         []           []  \n",
       "34         []           []         []           []  \n",
       "35       None         None       None         None  \n",
       "36       None         None       None         None  \n",
       "37         []           []         []           []  \n",
       "38         []           []         []           []  \n",
       "39       None         None       None         None  \n",
       "40       None         None       None         None  \n",
       "41       None         None       None         None  \n",
       "42         []           []         []           []  \n",
       "43       None         None       None         None  \n",
       "44       None         None       None         None  \n",
       "45       None         None       None         None  \n",
       "46         []           []         []           []  \n",
       "47       None         None       None         None  \n",
       "48         []           []         []           []  \n",
       "49         []           []         []           []  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "uniprot.parse_stream_response(\n",
    "    query=query,\n",
    "    response=response\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "301959e5",
   "metadata": {},
   "source": [
    "## Activity search"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "06c29588",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59f9589e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['uniprot_celiac-toxic.csv',\n",
       " 'uniprot_embryotoxic.csv',\n",
       " 'uniprot_ace-inhibitor.csv',\n",
       " 'uniprot_anuran-defense.csv',\n",
       " 'uniprot_campde-inhibitor.csv',\n",
       " 'uniprot_anti-neurotensive.csv',\n",
       " 'uniprot_antitrypanosomic.csv',\n",
       " 'uniprot_anticancer.csv',\n",
       " 'uniprot_anorectic.csv',\n",
       " 'uniprot_chemotactic.csv',\n",
       " 'uniprot_targeting-GP.csv',\n",
       " 'uniprot_Blood-Brain-Barrier.csv',\n",
       " 'uniprot_antitumor.csv',\n",
       " 'uniprot_cytotoxic.csv',\n",
       " 'uniprot_activating-ubiquitin-mediated-proteolysis.csv',\n",
       " 'uniprot_protein-kinase-c-inhibitor.csv',\n",
       " 'uniprot_antihiv.csv',\n",
       " 'uniprot_antidiabetic.csv',\n",
       " 'uniprot_calpain-2-inhibitor.csv',\n",
       " 'uniprot_antileishmania.csv',\n",
       " 'uniprot_inhibitor.csv',\n",
       " 'uniprot_antimicrobial.csv',\n",
       " 'uniprot_antituberculosis.csv',\n",
       " 'uniprot_antiviral.csv',\n",
       " 'uniprot_wound-healing.csv',\n",
       " 'uniprot_targeting-GN.csv',\n",
       " 'uniprot_hmg-coa-reductase-inhibitor.csv',\n",
       " 'uniprot_opioid-agonist.csv',\n",
       " 'uniprot_toxicology.csv',\n",
       " 'uniprot_antibacterial.csv',\n",
       " 'uniprot_antibiofilm.csv',\n",
       " 'uniprot_hypocholesterolemic.csv',\n",
       " 'uniprot_bacterial-permease-ligand.csv',\n",
       " 'uniprot_protease-inhibitor.csv',\n",
       " 'uniprot_antitoxin.csv',\n",
       " 'uniprot_polystyrene-surface-binding.csv',\n",
       " 'uniprot_stimulating.csv',\n",
       " 'uniprot_immunomodulating.csv',\n",
       " 'uniprot_opioid.csv',\n",
       " 'uniprot_dipeptidyl-peptidase-iii-inhibitor.csv',\n",
       " 'uniprot_tumor-homing.csv',\n",
       " 'uniprot_heparin-binding.csv',\n",
       " 'uniprot_anticoronavirus.csv',\n",
       " 'uniprot_targeting-mammals.csv',\n",
       " 'uniprot_antiplasmodial.csv',\n",
       " 'uniprot_uterotonic.csv',\n",
       " 'uniprot_cell-cell-comunication.csv',\n",
       " 'uniprot_binding.csv',\n",
       " 'uniprot_peptide-hormones.csv',\n",
       " 'uniprot_drug-delivery.csv',\n",
       " 'uniprot_anti-MRSA.csv',\n",
       " 'uniprot_antihypertensive.csv',\n",
       " 'uniprot_renin-inhibitor.csv',\n",
       " 'uniprot_neuroactive.csv',\n",
       " 'uniprot_surface-binding.csv',\n",
       " 'uniprot_antiantifreeze.csv',\n",
       " 'uniprot_insecticidal.csv',\n",
       " 'uniprot_antiprotozoal.csv',\n",
       " 'uniprot_alpha-amylase-inhibitor.csv',\n",
       " 'uniprot_antiparasitic.csv',\n",
       " 'uniprot_alpha-glucosidase-inhibitor.csv',\n",
       " 'uniprot_neuropeptide.csv',\n",
       " 'uniprot_trypsin-inhibitory.csv',\n",
       " 'uniprot_orphan-receptor-gpr14-agonist.csv',\n",
       " 'uniprot_antithrombotic.csv',\n",
       " 'uniprot_antiamnestic.csv',\n",
       " 'uniprot_antifungal.csv',\n",
       " 'uniprot_antiangiogenic.csv',\n",
       " 'uniprot_antioxidative.csv',\n",
       " 'uniprot_natriuretic.csv',\n",
       " 'uniprot_cell-penetrating.csv',\n",
       " 'uniprot_quorum-sensing.csv',\n",
       " 'uniprot_dipeptidyl-peptidaseiv.csv',\n",
       " 'uniprot_regulating.csv',\n",
       " 'uniprot_antioncogenic.csv',\n",
       " 'uniprot_contracting.csv',\n",
       " 'uniprot_immunostimulating.csv',\n",
       " 'uniprot_membrane-active-peptide.csv',\n",
       " 'uniprot_hypolipidemic.csv',\n",
       " 'uniprot_antiinflammatory.csv',\n",
       " 'uniprot_vasoconstrictor.csv',\n",
       " 'uniprot_opioid-antagonist.csv',\n",
       " 'uniprot_targeting-GV.csv',\n",
       " 'uniprot_hypotensive.csv',\n",
       " 'uniprot_edema-inducer.csv',\n",
       " 'uniprot_antiendotoxin.csv',\n",
       " 'uniprot_hemolytic.csv',\n",
       " 'uniprot_camkii-inhibitor.csv',\n",
       " 'uniprot_cytokine-growth-factor.csv']"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "uniprot_search_files = os.listdir(\"uniprot_search\")\n",
    "uniprot_search_files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ecb7ce2c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>activity</th>\n",
       "      <th>sequence_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Binding</td>\n",
       "      <td>411281</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Inhibitor</td>\n",
       "      <td>18950</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Surface binding</td>\n",
       "      <td>15663</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Regulating</td>\n",
       "      <td>11361</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Antimicrobial</td>\n",
       "      <td>5945</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>Dipeptidyl peptidaseiv</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>Antiamnestic</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>Edema inducer</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>Antiendotoxin</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>Total</td>\n",
       "      <td>502774</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>90 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  activity  sequence_count\n",
       "0                  Binding          411281\n",
       "1                Inhibitor           18950\n",
       "2          Surface binding           15663\n",
       "3               Regulating           11361\n",
       "4            Antimicrobial            5945\n",
       "..                     ...             ...\n",
       "85  Dipeptidyl peptidaseiv               0\n",
       "86            Antiamnestic               0\n",
       "87           Edema inducer               0\n",
       "88           Antiendotoxin               0\n",
       "89                   Total          502774\n",
       "\n",
       "[90 rows x 2 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create an empty list to store the results\n",
    "activity_data = []\n",
    "\n",
    "# Iterate over each file in uniprot_search_files\n",
    "for file in uniprot_search_files:\n",
    "    # Load the CSV file into a DataFrame\n",
    "    df = pd.read_csv(f\"uniprot_search/{file}\")\n",
    "    \n",
    "    # Extract the activity name from the file name\n",
    "    activity = file.replace(\"uniprot_\", \"\").replace(\".csv\", \"\").replace(\"-\", \" \").capitalize()\n",
    "    \n",
    "    # Count the number of sequences in the DataFrame\n",
    "    sequence_count = len(df)\n",
    "    \n",
    "    # Append the activity and sequence count to the list\n",
    "    activity_data.append({\"activity\": activity, \"sequence_count\": sequence_count})\n",
    "\n",
    "# Create a DataFrame from the activity data\n",
    "activity_df = pd.DataFrame(activity_data)\n",
    "\n",
    "# Sort the DataFrame by sequence count in descending order\n",
    "activity_df = activity_df.sort_values(by=\"sequence_count\", ascending=False)\n",
    "\n",
    "# Calculate the total number of sequences\n",
    "total_sequences = activity_df[\"sequence_count\"].sum()\n",
    "\n",
    "# Add a row for the total sequences\n",
    "activity_df = pd.concat([activity_df, pd.DataFrame([{\"activity\": \"Total\", \"sequence_count\": total_sequences}])], ignore_index=True)\n",
    "\n",
    "# Display the resulting DataFrame\n",
    "activity_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4abc5ec2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>activity</th>\n",
       "      <th>sequence_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Antimicrobial</td>\n",
       "      <td>5945</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Antiviral</td>\n",
       "      <td>3835</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Antibacterial</td>\n",
       "      <td>2366</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Antifungal</td>\n",
       "      <td>1209</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Antitoxin</td>\n",
       "      <td>832</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Antitumor</td>\n",
       "      <td>479</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Anticancer</td>\n",
       "      <td>393</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>Antiparasitic</td>\n",
       "      <td>140</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>Antiangiogenic</td>\n",
       "      <td>86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>Antioxidative</td>\n",
       "      <td>55</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>Antithrombotic</td>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>Antidiabetic</td>\n",
       "      <td>42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>Antiprotozoal</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>Anti mrsa</td>\n",
       "      <td>31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>Antiinflammatory</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>Antihypertensive</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>Antituberculosis</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>Antibiofilm</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>Antiplasmodial</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>Antioncogenic</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>Antileishmania</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>Antihiv</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>Antitrypanosomic</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>Anti neurotensive</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>Antiantifreeze</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>Anticoronavirus</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>Antiamnestic</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>Antiendotoxin</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             activity  sequence_count\n",
       "4       Antimicrobial            5945\n",
       "6           Antiviral            3835\n",
       "10      Antibacterial            2366\n",
       "17         Antifungal            1209\n",
       "19          Antitoxin             832\n",
       "24          Antitumor             479\n",
       "25         Anticancer             393\n",
       "32      Antiparasitic             140\n",
       "37     Antiangiogenic              86\n",
       "41      Antioxidative              55\n",
       "42     Antithrombotic              50\n",
       "45       Antidiabetic              42\n",
       "48      Antiprotozoal              33\n",
       "49          Anti mrsa              31\n",
       "52   Antiinflammatory              27\n",
       "56   Antihypertensive              21\n",
       "60   Antituberculosis              15\n",
       "66        Antibiofilm               9\n",
       "67     Antiplasmodial               8\n",
       "71      Antioncogenic               2\n",
       "73     Antileishmania               0\n",
       "74            Antihiv               0\n",
       "79   Antitrypanosomic               0\n",
       "80  Anti neurotensive               0\n",
       "81     Antiantifreeze               0\n",
       "83    Anticoronavirus               0\n",
       "86       Antiamnestic               0\n",
       "88      Antiendotoxin               0"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "anti_activities_df = activity_df[activity_df['activity'].str.startswith('Anti', na=False)]\n",
    "anti_activities_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "76fc2d02",
   "metadata": {},
   "source": [
    "## ALphafold"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "2e172ef8",
   "metadata": {},
   "outputs": [],
   "source": [
    "fields = {\n",
    "    \"entry\": \"entryId\",\n",
    "    \"gene\": \"gene\",\n",
    "    \"tax_id\": \"taxId\",\n",
    "    \"sequence\": \"uniprotSequence\",\n",
    "    \"model_created\": \"modelCreatedDate\",\n",
    "    \"latest_version\": \"latestVersion\",\n",
    "    \"all_versions\": \"allVersions\"\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "ada4900c",
   "metadata": {},
   "outputs": [],
   "source": [
    "from src.alphafold import AlphafoldInterface\n",
    "from src.utils import get_feature_keys"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "37b81082",
   "metadata": {},
   "outputs": [],
   "source": [
    "instance = AlphafoldInterface(\n",
    "    max_workers=2, \n",
    "    structures=['pdb'], \n",
    "    fields_to_extract=fields\n",
    ")\n",
    "uniprot_ids = [\"P02666\", \"Q9TSI0\", \"P33048\", \"P11839\", \"O15552\", \"P76011\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "64d5bb9a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'entryId': 'str',\n",
       " 'gene': 'str',\n",
       " 'sequenceChecksum': 'str',\n",
       " 'sequenceVersionDate': 'str',\n",
       " 'uniprotAccession': 'str',\n",
       " 'uniprotId': 'str',\n",
       " 'uniprotDescription': 'str',\n",
       " 'taxId': 'int',\n",
       " 'organismScientificName': 'str',\n",
       " 'uniprotStart': 'int',\n",
       " 'uniprotEnd': 'int',\n",
       " 'uniprotSequence': 'str',\n",
       " 'modelCreatedDate': 'str',\n",
       " 'latestVersion': 'int',\n",
       " 'allVersions': 'list(int)',\n",
       " 'bcifUrl': 'str',\n",
       " 'cifUrl': 'str',\n",
       " 'pdbUrl': 'str',\n",
       " 'paeImageUrl': 'str',\n",
       " 'paeDocUrl': 'str',\n",
       " 'amAnnotationsUrl': 'NoneType',\n",
       " 'amAnnotationsHg19Url': 'NoneType',\n",
       " 'amAnnotationsHg38Url': 'NoneType',\n",
       " 'isReviewed': 'bool',\n",
       " 'isReferenceProteome': 'bool'}"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_feature_keys(instance.fetch_prediction(\"P02666\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cee11808",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetching prediction for P02666\n",
      "Fetching prediction for Q9TSI0\n",
      "Fetching prediction for P33048\n",
      "Fetching prediction for P11839\n",
      "Fetching prediction for O15552\n",
      "Fetching prediction for P76011\n"
     ]
    }
   ],
   "source": [
    "instance.download_from_uniprot_ids(uniprot_ids).to_csv(\"results/alphafold.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9837e24a",
   "metadata": {},
   "source": [
    "## PDB"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ad092f1",
   "metadata": {},
   "source": [
    "Puedo usar rcbapi.search para hacer busquedas como en uniprot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d9d8b4a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "from src.proteindatabank import PDBInterface"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2e12cffe",
   "metadata": {},
   "outputs": [],
   "source": [
    "fields = {\n",
    "    \"rcsb_id\": \"rcsb_id\",\n",
    "    \"2\": \"rcsb_comp_model_provenance\",\n",
    "    \"3\": \"rcsb_entry_info.branched_molecular_weight_minimum\",\n",
    "    \"4\": \"rcsb_entry_info.resolution_combined\",\n",
    "    \"5\": \"rcsb_entry_info.experimental_method\",\n",
    "    \"6\": \"rcsb_entry_info.diffrn_resolution_high.value\",\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "795d7a99",
   "metadata": {},
   "outputs": [],
   "source": [
    "fields = [\"rcsb_id\", \"rcsb_entry_info.experimental_method\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0c719612",
   "metadata": {},
   "outputs": [],
   "source": [
    "instance = PDBInterface(\n",
    "    max_workers=2,\n",
    "    fields_to_extract=fields, \n",
    "    return_data_list=[\"rcsb_id\", \"rcsb_comp_model_provenance\", \"rcsb_entry_info\"],\n",
    "    output_dir=\"test\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3b3f66ed",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 1/1 [00:00<00:00,  1.29it/s]\n"
     ]
    }
   ],
   "source": [
    "instance.fetch_to_datafame(\n",
    "    [\"4HHB\", \"1A8I\", \"1A8J\", \"1A8K\", \"1A8L\", \"1A8M\"]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5337607d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Downloading 4HHB in pdb format...\n",
      "Downloading 1A8I in pdb format...\n",
      "Downloading 1A8J in pdb format...\n",
      "Downloading 1A8K in pdb format...\n",
      "Downloading 1A8L in pdb format...\n",
      "Downloading 1A8M in pdb format...\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['/home/diegof/Documents/BioSeqDownloader/test/4HHB.pdb',\n",
       " '/home/diegof/Documents/BioSeqDownloader/test/1A8I.pdb',\n",
       " '/home/diegof/Documents/BioSeqDownloader/test/1A8J.pdb',\n",
       " '/home/diegof/Documents/BioSeqDownloader/test/1A8K.pdb',\n",
       " '/home/diegof/Documents/BioSeqDownloader/test/1A8L.pdb',\n",
       " '/home/diegof/Documents/BioSeqDownloader/test/1A8M.pdb']"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "instance.download_structures(\n",
    "    [\"4HHB\", \"1A8I\", \"1A8J\", \"1A8K\", \"1A8L\", \"1A8M\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ed86b941",
   "metadata": {},
   "source": [
    "## Refseq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "700d5f92",
   "metadata": {},
   "outputs": [],
   "source": [
    "from src.refseq import RefSeqInterface\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "66badd75",
   "metadata": {},
   "outputs": [],
   "source": [
    "features = {\n",
    "    \"1\": \"GBSeq_locus\",\n",
    "    \"2\": \"GBSeq_length\",\n",
    "    \"3\": \"GBSeq_keywords\",\n",
    "    \"4\": \"GBSeq_feature-table.GBFeature_intervals\",\n",
    "    \"5\": \"GBSeq_feature-table.GBFeature_intervals.GBInterval_from\"\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "06bb471b",
   "metadata": {},
   "outputs": [],
   "source": [
    "features = [\n",
    "    \"GBSeq_locus\",\n",
    "    \"GBSeq_length\",\n",
    "    \"GBSeq_keywords\",\n",
    "    \"GBSeq_feature-table.GBFeature_intervals\",\n",
    "    \"GBSeq_feature-table.GBFeature_intervals.GBInterval_from\"\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "82f71c4c",
   "metadata": {},
   "outputs": [],
   "source": [
    "instance = RefSeqInterface(\n",
    "    max_workers=2,\n",
    "    #fields_to_extract=features,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63f53aa2",
   "metadata": {},
   "outputs": [],
   "source": [
    "instance.download_from_refseq_ids(\n",
    "    [\"XP_010804480.1\", \"XP_010804481.1\", \"XP_010804482.1\"]\n",
    ").to_csv(\"results/refseq.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "565e82a3",
   "metadata": {},
   "source": [
    "## Brenda"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a2220e9d",
   "metadata": {},
   "outputs": [],
   "source": [
    "from src.brenda import BrendaInstance\n",
    "from src.utils import get_feature_keys"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63c9ac2a",
   "metadata": {},
   "outputs": [],
   "source": [
    "instance = BrendaInstance(\n",
    "    email=\"diego.fernandez@umag.cl\",\n",
    "    password=\"\",\n",
    "    functions=[\"getPhStability\", \"getKmValue\", \"getIc50Value\"],\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c1fa871f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Executing getPhStability\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'phStabilityMaximum': 'str',\n",
       " 'phStability': 'str',\n",
       " 'literature': 'list(int)',\n",
       " 'commentary': 'str',\n",
       " 'organism': 'str',\n",
       " 'ecNumber': 'str'}"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_feature_keys(instance.fetch(\"1.1.1.1\", \"Homo sapiens\", \"getPhStability\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "1018f6e4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Executing getPhStability\n",
      "Executing getKmValue\n",
      "Executing getIc50Value\n",
      "Error fetching data for 1.1.1.1 in Homo sapiens: ('Connection aborted.', RemoteDisconnected('Remote end closed connection without response'))\n",
      "Saved getKmValue data for 1.1.1.1 in Homo sapiens to cache.\n"
     ]
    }
   ],
   "source": [
    "instance.fetch_to_dataframe([(\"1.1.1.1\", \"Homo sapiens\")])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6b8dec67",
   "metadata": {},
   "source": [
    "## Reactome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ac49e8e8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'dbId': 'int',\n",
       " 'displayName': 'str',\n",
       " 'stId': 'str',\n",
       " 'stIdVersion': 'str',\n",
       " 'isInDisease': 'bool',\n",
       " 'isInferred': 'bool',\n",
       " 'name': 'list(str)',\n",
       " 'releaseDate': 'str',\n",
       " 'speciesName': 'str',\n",
       " 'category': 'str',\n",
       " 'className': 'str',\n",
       " 'schemaClass': 'str'}"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from src.reactome import ReactomeInstance\n",
    "from src.utils import get_feature_keys\n",
    "import pandas as pd\n",
    "\n",
    "instance = ReactomeInstance(\n",
    "    max_workers=2\n",
    ")\n",
    "get_feature_keys(instance.fetch(\"R-DME-1834941\")[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e9b3601e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dbId</th>\n",
       "      <th>displayName</th>\n",
       "      <th>stId</th>\n",
       "      <th>stIdVersion</th>\n",
       "      <th>isInDisease</th>\n",
       "      <th>isInferred</th>\n",
       "      <th>name</th>\n",
       "      <th>releaseDate</th>\n",
       "      <th>speciesName</th>\n",
       "      <th>category</th>\n",
       "      <th>className</th>\n",
       "      <th>schemaClass</th>\n",
       "      <th>hasDiagram</th>\n",
       "      <th>hasEHLD</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10790592</td>\n",
       "      <td>STING binds cyclic GMP-AMP</td>\n",
       "      <td>R-DME-3244643</td>\n",
       "      <td>R-DME-3244643.1</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[STING binds cyclic GMP-AMP]</td>\n",
       "      <td>2025-03-26</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>omitted</td>\n",
       "      <td>Reaction</td>\n",
       "      <td>BlackBoxEvent</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>10819344</td>\n",
       "      <td>DDX41 binds bacterial c-di-AMP, c-di-GMP</td>\n",
       "      <td>R-DME-9013869</td>\n",
       "      <td>R-DME-9013869.1</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[DDX41 binds bacterial c-di-AMP, c-di-GMP  ]</td>\n",
       "      <td>2025-03-26</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>binding</td>\n",
       "      <td>Reaction</td>\n",
       "      <td>Reaction</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>10790177</td>\n",
       "      <td>STING dimerization</td>\n",
       "      <td>R-DME-3134800</td>\n",
       "      <td>R-DME-3134800.1</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[STING dimerization]</td>\n",
       "      <td>2025-03-26</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>binding</td>\n",
       "      <td>Reaction</td>\n",
       "      <td>Reaction</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>10788129</td>\n",
       "      <td>STING binds c-di-GMP</td>\n",
       "      <td>R-DME-2396009</td>\n",
       "      <td>R-DME-2396009.1</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[STING binds c-di-GMP]</td>\n",
       "      <td>2025-03-26</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>omitted</td>\n",
       "      <td>Reaction</td>\n",
       "      <td>BlackBoxEvent</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>10829117</td>\n",
       "      <td>STAT6-mediated induction of chemokines</td>\n",
       "      <td>R-DME-3249367</td>\n",
       "      <td>R-DME-3249367.1</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[STAT6-mediated induction of chemokines]</td>\n",
       "      <td>2025-03-26</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Pathway</td>\n",
       "      <td>Pathway</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>10790821</td>\n",
       "      <td>STING recruits TBK1 and STAT6</td>\n",
       "      <td>R-DME-3249378</td>\n",
       "      <td>R-DME-3249378.1</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[STING recruits TBK1 and STAT6]</td>\n",
       "      <td>2025-03-26</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>binding</td>\n",
       "      <td>Reaction</td>\n",
       "      <td>Reaction</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>10790823</td>\n",
       "      <td>p-S407-STAT6 is phosphorylated at Tyr641</td>\n",
       "      <td>R-DME-3249379</td>\n",
       "      <td>R-DME-3249379.1</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[p-S407-STAT6 is phosphorylated at Tyr641]</td>\n",
       "      <td>2025-03-26</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>transition</td>\n",
       "      <td>Reaction</td>\n",
       "      <td>Reaction</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>10790813</td>\n",
       "      <td>TBK1 phosphorylates STAT6 at Ser407</td>\n",
       "      <td>R-DME-3249371</td>\n",
       "      <td>R-DME-3249371.1</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[TBK1 phosphorylates STAT6 at Ser407]</td>\n",
       "      <td>2025-03-26</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>transition</td>\n",
       "      <td>Reaction</td>\n",
       "      <td>Reaction</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>10790798</td>\n",
       "      <td>p-S407,Y641-STAT6 dimer migrates to the nucleus</td>\n",
       "      <td>R-DME-3249370</td>\n",
       "      <td>R-DME-3249370.1</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[p-S407,Y641-STAT6 dimer migrates to the nucleus]</td>\n",
       "      <td>2025-03-26</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>transition</td>\n",
       "      <td>Reaction</td>\n",
       "      <td>Reaction</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>10790815</td>\n",
       "      <td>p-S407,Y641-STAT6 is dimerized</td>\n",
       "      <td>R-DME-3249372</td>\n",
       "      <td>R-DME-3249372.1</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[p-S407,Y641-STAT6 is dimerized]</td>\n",
       "      <td>2025-03-26</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>binding</td>\n",
       "      <td>Reaction</td>\n",
       "      <td>Reaction</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>10790825</td>\n",
       "      <td>TBK1 is phosphorylated within STING:TBK1:STAT6...</td>\n",
       "      <td>R-DME-3249390</td>\n",
       "      <td>R-DME-3249390.1</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[TBK1 is phosphorylated within STING:TBK1:STAT...</td>\n",
       "      <td>2025-03-26</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>transition</td>\n",
       "      <td>Reaction</td>\n",
       "      <td>Reaction</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        dbId                                        displayName  \\\n",
       "0   10790592                         STING binds cyclic GMP-AMP   \n",
       "1   10819344         DDX41 binds bacterial c-di-AMP, c-di-GMP     \n",
       "2   10790177                                 STING dimerization   \n",
       "3   10788129                               STING binds c-di-GMP   \n",
       "4   10829117             STAT6-mediated induction of chemokines   \n",
       "5   10790821                      STING recruits TBK1 and STAT6   \n",
       "6   10790823           p-S407-STAT6 is phosphorylated at Tyr641   \n",
       "7   10790813                TBK1 phosphorylates STAT6 at Ser407   \n",
       "8   10790798    p-S407,Y641-STAT6 dimer migrates to the nucleus   \n",
       "9   10790815                     p-S407,Y641-STAT6 is dimerized   \n",
       "10  10790825  TBK1 is phosphorylated within STING:TBK1:STAT6...   \n",
       "\n",
       "             stId      stIdVersion  isInDisease  isInferred  \\\n",
       "0   R-DME-3244643  R-DME-3244643.1        False        True   \n",
       "1   R-DME-9013869  R-DME-9013869.1        False        True   \n",
       "2   R-DME-3134800  R-DME-3134800.1        False        True   \n",
       "3   R-DME-2396009  R-DME-2396009.1        False        True   \n",
       "4   R-DME-3249367  R-DME-3249367.1        False        True   \n",
       "5   R-DME-3249378  R-DME-3249378.1        False        True   \n",
       "6   R-DME-3249379  R-DME-3249379.1        False        True   \n",
       "7   R-DME-3249371  R-DME-3249371.1        False        True   \n",
       "8   R-DME-3249370  R-DME-3249370.1        False        True   \n",
       "9   R-DME-3249372  R-DME-3249372.1        False        True   \n",
       "10  R-DME-3249390  R-DME-3249390.1        False        True   \n",
       "\n",
       "                                                 name releaseDate  \\\n",
       "0                        [STING binds cyclic GMP-AMP]  2025-03-26   \n",
       "1        [DDX41 binds bacterial c-di-AMP, c-di-GMP  ]  2025-03-26   \n",
       "2                                [STING dimerization]  2025-03-26   \n",
       "3                              [STING binds c-di-GMP]  2025-03-26   \n",
       "4            [STAT6-mediated induction of chemokines]  2025-03-26   \n",
       "5                     [STING recruits TBK1 and STAT6]  2025-03-26   \n",
       "6          [p-S407-STAT6 is phosphorylated at Tyr641]  2025-03-26   \n",
       "7               [TBK1 phosphorylates STAT6 at Ser407]  2025-03-26   \n",
       "8   [p-S407,Y641-STAT6 dimer migrates to the nucleus]  2025-03-26   \n",
       "9                    [p-S407,Y641-STAT6 is dimerized]  2025-03-26   \n",
       "10  [TBK1 is phosphorylated within STING:TBK1:STAT...  2025-03-26   \n",
       "\n",
       "                speciesName    category className    schemaClass hasDiagram  \\\n",
       "0   Drosophila melanogaster     omitted  Reaction  BlackBoxEvent        NaN   \n",
       "1   Drosophila melanogaster     binding  Reaction       Reaction        NaN   \n",
       "2   Drosophila melanogaster     binding  Reaction       Reaction        NaN   \n",
       "3   Drosophila melanogaster     omitted  Reaction  BlackBoxEvent        NaN   \n",
       "4   Drosophila melanogaster         NaN   Pathway        Pathway      False   \n",
       "5   Drosophila melanogaster     binding  Reaction       Reaction        NaN   \n",
       "6   Drosophila melanogaster  transition  Reaction       Reaction        NaN   \n",
       "7   Drosophila melanogaster  transition  Reaction       Reaction        NaN   \n",
       "8   Drosophila melanogaster  transition  Reaction       Reaction        NaN   \n",
       "9   Drosophila melanogaster     binding  Reaction       Reaction        NaN   \n",
       "10  Drosophila melanogaster  transition  Reaction       Reaction        NaN   \n",
       "\n",
       "   hasEHLD  \n",
       "0      NaN  \n",
       "1      NaN  \n",
       "2      NaN  \n",
       "3      NaN  \n",
       "4    False  \n",
       "5      NaN  \n",
       "6      NaN  \n",
       "7      NaN  \n",
       "8      NaN  \n",
       "9      NaN  \n",
       "10     NaN  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "instance = ReactomeInstance()\n",
    "response = instance.fetch(\"R-DME-1834941\")\n",
    "pd.DataFrame(instance.parse(response))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e6a4a0ef",
   "metadata": {},
   "source": [
    "## Interpro"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "735933bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "from src.interpro import InterproInstance\n",
    "from src.utils import get_feature_keys\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "9bc15d9f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "17 records found\n"
     ]
    }
   ],
   "source": [
    "instance = InterproInstance(\n",
    "    #fields_to_extract={\n",
    "    #    \"1\": \"metadata.accession\",\n",
    "    #    \"2\": \"metadata.name\"\n",
    "    #}\n",
    ")\n",
    "test = instance.fetch(\n",
    "    type = \"entry\",\n",
    "    db = \"InterPro\",\n",
    "    entry_integration = \"\",\n",
    "    modifiers = {},\n",
    "    filter_type = \"protein\",\n",
    "    filter_db = \"UniProt\",\n",
    "    filter_value = \"P05067\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "40136e42",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'metadata.accession': 'dict(str)',\n",
       " 'metadata.name': 'dict(str)',\n",
       " 'metadata.source_database': 'dict(str)',\n",
       " 'metadata.type': 'dict(str)',\n",
       " 'metadata.integrated': 'dict(NoneType)',\n",
       " 'metadata.member_databases.profile.PS50279': 'dict(dict(dict(str)))',\n",
       " 'metadata.member_databases.profile': 'dict(dict(dict))',\n",
       " 'metadata.member_databases.pfam.PF00014': 'dict(dict(dict(str)))',\n",
       " 'metadata.member_databases.pfam': 'dict(dict(dict))',\n",
       " 'metadata.member_databases.smart.SM00131': 'dict(dict(dict(str)))',\n",
       " 'metadata.member_databases.smart': 'dict(dict(dict))',\n",
       " 'metadata.member_databases.prints.PR00759': 'dict(dict(dict(str)))',\n",
       " 'metadata.member_databases.prints': 'dict(dict(dict))',\n",
       " 'metadata.member_databases': 'dict(dict)',\n",
       " 'metadata.go_terms.identifier': 'dict(str)',\n",
       " 'metadata.go_terms.name': 'dict(str)',\n",
       " 'metadata.go_terms.category.code': 'dict(dict(str))',\n",
       " 'metadata.go_terms.category.name': 'dict(dict(str))',\n",
       " 'metadata.go_terms.category': 'dict(dict)',\n",
       " 'metadata': 'dict',\n",
       " 'proteins.accession': 'str',\n",
       " 'proteins.protein_length': 'int',\n",
       " 'proteins.source_database': 'str',\n",
       " 'proteins.organism': 'str',\n",
       " 'proteins.in_alphafold': 'bool',\n",
       " 'proteins.entry_protein_locations.fragments.start': 'int',\n",
       " 'proteins.entry_protein_locations.fragments.end': 'int',\n",
       " 'proteins.entry_protein_locations.fragments.dc-status': 'str',\n",
       " 'proteins.entry_protein_locations.representative': 'bool',\n",
       " 'proteins.entry_protein_locations.model': 'NoneType',\n",
       " 'proteins.entry_protein_locations.score': 'NoneType'}"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_feature_keys(test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "57978983",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "17 records found\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>metadata</th>\n",
       "      <th>proteins</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>{'accession': 'IPR002223', 'name': 'Pancreatic...</td>\n",
       "      <td>[{'accession': 'p05067', 'protein_length': 770...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>{'accession': 'IPR008154', 'name': 'Amyloidoge...</td>\n",
       "      <td>[{'accession': 'p05067', 'protein_length': 770...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>{'accession': 'IPR008155', 'name': 'Amyloidoge...</td>\n",
       "      <td>[{'accession': 'p05067', 'protein_length': 770...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>{'accession': 'IPR011178', 'name': 'Amyloidoge...</td>\n",
       "      <td>[{'accession': 'p05067', 'protein_length': 770...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>{'accession': 'IPR011993', 'name': 'PH-like do...</td>\n",
       "      <td>[{'accession': 'p05067', 'protein_length': 770...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>{'accession': 'IPR013803', 'name': 'Amyloidoge...</td>\n",
       "      <td>[{'accession': 'p05067', 'protein_length': 770...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>{'accession': 'IPR015849', 'name': 'Amyloidoge...</td>\n",
       "      <td>[{'accession': 'p05067', 'protein_length': 770...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>{'accession': 'IPR019543', 'name': 'Beta-amylo...</td>\n",
       "      <td>[{'accession': 'p05067', 'protein_length': 770...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>{'accession': 'IPR019744', 'name': 'Amyloidoge...</td>\n",
       "      <td>[{'accession': 'p05067', 'protein_length': 770...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>{'accession': 'IPR019745', 'name': 'Amyloidoge...</td>\n",
       "      <td>[{'accession': 'p05067', 'protein_length': 770...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>{'accession': 'IPR020901', 'name': 'Proteinase...</td>\n",
       "      <td>[{'accession': 'p05067', 'protein_length': 770...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>{'accession': 'IPR024329', 'name': 'Amyloidoge...</td>\n",
       "      <td>[{'accession': 'p05067', 'protein_length': 770...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>{'accession': 'IPR036176', 'name': 'E2 domain ...</td>\n",
       "      <td>[{'accession': 'p05067', 'protein_length': 770...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>{'accession': 'IPR036454', 'name': 'Amyloidoge...</td>\n",
       "      <td>[{'accession': 'p05067', 'protein_length': 770...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>{'accession': 'IPR036669', 'name': 'Amyloidoge...</td>\n",
       "      <td>[{'accession': 'p05067', 'protein_length': 770...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>{'accession': 'IPR036880', 'name': 'Pancreatic...</td>\n",
       "      <td>[{'accession': 'p05067', 'protein_length': 770...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>{'accession': 'IPR037071', 'name': 'Amyloidoge...</td>\n",
       "      <td>[{'accession': 'p05067', 'protein_length': 770...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             metadata  \\\n",
       "0   {'accession': 'IPR002223', 'name': 'Pancreatic...   \n",
       "1   {'accession': 'IPR008154', 'name': 'Amyloidoge...   \n",
       "2   {'accession': 'IPR008155', 'name': 'Amyloidoge...   \n",
       "3   {'accession': 'IPR011178', 'name': 'Amyloidoge...   \n",
       "4   {'accession': 'IPR011993', 'name': 'PH-like do...   \n",
       "5   {'accession': 'IPR013803', 'name': 'Amyloidoge...   \n",
       "6   {'accession': 'IPR015849', 'name': 'Amyloidoge...   \n",
       "7   {'accession': 'IPR019543', 'name': 'Beta-amylo...   \n",
       "8   {'accession': 'IPR019744', 'name': 'Amyloidoge...   \n",
       "9   {'accession': 'IPR019745', 'name': 'Amyloidoge...   \n",
       "10  {'accession': 'IPR020901', 'name': 'Proteinase...   \n",
       "11  {'accession': 'IPR024329', 'name': 'Amyloidoge...   \n",
       "12  {'accession': 'IPR036176', 'name': 'E2 domain ...   \n",
       "13  {'accession': 'IPR036454', 'name': 'Amyloidoge...   \n",
       "14  {'accession': 'IPR036669', 'name': 'Amyloidoge...   \n",
       "15  {'accession': 'IPR036880', 'name': 'Pancreatic...   \n",
       "16  {'accession': 'IPR037071', 'name': 'Amyloidoge...   \n",
       "\n",
       "                                             proteins  \n",
       "0   [{'accession': 'p05067', 'protein_length': 770...  \n",
       "1   [{'accession': 'p05067', 'protein_length': 770...  \n",
       "2   [{'accession': 'p05067', 'protein_length': 770...  \n",
       "3   [{'accession': 'p05067', 'protein_length': 770...  \n",
       "4   [{'accession': 'p05067', 'protein_length': 770...  \n",
       "5   [{'accession': 'p05067', 'protein_length': 770...  \n",
       "6   [{'accession': 'p05067', 'protein_length': 770...  \n",
       "7   [{'accession': 'p05067', 'protein_length': 770...  \n",
       "8   [{'accession': 'p05067', 'protein_length': 770...  \n",
       "9   [{'accession': 'p05067', 'protein_length': 770...  \n",
       "10  [{'accession': 'p05067', 'protein_length': 770...  \n",
       "11  [{'accession': 'p05067', 'protein_length': 770...  \n",
       "12  [{'accession': 'p05067', 'protein_length': 770...  \n",
       "13  [{'accession': 'p05067', 'protein_length': 770...  \n",
       "14  [{'accession': 'p05067', 'protein_length': 770...  \n",
       "15  [{'accession': 'p05067', 'protein_length': 770...  \n",
       "16  [{'accession': 'p05067', 'protein_length': 770...  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "instance.fetch_to_dataframe(\n",
    "    [\n",
    "        {\n",
    "            \"type\": \"entry\",\n",
    "            \"db\": \"InterPro\",\n",
    "            \"entry_integration\": \"\",\n",
    "            \"modifiers\": {\n",
    "                \"go_term\": \"GO:0004867\"\n",
    "            },\n",
    "            \"filter_type\": \"protein\",\n",
    "            \"filter_db\": \"UniProt\",\n",
    "            \"filter_value\": \"P05067\"\n",
    "        }\n",
    "    ]\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "57034340",
   "metadata": {},
   "source": [
    "## STRING"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6ed9cccf",
   "metadata": {},
   "outputs": [],
   "source": [
    "from src.stringdb import StringInterface"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "dea4ee3d",
   "metadata": {},
   "outputs": [],
   "source": [
    "instance = StringInterface()\n",
    "test = instance.fetch(\n",
    "    outfmt=\"json\",\n",
    "    method=\"interaction_partners\",\n",
    "    params={\n",
    "        \"identifiers\": [\"p53\"],\n",
    "    }\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c46d8ff1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'stringId_A': '7227.FBpp0083753',\n",
       "  'stringId_B': '7227.FBpp0110174',\n",
       "  'preferredName_A': 'p53',\n",
       "  'preferredName_B': 'tefu',\n",
       "  'ncbiTaxonId': 7227,\n",
       "  'score': 0.985,\n",
       "  'nscore': 0,\n",
       "  'fscore': 0,\n",
       "  'pscore': 0,\n",
       "  'ascore': 0,\n",
       "  'escore': 0.6,\n",
       "  'dscore': 0.9,\n",
       "  'tscore': 0.659},\n",
       " {'stringId_A': '7227.FBpp0083753',\n",
       "  'stringId_B': '7227.FBpp0074047',\n",
       "  'preferredName_A': 'p53',\n",
       "  'preferredName_B': 'mei-41',\n",
       "  'ncbiTaxonId': 7227,\n",
       "  'score': 0.962,\n",
       "  'nscore': 0,\n",
       "  'fscore': 0,\n",
       "  'pscore': 0,\n",
       "  'ascore': 0.071,\n",
       "  'escore': 0,\n",
       "  'dscore': 0.9,\n",
       "  'tscore': 0.627},\n",
       " {'stringId_A': '7227.FBpp0083753',\n",
       "  'stringId_B': '7227.FBpp0304253',\n",
       "  'preferredName_A': 'p53',\n",
       "  'preferredName_B': 'hpo',\n",
       "  'ncbiTaxonId': 7227,\n",
       "  'score': 0.94,\n",
       "  'nscore': 0,\n",
       "  'fscore': 0,\n",
       "  'pscore': 0,\n",
       "  'ascore': 0.062,\n",
       "  'escore': 0.225,\n",
       "  'dscore': 0.9,\n",
       "  'tscore': 0.28},\n",
       " {'stringId_A': '7227.FBpp0083753',\n",
       "  'stringId_B': '7227.FBpp0080860',\n",
       "  'preferredName_A': 'p53',\n",
       "  'preferredName_B': 'lok',\n",
       "  'ncbiTaxonId': 7227,\n",
       "  'score': 0.932,\n",
       "  'nscore': 0,\n",
       "  'fscore': 0,\n",
       "  'pscore': 0,\n",
       "  'ascore': 0.076,\n",
       "  'escore': 0.6,\n",
       "  'dscore': 0.75,\n",
       "  'tscore': 0.36},\n",
       " {'stringId_A': '7227.FBpp0083753',\n",
       "  'stringId_B': '7227.FBpp0076806',\n",
       "  'preferredName_A': 'p53',\n",
       "  'preferredName_B': 'scny',\n",
       "  'ncbiTaxonId': 7227,\n",
       "  'score': 0.9,\n",
       "  'nscore': 0,\n",
       "  'fscore': 0,\n",
       "  'pscore': 0,\n",
       "  'ascore': 0,\n",
       "  'escore': 0,\n",
       "  'dscore': 0.9,\n",
       "  'tscore': 0},\n",
       " {'stringId_A': '7227.FBpp0083753',\n",
       "  'stringId_B': '7227.FBpp0080443',\n",
       "  'preferredName_A': 'p53',\n",
       "  'preferredName_B': 'grp',\n",
       "  'ncbiTaxonId': 7227,\n",
       "  'score': 0.854,\n",
       "  'nscore': 0,\n",
       "  'fscore': 0,\n",
       "  'pscore': 0,\n",
       "  'ascore': 0.059,\n",
       "  'escore': 0,\n",
       "  'dscore': 0.75,\n",
       "  'tscore': 0.432},\n",
       " {'stringId_A': '7227.FBpp0083753',\n",
       "  'stringId_B': '7227.FBpp0074886',\n",
       "  'preferredName_A': 'p53',\n",
       "  'preferredName_B': 'rpr',\n",
       "  'ncbiTaxonId': 7227,\n",
       "  'score': 0.838,\n",
       "  'nscore': 0,\n",
       "  'fscore': 0,\n",
       "  'pscore': 0,\n",
       "  'ascore': 0,\n",
       "  'escore': 0,\n",
       "  'dscore': 0,\n",
       "  'tscore': 0.838},\n",
       " {'stringId_A': '7227.FBpp0083753',\n",
       "  'stringId_B': '7227.FBpp0402906',\n",
       "  'preferredName_A': 'p53',\n",
       "  'preferredName_B': 'yki',\n",
       "  'ncbiTaxonId': 7227,\n",
       "  'score': 0.833,\n",
       "  'nscore': 0,\n",
       "  'fscore': 0,\n",
       "  'pscore': 0,\n",
       "  'ascore': 0,\n",
       "  'escore': 0.292,\n",
       "  'dscore': 0.7,\n",
       "  'tscore': 0.281},\n",
       " {'stringId_A': '7227.FBpp0083753',\n",
       "  'stringId_B': '7227.FBpp0075807',\n",
       "  'preferredName_A': 'p53',\n",
       "  'preferredName_B': 'CycA',\n",
       "  'ncbiTaxonId': 7227,\n",
       "  'score': 0.832,\n",
       "  'nscore': 0,\n",
       "  'fscore': 0,\n",
       "  'pscore': 0,\n",
       "  'ascore': 0,\n",
       "  'escore': 0,\n",
       "  'dscore': 0.75,\n",
       "  'tscore': 0.356},\n",
       " {'stringId_A': '7227.FBpp0083753',\n",
       "  'stringId_B': '7227.FBpp0289675',\n",
       "  'preferredName_A': 'p53',\n",
       "  'preferredName_B': 'Cdk2',\n",
       "  'ncbiTaxonId': 7227,\n",
       "  'score': 0.827,\n",
       "  'nscore': 0,\n",
       "  'fscore': 0,\n",
       "  'pscore': 0,\n",
       "  'ascore': 0.079,\n",
       "  'escore': 0,\n",
       "  'dscore': 0.75,\n",
       "  'tscore': 0.311}]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "instance.parse(test, fmt=\"json\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "9635fbc2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>stringId_A</th>\n",
       "      <th>stringId_B</th>\n",
       "      <th>preferredName_A</th>\n",
       "      <th>preferredName_B</th>\n",
       "      <th>ncbiTaxonId</th>\n",
       "      <th>score</th>\n",
       "      <th>nscore</th>\n",
       "      <th>fscore</th>\n",
       "      <th>pscore</th>\n",
       "      <th>ascore</th>\n",
       "      <th>escore</th>\n",
       "      <th>dscore</th>\n",
       "      <th>tscore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9606.ENSP00000266970</td>\n",
       "      <td>9606.ENSP00000481380</td>\n",
       "      <td>CDK2</td>\n",
       "      <td>CCNA2</td>\n",
       "      <td>9606</td>\n",
       "      <td>0.999</td>\n",
       "      <td>0</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.453</td>\n",
       "      <td>0.999</td>\n",
       "      <td>0.9</td>\n",
       "      <td>0.999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9606.ENSP00000266970</td>\n",
       "      <td>9606.ENSP00000413720</td>\n",
       "      <td>CDK2</td>\n",
       "      <td>CDKN1C</td>\n",
       "      <td>9606</td>\n",
       "      <td>0.999</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.085</td>\n",
       "      <td>0.859</td>\n",
       "      <td>0.9</td>\n",
       "      <td>0.970</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>9606.ENSP00000266970</td>\n",
       "      <td>9606.ENSP00000228872</td>\n",
       "      <td>CDK2</td>\n",
       "      <td>CDKN1B</td>\n",
       "      <td>9606</td>\n",
       "      <td>0.999</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.085</td>\n",
       "      <td>0.999</td>\n",
       "      <td>0.9</td>\n",
       "      <td>0.999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9606.ENSP00000266970</td>\n",
       "      <td>9606.ENSP00000429089</td>\n",
       "      <td>CDK2</td>\n",
       "      <td>CCNE2</td>\n",
       "      <td>9606</td>\n",
       "      <td>0.999</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.200</td>\n",
       "      <td>0.995</td>\n",
       "      <td>0.9</td>\n",
       "      <td>0.996</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9606.ENSP00000266970</td>\n",
       "      <td>9606.ENSP00000255465</td>\n",
       "      <td>CDK2</td>\n",
       "      <td>CCNA1</td>\n",
       "      <td>9606</td>\n",
       "      <td>0.999</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.292</td>\n",
       "      <td>0.942</td>\n",
       "      <td>0.9</td>\n",
       "      <td>0.999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2956</th>\n",
       "      <td>9606.ENSP00000269305</td>\n",
       "      <td>9606.ENSP00000310928</td>\n",
       "      <td>TP53</td>\n",
       "      <td>PPARD</td>\n",
       "      <td>9606</td>\n",
       "      <td>0.400</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2957</th>\n",
       "      <td>9606.ENSP00000269305</td>\n",
       "      <td>9606.ENSP00000419945</td>\n",
       "      <td>TP53</td>\n",
       "      <td>ERVW-1</td>\n",
       "      <td>9606</td>\n",
       "      <td>0.400</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2958</th>\n",
       "      <td>9606.ENSP00000269305</td>\n",
       "      <td>9606.ENSP00000462980</td>\n",
       "      <td>TP53</td>\n",
       "      <td>TAF4B</td>\n",
       "      <td>9606</td>\n",
       "      <td>0.400</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.4</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2959</th>\n",
       "      <td>9606.ENSP00000269305</td>\n",
       "      <td>9606.ENSP00000431885</td>\n",
       "      <td>TP53</td>\n",
       "      <td>TYK2</td>\n",
       "      <td>9606</td>\n",
       "      <td>0.400</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.083</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.372</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2960</th>\n",
       "      <td>9606.ENSP00000269305</td>\n",
       "      <td>9606.ENSP00000356247</td>\n",
       "      <td>TP53</td>\n",
       "      <td>LGR6</td>\n",
       "      <td>9606</td>\n",
       "      <td>0.400</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.165</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.311</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2961 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                stringId_A            stringId_B preferredName_A  \\\n",
       "0     9606.ENSP00000266970  9606.ENSP00000481380            CDK2   \n",
       "1     9606.ENSP00000266970  9606.ENSP00000413720            CDK2   \n",
       "2     9606.ENSP00000266970  9606.ENSP00000228872            CDK2   \n",
       "3     9606.ENSP00000266970  9606.ENSP00000429089            CDK2   \n",
       "4     9606.ENSP00000266970  9606.ENSP00000255465            CDK2   \n",
       "...                    ...                   ...             ...   \n",
       "2956  9606.ENSP00000269305  9606.ENSP00000310928            TP53   \n",
       "2957  9606.ENSP00000269305  9606.ENSP00000419945            TP53   \n",
       "2958  9606.ENSP00000269305  9606.ENSP00000462980            TP53   \n",
       "2959  9606.ENSP00000269305  9606.ENSP00000431885            TP53   \n",
       "2960  9606.ENSP00000269305  9606.ENSP00000356247            TP53   \n",
       "\n",
       "     preferredName_B  ncbiTaxonId  score  nscore  fscore  pscore  ascore  \\\n",
       "0              CCNA2         9606  0.999       0   0.003     0.0   0.453   \n",
       "1             CDKN1C         9606  0.999       0   0.000     0.0   0.085   \n",
       "2             CDKN1B         9606  0.999       0   0.000     0.0   0.085   \n",
       "3              CCNE2         9606  0.999       0   0.000     0.0   0.200   \n",
       "4              CCNA1         9606  0.999       0   0.000     0.0   0.292   \n",
       "...              ...          ...    ...     ...     ...     ...     ...   \n",
       "2956           PPARD         9606  0.400       0   0.000     0.0   0.000   \n",
       "2957          ERVW-1         9606  0.400       0   0.000     0.0   0.000   \n",
       "2958           TAF4B         9606  0.400       0   0.000     0.0   0.000   \n",
       "2959            TYK2         9606  0.400       0   0.000     0.0   0.083   \n",
       "2960            LGR6         9606  0.400       0   0.000     0.0   0.000   \n",
       "\n",
       "      escore  dscore  tscore  \n",
       "0      0.999     0.9   0.999  \n",
       "1      0.859     0.9   0.970  \n",
       "2      0.999     0.9   0.999  \n",
       "3      0.995     0.9   0.996  \n",
       "4      0.942     0.9   0.999  \n",
       "...      ...     ...     ...  \n",
       "2956   0.000     0.0   0.400  \n",
       "2957   0.000     0.0   0.400  \n",
       "2958   0.000     0.4   0.000  \n",
       "2959   0.000     0.0   0.372  \n",
       "2960   0.165     0.0   0.311  \n",
       "\n",
       "[2961 rows x 13 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "instance.fetch_to_dataframe(\n",
    "    outfmt=\"json\",\n",
    "    method=\"interaction_partners\",\n",
    "    params={\n",
    "        \"identifiers\": [\"p53\", \"cdk2\"],\n",
    "        \"species\": 9606,\n",
    "    }\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cf00f1af",
   "metadata": {},
   "source": [
    "## KEGG"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "38d0fe83",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Response [200]>"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from src.kegg import KEGGInterface\n",
    "\n",
    "instance = KEGGInterface(\n",
    "    fields_to_extract=[\"ENTRY\", \"NAME\", \"PATHWAY\"]\n",
    ")\n",
    "instance.fetch(\n",
    "    method=\"get\",\n",
    "    query=[\"hsa:10458\", \"ece:Z5100\"]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "b56a8083",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ENTRY</th>\n",
       "      <th>NAME</th>\n",
       "      <th>PATHWAY</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10458             CDS       T01001</td>\n",
       "      <td>(RefSeq) BAR/IMD domain containing adaptor pro...</td>\n",
       "      <td>hsa04520  Adherens junction hsa04810  Regulati...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Z5100             CDS       T00044</td>\n",
       "      <td>(GenBank) espF</td>\n",
       "      <td>ece05130  Pathogenic Escherichia coli infection</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                ENTRY  \\\n",
       "0  10458             CDS       T01001   \n",
       "1  Z5100             CDS       T00044   \n",
       "\n",
       "                                                NAME  \\\n",
       "0  (RefSeq) BAR/IMD domain containing adaptor pro...   \n",
       "1                                     (GenBank) espF   \n",
       "\n",
       "                                             PATHWAY  \n",
       "0  hsa04520  Adherens junction hsa04810  Regulati...  \n",
       "1    ece05130  Pathogenic Escherichia coli infection  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "instance.fetch_to_dataframe(\n",
    "    method=\"get\",\n",
    "    query=[\"hsa:10458\", \"ece:Z5100\"]\n",
    "    \n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b9fd8a37",
   "metadata": {},
   "source": [
    "## Gen Ontology"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "e147770a",
   "metadata": {},
   "outputs": [],
   "source": [
    "from src.genontology import GenOntologyInterface\n",
    "instance = GenOntologyInterface(\n",
    "    max_workers=2,\n",
    "    fields_to_extract = [\"goid\", \"label\"]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "36209712",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetching data from https://api.geneontology.org/api/ontology/term/GO%3A0008150\n",
      "Fetching data from https://api.geneontology.org/api/ontology/term/GO%3A0008150/graph\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>goid</th>\n",
       "      <th>label</th>\n",
       "      <th>relationships</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GO:0008150</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>[GO:0044848, GO:0044419, GO:0008150, GO:006500...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         goid               label  \\\n",
       "0  GO:0008150  biological_process   \n",
       "\n",
       "                                       relationships  \n",
       "0  [GO:0044848, GO:0044419, GO:0008150, GO:006500...  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "instance.fetch_to_dataframe(\n",
    "    method= \"ontology-term\",\n",
    "    query= \"GO:0008150\", \n",
    "    option=None,\n",
    "    look_for_relationships=True\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e13128ee",
   "metadata": {},
   "source": [
    "### Show graph test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "65e0ff5b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "✅ Grafo guardado en: mi_grafo.html\n"
     ]
    }
   ],
   "source": [
    "import json\n",
    "\n",
    "# Convierte a formato Cytoscape.js\n",
    "def convert_to_cytoscape_format(graph_json):\n",
    "    elements = []\n",
    "\n",
    "    for node in graph_json['nodes']:\n",
    "        elements.append({\n",
    "            \"data\": {\n",
    "                \"id\": node[\"id\"],\n",
    "                \"label\": node[\"lbl\"]\n",
    "            }\n",
    "        })\n",
    "\n",
    "    for edge in graph_json['edges']:\n",
    "        elements.append({\n",
    "            \"data\": {\n",
    "                \"source\": edge[\"sub\"],\n",
    "                \"target\": edge[\"obj\"],\n",
    "                \"label\": edge[\"pred\"]\n",
    "            }\n",
    "        })\n",
    "\n",
    "    return elements\n",
    "\n",
    "# Genera el HTML\n",
    "def create_cytoscape_html(graph_json):\n",
    "    elements = convert_to_cytoscape_format(graph_json)\n",
    "\n",
    "    html = f\"\"\"\n",
    "    <!DOCTYPE html>\n",
    "    <html>\n",
    "    <head>\n",
    "        <meta charset=\"utf-8\">\n",
    "        <title>Grafo GO Interactivo</title>\n",
    "        <script src=\"https://unpkg.com/cytoscape@3.19.0/dist/cytoscape.min.js\"></script>\n",
    "        <script src=\"https://unpkg.com/dagre@0.8.5/dist/dagre.min.js\"></script>\n",
    "        <script src=\"https://unpkg.com/cytoscape-dagre@2.3.2/cytoscape-dagre.js\"></script>\n",
    "        <style>\n",
    "            html, body {{\n",
    "                margin: 0;\n",
    "                padding: 0;\n",
    "                height: 100%;\n",
    "                width: 100%;\n",
    "                font-family: Arial, sans-serif;\n",
    "            }}\n",
    "            #cy {{\n",
    "                height: 100%;\n",
    "                width: 100%;\n",
    "                display: block;\n",
    "            }}\n",
    "        </style>\n",
    "    </head>\n",
    "    <body>\n",
    "        <div id=\"cy\"></div>\n",
    "        <script>\n",
    "            cytoscape.use(cytoscapeDagre);\n",
    "\n",
    "            var cy = cytoscape({{\n",
    "                container: document.getElementById('cy'),\n",
    "                elements: {json.dumps(elements)},\n",
    "                style: [\n",
    "                    {{\n",
    "                        selector: 'node',\n",
    "                        style: {{\n",
    "                            'shape': 'roundrectangle',\n",
    "                            'label': 'data(label)',\n",
    "                            'text-valign': 'center',\n",
    "                            'text-halign': 'center',\n",
    "                            'background-color': '#AED6F1',\n",
    "                            'color': '#1B2631',\n",
    "                            'font-size': '8px',\n",
    "                            'width': 'label',\n",
    "                            'height': 'label',\n",
    "                            'padding': '6px',\n",
    "                            'border-width': 1,\n",
    "                            'border-color': '#2980B9'\n",
    "                        }}\n",
    "                    }},\n",
    "                    {{\n",
    "                        selector: 'edge',\n",
    "                        style: {{\n",
    "                            'width': 2,\n",
    "                            'label': 'data(label)',\n",
    "                            'line-color': '#B2BABB',\n",
    "                            'target-arrow-color': '#B2BABB',\n",
    "                            'target-arrow-shape': 'triangle',\n",
    "                            'curve-style': 'bezier',\n",
    "                            'font-size': '7px',\n",
    "                            'color': '#5D6D7E',\n",
    "                            'text-background-opacity': 1,\n",
    "                            'text-background-color': '#fff',\n",
    "                            'text-background-shape': 'roundrectangle',\n",
    "                            'text-background-padding': 2\n",
    "                        }}\n",
    "                    }}\n",
    "                ],\n",
    "                layout: {{\n",
    "                    name: 'dagre',\n",
    "                    rankDir: 'TB',\n",
    "                    nodeSep: 30,\n",
    "                    edgeSep: 10,\n",
    "                    rankSep: 50\n",
    "                }},\n",
    "                zoomingEnabled: true,\n",
    "                userZoomingEnabled: true,\n",
    "                boxSelectionEnabled: false\n",
    "            }});\n",
    "        </script>\n",
    "    </body>\n",
    "    </html>\n",
    "    \"\"\"\n",
    "    return html\n",
    "\n",
    "# Guardar archivo HTML\n",
    "def save_graph_to_html(graph_json, output_file='cytoscape_graph.html'):\n",
    "    html = create_cytoscape_html(graph_json)\n",
    "    with open(output_file, 'w', encoding='utf-8') as f:\n",
    "        f.write(html)\n",
    "    print(f\"✅ Grafo guardado en: {output_file}\")\n",
    "\n",
    "# Ejecutar\n",
    "save_graph_to_html(instance.parse(response)[\"topology_graph_json\"], 'mi_grafo.html')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c5b1a3b6",
   "metadata": {},
   "source": [
    "## BioGRID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "33cd0d6d",
   "metadata": {},
   "outputs": [],
   "source": [
    "from src.biogrid import BioGRIDInterface"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0f500860",
   "metadata": {},
   "outputs": [],
   "source": [
    "instance = BioGRIDInterface(\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f0a22d47",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetching data from https://webservice.thebiogrid.org/interactions?accessKey=87aee10241ab2fcc6e497352e5024380&geneList=P53|CDK2|BRCA1&format=json\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BIOGRID_INTERACTION_ID</th>\n",
       "      <th>ENTREZ_GENE_A</th>\n",
       "      <th>ENTREZ_GENE_B</th>\n",
       "      <th>BIOGRID_ID_A</th>\n",
       "      <th>BIOGRID_ID_B</th>\n",
       "      <th>SYSTEMATIC_NAME_A</th>\n",
       "      <th>SYSTEMATIC_NAME_B</th>\n",
       "      <th>OFFICIAL_SYMBOL_A</th>\n",
       "      <th>OFFICIAL_SYMBOL_B</th>\n",
       "      <th>SYNONYMS_A</th>\n",
       "      <th>...</th>\n",
       "      <th>PUBMED_ID</th>\n",
       "      <th>ORGANISM_A</th>\n",
       "      <th>ORGANISM_B</th>\n",
       "      <th>THROUGHPUT</th>\n",
       "      <th>QUANTITATION</th>\n",
       "      <th>MODIFICATION</th>\n",
       "      <th>ONTOLOGY_TERMS</th>\n",
       "      <th>QUALIFICATIONS</th>\n",
       "      <th>TAGS</th>\n",
       "      <th>SOURCEDB</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2368</td>\n",
       "      <td>672</td>\n",
       "      <td>466</td>\n",
       "      <td>107140</td>\n",
       "      <td>106956</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>ATF1</td>\n",
       "      <td>BRCAI|BRCC1|BROVCA1|FANCS|IRIS|PNCA4|PPP1R53|P...</td>\n",
       "      <td>...</td>\n",
       "      <td>10945975</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "      <td>Low Throughput</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>{}</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>BIOGRID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2398</td>\n",
       "      <td>672</td>\n",
       "      <td>4436</td>\n",
       "      <td>107140</td>\n",
       "      <td>110573</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>MSH2</td>\n",
       "      <td>BRCAI|BRCC1|BROVCA1|FANCS|IRIS|PNCA4|PPP1R53|P...</td>\n",
       "      <td>...</td>\n",
       "      <td>11498787</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "      <td>Low Throughput</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>{}</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>BIOGRID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2411</td>\n",
       "      <td>672</td>\n",
       "      <td>580</td>\n",
       "      <td>107140</td>\n",
       "      <td>107056</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BARD1</td>\n",
       "      <td>BRCAI|BRCC1|BROVCA1|FANCS|IRIS|PNCA4|PPP1R53|P...</td>\n",
       "      <td>...</td>\n",
       "      <td>8944023</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "      <td>Low Throughput</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>{}</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>BIOGRID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2424</td>\n",
       "      <td>672</td>\n",
       "      <td>2956</td>\n",
       "      <td>107140</td>\n",
       "      <td>109211</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>MSH6</td>\n",
       "      <td>BRCAI|BRCC1|BROVCA1|FANCS|IRIS|PNCA4|PPP1R53|P...</td>\n",
       "      <td>...</td>\n",
       "      <td>11498787</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "      <td>Low Throughput</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>{}</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>BIOGRID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3546</td>\n",
       "      <td>1033</td>\n",
       "      <td>1017</td>\n",
       "      <td>107467</td>\n",
       "      <td>107452</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>CDKN3</td>\n",
       "      <td>CDK2</td>\n",
       "      <td>CDI1|CIP2|KAP|KAP1</td>\n",
       "      <td>...</td>\n",
       "      <td>12745075</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "      <td>Low Throughput</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>{}</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>BIOGRID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9995</th>\n",
       "      <td>3390865</td>\n",
       "      <td>4193</td>\n",
       "      <td>7157</td>\n",
       "      <td>110358</td>\n",
       "      <td>113010</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>MDM2</td>\n",
       "      <td>TP53</td>\n",
       "      <td>ACTFS|HDMX|hdm2</td>\n",
       "      <td>...</td>\n",
       "      <td>35025136</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "      <td>Low Throughput</td>\n",
       "      <td>-</td>\n",
       "      <td>Ubiquitination</td>\n",
       "      <td>{}</td>\n",
       "      <td>E2: UbcH5b</td>\n",
       "      <td>-</td>\n",
       "      <td>BIOGRID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9996</th>\n",
       "      <td>3390899</td>\n",
       "      <td>7157</td>\n",
       "      <td>5111</td>\n",
       "      <td>113010</td>\n",
       "      <td>111142</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>TP53</td>\n",
       "      <td>PCNA</td>\n",
       "      <td>BCC7|LFS1|P53|TRP53</td>\n",
       "      <td>...</td>\n",
       "      <td>34645785</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "      <td>Low Throughput</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>{}</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>BIOGRID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9997</th>\n",
       "      <td>3390900</td>\n",
       "      <td>5111</td>\n",
       "      <td>7157</td>\n",
       "      <td>111142</td>\n",
       "      <td>113010</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>PCNA</td>\n",
       "      <td>TP53</td>\n",
       "      <td>ATLD2</td>\n",
       "      <td>...</td>\n",
       "      <td>34645785</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "      <td>Low Throughput</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>{}</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>BIOGRID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9998</th>\n",
       "      <td>3390901</td>\n",
       "      <td>11201</td>\n",
       "      <td>7157</td>\n",
       "      <td>116370</td>\n",
       "      <td>113010</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>POLI</td>\n",
       "      <td>TP53</td>\n",
       "      <td>RAD30B|RAD3OB</td>\n",
       "      <td>...</td>\n",
       "      <td>34645785</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "      <td>Low Throughput</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>{}</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>BIOGRID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9999</th>\n",
       "      <td>3394128</td>\n",
       "      <td>472</td>\n",
       "      <td>7157</td>\n",
       "      <td>106962</td>\n",
       "      <td>113010</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ATM</td>\n",
       "      <td>TP53</td>\n",
       "      <td>AT1|ATA|ATC|ATD|ATDC|ATE|TEL1|TELO1</td>\n",
       "      <td>...</td>\n",
       "      <td>12645530</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "      <td>Low Throughput</td>\n",
       "      <td>-</td>\n",
       "      <td>Phosphorylation</td>\n",
       "      <td>{}</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>BIOGRID</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10000 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      BIOGRID_INTERACTION_ID ENTREZ_GENE_A ENTREZ_GENE_B  BIOGRID_ID_A  \\\n",
       "0                       2368           672           466        107140   \n",
       "1                       2398           672          4436        107140   \n",
       "2                       2411           672           580        107140   \n",
       "3                       2424           672          2956        107140   \n",
       "4                       3546          1033          1017        107467   \n",
       "...                      ...           ...           ...           ...   \n",
       "9995                 3390865          4193          7157        110358   \n",
       "9996                 3390899          7157          5111        113010   \n",
       "9997                 3390900          5111          7157        111142   \n",
       "9998                 3390901         11201          7157        116370   \n",
       "9999                 3394128           472          7157        106962   \n",
       "\n",
       "      BIOGRID_ID_B SYSTEMATIC_NAME_A SYSTEMATIC_NAME_B OFFICIAL_SYMBOL_A  \\\n",
       "0           106956                 -                 -             BRCA1   \n",
       "1           110573                 -                 -             BRCA1   \n",
       "2           107056                 -                 -             BRCA1   \n",
       "3           109211                 -                 -             BRCA1   \n",
       "4           107452                 -                 -             CDKN3   \n",
       "...            ...               ...               ...               ...   \n",
       "9995        113010                 -                 -              MDM2   \n",
       "9996        111142                 -                 -              TP53   \n",
       "9997        113010                 -                 -              PCNA   \n",
       "9998        113010                 -                 -              POLI   \n",
       "9999        113010                 -                 -               ATM   \n",
       "\n",
       "     OFFICIAL_SYMBOL_B                                         SYNONYMS_A  \\\n",
       "0                 ATF1  BRCAI|BRCC1|BROVCA1|FANCS|IRIS|PNCA4|PPP1R53|P...   \n",
       "1                 MSH2  BRCAI|BRCC1|BROVCA1|FANCS|IRIS|PNCA4|PPP1R53|P...   \n",
       "2                BARD1  BRCAI|BRCC1|BROVCA1|FANCS|IRIS|PNCA4|PPP1R53|P...   \n",
       "3                 MSH6  BRCAI|BRCC1|BROVCA1|FANCS|IRIS|PNCA4|PPP1R53|P...   \n",
       "4                 CDK2                                 CDI1|CIP2|KAP|KAP1   \n",
       "...                ...                                                ...   \n",
       "9995              TP53                                    ACTFS|HDMX|hdm2   \n",
       "9996              PCNA                                BCC7|LFS1|P53|TRP53   \n",
       "9997              TP53                                              ATLD2   \n",
       "9998              TP53                                      RAD30B|RAD3OB   \n",
       "9999              TP53                AT1|ATA|ATC|ATD|ATDC|ATE|TEL1|TELO1   \n",
       "\n",
       "      ... PUBMED_ID ORGANISM_A ORGANISM_B      THROUGHPUT  QUANTITATION  \\\n",
       "0     ...  10945975       9606       9606  Low Throughput             -   \n",
       "1     ...  11498787       9606       9606  Low Throughput             -   \n",
       "2     ...   8944023       9606       9606  Low Throughput             -   \n",
       "3     ...  11498787       9606       9606  Low Throughput             -   \n",
       "4     ...  12745075       9606       9606  Low Throughput             -   \n",
       "...   ...       ...        ...        ...             ...           ...   \n",
       "9995  ...  35025136       9606       9606  Low Throughput             -   \n",
       "9996  ...  34645785       9606       9606  Low Throughput             -   \n",
       "9997  ...  34645785       9606       9606  Low Throughput             -   \n",
       "9998  ...  34645785       9606       9606  Low Throughput             -   \n",
       "9999  ...  12645530       9606       9606  Low Throughput             -   \n",
       "\n",
       "         MODIFICATION  ONTOLOGY_TERMS QUALIFICATIONS TAGS SOURCEDB  \n",
       "0                   -              {}              -    -  BIOGRID  \n",
       "1                   -              {}              -    -  BIOGRID  \n",
       "2                   -              {}              -    -  BIOGRID  \n",
       "3                   -              {}              -    -  BIOGRID  \n",
       "4                   -              {}              -    -  BIOGRID  \n",
       "...               ...             ...            ...  ...      ...  \n",
       "9995   Ubiquitination              {}     E2: UbcH5b    -  BIOGRID  \n",
       "9996                -              {}              -    -  BIOGRID  \n",
       "9997                -              {}              -    -  BIOGRID  \n",
       "9998                -              {}              -    -  BIOGRID  \n",
       "9999  Phosphorylation              {}              -    -  BIOGRID  \n",
       "\n",
       "[10000 rows x 24 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "instance.fetch_to_dataframe(\n",
    "    method=\"interactions\",\n",
    "    query={\n",
    "        \"accessKey\": \"\",\n",
    "        \"geneList\": [\"P53\", \"CDK2\", \"BRCA1\"],\n",
    "        \"format\": \"json\"\n",
    "    }\n",
    ")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bioseqdownloader",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
