{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "887ed803-c6dd-4da6-9845-3afb0c5f8392",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "import sys, os, json\n",
    "sys.path.append('../')\n",
    "from scripts.uniprot import UniprotInterface\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "7e4834c7",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "def replace_char_at_index(s, i, new_char):\n",
    "    if i < 0 or i >= len(s):\n",
    "        raise IndexError(\"Index out of range.\")\n",
    "    return s[:i] + new_char + s[i+1:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "b318b89f",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "ids = [\"Q75UA4\"]\n",
    "from_db = 'UniProtKB_AC-ID'\n",
    "to_db = 'UniProtKB'\n",
    "disease = \"CRC\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "0a647744",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "downloader = UniprotInterface()\n",
    "\n",
    "job_id = downloader.submit_id_mapping(from_db=from_db, to_db=to_db, ids=ids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "eef323a8",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetched: 1 / 1\n"
     ]
    }
   ],
   "source": [
    "if downloader.check_id_mapping_results_ready(job_id):\n",
    "    link = downloader.get_id_mapping_results_link(job_id)\n",
    "    results = downloader.get_id_mapping_results_search(link)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "b9396a36",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "with open(\"results.json\", \"w\") as f:\n",
    "    json.dump(results, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "243e35c2",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'MKFGKFVLLAASTALAVVGLGGPAAADSTPQAQPSIIGGSNATSGPWAARLFVNGRQNCTATIIAPQYILTAKHCVSSSGTYTFRIGSLDQTSGGTMATGSTITRYPGSADLAIVRLTTSVNATYSPLGSVGDVSVGQNVSVYGWGATSQCGSEINCQSRYLKVATVRVNSISCSDYTGGVAVCANRVNGITAGGDSGGPMFASGRQVGVASTSDRVNNTAYTNITRYRSWISQVAGV'"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results['results'][0]['to']['sequence']['value']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "0d06bc8c",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Q75UA4\n"
     ]
    }
   ],
   "source": [
    "for result in results['results']:\n",
    "    print(result['from'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f536ba57",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "export_data = []\n",
    "sequence = results['results'][0]['to']['sequence']['value']\n",
    "for feature in results['results'][0]['to']['features']:\n",
    "    row = []\n",
    "    if feature['type'] == 'Natural variant' and disease in feature['description']:     \n",
    "        row.append(feature['featureId'])\n",
    "        location_start = feature['location']['start']['value']\n",
    "        location_end = feature['location']['end']['value']\n",
    "        if location_start == location_end:\n",
    "            row.append(location_start)\n",
    "            original_sequence = feature['alternativeSequence']['originalSequence']\n",
    "            new_sequence = feature['alternativeSequence']['alternativeSequences'][0]\n",
    "            row.append(f\"{original_sequence}->{new_sequence}\")\n",
    "            row.append(replace_char_at_index(sequence, int(location_start)-1, new_sequence))\n",
    "        else:\n",
    "            row.append(f\"{location_start}-{location_end}\")\n",
    "            row.append(\"missing\")\n",
    "            row.append(sequence[:int(location_start)-1] + sequence[int(location_end)-1:])\n",
    "        export_data.append(row)\n",
    "export_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c43c304e",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "df = pd.DataFrame(export_data, columns=[\"variant id\", \"position\", \"change\", \"sequence\"])\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0b712ab6",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "df.to_csv(\"results.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "41389a75",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[{'database': 'PubMed', 'id': '11133465'}, {'database': 'DOI', 'id': '10.1128/AEM.67.1.345-353.2001'}]\n",
      "[{'database': 'PubMed', 'id': '16237016'}, {'database': 'DOI', 'id': '10.1128/JB.187.21.7333-7340.2005'}]\n"
     ]
    }
   ],
   "source": [
    "result = results['results'][0]\n",
    "for reference in result['to']['references']:\n",
    "    print(reference['citation']['citationCrossReferences'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "3e9e1647-5992-4b1e-a28f-c1d73fd8af0f",
   "metadata": {},
   "outputs": [],
   "source": [
    "references_list = []\n",
    "result = results['results'][0]                 \n",
    "\n",
    "try:\n",
    "    for r in result['to']['references']:\n",
    "        tmp = {}\n",
    "        tmp[\"citacionCrossReferences\"] = r['citation']['citationCrossReferences']\n",
    "        tmp.update({\"title\": r['citation']['title']})\n",
    "        references_list.append(tmp)\n",
    "except KeyError:\n",
    "    pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "28767add-3dbd-45fd-9c7e-8a519ff59da0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'citacionCrossReferences': [{'database': 'PubMed', 'id': '11133465'},\n",
       "   {'database': 'DOI', 'id': '10.1128/AEM.67.1.345-353.2001'}],\n",
       "  'title': 'Purification and characterization of an extracellular poly(L-lactic acid) depolymerase from a soil isolate, Amycolatopsis sp. strain K104-1.'},\n",
       " {'citacionCrossReferences': [{'database': 'PubMed', 'id': '16237016'},\n",
       "   {'database': 'DOI', 'id': '10.1128/JB.187.21.7333-7340.2005'}],\n",
       "  'title': 'Gene cloning and molecular characterization of an extracellular poly(L-lactic acid) depolymerase from Amycolatopsis sp. strain K104-1.'}]"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "references_list"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1aa7cd1e",
   "metadata": {},
   "source": [
    "## Blast"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "1102bd78",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, argparse\n",
    "import shutil\n",
    "import subprocess\n",
    "import tarfile\n",
    "from pathlib import Path\n",
    "from urllib.request import urlopen\n",
    "import re\n",
    "from typing import List\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "DB_DIR = os.path.join(\"scripts\", \"db\")\n",
    "BLAST_BASE_URL = \"https://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/LATEST/\"\n",
    "UNIPROT_BASE_URL = \"https://ftp.uniprot.org/pub/databases/uniprot/current_release\"\n",
    "BLAST_DIR = Path(\"blast_bin\")\n",
    "#https://ftp.uniprot.org/pub/databases/uniprot/uniref/uniref100/uniref100.fasta.gz\n",
    "#https://ftp.uniprot.org/pub/databases/uniprot/current_release/uniref/uniref100/uniref.xsd\n",
    "#https://ftp.uniprot.org/pub/databases/uniprot/uniref/uniref100/uniref100.xml.gz\n",
    "#https://ftp.uniprot.org/pub/databases/uniprot/uniref/uniref90/uniref90.xml.gz\n",
    "#https://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/complete/uniprot_sprot.xml.gz\n",
    "#https://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/complete/uniprot_trembl.fasta.gz\n",
    "\n",
    "databases = {\n",
    "    \"uniprotkb_reviewed\": \"knowledgebase/complete/uniprot_sprot\",\n",
    "    \"uniprotkb_unreviewed\": \"knowledgebase/complete/uniprot_trembl\",\n",
    "    \"uniref100\": \"uniref/niref100/uniref100\",\n",
    "    \"uniref90\": \"uniref/uniref90/uniref90\",\n",
    "    \"uniref50\": \"uniref/uniref50/uniref50\",\n",
    "}\n",
    "\n",
    "def download_uniprot_database(db_name: str, extension: str = \"xml\"):\n",
    "    \"\"\" Download a Uniprot database from the Uniprot FTP server.\n",
    "    Args:\n",
    "        db_name (str): Name of the database to download.\n",
    "        extension (str): File extension of the database. Default is \"xml\".\n",
    "    \"\"\"\n",
    "\n",
    "    if db_name not in databases:\n",
    "        raise ValueError(f\"Database {db_name} is not supported. Supported databases are: {', '.join(databases.keys())}.\")\n",
    "    \n",
    "    db_path = os.path.join(DB_DIR, f\"{db_name}.{extension}\")\n",
    "    \n",
    "    if not os.path.exists(db_path):\n",
    "        os.makedirs(DB_DIR, exist_ok=True)\n",
    "        url = f\"{UNIPROT_BASE_URL}/{databases[db_name]}.{extension}.gz\"\n",
    "        os.system(f\"wget {url} -O {db_path}.gz\")\n",
    "        print(f\"Unzipping {db_path}...\")\n",
    "        subprocess.run([\"gunzip\", db_path], check=True)\n",
    "    else:\n",
    "        print(f\"Database {db_name} already exists at {db_path}.\")\n",
    "\n",
    "def get_latest_version_url():\n",
    "    \"\"\"Retrieve the latest BLAST+ tarball URL from the NCBI FTP site.\"\"\"\n",
    "    with urlopen(BLAST_BASE_URL) as response:\n",
    "        html = response.read().decode(\"utf-8\")\n",
    "    # Look for something like: ncbi-blast-2.16.0+-x64-linux.tar.gz\n",
    "    match = re.search(r'ncbi-blast-(\\d+\\.\\d+\\.\\d+\\+)-x64-linux\\.tar\\.gz', html)\n",
    "    if match:\n",
    "        version = match.group(1)\n",
    "        tar_name = f\"ncbi-blast-{version}-x64-linux.tar.gz\"\n",
    "        return version, BLAST_BASE_URL + tar_name\n",
    "    else:\n",
    "        raise RuntimeError(\"Could not find the latest BLAST version from NCBI.\")\n",
    "\n",
    "def is_blast_installed():\n",
    "    \"\"\"Check if 'blastp' is available in the system PATH.\"\"\"\n",
    "    try:\n",
    "        subprocess.run([\"blastp\", \"-version\"], check=True, stdout=subprocess.DEVNULL)\n",
    "        return True\n",
    "    except (subprocess.CalledProcessError, FileNotFoundError):\n",
    "        return False\n",
    "\n",
    "\n",
    "def download_and_extract_blast(version: str, url: str):\n",
    "    \"\"\"Download and extract the BLAST+ tarball.\"\"\"\n",
    "    tarball_name = url.split(\"/\")[-1]\n",
    "    if not Path(tarball_name).exists():\n",
    "        print(f\"Downloading BLAST+ {version}...\")\n",
    "        subprocess.run([\"wget\", url], check=True)\n",
    "\n",
    "    print(\"Extracting BLAST+...\")\n",
    "    with tarfile.open(tarball_name, \"r:gz\") as tar:\n",
    "        tar.extractall(BLAST_DIR)\n",
    "    print(f\"BLAST extracted to: {BLAST_DIR.resolve()}\")\n",
    "\n",
    "\n",
    "def get_local_blastp_path(version: str):\n",
    "    \"\"\"Return the path to local blastp binary.\"\"\"\n",
    "    return BLAST_DIR / f\"ncbi-blast-{version}\" / \"bin\" / \"blastp\"\n",
    "\n",
    "\n",
    "def check_blast():\n",
    "    \"\"\"Ensure BLAST is installed. Return path to `blastp` binary.\"\"\"\n",
    "    if is_blast_installed():\n",
    "        print(\"System-wide BLAST is installed.\")\n",
    "        return shutil.which(\"blastp\")\n",
    "    else:\n",
    "        version, url = get_latest_version_url()\n",
    "        local_blastp = get_local_blastp_path(version)\n",
    "        if not local_blastp.exists():\n",
    "            print(f\"BLAST {version} not found locally. Installing...\")\n",
    "            BLAST_DIR.mkdir(exist_ok=True)\n",
    "            download_and_extract_blast(version, url)\n",
    "        else:\n",
    "            print(f\"Using already downloaded BLAST {version}.\")\n",
    "        return str(local_blastp)\n",
    "\n",
    "def make_blast_database(db_name: str, db_type: str = \"prot\", extension: str = \"xml\"):\n",
    "    \"\"\"Create a BLAST database from the Uniprot database.\"\"\"\n",
    "    db_path = os.path.join(DB_DIR, f\"{db_name}.{extension}\")\n",
    "    if not os.path.exists(db_path):\n",
    "        raise FileNotFoundError(f\"Database {db_name} not found at {db_path}. Please download it first.\")\n",
    "    \n",
    "    # Check if the database is already created\n",
    "    blast_db_path = os.path.join(DB_DIR, db_name)\n",
    "    extensions = [\".pdb\", \".phr\", \".pin\", \".psq\", \".pot\", \".psq\", \".ptf\", \".pto\"]\n",
    "    makedb = False\n",
    "    # For all extensions check if exists if there is one failing makedb again\n",
    "    for ext in extensions:\n",
    "        if not os.path.exists(blast_db_path + \"/db\" + ext):\n",
    "            makedb = True\n",
    "            break\n",
    "    if makedb:\n",
    "        print(f\"Creating BLAST database for {db_name}...\")\n",
    "        blast_db_cmd = [\n",
    "            \"makeblastdb\",\n",
    "            \"-in\", db_path,\n",
    "            \"-dbtype\", db_type,\n",
    "            \"-out\", os.path.join(DB_DIR, db_name) + \"/db\",\n",
    "        ]\n",
    "    \n",
    "        subprocess.run(blast_db_cmd, check=True)\n",
    "        print(f\"BLAST database created at: {os.path.join(DB_DIR, databases[db_name])}\")\n",
    "    else:\n",
    "        print(f\"BLAST database already exists at {blast_db_path}. No need to create it again.\")\n",
    "\n",
    "def run_blast(sequences: List[str], db_name: str, blast_type: str = \"blastp\", evalue: float = 0.001):\n",
    "    \"\"\"Run BLAST search.\"\"\"\n",
    "    blast_db_path = os.path.join(DB_DIR, db_name)\n",
    "    if not os.path.exists(blast_db_path):\n",
    "        raise FileNotFoundError(f\"Database {db_name} not found at {blast_db_path}. Please download it first.\")\n",
    "\n",
    "    # Make tmp directory if it does not exist\n",
    "    os.makedirs(\"tmp\", exist_ok=True)\n",
    "\n",
    "    # Write sequences to a temporary file\n",
    "    with open(\"tmp/sequences.fasta\", \"w\") as f:\n",
    "        for i, seq in enumerate(sequences):\n",
    "            f.write(f\">{i}\\n{seq}\\n\")\n",
    "    \n",
    "    blast_cmd = [\n",
    "        blast_type,\n",
    "        \"-query\", \"tmp/sequences.fasta\",\n",
    "        \"-db\", blast_db_path + \"/db\",\n",
    "        \"-outfmt\", \"6\",\n",
    "        \"-evalue\", str(evalue),\n",
    "    ]\n",
    "    \n",
    "    print(f\"Running BLAST search...\")\n",
    "    with open(\"tmp/blast_results.txt\", \"w\") as f:\n",
    "        subprocess.run(blast_cmd, stdout=f, check=True)\n",
    "    print(f\"BLAST results saved to tmp/blast_results.txt\")\n",
    "    # Clean up temporary file\n",
    "    os.remove(\"tmp/sequences.fasta\")\n",
    "\n",
    "def parse_blast_results(file_path: str, identity_threshold: float = 90.0):\n",
    "    \"\"\"Parse BLAST results from a file.\"\"\"\n",
    "    with open(file_path, \"r\") as f:\n",
    "        results = f.readlines()\n",
    "    \n",
    "    parsed_results = []\n",
    "    for line in results:\n",
    "        fields = line.strip().split(\"\\t\")\n",
    "        identity = float(fields[2])\n",
    "        if identity >= identity_threshold:\n",
    "            parsed_results.append({\n",
    "                \"query\": fields[0],\n",
    "                \"subject\": fields[1],\n",
    "                \"identity\": fields[2],\n",
    "                \"alignment_length\": fields[3],\n",
    "                \"evalue\": fields[4],\n",
    "                \"bit_score\": fields[5],\n",
    "            })\n",
    "    \n",
    "    return parsed_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "de905467",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Database uniprotkb_reviewed already exists at scripts/db/uniprotkb_reviewed.fasta.\n",
      "System-wide BLAST is installed.\n",
      "Using blastp at: /home/diego/micromamba/envs/bioseqdownloader/bin/blastp\n",
      "BLAST database already exists at scripts/db/uniprotkb_reviewed. No need to create it again.\n",
      "Running BLAST search...\n",
      "BLAST results saved to tmp/blast_results.txt\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequences</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MARPLLGKTSSVRRRLESLSACSIFFFLRKFCQKMASLVFLNSPVY...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MSFKVYDPIAELIATQFPTSNPDLQIINNDVLVVSPHKITLPMGPQ...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NYTETAQAIARSWRAGSHDRLKARGEAVAVTVHRLVAVPRGRDTPR...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           sequences  id\n",
       "0  MARPLLGKTSSVRRRLESLSACSIFFFLRKFCQKMASLVFLNSPVY...   0\n",
       "1  MSFKVYDPIAELIATQFPTSNPDLQIINNDVLVVSPHKITLPMGPQ...   1\n",
       "2  NYTETAQAIARSWRAGSHDRLKARGEAVAVTVHRLVAVPRGRDTPR...   2"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv(\"data/test.csv\")\n",
    "sequences = df[\"sequences\"].dropna().unique().tolist()\n",
    "    \n",
    "download_uniprot_database(\"uniprotkb_reviewed\", \"fasta\")\n",
    "    \n",
    "blastp_path = check_blast()\n",
    "print(f\"Using blastp at: {blastp_path}\")\n",
    "\n",
    "make_blast_database(\"uniprotkb_reviewed\", extension=\"fasta\")\n",
    "\n",
    "run_blast(sequences, \"uniprotkb_reviewed\", blast_type=\"blastp\", evalue=0.0001)\n",
    "\n",
    "results = parse_blast_results(\"tmp/blast_results.txt\")\n",
    "\n",
    "# Convert to DataFrame\n",
    "sequences_df = pd.DataFrame(sequences, columns=[\"sequences\"])\n",
    "sequences_df[\"id\"] = sequences_df.index\n",
    "\n",
    "sequences_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "4009c945",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_blast = pd.DataFrame(results)\n",
    "\n",
    "df_blast = df_blast.rename(columns={\"query\": \"id\", \"subject\": \"subject_id\"})\n",
    "df_blast[\"id\"] = df_blast[\"id\"].astype(int)\n",
    "df_blast = df_blast.merge(sequences_df, on=\"id\", how=\"left\")\n",
    "df_blast = df_blast.drop(columns=[\"id\"])\n",
    "df_blast = df_blast.rename(columns={\"sequences\": \"sequence\"})\n",
    "\n",
    "# Separate subject into source, accession, entry_name\n",
    "df_blast[\"source\"] = df_blast[\"subject_id\"].apply(lambda x: x.split(\"|\")[0])\n",
    "df_blast[\"accession\"] = df_blast[\"subject_id\"].apply(lambda x: x.split(\"|\")[1])\n",
    "df_blast[\"entry_name\"] = df_blast[\"subject_id\"].apply(lambda x: x.split(\"|\")[2])\n",
    "df_blast = df_blast.drop(columns=[\"subject_id\"])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "a32f8e17",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identity</th>\n",
       "      <th>alignment_length</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bit_score</th>\n",
       "      <th>sequence</th>\n",
       "      <th>source</th>\n",
       "      <th>accession</th>\n",
       "      <th>entry_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>100.000</td>\n",
       "      <td>438</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>MARPLLGKTSSVRRRLESLSACSIFFFLRKFCQKMASLVFLNSPVY...</td>\n",
       "      <td>sp</td>\n",
       "      <td>Q6GZX2</td>\n",
       "      <td>003R_FRG3G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>100.000</td>\n",
       "      <td>180</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>MSFKVYDPIAELIATQFPTSNPDLQIINNDVLVVSPHKITLPMGPQ...</td>\n",
       "      <td>sp</td>\n",
       "      <td>Q197F2</td>\n",
       "      <td>008L_IIV3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>100.000</td>\n",
       "      <td>50</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>MSFKVYDPIAELIATQFPTSNPDLQIINNDVLVVSPHKITLPMGPQ...</td>\n",
       "      <td>sp</td>\n",
       "      <td>Q6GZW6</td>\n",
       "      <td>009L_FRG3G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>100.000</td>\n",
       "      <td>345</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NYTETAQAIARSWRAGSHDRLKARGEAVAVTVHRLVAVPRGRDTPR...</td>\n",
       "      <td>sp</td>\n",
       "      <td>Q6GZW6</td>\n",
       "      <td>009L_FRG3G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  identity alignment_length evalue bit_score  \\\n",
       "0  100.000              438      0         0   \n",
       "1  100.000              180      0         0   \n",
       "2  100.000               50      0         0   \n",
       "3  100.000              345      0         0   \n",
       "\n",
       "                                            sequence source accession  \\\n",
       "0  MARPLLGKTSSVRRRLESLSACSIFFFLRKFCQKMASLVFLNSPVY...     sp    Q6GZX2   \n",
       "1  MSFKVYDPIAELIATQFPTSNPDLQIINNDVLVVSPHKITLPMGPQ...     sp    Q197F2   \n",
       "2  MSFKVYDPIAELIATQFPTSNPDLQIINNDVLVVSPHKITLPMGPQ...     sp    Q6GZW6   \n",
       "3  NYTETAQAIARSWRAGSHDRLKARGEAVAVTVHRLVAVPRGRDTPR...     sp    Q6GZW6   \n",
       "\n",
       "   entry_name  \n",
       "0  003R_FRG3G  \n",
       "1   008L_IIV3  \n",
       "2  009L_FRG3G  \n",
       "3  009L_FRG3G  "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_blast"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d944e22",
   "metadata": {},
   "source": [
    "# GO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "5899c06f",
   "metadata": {},
   "outputs": [],
   "source": [
    "from scripts.description_go import *\n",
    "import os, ast\n",
    "import pandas as pd\n",
    "from tqdm import tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "b6b45cd2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[DESCRIPTION_GO] Getting Gen Ontology\n",
      "Docker version 28.0.0, build f9ced58158\n",
      "[DESCRIPTION_GO] Metastudent found.\n",
      "[DESCRIPTION_GO] Go terms found in input data.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>entry_type</th>\n",
       "      <th>protein_name</th>\n",
       "      <th>ec_numbers</th>\n",
       "      <th>organism</th>\n",
       "      <th>taxon_id</th>\n",
       "      <th>sequence</th>\n",
       "      <th>length</th>\n",
       "      <th>go_terms</th>\n",
       "      <th>pfam_ids</th>\n",
       "      <th>references</th>\n",
       "      <th>features</th>\n",
       "      <th>keywords</th>\n",
       "      <th>source_db</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [uniprot_id, entry_type, protein_name, ec_numbers, organism, taxon_id, sequence, length, go_terms, pfam_ids, references, features, keywords, source_db]\n",
       "Index: []"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DOCKER_IMAGE_NAME = \"metastudent\"\n",
    "DOCKER_CONTAINER_NAME = \"metastudent_container\"\n",
    "HOST_INPUT_FILE = os.path.abspath(\"tmp/sequences.fasta\")\n",
    "HOST_OUTPUT_DIR = os.path.abspath(\"tmp/\")\n",
    "CONTAINER_INPUT_FILE = \"/app/input.fasta\"\n",
    "CONTAINER_OUTPUT_FILE = \"/app/output.result\"\n",
    "\n",
    "\n",
    "print(\"[DESCRIPTION_GO] Getting Gen Ontology\")\n",
    "tqdm.pandas()\n",
    "\n",
    "if not check_dependencies(DOCKER_IMAGE_NAME):\n",
    "    print(\"[DESCRIPTION_GO] Metastudent not found. Installing...\")\n",
    "    install_dependencies(DOCKER_IMAGE_NAME)\n",
    "else:\n",
    "    print(\"[DESCRIPTION_GO] Metastudent found.\")\n",
    "\n",
    "input_df = pd.read_csv(\"results/umami_uniprot.csv\")\n",
    "obsolete_df = pd.read_csv(\"scripts/resources/amiGO_data.csv\", sep=\"\\t\", names=[\"id_go\", \"description\", \"is_obsolete\"])\n",
    "input_df['go_terms'] = input_df['go_terms'].apply(lambda x: ast.literal_eval(x) if isinstance(x, str) else x)\n",
    "\n",
    "parsed_df = pd.DataFrame()\n",
    "if os.path.isfile(f\"{HOST_OUTPUT_DIR}/output.BPO.txt\") and \\\n",
    "        os.path.isfile(f\"{HOST_OUTPUT_DIR}/output.CCO.txt\") and \\\n",
    "        os.path.isfile(f\"{HOST_OUTPUT_DIR}/output.MFO.txt\"):\n",
    "    print(\"[DESCRIPTION_GO] Metastudent results found.\")\n",
    "    parsed_df = parse_outputs(\"uniprot_id\")\n",
    "\n",
    "# Filter input_df with go_terms ~= null\n",
    "input_df_with_go_terms = input_df[input_df[\"go_terms\"].apply(lambda x: isinstance(x, list) and len(x) > 0)]\n",
    "input_df = input_df[input_df[\"go_terms\"].apply(lambda x: isinstance(x, list) and len(x) == 0)]\n",
    "\n",
    "if not input_df_with_go_terms.empty:\n",
    "    print(\"[DESCRIPTION_GO] Go terms found in input data.\")\n",
    "    input_df_with_go_terms = input_df_with_go_terms[[\"uniprot_id\", \"go_terms\"]]\n",
    "    input_df_with_go_terms = input_df_with_go_terms.explode(\"go_terms\")\n",
    "    parsed_df = pd.concat(\n",
    "        [\n",
    "            parsed_df,\n",
    "            pd.merge(\n",
    "                input_df_with_go_terms, \n",
    "                obsolete_df, \n",
    "                left_on=\"go_terms\", \n",
    "                right_on=\"id_go\", \n",
    "                how=\"left\"\n",
    "            )\n",
    "            .drop(columns=[\"go_terms\"])\n",
    "            .rename(columns={\"id_go\": \"go\"})  \n",
    "        ]\n",
    "    )\n",
    "\n",
    "input_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "6a2a9d8e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[DESCRIPTION_GO] 0 sequences have not been processed.\n"
     ]
    }
   ],
   "source": [
    "if not parsed_df.empty:\n",
    "    # Check if all sequences have been processed\n",
    "    parsed_ids = parsed_df[\"uniprot_id\"].unique()\n",
    "    input_ids = input_df[\"uniprot_id\"].unique()\n",
    "    if len(parsed_ids) == len(input_ids):\n",
    "        print(\"[DESCRIPTION_GO] All sequences have been processed.\")\n",
    "        input_df = pd.DataFrame()\n",
    "    else:\n",
    "        input_df = input_df[~input_df[\"uniprot_id\"].isin(parsed_ids)]\n",
    "        print(f\"[DESCRIPTION_GO] {len(input_df)} sequences have not been processed.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "d815e6cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.makedirs(HOST_OUTPUT_DIR, exist_ok=True)\n",
    "if not input_df.empty:\n",
    "    print(\"[DESCRIPTION_GO] Running in batches of 50...\")\n",
    "    for i in tqdm(range(0, len(input_df), 50)):\n",
    "        run_in_batches(input_df[i:i+50], HOST_OUTPUT_DIR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7cffac13",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File '/home/diego/Documents/PythonProjects/BioSeqDownloader/tmp/output.BPO.txt' not found.\n"
     ]
    },
    {
     "ename": "NameError",
     "evalue": "name 'exit' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mFileNotFoundError\u001b[39m                         Traceback (most recent call last)",
      "\u001b[36mFile \u001b[39m\u001b[32m~/Documents/PythonProjects/BioSeqDownloader/scripts/description_go.py:27\u001b[39m, in \u001b[36mparse_outputs\u001b[39m\u001b[34m(id_column)\u001b[39m\n\u001b[32m     26\u001b[39m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[32m---> \u001b[39m\u001b[32m27\u001b[39m     df_go = \u001b[43mpd\u001b[49m\u001b[43m.\u001b[49m\u001b[43mread_csv\u001b[49m\u001b[43m(\u001b[49m\n\u001b[32m     28\u001b[39m \u001b[43m            \u001b[49m\u001b[43mfile\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\n\u001b[32m     29\u001b[39m \u001b[43m            \u001b[49m\u001b[43mheader\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\n\u001b[32m     30\u001b[39m \u001b[43m            \u001b[49m\u001b[43msep\u001b[49m\u001b[43m=\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[38;5;130;43;01m\\t\u001b[39;49;00m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\n\u001b[32m     31\u001b[39m \u001b[43m            \u001b[49m\u001b[43mnames\u001b[49m\u001b[43m=\u001b[49m\u001b[43m[\u001b[49m\u001b[43mid_column\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mgo\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mprobability\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mterm\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m     32\u001b[39m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m     33\u001b[39m     df_go[\u001b[33m'\u001b[39m\u001b[33mgo_type\u001b[39m\u001b[33m'\u001b[39m] = file_type\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/micromamba/envs/bioseqdownloader/lib/python3.13/site-packages/pandas/io/parsers/readers.py:1026\u001b[39m, in \u001b[36mread_csv\u001b[39m\u001b[34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, date_format, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, encoding_errors, dialect, on_bad_lines, delim_whitespace, low_memory, memory_map, float_precision, storage_options, dtype_backend)\u001b[39m\n\u001b[32m   1024\u001b[39m kwds.update(kwds_defaults)\n\u001b[32m-> \u001b[39m\u001b[32m1026\u001b[39m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_read\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilepath_or_buffer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/micromamba/envs/bioseqdownloader/lib/python3.13/site-packages/pandas/io/parsers/readers.py:620\u001b[39m, in \u001b[36m_read\u001b[39m\u001b[34m(filepath_or_buffer, kwds)\u001b[39m\n\u001b[32m    619\u001b[39m \u001b[38;5;66;03m# Create the parser.\u001b[39;00m\n\u001b[32m--> \u001b[39m\u001b[32m620\u001b[39m parser = \u001b[43mTextFileReader\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilepath_or_buffer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m*\u001b[49m\u001b[43m*\u001b[49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    622\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m chunksize \u001b[38;5;129;01mor\u001b[39;00m iterator:\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/micromamba/envs/bioseqdownloader/lib/python3.13/site-packages/pandas/io/parsers/readers.py:1620\u001b[39m, in \u001b[36mTextFileReader.__init__\u001b[39m\u001b[34m(self, f, engine, **kwds)\u001b[39m\n\u001b[32m   1619\u001b[39m \u001b[38;5;28mself\u001b[39m.handles: IOHandles | \u001b[38;5;28;01mNone\u001b[39;00m = \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[32m-> \u001b[39m\u001b[32m1620\u001b[39m \u001b[38;5;28mself\u001b[39m._engine = \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43m_make_engine\u001b[49m\u001b[43m(\u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43mengine\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/micromamba/envs/bioseqdownloader/lib/python3.13/site-packages/pandas/io/parsers/readers.py:1880\u001b[39m, in \u001b[36mTextFileReader._make_engine\u001b[39m\u001b[34m(self, f, engine)\u001b[39m\n\u001b[32m   1879\u001b[39m         mode += \u001b[33m\"\u001b[39m\u001b[33mb\u001b[39m\u001b[33m\"\u001b[39m\n\u001b[32m-> \u001b[39m\u001b[32m1880\u001b[39m \u001b[38;5;28mself\u001b[39m.handles = \u001b[43mget_handle\u001b[49m\u001b[43m(\u001b[49m\n\u001b[32m   1881\u001b[39m \u001b[43m    \u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1882\u001b[39m \u001b[43m    \u001b[49m\u001b[43mmode\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1883\u001b[39m \u001b[43m    \u001b[49m\u001b[43mencoding\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43moptions\u001b[49m\u001b[43m.\u001b[49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mencoding\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1884\u001b[39m \u001b[43m    \u001b[49m\u001b[43mcompression\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43moptions\u001b[49m\u001b[43m.\u001b[49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mcompression\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1885\u001b[39m \u001b[43m    \u001b[49m\u001b[43mmemory_map\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43moptions\u001b[49m\u001b[43m.\u001b[49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mmemory_map\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1886\u001b[39m \u001b[43m    \u001b[49m\u001b[43mis_text\u001b[49m\u001b[43m=\u001b[49m\u001b[43mis_text\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1887\u001b[39m \u001b[43m    \u001b[49m\u001b[43merrors\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43moptions\u001b[49m\u001b[43m.\u001b[49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mencoding_errors\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mstrict\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1888\u001b[39m \u001b[43m    \u001b[49m\u001b[43mstorage_options\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43moptions\u001b[49m\u001b[43m.\u001b[49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mstorage_options\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1889\u001b[39m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m   1890\u001b[39m \u001b[38;5;28;01massert\u001b[39;00m \u001b[38;5;28mself\u001b[39m.handles \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/micromamba/envs/bioseqdownloader/lib/python3.13/site-packages/pandas/io/common.py:873\u001b[39m, in \u001b[36mget_handle\u001b[39m\u001b[34m(path_or_buf, mode, encoding, compression, memory_map, is_text, errors, storage_options)\u001b[39m\n\u001b[32m    871\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m ioargs.encoding \u001b[38;5;129;01mand\u001b[39;00m \u001b[33m\"\u001b[39m\u001b[33mb\u001b[39m\u001b[33m\"\u001b[39m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m ioargs.mode:\n\u001b[32m    872\u001b[39m     \u001b[38;5;66;03m# Encoding\u001b[39;00m\n\u001b[32m--> \u001b[39m\u001b[32m873\u001b[39m     handle = \u001b[38;5;28;43mopen\u001b[39;49m\u001b[43m(\u001b[49m\n\u001b[32m    874\u001b[39m \u001b[43m        \u001b[49m\u001b[43mhandle\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    875\u001b[39m \u001b[43m        \u001b[49m\u001b[43mioargs\u001b[49m\u001b[43m.\u001b[49m\u001b[43mmode\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    876\u001b[39m \u001b[43m        \u001b[49m\u001b[43mencoding\u001b[49m\u001b[43m=\u001b[49m\u001b[43mioargs\u001b[49m\u001b[43m.\u001b[49m\u001b[43mencoding\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    877\u001b[39m \u001b[43m        \u001b[49m\u001b[43merrors\u001b[49m\u001b[43m=\u001b[49m\u001b[43merrors\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    878\u001b[39m \u001b[43m        \u001b[49m\u001b[43mnewline\u001b[49m\u001b[43m=\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[32m    879\u001b[39m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    880\u001b[39m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[32m    881\u001b[39m     \u001b[38;5;66;03m# Binary mode\u001b[39;00m\n",
      "\u001b[31mFileNotFoundError\u001b[39m: [Errno 2] No such file or directory: '/home/diego/Documents/PythonProjects/BioSeqDownloader/tmp/output.BPO.txt'",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[31mNameError\u001b[39m                                 Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[23]\u001b[39m\u001b[32m, line 4\u001b[39m\n\u001b[32m      1\u001b[39m test = pd.concat(\n\u001b[32m      2\u001b[39m     [\n\u001b[32m      3\u001b[39m         parsed_df,\n\u001b[32m----> \u001b[39m\u001b[32m4\u001b[39m         \u001b[43mparse_outputs\u001b[49m\u001b[43m(\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43muniprot_id\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[32m      5\u001b[39m     ]\n\u001b[32m      6\u001b[39m )\n\u001b[32m      8\u001b[39m test = test.sort_values(by=\u001b[33m\"\u001b[39m\u001b[33muniprot_id\u001b[39m\u001b[33m\"\u001b[39m)\n\u001b[32m      9\u001b[39m test = test.merge(obsolete_df, left_on=\u001b[33m\"\u001b[39m\u001b[33mgo\u001b[39m\u001b[33m\"\u001b[39m, right_on=\u001b[33m\"\u001b[39m\u001b[33mid_go\u001b[39m\u001b[33m\"\u001b[39m, how=\u001b[33m\"\u001b[39m\u001b[33mleft\u001b[39m\u001b[33m\"\u001b[39m)\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/Documents/PythonProjects/BioSeqDownloader/scripts/description_go.py:40\u001b[39m, in \u001b[36mparse_outputs\u001b[39m\u001b[34m(id_column)\u001b[39m\n\u001b[32m     38\u001b[39m     \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mFileNotFoundError\u001b[39;00m:\n\u001b[32m     39\u001b[39m         \u001b[38;5;28mprint\u001b[39m(\u001b[33mf\u001b[39m\u001b[33m\"\u001b[39m\u001b[33mFile \u001b[39m\u001b[33m'\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mfile\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m'\u001b[39m\u001b[33m not found.\u001b[39m\u001b[33m\"\u001b[39m)\n\u001b[32m---> \u001b[39m\u001b[32m40\u001b[39m         \u001b[43mexit\u001b[49m(\u001b[32m1\u001b[39m)\n\u001b[32m     42\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m results:\n\u001b[32m     43\u001b[39m     \u001b[38;5;66;03m# df = pd.DataFrame({\u001b[39;00m\n\u001b[32m     44\u001b[39m     \u001b[38;5;66;03m#     \"id_seq\": [\"None\"],\u001b[39;00m\n\u001b[32m   (...)\u001b[39m\u001b[32m     48\u001b[39m     \u001b[38;5;66;03m#     \"go\": [\"No results found\"]\u001b[39;00m\n\u001b[32m     49\u001b[39m     \u001b[38;5;66;03m# })\u001b[39;00m\n\u001b[32m     50\u001b[39m     df = pd.DataFrame(columns=[id_column, \u001b[33m\"\u001b[39m\u001b[33mgo\u001b[39m\u001b[33m\"\u001b[39m, \u001b[33m\"\u001b[39m\u001b[33mprobability\u001b[39m\u001b[33m\"\u001b[39m, \u001b[33m\"\u001b[39m\u001b[33mterm\u001b[39m\u001b[33m\"\u001b[39m, \u001b[33m\"\u001b[39m\u001b[33mgo_type\u001b[39m\u001b[33m\"\u001b[39m])\n",
      "\u001b[31mNameError\u001b[39m: name 'exit' is not defined"
     ]
    }
   ],
   "source": [
    "test = pd.concat(\n",
    "    [\n",
    "        parsed_df,\n",
    "        parse_outputs(\"uniprot_id\")\n",
    "    ]\n",
    ")\n",
    "    \n",
    "test = test.sort_values(by=\"uniprot_id\")\n",
    "test = test.merge(obsolete_df, left_on=\"go\", right_on=\"id_go\", how=\"left\")\n",
    "test = test.drop(columns=[\"id_go\"])\n",
    "\n",
    "test"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "504746e9",
   "metadata": {},
   "source": [
    "## Uniprot query"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9df8cf72",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Current working directory: /home/diego/Documents/PythonProjects/BioSeqDownloader/src\n"
     ]
    }
   ],
   "source": [
    "import os, json\n",
    "os.chdir(\"src/\")\n",
    "print(f\"Current working directory: {os.getcwd()}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "ec84c845",
   "metadata": {},
   "outputs": [],
   "source": [
    "query=\"organism_name:homo sapiens (human) AND length:[15 TO 30] AND reviewed:true\"\n",
    "fields=\"accession,protein_name,sequence,ec,lineage,organism_name,xref_pfam,xref_alphafolddb,xref_pdb,go_id\"\n",
    "sort=\"accession asc\"\n",
    "download=True\n",
    "format=\"json\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "6f64e04b",
   "metadata": {},
   "outputs": [],
   "source": [
    "from uniprot import UniprotInterface\n",
    "\n",
    "uniprot = UniprotInterface()\n",
    "response = uniprot.submit_stream(\n",
    "    query=query,\n",
    "    fields=fields,\n",
    "    sort=sort,\n",
    "    include_isoform=True,\n",
    "    download=download,\n",
    "    format=format\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "02c0f301",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query</th>\n",
       "      <th>accession</th>\n",
       "      <th>protein_name</th>\n",
       "      <th>organism_name</th>\n",
       "      <th>taxon_id</th>\n",
       "      <th>ineage</th>\n",
       "      <th>sequence</th>\n",
       "      <th>length</th>\n",
       "      <th>go_terms</th>\n",
       "      <th>pfam_ids</th>\n",
       "      <th>alphafold_ids</th>\n",
       "      <th>pdb_ids</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>A0A075B6S0</td>\n",
       "      <td>T cell receptor gamma joining 1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>NYYKKLFGSGTTLVVT</td>\n",
       "      <td>16</td>\n",
       "      <td>[GO:0042101, GO:0002250]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[A0A075B6S0]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>A0A075B6Y3</td>\n",
       "      <td>T cell receptor alpha joining 3</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>GYSSASKIIFGSGTRLSIRP</td>\n",
       "      <td>20</td>\n",
       "      <td>[GO:0042101, GO:0002250]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[A0A075B6Y3]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>A0A075B6Y9</td>\n",
       "      <td>T cell receptor alpha joining 42</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>YGGSQGNLIFGKGTKLSVKP</td>\n",
       "      <td>20</td>\n",
       "      <td>[GO:0005886, GO:0002250]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[A0A075B6Y9]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>A0A075B700</td>\n",
       "      <td>T cell receptor alpha joining 31</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>NNNARLMFGDGTQLVVKP</td>\n",
       "      <td>18</td>\n",
       "      <td>[GO:0005886, GO:0002250]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[A0A075B700]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>A0A075B706</td>\n",
       "      <td>T cell receptor delta joining 1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>TDKLIFGKGTRVTVEP</td>\n",
       "      <td>16</td>\n",
       "      <td>[GO:0042101, GO:0002250]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[A0A075B706]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>A0A0A0MT70</td>\n",
       "      <td>T cell receptor beta joining 2-6</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>SGANVLTFGAGSRLTVL</td>\n",
       "      <td>17</td>\n",
       "      <td>[GO:0042101, GO:0002250]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[A0A0A0MT70]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>A0A0A0MT78</td>\n",
       "      <td>T cell receptor beta joining 2-7</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>SYEQYFGPGTRLTVT</td>\n",
       "      <td>15</td>\n",
       "      <td>[GO:0042101, GO:0002250]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>A0A0A0MT87</td>\n",
       "      <td>T cell receptor beta joining 2-4</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>AKNIQYFGAGTRLSVL</td>\n",
       "      <td>16</td>\n",
       "      <td>[GO:0042101, GO:0002250]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[A0A0A0MT87]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>A0A0A0MT94</td>\n",
       "      <td>T cell receptor beta joining 2-2</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>NTGELFFGEGSRLTVL</td>\n",
       "      <td>16</td>\n",
       "      <td>[GO:0042101, GO:0002250]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[A0A0A0MT94]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>A0A0A0MTA4</td>\n",
       "      <td>T cell receptor beta joining 2-5</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>QETQYFGPGTRLLVL</td>\n",
       "      <td>15</td>\n",
       "      <td>[GO:0042101, GO:0002250]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>A0A0A0MTA7</td>\n",
       "      <td>T cell receptor beta joining 2-1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>SYNEQFFGPGTRLTVL</td>\n",
       "      <td>16</td>\n",
       "      <td>[GO:0042101, GO:0002250]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[A0A0A0MTA7]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>A0A0B4J200</td>\n",
       "      <td>T cell receptor beta joining 2-3</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>STDTQYFGPGTRLTVL</td>\n",
       "      <td>16</td>\n",
       "      <td>[GO:0042101, GO:0002250]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[A0A0B4J200]</td>\n",
       "      <td>[7NDT]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>A0A0C4DH62</td>\n",
       "      <td>Immunoglobulin heavy joining 1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>AEYFQHWGQGTLVTVSS</td>\n",
       "      <td>17</td>\n",
       "      <td>[GO:0005576, GO:0019814, GO:0005886, GO:0002250]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[A0A0C4DH62]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>A0A0C5B5G6</td>\n",
       "      <td>Mitochondrial-derived peptide MOTS-c</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>MRWQEMGYIFYPRKLR</td>\n",
       "      <td>16</td>\n",
       "      <td>[GO:0005615, GO:0005739, GO:0005634, GO:000367...</td>\n",
       "      <td>[PF21945]</td>\n",
       "      <td>[A0A0C5B5G6]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>A0A0J9YWP8</td>\n",
       "      <td>T cell receptor beta joining 1-3</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>SGNTIYFGEGSWLTVV</td>\n",
       "      <td>16</td>\n",
       "      <td>[GO:0042101, GO:0002250]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[A0A0J9YWP8]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>A0A0J9YWX3</td>\n",
       "      <td>T cell receptor beta joining 1-6</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>SYNSPLHFGNGTRLTVT</td>\n",
       "      <td>17</td>\n",
       "      <td>[GO:0042101, GO:0002250]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[A0A0J9YWX3]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>A0A0J9YX06</td>\n",
       "      <td>T cell receptor beta joining 1-2</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>NYGYTFGSGTRLTVV</td>\n",
       "      <td>15</td>\n",
       "      <td>[GO:0042101, GO:0002250]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[7NDQ, 7NDU]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>A0A0J9YXA8</td>\n",
       "      <td>T cell receptor beta joining 1-1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>NTEAFFGQGTRLTVV</td>\n",
       "      <td>15</td>\n",
       "      <td>[GO:0042101, GO:0002250]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>A0A0J9YXG5</td>\n",
       "      <td>T cell receptor beta joining 1-4</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>TNEKLFFGSGTQLSVL</td>\n",
       "      <td>16</td>\n",
       "      <td>[GO:0042101, GO:0002250]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[A0A0J9YXG5]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>A0A0J9YXM7</td>\n",
       "      <td>T cell receptor beta joining 1-5</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>SNQPQHFGDGTRLSIL</td>\n",
       "      <td>16</td>\n",
       "      <td>[GO:0042101, GO:0002250]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[A0A0J9YXM7]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>A0A1B0GWH6</td>\n",
       "      <td>LASP1 neighbor protein</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>MLDIFILMFFAIIGLVILSYIIYLL</td>\n",
       "      <td>25</td>\n",
       "      <td>[GO:0016020]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[A0A1B0GWH6]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>C0HM02</td>\n",
       "      <td>PRKCH upstream open reading frame 2</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>MASRGALRRCLSPGLPRLLHLSRGLA</td>\n",
       "      <td>26</td>\n",
       "      <td>[GO:0120283, GO:0030291, GO:0045182, GO:003419...</td>\n",
       "      <td>[PF21952]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>C0HMA1</td>\n",
       "      <td>HSPA8-interacting micropeptide miPEP155</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>MEMALMVAQTRKGKSVV</td>\n",
       "      <td>17</td>\n",
       "      <td>[GO:0005737, GO:0005634, GO:0030544, GO:000237...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>D0EPY3</td>\n",
       "      <td>Small integral membrane protein 47</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>MNFQENVTLAMALFTILTSIYFFNKAQQ</td>\n",
       "      <td>28</td>\n",
       "      <td>[GO:0016020]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[D0EPY3]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>P0CJ68</td>\n",
       "      <td>Humanin-like 1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>MAPRGFSCLLLSTSEIDLPVKRRT</td>\n",
       "      <td>24</td>\n",
       "      <td>[GO:0005737, GO:0005576, GO:0048019, GO:1900118]</td>\n",
       "      <td>[PF15040]</td>\n",
       "      <td>[P0CJ68]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>P0CJ69</td>\n",
       "      <td>Humanin-like 2</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>MAPRGFSCLLLSTSEIDLPVKRLLSSVF</td>\n",
       "      <td>28</td>\n",
       "      <td>[GO:0005737, GO:0005576, GO:0048019, GO:1900118]</td>\n",
       "      <td>[PF15040]</td>\n",
       "      <td>[P0CJ69]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>P0CJ70</td>\n",
       "      <td>Humanin-like 3</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>MATRRFSCLLLSTSEIDLSVKRRI</td>\n",
       "      <td>24</td>\n",
       "      <td>[GO:0005737, GO:0005576, GO:0048019, GO:1900118]</td>\n",
       "      <td>[PF15040]</td>\n",
       "      <td>[P0CJ70]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>P0CJ71</td>\n",
       "      <td>Humanin-like 4</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>MATQGFSCLLLSVSEIDLSMKRQYKQIR</td>\n",
       "      <td>28</td>\n",
       "      <td>[GO:0005737, GO:0005576, GO:0048019, GO:1900118]</td>\n",
       "      <td>[PF15040]</td>\n",
       "      <td>[P0CJ71]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>P0CJ72</td>\n",
       "      <td>Humanin-like 5</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>MATPGFSCLLLSTSEIDLPMKRRV</td>\n",
       "      <td>24</td>\n",
       "      <td>[GO:0005576, GO:0005739, GO:0048019, GO:007273...</td>\n",
       "      <td>[PF15040]</td>\n",
       "      <td>[P0CJ72]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>P0CJ73</td>\n",
       "      <td>Humanin-like 6</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>MTPRGFSCLLLPTSETDLPVKRRT</td>\n",
       "      <td>24</td>\n",
       "      <td>[GO:0005737, GO:0005576, GO:0048019, GO:1900118]</td>\n",
       "      <td>[PF15040]</td>\n",
       "      <td>[P0CJ73]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>P0CJ74</td>\n",
       "      <td>Humanin-like 7</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>MATGGFGCLLLLIREIDLSVKRQI</td>\n",
       "      <td>24</td>\n",
       "      <td>[GO:0005737, GO:0005576, GO:0048019, GO:1900118]</td>\n",
       "      <td>[PF15040]</td>\n",
       "      <td>[P0CJ74]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>P0CJ75</td>\n",
       "      <td>Humanin-like 8</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>MAPRGFSCLLLSTSEIDLPVKRRA</td>\n",
       "      <td>24</td>\n",
       "      <td>[GO:0005737, GO:0005576, GO:0048019, GO:1900118]</td>\n",
       "      <td>[PF15040]</td>\n",
       "      <td>[P0CJ75]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>P0CJ76</td>\n",
       "      <td>Humanin-like 9</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>MARRGFSCLLLSTTATDLPVKRRT</td>\n",
       "      <td>24</td>\n",
       "      <td>[GO:0005737, GO:0005576, GO:0048019, GO:1900118]</td>\n",
       "      <td>[PF15040]</td>\n",
       "      <td>[P0CJ76]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>P0CJ77</td>\n",
       "      <td>Humanin-like 10</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>MTTRGFSCLLLLIREIDLSAKRRI</td>\n",
       "      <td>24</td>\n",
       "      <td>[GO:0005737, GO:0005576, GO:0048019, GO:1900118]</td>\n",
       "      <td>[PF15040]</td>\n",
       "      <td>[P0CJ77]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>P0DMP1</td>\n",
       "      <td>Humanin-like 12</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>MAPRGFSCLLLSTSEIDLPVKRRA</td>\n",
       "      <td>24</td>\n",
       "      <td>[GO:0005737, GO:0005576]</td>\n",
       "      <td>[PF15040]</td>\n",
       "      <td>[P0DMP1]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>P16070-2</td>\n",
       "      <td>Isoform 2 of CD44 antigen</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>MDKFWWHAAWGLCLVPLSLAQIGVGRRKS</td>\n",
       "      <td>29</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>P59103-2</td>\n",
       "      <td>Isoform 2 of D-amino acid oxidase regulator</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>MLEKLMGADSLQLFRV</td>\n",
       "      <td>16</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>P62945</td>\n",
       "      <td>Small ribosomal subunit protein eS32</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>MRAKWRKKRMRRLKRKRRKMRQRSK</td>\n",
       "      <td>25</td>\n",
       "      <td>[GO:0005737, GO:0005829, GO:0022625, GO:000578...</td>\n",
       "      <td>[PF05162]</td>\n",
       "      <td>[P62945]</td>\n",
       "      <td>[4UG0, 4V6X, 5A2Q, 5AJ0, 5LKS, 5OA3, 5VYC, 6G5...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>P84351</td>\n",
       "      <td>Osteocalcin</td>\n",
       "      <td>Homo sapiens neanderthalensis</td>\n",
       "      <td>63221</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>REVCELNPDCDELADHIGFQEAYR</td>\n",
       "      <td>24</td>\n",
       "      <td>[GO:0005576, GO:0005509, GO:0005179, GO:004684...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[P84351]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>Q5TGZ0-2</td>\n",
       "      <td>Isoform 2 of MICOS complex subunit MIC10</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>MSESELGRKWDRCLADAVVKIESW</td>\n",
       "      <td>24</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>Q68G75-2</td>\n",
       "      <td>Isoform 2 of LEM domain-containing protein 1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>MVDVKCLSDCKLQNQLEKLGFSPGPILLA</td>\n",
       "      <td>29</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>Q6Q0C1-2</td>\n",
       "      <td>Isoform 2 of Solute carrier family 25 member 47</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>MDFVAGAIGGVCGVAVGYPLDTVKGLLALP</td>\n",
       "      <td>30</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>Q8IVG9</td>\n",
       "      <td>Humanin</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>MAPRGFSCLLLLTSEIDLPVKRRA</td>\n",
       "      <td>24</td>\n",
       "      <td>[GO:0005737, GO:0005576, GO:0005615, GO:000573...</td>\n",
       "      <td>[PF15040]</td>\n",
       "      <td>[Q8IVG9]</td>\n",
       "      <td>[1Y32, 2GD3, 5GIW, 7WVX]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>Q8IZ96-15</td>\n",
       "      <td>Isoform 16 of CKLF-like MARVEL transmembrane d...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>MDPEHAKPESSEAPSGNLKQPETAAALR</td>\n",
       "      <td>28</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>Q8N688-2</td>\n",
       "      <td>Isoform 2 of Beta-defensin 123</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>MKLLLLTLTVLLLLSQLTPDLSNPESTC</td>\n",
       "      <td>28</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>Q8N769-2</td>\n",
       "      <td>Isoform 2 of Putative uncharacterized protein ...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>MGREMKKTGTPRPFRIEDPNQQPT</td>\n",
       "      <td>24</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>Q9NRI7</td>\n",
       "      <td>Putative pancreatic polypeptide 2</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>MAAACRCLSLLLLSTCVALLL</td>\n",
       "      <td>21</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[Q9NRI7]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>Q9NZM6-4</td>\n",
       "      <td>Isoform 4 of Polycystin-2-like protein 2</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>MAEASRWHRGGYVISIFGHFCAW</td>\n",
       "      <td>23</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>S4R3P1</td>\n",
       "      <td>Humanin-like 13</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>MDTQGFSCLLLLISEIDLSVKRRI</td>\n",
       "      <td>24</td>\n",
       "      <td>[GO:0005737, GO:0005576, GO:0048019, GO:1900118]</td>\n",
       "      <td>[PF15040]</td>\n",
       "      <td>[S4R3P1]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>S4R3Y5</td>\n",
       "      <td>Humanin-like 11</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>[Eukaryota, Metazoa, Chordata, Craniata, Verte...</td>\n",
       "      <td>MATRGFSCLLLVISEIDLSVKRWV</td>\n",
       "      <td>24</td>\n",
       "      <td>[GO:0005737, GO:0005576, GO:0048019, GO:1900118]</td>\n",
       "      <td>[PF15040]</td>\n",
       "      <td>[S4R3Y5]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                query   accession  \\\n",
       "0   organism_name:homo sapiens (human) AND length:...  A0A075B6S0   \n",
       "1   organism_name:homo sapiens (human) AND length:...  A0A075B6Y3   \n",
       "2   organism_name:homo sapiens (human) AND length:...  A0A075B6Y9   \n",
       "3   organism_name:homo sapiens (human) AND length:...  A0A075B700   \n",
       "4   organism_name:homo sapiens (human) AND length:...  A0A075B706   \n",
       "5   organism_name:homo sapiens (human) AND length:...  A0A0A0MT70   \n",
       "6   organism_name:homo sapiens (human) AND length:...  A0A0A0MT78   \n",
       "7   organism_name:homo sapiens (human) AND length:...  A0A0A0MT87   \n",
       "8   organism_name:homo sapiens (human) AND length:...  A0A0A0MT94   \n",
       "9   organism_name:homo sapiens (human) AND length:...  A0A0A0MTA4   \n",
       "10  organism_name:homo sapiens (human) AND length:...  A0A0A0MTA7   \n",
       "11  organism_name:homo sapiens (human) AND length:...  A0A0B4J200   \n",
       "12  organism_name:homo sapiens (human) AND length:...  A0A0C4DH62   \n",
       "13  organism_name:homo sapiens (human) AND length:...  A0A0C5B5G6   \n",
       "14  organism_name:homo sapiens (human) AND length:...  A0A0J9YWP8   \n",
       "15  organism_name:homo sapiens (human) AND length:...  A0A0J9YWX3   \n",
       "16  organism_name:homo sapiens (human) AND length:...  A0A0J9YX06   \n",
       "17  organism_name:homo sapiens (human) AND length:...  A0A0J9YXA8   \n",
       "18  organism_name:homo sapiens (human) AND length:...  A0A0J9YXG5   \n",
       "19  organism_name:homo sapiens (human) AND length:...  A0A0J9YXM7   \n",
       "20  organism_name:homo sapiens (human) AND length:...  A0A1B0GWH6   \n",
       "21  organism_name:homo sapiens (human) AND length:...      C0HM02   \n",
       "22  organism_name:homo sapiens (human) AND length:...      C0HMA1   \n",
       "23  organism_name:homo sapiens (human) AND length:...      D0EPY3   \n",
       "24  organism_name:homo sapiens (human) AND length:...      P0CJ68   \n",
       "25  organism_name:homo sapiens (human) AND length:...      P0CJ69   \n",
       "26  organism_name:homo sapiens (human) AND length:...      P0CJ70   \n",
       "27  organism_name:homo sapiens (human) AND length:...      P0CJ71   \n",
       "28  organism_name:homo sapiens (human) AND length:...      P0CJ72   \n",
       "29  organism_name:homo sapiens (human) AND length:...      P0CJ73   \n",
       "30  organism_name:homo sapiens (human) AND length:...      P0CJ74   \n",
       "31  organism_name:homo sapiens (human) AND length:...      P0CJ75   \n",
       "32  organism_name:homo sapiens (human) AND length:...      P0CJ76   \n",
       "33  organism_name:homo sapiens (human) AND length:...      P0CJ77   \n",
       "34  organism_name:homo sapiens (human) AND length:...      P0DMP1   \n",
       "35  organism_name:homo sapiens (human) AND length:...    P16070-2   \n",
       "36  organism_name:homo sapiens (human) AND length:...    P59103-2   \n",
       "37  organism_name:homo sapiens (human) AND length:...      P62945   \n",
       "38  organism_name:homo sapiens (human) AND length:...      P84351   \n",
       "39  organism_name:homo sapiens (human) AND length:...    Q5TGZ0-2   \n",
       "40  organism_name:homo sapiens (human) AND length:...    Q68G75-2   \n",
       "41  organism_name:homo sapiens (human) AND length:...    Q6Q0C1-2   \n",
       "42  organism_name:homo sapiens (human) AND length:...      Q8IVG9   \n",
       "43  organism_name:homo sapiens (human) AND length:...   Q8IZ96-15   \n",
       "44  organism_name:homo sapiens (human) AND length:...    Q8N688-2   \n",
       "45  organism_name:homo sapiens (human) AND length:...    Q8N769-2   \n",
       "46  organism_name:homo sapiens (human) AND length:...      Q9NRI7   \n",
       "47  organism_name:homo sapiens (human) AND length:...    Q9NZM6-4   \n",
       "48  organism_name:homo sapiens (human) AND length:...      S4R3P1   \n",
       "49  organism_name:homo sapiens (human) AND length:...      S4R3Y5   \n",
       "\n",
       "                                         protein_name  \\\n",
       "0                     T cell receptor gamma joining 1   \n",
       "1                     T cell receptor alpha joining 3   \n",
       "2                    T cell receptor alpha joining 42   \n",
       "3                    T cell receptor alpha joining 31   \n",
       "4                     T cell receptor delta joining 1   \n",
       "5                    T cell receptor beta joining 2-6   \n",
       "6                    T cell receptor beta joining 2-7   \n",
       "7                    T cell receptor beta joining 2-4   \n",
       "8                    T cell receptor beta joining 2-2   \n",
       "9                    T cell receptor beta joining 2-5   \n",
       "10                   T cell receptor beta joining 2-1   \n",
       "11                   T cell receptor beta joining 2-3   \n",
       "12                     Immunoglobulin heavy joining 1   \n",
       "13               Mitochondrial-derived peptide MOTS-c   \n",
       "14                   T cell receptor beta joining 1-3   \n",
       "15                   T cell receptor beta joining 1-6   \n",
       "16                   T cell receptor beta joining 1-2   \n",
       "17                   T cell receptor beta joining 1-1   \n",
       "18                   T cell receptor beta joining 1-4   \n",
       "19                   T cell receptor beta joining 1-5   \n",
       "20                             LASP1 neighbor protein   \n",
       "21                PRKCH upstream open reading frame 2   \n",
       "22            HSPA8-interacting micropeptide miPEP155   \n",
       "23                 Small integral membrane protein 47   \n",
       "24                                     Humanin-like 1   \n",
       "25                                     Humanin-like 2   \n",
       "26                                     Humanin-like 3   \n",
       "27                                     Humanin-like 4   \n",
       "28                                     Humanin-like 5   \n",
       "29                                     Humanin-like 6   \n",
       "30                                     Humanin-like 7   \n",
       "31                                     Humanin-like 8   \n",
       "32                                     Humanin-like 9   \n",
       "33                                    Humanin-like 10   \n",
       "34                                    Humanin-like 12   \n",
       "35                          Isoform 2 of CD44 antigen   \n",
       "36        Isoform 2 of D-amino acid oxidase regulator   \n",
       "37               Small ribosomal subunit protein eS32   \n",
       "38                                        Osteocalcin   \n",
       "39           Isoform 2 of MICOS complex subunit MIC10   \n",
       "40       Isoform 2 of LEM domain-containing protein 1   \n",
       "41    Isoform 2 of Solute carrier family 25 member 47   \n",
       "42                                            Humanin   \n",
       "43  Isoform 16 of CKLF-like MARVEL transmembrane d...   \n",
       "44                     Isoform 2 of Beta-defensin 123   \n",
       "45  Isoform 2 of Putative uncharacterized protein ...   \n",
       "46                  Putative pancreatic polypeptide 2   \n",
       "47           Isoform 4 of Polycystin-2-like protein 2   \n",
       "48                                    Humanin-like 13   \n",
       "49                                    Humanin-like 11   \n",
       "\n",
       "                    organism_name  taxon_id  \\\n",
       "0                    Homo sapiens      9606   \n",
       "1                    Homo sapiens      9606   \n",
       "2                    Homo sapiens      9606   \n",
       "3                    Homo sapiens      9606   \n",
       "4                    Homo sapiens      9606   \n",
       "5                    Homo sapiens      9606   \n",
       "6                    Homo sapiens      9606   \n",
       "7                    Homo sapiens      9606   \n",
       "8                    Homo sapiens      9606   \n",
       "9                    Homo sapiens      9606   \n",
       "10                   Homo sapiens      9606   \n",
       "11                   Homo sapiens      9606   \n",
       "12                   Homo sapiens      9606   \n",
       "13                   Homo sapiens      9606   \n",
       "14                   Homo sapiens      9606   \n",
       "15                   Homo sapiens      9606   \n",
       "16                   Homo sapiens      9606   \n",
       "17                   Homo sapiens      9606   \n",
       "18                   Homo sapiens      9606   \n",
       "19                   Homo sapiens      9606   \n",
       "20                   Homo sapiens      9606   \n",
       "21                   Homo sapiens      9606   \n",
       "22                   Homo sapiens      9606   \n",
       "23                   Homo sapiens      9606   \n",
       "24                   Homo sapiens      9606   \n",
       "25                   Homo sapiens      9606   \n",
       "26                   Homo sapiens      9606   \n",
       "27                   Homo sapiens      9606   \n",
       "28                   Homo sapiens      9606   \n",
       "29                   Homo sapiens      9606   \n",
       "30                   Homo sapiens      9606   \n",
       "31                   Homo sapiens      9606   \n",
       "32                   Homo sapiens      9606   \n",
       "33                   Homo sapiens      9606   \n",
       "34                   Homo sapiens      9606   \n",
       "35                   Homo sapiens      9606   \n",
       "36                   Homo sapiens      9606   \n",
       "37                   Homo sapiens      9606   \n",
       "38  Homo sapiens neanderthalensis     63221   \n",
       "39                   Homo sapiens      9606   \n",
       "40                   Homo sapiens      9606   \n",
       "41                   Homo sapiens      9606   \n",
       "42                   Homo sapiens      9606   \n",
       "43                   Homo sapiens      9606   \n",
       "44                   Homo sapiens      9606   \n",
       "45                   Homo sapiens      9606   \n",
       "46                   Homo sapiens      9606   \n",
       "47                   Homo sapiens      9606   \n",
       "48                   Homo sapiens      9606   \n",
       "49                   Homo sapiens      9606   \n",
       "\n",
       "                                               ineage  \\\n",
       "0   [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "1   [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "2   [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "3   [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "4   [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "5   [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "6   [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "7   [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "8   [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "9   [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "10  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "11  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "12  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "13  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "14  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "15  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "16  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "17  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "18  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "19  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "20  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "21  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "22  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "23  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "24  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "25  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "26  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "27  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "28  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "29  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "30  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "31  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "32  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "33  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "34  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "35  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "36  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "37  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "38  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "39  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "40  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "41  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "42  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "43  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "44  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "45  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "46  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "47  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "48  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "49  [Eukaryota, Metazoa, Chordata, Craniata, Verte...   \n",
       "\n",
       "                          sequence  length  \\\n",
       "0                 NYYKKLFGSGTTLVVT      16   \n",
       "1             GYSSASKIIFGSGTRLSIRP      20   \n",
       "2             YGGSQGNLIFGKGTKLSVKP      20   \n",
       "3               NNNARLMFGDGTQLVVKP      18   \n",
       "4                 TDKLIFGKGTRVTVEP      16   \n",
       "5                SGANVLTFGAGSRLTVL      17   \n",
       "6                  SYEQYFGPGTRLTVT      15   \n",
       "7                 AKNIQYFGAGTRLSVL      16   \n",
       "8                 NTGELFFGEGSRLTVL      16   \n",
       "9                  QETQYFGPGTRLLVL      15   \n",
       "10                SYNEQFFGPGTRLTVL      16   \n",
       "11                STDTQYFGPGTRLTVL      16   \n",
       "12               AEYFQHWGQGTLVTVSS      17   \n",
       "13                MRWQEMGYIFYPRKLR      16   \n",
       "14                SGNTIYFGEGSWLTVV      16   \n",
       "15               SYNSPLHFGNGTRLTVT      17   \n",
       "16                 NYGYTFGSGTRLTVV      15   \n",
       "17                 NTEAFFGQGTRLTVV      15   \n",
       "18                TNEKLFFGSGTQLSVL      16   \n",
       "19                SNQPQHFGDGTRLSIL      16   \n",
       "20       MLDIFILMFFAIIGLVILSYIIYLL      25   \n",
       "21      MASRGALRRCLSPGLPRLLHLSRGLA      26   \n",
       "22               MEMALMVAQTRKGKSVV      17   \n",
       "23    MNFQENVTLAMALFTILTSIYFFNKAQQ      28   \n",
       "24        MAPRGFSCLLLSTSEIDLPVKRRT      24   \n",
       "25    MAPRGFSCLLLSTSEIDLPVKRLLSSVF      28   \n",
       "26        MATRRFSCLLLSTSEIDLSVKRRI      24   \n",
       "27    MATQGFSCLLLSVSEIDLSMKRQYKQIR      28   \n",
       "28        MATPGFSCLLLSTSEIDLPMKRRV      24   \n",
       "29        MTPRGFSCLLLPTSETDLPVKRRT      24   \n",
       "30        MATGGFGCLLLLIREIDLSVKRQI      24   \n",
       "31        MAPRGFSCLLLSTSEIDLPVKRRA      24   \n",
       "32        MARRGFSCLLLSTTATDLPVKRRT      24   \n",
       "33        MTTRGFSCLLLLIREIDLSAKRRI      24   \n",
       "34        MAPRGFSCLLLSTSEIDLPVKRRA      24   \n",
       "35   MDKFWWHAAWGLCLVPLSLAQIGVGRRKS      29   \n",
       "36                MLEKLMGADSLQLFRV      16   \n",
       "37       MRAKWRKKRMRRLKRKRRKMRQRSK      25   \n",
       "38        REVCELNPDCDELADHIGFQEAYR      24   \n",
       "39        MSESELGRKWDRCLADAVVKIESW      24   \n",
       "40   MVDVKCLSDCKLQNQLEKLGFSPGPILLA      29   \n",
       "41  MDFVAGAIGGVCGVAVGYPLDTVKGLLALP      30   \n",
       "42        MAPRGFSCLLLLTSEIDLPVKRRA      24   \n",
       "43    MDPEHAKPESSEAPSGNLKQPETAAALR      28   \n",
       "44    MKLLLLTLTVLLLLSQLTPDLSNPESTC      28   \n",
       "45        MGREMKKTGTPRPFRIEDPNQQPT      24   \n",
       "46           MAAACRCLSLLLLSTCVALLL      21   \n",
       "47         MAEASRWHRGGYVISIFGHFCAW      23   \n",
       "48        MDTQGFSCLLLLISEIDLSVKRRI      24   \n",
       "49        MATRGFSCLLLVISEIDLSVKRWV      24   \n",
       "\n",
       "                                             go_terms   pfam_ids  \\\n",
       "0                            [GO:0042101, GO:0002250]         []   \n",
       "1                            [GO:0042101, GO:0002250]         []   \n",
       "2                            [GO:0005886, GO:0002250]         []   \n",
       "3                            [GO:0005886, GO:0002250]         []   \n",
       "4                            [GO:0042101, GO:0002250]         []   \n",
       "5                            [GO:0042101, GO:0002250]         []   \n",
       "6                            [GO:0042101, GO:0002250]         []   \n",
       "7                            [GO:0042101, GO:0002250]         []   \n",
       "8                            [GO:0042101, GO:0002250]         []   \n",
       "9                            [GO:0042101, GO:0002250]         []   \n",
       "10                           [GO:0042101, GO:0002250]         []   \n",
       "11                           [GO:0042101, GO:0002250]         []   \n",
       "12   [GO:0005576, GO:0019814, GO:0005886, GO:0002250]         []   \n",
       "13  [GO:0005615, GO:0005739, GO:0005634, GO:000367...  [PF21945]   \n",
       "14                           [GO:0042101, GO:0002250]         []   \n",
       "15                           [GO:0042101, GO:0002250]         []   \n",
       "16                           [GO:0042101, GO:0002250]         []   \n",
       "17                           [GO:0042101, GO:0002250]         []   \n",
       "18                           [GO:0042101, GO:0002250]         []   \n",
       "19                           [GO:0042101, GO:0002250]         []   \n",
       "20                                       [GO:0016020]         []   \n",
       "21  [GO:0120283, GO:0030291, GO:0045182, GO:003419...  [PF21952]   \n",
       "22  [GO:0005737, GO:0005634, GO:0030544, GO:000237...         []   \n",
       "23                                       [GO:0016020]         []   \n",
       "24   [GO:0005737, GO:0005576, GO:0048019, GO:1900118]  [PF15040]   \n",
       "25   [GO:0005737, GO:0005576, GO:0048019, GO:1900118]  [PF15040]   \n",
       "26   [GO:0005737, GO:0005576, GO:0048019, GO:1900118]  [PF15040]   \n",
       "27   [GO:0005737, GO:0005576, GO:0048019, GO:1900118]  [PF15040]   \n",
       "28  [GO:0005576, GO:0005739, GO:0048019, GO:007273...  [PF15040]   \n",
       "29   [GO:0005737, GO:0005576, GO:0048019, GO:1900118]  [PF15040]   \n",
       "30   [GO:0005737, GO:0005576, GO:0048019, GO:1900118]  [PF15040]   \n",
       "31   [GO:0005737, GO:0005576, GO:0048019, GO:1900118]  [PF15040]   \n",
       "32   [GO:0005737, GO:0005576, GO:0048019, GO:1900118]  [PF15040]   \n",
       "33   [GO:0005737, GO:0005576, GO:0048019, GO:1900118]  [PF15040]   \n",
       "34                           [GO:0005737, GO:0005576]  [PF15040]   \n",
       "35                                               None       None   \n",
       "36                                               None       None   \n",
       "37  [GO:0005737, GO:0005829, GO:0022625, GO:000578...  [PF05162]   \n",
       "38  [GO:0005576, GO:0005509, GO:0005179, GO:004684...         []   \n",
       "39                                               None       None   \n",
       "40                                               None       None   \n",
       "41                                               None       None   \n",
       "42  [GO:0005737, GO:0005576, GO:0005615, GO:000573...  [PF15040]   \n",
       "43                                               None       None   \n",
       "44                                               None       None   \n",
       "45                                               None       None   \n",
       "46                                                 []         []   \n",
       "47                                               None       None   \n",
       "48   [GO:0005737, GO:0005576, GO:0048019, GO:1900118]  [PF15040]   \n",
       "49   [GO:0005737, GO:0005576, GO:0048019, GO:1900118]  [PF15040]   \n",
       "\n",
       "   alphafold_ids                                            pdb_ids  \n",
       "0   [A0A075B6S0]                                                 []  \n",
       "1   [A0A075B6Y3]                                                 []  \n",
       "2   [A0A075B6Y9]                                                 []  \n",
       "3   [A0A075B700]                                                 []  \n",
       "4   [A0A075B706]                                                 []  \n",
       "5   [A0A0A0MT70]                                                 []  \n",
       "6             []                                                 []  \n",
       "7   [A0A0A0MT87]                                                 []  \n",
       "8   [A0A0A0MT94]                                                 []  \n",
       "9             []                                                 []  \n",
       "10  [A0A0A0MTA7]                                                 []  \n",
       "11  [A0A0B4J200]                                             [7NDT]  \n",
       "12  [A0A0C4DH62]                                                 []  \n",
       "13  [A0A0C5B5G6]                                                 []  \n",
       "14  [A0A0J9YWP8]                                                 []  \n",
       "15  [A0A0J9YWX3]                                                 []  \n",
       "16            []                                       [7NDQ, 7NDU]  \n",
       "17            []                                                 []  \n",
       "18  [A0A0J9YXG5]                                                 []  \n",
       "19  [A0A0J9YXM7]                                                 []  \n",
       "20  [A0A1B0GWH6]                                                 []  \n",
       "21            []                                                 []  \n",
       "22            []                                                 []  \n",
       "23      [D0EPY3]                                                 []  \n",
       "24      [P0CJ68]                                                 []  \n",
       "25      [P0CJ69]                                                 []  \n",
       "26      [P0CJ70]                                                 []  \n",
       "27      [P0CJ71]                                                 []  \n",
       "28      [P0CJ72]                                                 []  \n",
       "29      [P0CJ73]                                                 []  \n",
       "30      [P0CJ74]                                                 []  \n",
       "31      [P0CJ75]                                                 []  \n",
       "32      [P0CJ76]                                                 []  \n",
       "33      [P0CJ77]                                                 []  \n",
       "34      [P0DMP1]                                                 []  \n",
       "35          None                                               None  \n",
       "36          None                                               None  \n",
       "37      [P62945]  [4UG0, 4V6X, 5A2Q, 5AJ0, 5LKS, 5OA3, 5VYC, 6G5...  \n",
       "38      [P84351]                                                 []  \n",
       "39          None                                               None  \n",
       "40          None                                               None  \n",
       "41          None                                               None  \n",
       "42      [Q8IVG9]                           [1Y32, 2GD3, 5GIW, 7WVX]  \n",
       "43          None                                               None  \n",
       "44          None                                               None  \n",
       "45          None                                               None  \n",
       "46      [Q9NRI7]                                                 []  \n",
       "47          None                                               None  \n",
       "48      [S4R3P1]                                                 []  \n",
       "49      [S4R3Y5]                                                 []  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "uniprot.parse_stream_response(\n",
    "    query=query,\n",
    "    response=response\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "301959e5",
   "metadata": {},
   "source": [
    "## Activity search"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "06c29588",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "59f9589e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['uniprot_celiac-toxic.csv',\n",
       " 'uniprot_embryotoxic.csv',\n",
       " 'uniprot_ace-inhibitor.csv',\n",
       " 'uniprot_anuran-defense.csv',\n",
       " 'uniprot_campde-inhibitor.csv',\n",
       " 'uniprot_anti-neurotensive.csv',\n",
       " 'uniprot_antitrypanosomic.csv',\n",
       " 'uniprot_anticancer.csv',\n",
       " 'uniprot_anorectic.csv',\n",
       " 'uniprot_chemotactic.csv',\n",
       " 'uniprot_targeting-GP.csv',\n",
       " 'uniprot_Blood-Brain-Barrier.csv',\n",
       " 'uniprot_antitumor.csv',\n",
       " 'uniprot_cytotoxic.csv',\n",
       " 'uniprot_activating-ubiquitin-mediated-proteolysis.csv',\n",
       " 'uniprot_protein-kinase-c-inhibitor.csv',\n",
       " 'uniprot_antihiv.csv',\n",
       " 'uniprot_antidiabetic.csv',\n",
       " 'uniprot_calpain-2-inhibitor.csv',\n",
       " 'uniprot_antileishmania.csv',\n",
       " 'uniprot_inhibitor.csv',\n",
       " 'uniprot_antimicrobial.csv',\n",
       " 'uniprot_antituberculosis.csv',\n",
       " 'uniprot_antiviral.csv',\n",
       " 'uniprot_wound-healing.csv',\n",
       " 'uniprot_targeting-GN.csv',\n",
       " 'uniprot_hmg-coa-reductase-inhibitor.csv',\n",
       " 'uniprot_opioid-agonist.csv',\n",
       " 'uniprot_toxicology.csv',\n",
       " 'uniprot_antibacterial.csv',\n",
       " 'uniprot_antibiofilm.csv',\n",
       " 'uniprot_hypocholesterolemic.csv',\n",
       " 'uniprot_bacterial-permease-ligand.csv',\n",
       " 'uniprot_protease-inhibitor.csv',\n",
       " 'uniprot_antitoxin.csv',\n",
       " 'uniprot_polystyrene-surface-binding.csv',\n",
       " 'uniprot_stimulating.csv',\n",
       " 'uniprot_immunomodulating.csv',\n",
       " 'uniprot_opioid.csv',\n",
       " 'uniprot_dipeptidyl-peptidase-iii-inhibitor.csv',\n",
       " 'uniprot_tumor-homing.csv',\n",
       " 'uniprot_heparin-binding.csv',\n",
       " 'uniprot_anticoronavirus.csv',\n",
       " 'uniprot_targeting-mammals.csv',\n",
       " 'uniprot_antiplasmodial.csv',\n",
       " 'uniprot_uterotonic.csv',\n",
       " 'uniprot_cell-cell-comunication.csv',\n",
       " 'uniprot_binding.csv',\n",
       " 'uniprot_peptide-hormones.csv',\n",
       " 'uniprot_drug-delivery.csv',\n",
       " 'uniprot_anti-MRSA.csv',\n",
       " 'uniprot_antihypertensive.csv',\n",
       " 'uniprot_renin-inhibitor.csv',\n",
       " 'uniprot_neuroactive.csv',\n",
       " 'uniprot_surface-binding.csv',\n",
       " 'uniprot_antiantifreeze.csv',\n",
       " 'uniprot_insecticidal.csv',\n",
       " 'uniprot_antiprotozoal.csv',\n",
       " 'uniprot_alpha-amylase-inhibitor.csv',\n",
       " 'uniprot_antiparasitic.csv',\n",
       " 'uniprot_alpha-glucosidase-inhibitor.csv',\n",
       " 'uniprot_neuropeptide.csv',\n",
       " 'uniprot_trypsin-inhibitory.csv',\n",
       " 'uniprot_orphan-receptor-gpr14-agonist.csv',\n",
       " 'uniprot_antithrombotic.csv',\n",
       " 'uniprot_antiamnestic.csv',\n",
       " 'uniprot_antifungal.csv',\n",
       " 'uniprot_antiangiogenic.csv',\n",
       " 'uniprot_antioxidative.csv',\n",
       " 'uniprot_natriuretic.csv',\n",
       " 'uniprot_cell-penetrating.csv',\n",
       " 'uniprot_quorum-sensing.csv',\n",
       " 'uniprot_dipeptidyl-peptidaseiv.csv',\n",
       " 'uniprot_regulating.csv',\n",
       " 'uniprot_antioncogenic.csv',\n",
       " 'uniprot_contracting.csv',\n",
       " 'uniprot_immunostimulating.csv',\n",
       " 'uniprot_membrane-active-peptide.csv',\n",
       " 'uniprot_hypolipidemic.csv',\n",
       " 'uniprot_antiinflammatory.csv',\n",
       " 'uniprot_vasoconstrictor.csv',\n",
       " 'uniprot_opioid-antagonist.csv',\n",
       " 'uniprot_targeting-GV.csv',\n",
       " 'uniprot_hypotensive.csv',\n",
       " 'uniprot_edema-inducer.csv',\n",
       " 'uniprot_antiendotoxin.csv',\n",
       " 'uniprot_hemolytic.csv',\n",
       " 'uniprot_camkii-inhibitor.csv',\n",
       " 'uniprot_cytokine-growth-factor.csv']"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "uniprot_search_files = os.listdir(\"uniprot_search\")\n",
    "uniprot_search_files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ecb7ce2c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>activity</th>\n",
       "      <th>sequence_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Binding</td>\n",
       "      <td>411281</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Inhibitor</td>\n",
       "      <td>18950</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Surface binding</td>\n",
       "      <td>15663</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Regulating</td>\n",
       "      <td>11361</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Antimicrobial</td>\n",
       "      <td>5945</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>Dipeptidyl peptidaseiv</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>Antiamnestic</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>Edema inducer</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>Antiendotoxin</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>Total</td>\n",
       "      <td>502774</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>90 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  activity  sequence_count\n",
       "0                  Binding          411281\n",
       "1                Inhibitor           18950\n",
       "2          Surface binding           15663\n",
       "3               Regulating           11361\n",
       "4            Antimicrobial            5945\n",
       "..                     ...             ...\n",
       "85  Dipeptidyl peptidaseiv               0\n",
       "86            Antiamnestic               0\n",
       "87           Edema inducer               0\n",
       "88           Antiendotoxin               0\n",
       "89                   Total          502774\n",
       "\n",
       "[90 rows x 2 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create an empty list to store the results\n",
    "activity_data = []\n",
    "\n",
    "# Iterate over each file in uniprot_search_files\n",
    "for file in uniprot_search_files:\n",
    "    # Load the CSV file into a DataFrame\n",
    "    df = pd.read_csv(f\"uniprot_search/{file}\")\n",
    "    \n",
    "    # Extract the activity name from the file name\n",
    "    activity = file.replace(\"uniprot_\", \"\").replace(\".csv\", \"\").replace(\"-\", \" \").capitalize()\n",
    "    \n",
    "    # Count the number of sequences in the DataFrame\n",
    "    sequence_count = len(df)\n",
    "    \n",
    "    # Append the activity and sequence count to the list\n",
    "    activity_data.append({\"activity\": activity, \"sequence_count\": sequence_count})\n",
    "\n",
    "# Create a DataFrame from the activity data\n",
    "activity_df = pd.DataFrame(activity_data)\n",
    "\n",
    "# Sort the DataFrame by sequence count in descending order\n",
    "activity_df = activity_df.sort_values(by=\"sequence_count\", ascending=False)\n",
    "\n",
    "# Calculate the total number of sequences\n",
    "total_sequences = activity_df[\"sequence_count\"].sum()\n",
    "\n",
    "# Add a row for the total sequences\n",
    "activity_df = pd.concat([activity_df, pd.DataFrame([{\"activity\": \"Total\", \"sequence_count\": total_sequences}])], ignore_index=True)\n",
    "\n",
    "# Display the resulting DataFrame\n",
    "activity_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "4abc5ec2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>activity</th>\n",
       "      <th>sequence_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Antimicrobial</td>\n",
       "      <td>5945</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Antiviral</td>\n",
       "      <td>3835</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Antibacterial</td>\n",
       "      <td>2366</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Antifungal</td>\n",
       "      <td>1209</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Antitoxin</td>\n",
       "      <td>832</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Antitumor</td>\n",
       "      <td>479</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Anticancer</td>\n",
       "      <td>393</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>Antiparasitic</td>\n",
       "      <td>140</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>Antiangiogenic</td>\n",
       "      <td>86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>Antioxidative</td>\n",
       "      <td>55</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>Antithrombotic</td>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>Antidiabetic</td>\n",
       "      <td>42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>Antiprotozoal</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>Anti mrsa</td>\n",
       "      <td>31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>Antiinflammatory</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>Antihypertensive</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>Antituberculosis</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>Antibiofilm</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>Antiplasmodial</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>Antioncogenic</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>Antileishmania</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>Antihiv</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>Antitrypanosomic</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>Anti neurotensive</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>Antiantifreeze</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>Anticoronavirus</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>Antiamnestic</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>Antiendotoxin</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             activity  sequence_count\n",
       "4       Antimicrobial            5945\n",
       "6           Antiviral            3835\n",
       "10      Antibacterial            2366\n",
       "17         Antifungal            1209\n",
       "19          Antitoxin             832\n",
       "24          Antitumor             479\n",
       "25         Anticancer             393\n",
       "32      Antiparasitic             140\n",
       "37     Antiangiogenic              86\n",
       "41      Antioxidative              55\n",
       "42     Antithrombotic              50\n",
       "45       Antidiabetic              42\n",
       "48      Antiprotozoal              33\n",
       "49          Anti mrsa              31\n",
       "52   Antiinflammatory              27\n",
       "56   Antihypertensive              21\n",
       "60   Antituberculosis              15\n",
       "66        Antibiofilm               9\n",
       "67     Antiplasmodial               8\n",
       "71      Antioncogenic               2\n",
       "73     Antileishmania               0\n",
       "74            Antihiv               0\n",
       "79   Antitrypanosomic               0\n",
       "80  Anti neurotensive               0\n",
       "81     Antiantifreeze               0\n",
       "83    Anticoronavirus               0\n",
       "86       Antiamnestic               0\n",
       "88      Antiendotoxin               0"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "anti_activities_df = activity_df[activity_df['activity'].str.startswith('Anti', na=False)]\n",
    "anti_activities_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "76fc2d02",
   "metadata": {},
   "source": [
    "### ALphafold"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "4d233905",
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests, re, os, time, threading\n",
    "from requests.adapters import HTTPAdapter, Retry\n",
    "import pandas as pd\n",
    "from concurrent.futures import ThreadPoolExecutor, as_completed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "79892fa3",
   "metadata": {},
   "outputs": [],
   "source": [
    "API_URL = \"https://alphafold.ebi.ac.uk/api/prediction/\"\n",
    "POLLING_INTERVAL = 3\n",
    "#SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))\n",
    "CACHE_DIR = os.path.join(\"src\", \".alphafold_cache\")\n",
    "\n",
    "class AlphafoldInterface():\n",
    "    def __init__(self, total_retries=5):\n",
    "        self.retries = Retry(total=total_retries, backoff_factor=0.25, status_forcelist=[ 500, 502, 503, 504 ])\n",
    "        self.session = requests.Session()\n",
    "        self.session.mount('https://', HTTPAdapter(max_retries=self.retries))\n",
    "        self.session.headers.update({\"Content-Type\": \"application/json\"})\n",
    "        \n",
    "\n",
    "    def get_prediction(self, uniprot_id) -> dict:\n",
    "        \"\"\"\n",
    "        Get prediction for a given UniProt ID.\n",
    "        Args:\n",
    "            uniprot_id (str): UniProt ID to fetch prediction for.\n",
    "        Returns:\n",
    "            dict: Prediction data.\n",
    "        \"\"\"\n",
    "        url = f\"{API_URL}{uniprot_id}\"\n",
    "        \n",
    "        # Try and retry if necesary\n",
    "        try:\n",
    "            response = self.session.get(url)\n",
    "            response.raise_for_status()\n",
    "        except requests.exceptions.RequestException as e:\n",
    "            print(f\"Error fetching prediction for {uniprot_id}: {e}\")\n",
    "            return None\n",
    "        return response.json()\n",
    "    \n",
    "    def _camel_to_snake(self, name: str) -> str:\n",
    "        \"\"\"\n",
    "        Convert camelCase to snake_case.\n",
    "        Args:\n",
    "            name (str): String in camelCase.\n",
    "        Returns:\n",
    "            str: String in snake_case.\n",
    "        \"\"\"\n",
    "        return re.sub(r'(?<!^)(?=[A-Z])', '_', name).lower()\n",
    "    \n",
    "    def parse_prediction(self, prediction: dict, download_structure: bool = False) -> dict:\n",
    "        \"\"\"\n",
    "        Parse the prediction data.\n",
    "        Args:\n",
    "            prediction (dict): Prediction data.\n",
    "            download_structure (bool): Whether to download the structure.\n",
    "        Returns:\n",
    "            dict: Parsed prediction data.\n",
    "        \"\"\"\n",
    "        # Change keys from camelCase to snake_case\n",
    "        parsed_prediction = {}\n",
    "        for key, value in prediction.items():\n",
    "            if isinstance(value, dict):\n",
    "                parsed_prediction[key] = {self._camel_to_snake(k): v for k, v in value.items()}\n",
    "            else:\n",
    "                parsed_prediction[self._camel_to_snake(key)] = value\n",
    "        \n",
    "        # Download structure if requested\n",
    "        if download_structure:\n",
    "            # Get keys ending with \"_url\"\n",
    "            structure_url = [k for k in parsed_prediction.keys() if k.endswith(\"_url\") and parsed_prediction[k]]\n",
    "            if structure_url:\n",
    "                # Download the structure file\n",
    "                for url in structure_url:\n",
    "                    file_name = parsed_prediction[url].split(\"/\")[-1]\n",
    "                    # Check if file already exists\n",
    "                    if os.path.exists(CACHE_DIR + \"/\" + file_name):\n",
    "                        continue\n",
    "                    else:\n",
    "                        response = self.session.get(parsed_prediction[url])\n",
    "                        with open(CACHE_DIR + \"/\" + file_name, \"wb\") as f:\n",
    "                            f.write(response.content)\n",
    "        return parsed_prediction\n",
    "\n",
    "    def get_and_parse_predictions(self, uniprot_id: str, download_structure: bool = False) -> list:\n",
    "        \"\"\"\n",
    "        Get and parse predictions for a given UniProt ID.\n",
    "        Args:\n",
    "            uniprot_id (str): UniProt ID to fetch predictions for.\n",
    "            download_structure (bool): Whether to download structure files.\n",
    "        Returns:\n",
    "            list: List of parsed predictions.\n",
    "        \"\"\"\n",
    "        print(f\"Fetching prediction for {uniprot_id}\")\n",
    "        prediction = self.get_prediction(uniprot_id)\n",
    "        time.sleep(POLLING_INTERVAL)  # control spacing between calls\n",
    "\n",
    "        if not prediction:\n",
    "            return []\n",
    "        \n",
    "        return [self.parse_prediction(p, download_structure) for p in prediction]\n",
    "    \n",
    "    def get_predictions_for_ids(self, ids: list, download_structure: bool = False, max_workers: int = 5) -> pd.DataFrame:\n",
    "        \"\"\"\n",
    "        Get predictions in parallel for a list of UniProt IDs.\n",
    "        Args:\n",
    "            ids (list): List of UniProt IDs to fetch predictions for.\n",
    "            download_structure (bool): Whether to download structure files.\n",
    "            max_workers (int): Maximum number of parallel requests.\n",
    "        Returns:\n",
    "            pd.DataFrame: DataFrame containing parsed predictions.\n",
    "        \"\"\"\n",
    "        if not os.path.exists(CACHE_DIR):\n",
    "            os.makedirs(CACHE_DIR)\n",
    "\n",
    "        predictions = []\n",
    "        lock = threading.Lock()\n",
    "\n",
    "        with ThreadPoolExecutor(max_workers=max_workers) as executor:\n",
    "            futures = [executor.submit(self.get_and_parse_predictions, uid, download_structure) for uid in ids]\n",
    "            for future in as_completed(futures):\n",
    "                result = future.result()\n",
    "                if result:\n",
    "                    with lock:\n",
    "                        predictions.extend(result)\n",
    "\n",
    "        return pd.DataFrame(predictions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "37b81082",
   "metadata": {},
   "outputs": [],
   "source": [
    "instance = AlphafoldInterface()\n",
    "uniprot_ids = [\"P02666\", \"Q9TSI0\", \"P33048\", \"P11839\", \"O15552\", \"P76011\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cee11808",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[START] Fetching prediction for P02666\n",
      "[START] Fetching prediction for Q9TSI0\n",
      "[START] Fetching prediction for P33048\n",
      "[START] Fetching prediction for P11839\n",
      "[START] Fetching prediction for O15552\n",
      "[START] Fetching prediction for P76011\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>entry_id</th>\n",
       "      <th>gene</th>\n",
       "      <th>sequence_checksum</th>\n",
       "      <th>sequence_version_date</th>\n",
       "      <th>uniprot_accession</th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>uniprot_description</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>organism_scientific_name</th>\n",
       "      <th>uniprot_start</th>\n",
       "      <th>...</th>\n",
       "      <th>bcif_url</th>\n",
       "      <th>cif_url</th>\n",
       "      <th>pdb_url</th>\n",
       "      <th>pae_image_url</th>\n",
       "      <th>pae_doc_url</th>\n",
       "      <th>am_annotations_url</th>\n",
       "      <th>am_annotations_hg19_url</th>\n",
       "      <th>am_annotations_hg38_url</th>\n",
       "      <th>is_reviewed</th>\n",
       "      <th>is_reference_proteome</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AF-Q9TSI0-F1</td>\n",
       "      <td>CSN2</td>\n",
       "      <td>14FD3687DD17C5A9</td>\n",
       "      <td>2000-05-01</td>\n",
       "      <td>Q9TSI0</td>\n",
       "      <td>CASB_BUBBU</td>\n",
       "      <td>Beta-casein</td>\n",
       "      <td>89462</td>\n",
       "      <td>Bubalus bubalis</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>https://alphafold.ebi.ac.uk/files/AF-Q9TSI0-F1...</td>\n",
       "      <td>https://alphafold.ebi.ac.uk/files/AF-Q9TSI0-F1...</td>\n",
       "      <td>https://alphafold.ebi.ac.uk/files/AF-Q9TSI0-F1...</td>\n",
       "      <td>https://alphafold.ebi.ac.uk/files/AF-Q9TSI0-F1...</td>\n",
       "      <td>https://alphafold.ebi.ac.uk/files/AF-Q9TSI0-F1...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AF-P02666-F1</td>\n",
       "      <td>CSN2</td>\n",
       "      <td>F0BBDD8148A238AE</td>\n",
       "      <td>1989-07-01</td>\n",
       "      <td>P02666</td>\n",
       "      <td>CASB_BOVIN</td>\n",
       "      <td>Beta-casein</td>\n",
       "      <td>9913</td>\n",
       "      <td>Bos taurus</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>https://alphafold.ebi.ac.uk/files/AF-P02666-F1...</td>\n",
       "      <td>https://alphafold.ebi.ac.uk/files/AF-P02666-F1...</td>\n",
       "      <td>https://alphafold.ebi.ac.uk/files/AF-P02666-F1...</td>\n",
       "      <td>https://alphafold.ebi.ac.uk/files/AF-P02666-F1...</td>\n",
       "      <td>https://alphafold.ebi.ac.uk/files/AF-P02666-F1...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AF-P33048-F1</td>\n",
       "      <td>CSN2</td>\n",
       "      <td>96AE17746A01CD05</td>\n",
       "      <td>1993-10-01</td>\n",
       "      <td>P33048</td>\n",
       "      <td>CASB_CAPHI</td>\n",
       "      <td>Beta-casein</td>\n",
       "      <td>9925</td>\n",
       "      <td>Capra hircus</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>https://alphafold.ebi.ac.uk/files/AF-P33048-F1...</td>\n",
       "      <td>https://alphafold.ebi.ac.uk/files/AF-P33048-F1...</td>\n",
       "      <td>https://alphafold.ebi.ac.uk/files/AF-P33048-F1...</td>\n",
       "      <td>https://alphafold.ebi.ac.uk/files/AF-P33048-F1...</td>\n",
       "      <td>https://alphafold.ebi.ac.uk/files/AF-P33048-F1...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AF-P11839-F1</td>\n",
       "      <td>CSN2</td>\n",
       "      <td>061B4424DCB49EB1</td>\n",
       "      <td>1995-11-01</td>\n",
       "      <td>P11839</td>\n",
       "      <td>CASB_SHEEP</td>\n",
       "      <td>Beta-casein</td>\n",
       "      <td>9940</td>\n",
       "      <td>Ovis aries</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>https://alphafold.ebi.ac.uk/files/AF-P11839-F1...</td>\n",
       "      <td>https://alphafold.ebi.ac.uk/files/AF-P11839-F1...</td>\n",
       "      <td>https://alphafold.ebi.ac.uk/files/AF-P11839-F1...</td>\n",
       "      <td>https://alphafold.ebi.ac.uk/files/AF-P11839-F1...</td>\n",
       "      <td>https://alphafold.ebi.ac.uk/files/AF-P11839-F1...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AF-O15552-F1</td>\n",
       "      <td>FFAR2</td>\n",
       "      <td>F4A8AC6AFBDF1E90</td>\n",
       "      <td>1998-01-01</td>\n",
       "      <td>O15552</td>\n",
       "      <td>FFAR2_HUMAN</td>\n",
       "      <td>Free fatty acid receptor 2</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>https://alphafold.ebi.ac.uk/files/AF-O15552-F1...</td>\n",
       "      <td>https://alphafold.ebi.ac.uk/files/AF-O15552-F1...</td>\n",
       "      <td>https://alphafold.ebi.ac.uk/files/AF-O15552-F1...</td>\n",
       "      <td>https://alphafold.ebi.ac.uk/files/AF-O15552-F1...</td>\n",
       "      <td>https://alphafold.ebi.ac.uk/files/AF-O15552-F1...</td>\n",
       "      <td>https://alphafold.ebi.ac.uk/files/AF-O15552-F1...</td>\n",
       "      <td>https://alphafold.ebi.ac.uk/files/AF-O15552-F1...</td>\n",
       "      <td>https://alphafold.ebi.ac.uk/files/AF-O15552-F1...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>AF-P76011-F1</td>\n",
       "      <td>ymgE</td>\n",
       "      <td>78ED929220264E62</td>\n",
       "      <td>2001-07-11</td>\n",
       "      <td>P76011</td>\n",
       "      <td>YMGE_ECOLI</td>\n",
       "      <td>UPF0410 protein YmgE</td>\n",
       "      <td>83333</td>\n",
       "      <td>Escherichia coli (strain K12)</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>https://alphafold.ebi.ac.uk/files/AF-P76011-F1...</td>\n",
       "      <td>https://alphafold.ebi.ac.uk/files/AF-P76011-F1...</td>\n",
       "      <td>https://alphafold.ebi.ac.uk/files/AF-P76011-F1...</td>\n",
       "      <td>https://alphafold.ebi.ac.uk/files/AF-P76011-F1...</td>\n",
       "      <td>https://alphafold.ebi.ac.uk/files/AF-P76011-F1...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       entry_id   gene sequence_checksum sequence_version_date  \\\n",
       "0  AF-Q9TSI0-F1   CSN2  14FD3687DD17C5A9            2000-05-01   \n",
       "1  AF-P02666-F1   CSN2  F0BBDD8148A238AE            1989-07-01   \n",
       "2  AF-P33048-F1   CSN2  96AE17746A01CD05            1993-10-01   \n",
       "3  AF-P11839-F1   CSN2  061B4424DCB49EB1            1995-11-01   \n",
       "4  AF-O15552-F1  FFAR2  F4A8AC6AFBDF1E90            1998-01-01   \n",
       "5  AF-P76011-F1   ymgE  78ED929220264E62            2001-07-11   \n",
       "\n",
       "  uniprot_accession   uniprot_id         uniprot_description  tax_id  \\\n",
       "0            Q9TSI0   CASB_BUBBU                 Beta-casein   89462   \n",
       "1            P02666   CASB_BOVIN                 Beta-casein    9913   \n",
       "2            P33048   CASB_CAPHI                 Beta-casein    9925   \n",
       "3            P11839   CASB_SHEEP                 Beta-casein    9940   \n",
       "4            O15552  FFAR2_HUMAN  Free fatty acid receptor 2    9606   \n",
       "5            P76011   YMGE_ECOLI        UPF0410 protein YmgE   83333   \n",
       "\n",
       "        organism_scientific_name  uniprot_start  ...  \\\n",
       "0                Bubalus bubalis              1  ...   \n",
       "1                     Bos taurus              1  ...   \n",
       "2                   Capra hircus              1  ...   \n",
       "3                     Ovis aries              1  ...   \n",
       "4                   Homo sapiens              1  ...   \n",
       "5  Escherichia coli (strain K12)              1  ...   \n",
       "\n",
       "                                            bcif_url  \\\n",
       "0  https://alphafold.ebi.ac.uk/files/AF-Q9TSI0-F1...   \n",
       "1  https://alphafold.ebi.ac.uk/files/AF-P02666-F1...   \n",
       "2  https://alphafold.ebi.ac.uk/files/AF-P33048-F1...   \n",
       "3  https://alphafold.ebi.ac.uk/files/AF-P11839-F1...   \n",
       "4  https://alphafold.ebi.ac.uk/files/AF-O15552-F1...   \n",
       "5  https://alphafold.ebi.ac.uk/files/AF-P76011-F1...   \n",
       "\n",
       "                                             cif_url  \\\n",
       "0  https://alphafold.ebi.ac.uk/files/AF-Q9TSI0-F1...   \n",
       "1  https://alphafold.ebi.ac.uk/files/AF-P02666-F1...   \n",
       "2  https://alphafold.ebi.ac.uk/files/AF-P33048-F1...   \n",
       "3  https://alphafold.ebi.ac.uk/files/AF-P11839-F1...   \n",
       "4  https://alphafold.ebi.ac.uk/files/AF-O15552-F1...   \n",
       "5  https://alphafold.ebi.ac.uk/files/AF-P76011-F1...   \n",
       "\n",
       "                                             pdb_url  \\\n",
       "0  https://alphafold.ebi.ac.uk/files/AF-Q9TSI0-F1...   \n",
       "1  https://alphafold.ebi.ac.uk/files/AF-P02666-F1...   \n",
       "2  https://alphafold.ebi.ac.uk/files/AF-P33048-F1...   \n",
       "3  https://alphafold.ebi.ac.uk/files/AF-P11839-F1...   \n",
       "4  https://alphafold.ebi.ac.uk/files/AF-O15552-F1...   \n",
       "5  https://alphafold.ebi.ac.uk/files/AF-P76011-F1...   \n",
       "\n",
       "                                       pae_image_url  \\\n",
       "0  https://alphafold.ebi.ac.uk/files/AF-Q9TSI0-F1...   \n",
       "1  https://alphafold.ebi.ac.uk/files/AF-P02666-F1...   \n",
       "2  https://alphafold.ebi.ac.uk/files/AF-P33048-F1...   \n",
       "3  https://alphafold.ebi.ac.uk/files/AF-P11839-F1...   \n",
       "4  https://alphafold.ebi.ac.uk/files/AF-O15552-F1...   \n",
       "5  https://alphafold.ebi.ac.uk/files/AF-P76011-F1...   \n",
       "\n",
       "                                         pae_doc_url  \\\n",
       "0  https://alphafold.ebi.ac.uk/files/AF-Q9TSI0-F1...   \n",
       "1  https://alphafold.ebi.ac.uk/files/AF-P02666-F1...   \n",
       "2  https://alphafold.ebi.ac.uk/files/AF-P33048-F1...   \n",
       "3  https://alphafold.ebi.ac.uk/files/AF-P11839-F1...   \n",
       "4  https://alphafold.ebi.ac.uk/files/AF-O15552-F1...   \n",
       "5  https://alphafold.ebi.ac.uk/files/AF-P76011-F1...   \n",
       "\n",
       "                                  am_annotations_url  \\\n",
       "0                                               None   \n",
       "1                                               None   \n",
       "2                                               None   \n",
       "3                                               None   \n",
       "4  https://alphafold.ebi.ac.uk/files/AF-O15552-F1...   \n",
       "5                                               None   \n",
       "\n",
       "                             am_annotations_hg19_url  \\\n",
       "0                                               None   \n",
       "1                                               None   \n",
       "2                                               None   \n",
       "3                                               None   \n",
       "4  https://alphafold.ebi.ac.uk/files/AF-O15552-F1...   \n",
       "5                                               None   \n",
       "\n",
       "                             am_annotations_hg38_url is_reviewed  \\\n",
       "0                                               None        True   \n",
       "1                                               None        True   \n",
       "2                                               None        True   \n",
       "3                                               None        True   \n",
       "4  https://alphafold.ebi.ac.uk/files/AF-O15552-F1...        True   \n",
       "5                                               None        True   \n",
       "\n",
       "  is_reference_proteome  \n",
       "0                 False  \n",
       "1                  True  \n",
       "2                  True  \n",
       "3                  True  \n",
       "4                  True  \n",
       "5                  True  \n",
       "\n",
       "[6 rows x 25 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Extraer la calidad del modelo y solo el modelo\n",
    "instance.get_predictions_for_ids_parallel(uniprot_ids, download_structure=True, max_workers=2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9837e24a",
   "metadata": {},
   "source": [
    "## PDB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b3d6aff0",
   "metadata": {},
   "outputs": [],
   "source": [
    "!pip install --upgrade rcsb-api"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ad092f1",
   "metadata": {},
   "source": [
    "Puedo usar rcbapi.search para hacer busquedas como en uniprot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "46085caf",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "from rcsbapi.data import DataQuery as Query\n",
    "import requests\n",
    "import threading\n",
    "from concurrent.futures import ThreadPoolExecutor, as_completed\n",
    "# Check https://data.rcsb.org/rest/v1/core/entry/4HHB for more attributes\n",
    "# https://pdb101.rcsb.org/train/training-events/apis-python"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "88d50ed9",
   "metadata": {},
   "outputs": [],
   "source": [
    "query = Query(\n",
    "    input_type=\"entries\",\n",
    "    input_ids=[\"4HHB\"],  # CSM IDs can be used as well\n",
    "    return_data_list=[\"exptl.method\", \"rcsb_entry_info\", \"rcsb_entry_info\"]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "0e41e54a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 1/1 [00:00<00:00,  1.07it/s]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'data': {'entries': [{'rcsb_id': '4HHB',\n",
       "    'exptl': [{'method': 'X-RAY DIFFRACTION'}],\n",
       "    'rcsb_entry_info': {'polymer_monomer_count_maximum': 146,\n",
       "     'polymer_entity_count_nucleic_acid_hybrid': 0,\n",
       "     'deposited_unmodeled_polymer_monomer_count': 0,\n",
       "     'deposited_atom_count': 4779,\n",
       "     'experimental_method': 'X-ray',\n",
       "     'structure_determination_methodology': 'experimental',\n",
       "     'nonpolymer_bound_components': ['HEM'],\n",
       "     'resolution_combined': [1.74],\n",
       "     'na_polymer_entity_types': 'Other',\n",
       "     'diffrn_radiation_wavelength_minimum': None,\n",
       "     'assembly_count': 1,\n",
       "     'polymer_composition': 'heteromeric protein',\n",
       "     'cis_peptide_count': 0,\n",
       "     'branched_entity_count': 0,\n",
       "     'disulfide_bond_count': 0,\n",
       "     'diffrn_radiation_wavelength_maximum': None,\n",
       "     'structure_determination_methodology_priority': 10,\n",
       "     'polymer_entity_count_protein': 2,\n",
       "     'nonpolymer_molecular_weight_maximum': 0.62,\n",
       "     'deposited_modeled_polymer_monomer_count': 574,\n",
       "     'deposited_hydrogen_atom_count': 0,\n",
       "     'polymer_molecular_weight_maximum': 15.89,\n",
       "     'molecular_weight': 64.74,\n",
       "     'solvent_entity_count': 1,\n",
       "     'polymer_entity_count_DNA': 0,\n",
       "     'deposited_polymer_entity_instance_count': 4,\n",
       "     'entity_count': 5,\n",
       "     'polymer_monomer_count_minimum': 141,\n",
       "     'polymer_entity_count_nucleic_acid': 0,\n",
       "     'nonpolymer_entity_count': 2,\n",
       "     'inter_mol_covalent_bond_count': 0,\n",
       "     'deposited_model_count': 1,\n",
       "     'branched_molecular_weight_minimum': None,\n",
       "     'polymer_entity_taxonomy_count': 2,\n",
       "     'selected_polymer_entity_types': 'Protein (only)',\n",
       "     'ndb_struct_conf_na_feature_combined': None,\n",
       "     'deposited_solvent_atom_count': 221,\n",
       "     'inter_mol_metalic_bond_count': 4,\n",
       "     'software_programs_combined': None,\n",
       "     'polymer_entity_count': 2,\n",
       "     'branched_molecular_weight_maximum': None,\n",
       "     'polymer_molecular_weight_minimum': 15.15,\n",
       "     'diffrn_resolution_high': {'provenance_source': 'From refinement resolution cutoff',\n",
       "      'value': 1.74},\n",
       "     'nonpolymer_molecular_weight_minimum': 0.09,\n",
       "     'polymer_entity_count_RNA': 0,\n",
       "     'experimental_method_count': 1,\n",
       "     'deposited_polymer_monomer_count': 574,\n",
       "     'deposited_nonpolymer_entity_instance_count': 6,\n",
       "     'deposited_deuterated_water_count': 0}}]}}"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results = query.exec(\n",
    "    batch_size=5000,\n",
    "    progress_bar=True\n",
    ")\n",
    "results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "455c4a2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "PDB_CACHE_DIR = os.path.join(\"src\", \".pdb_cache\")\n",
    "class PdbInterface():\n",
    "    def __init__(self):\n",
    "        self.base_url = \"https://files.rcsb.org/download/\"\n",
    "    \n",
    "    def search_ids(self, input_ids: list, return_data_list: list = None, batch_size: int = 5000) -> dict:\n",
    "        if return_data_list is None:\n",
    "            return_data_list = [\"exptl.method\", \"rcsb_entry_info\", \"rcsb_entry_info\"]\n",
    "\n",
    "        query = Query(\n",
    "            input_type=\"entries\",\n",
    "            input_ids=input_ids,\n",
    "            return_data_list=return_data_list\n",
    "        )\n",
    "\n",
    "        results = query.exec(\n",
    "            batch_size=batch_size,\n",
    "            progress_bar=True\n",
    "        )\n",
    "\n",
    "        return results \n",
    "    \n",
    "    def download_structure(self, pdb_id: str, file_format: str = \"pdb\") -> str:\n",
    "        \"\"\"\n",
    "        Download the structure file for a given PDB ID.\n",
    "        Args:\n",
    "            pdb_id (str): PDB ID to download.\n",
    "            file_format (str): Format of the file to download. Default is \"pdb\".\n",
    "        Returns:\n",
    "            str: Path to the downloaded file.\n",
    "        \"\"\"\n",
    "        url = f\"{self.base_url}{pdb_id}.{file_format}\"\n",
    "        response = requests.get(url)\n",
    "        \n",
    "        if response.status_code == 200:\n",
    "            file_path = os.path.join(PDB_CACHE_DIR, f\"{pdb_id}.{file_format}\")\n",
    "            with open(file_path, \"wb\") as f:\n",
    "                f.write(response.content)\n",
    "            return file_path\n",
    "        else:\n",
    "            print(f\"Failed to download {pdb_id}: {response.status_code}\")\n",
    "            return None\n",
    "        \n",
    "    def download_structures_parallel(self, pdb_ids: list, file_format: str = \"pdb\", max_workers: int = 5) -> list:\n",
    "        \"\"\"\n",
    "        Download structure files in parallel\n",
    "        Args:\n",
    "            pdb_ids (list): List of PDB IDs to download.\n",
    "            file_format (str): Format of the files to download. Default is \"pdb\".\n",
    "            max_workers (int): Maximum number of parallel downloads.\n",
    "        Returns:\n",
    "\n",
    "            list: List of paths to the downloaded files.\n",
    "        \"\"\"\n",
    "        if not os.path.exists(PDB_CACHE_DIR):\n",
    "            os.makedirs(PDB_CACHE_DIR)\n",
    "\n",
    "        downloaded_files = []\n",
    "        lock = threading.Lock()\n",
    "\n",
    "        with ThreadPoolExecutor(max_workers=max_workers) as executor:\n",
    "            futures = [executor.submit(self.download_structure, pdb_id, file_format) for pdb_id in pdb_ids]\n",
    "            for future in as_completed(futures):\n",
    "                result = future.result()\n",
    "                if result:\n",
    "                    with lock:\n",
    "                        downloaded_files.append(result)\n",
    "\n",
    "        return downloaded_files\n",
    "    \n",
    "    def parse_search(self, results: dict) -> pd.DataFrame:\n",
    "        export_df = pd.DataFrame()\n",
    "        for entry in results[\"data\"][\"entries\"]:\n",
    "            df = pd.json_normalize(\n",
    "                entry, \n",
    "                sep=\"_\",\n",
    "                record_path=None\n",
    "            )\n",
    "\n",
    "            # Expand nested fields manually\n",
    "            df.insert(1,'method',df['exptl'].apply(lambda x: [y['method'] for y in x]))\n",
    "            df.drop(columns=['exptl'], inplace=True)\n",
    "\n",
    "            export_df = pd.concat([export_df, df], ignore_index=True)\n",
    "        return export_df\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f10e885",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['src/.pdb_cache/4HHB.pdb',\n",
       " 'src/.pdb_cache/1A8I.pdb',\n",
       " 'src/.pdb_cache/1A8J.pdb',\n",
       " 'src/.pdb_cache/1A8K.pdb',\n",
       " 'src/.pdb_cache/1A8L.pdb',\n",
       " 'src/.pdb_cache/1A8M.pdb']"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# TODO PDB Y resolucion\n",
    "instance = PdbInterface()\n",
    "# result = instance.search_ids(\n",
    "#     [\"4HHB\", \"1A8I\", \"1A8J\", \"1A8K\", \"1A8L\", \"1A8M\"], \n",
    "#     return_data_list=[\"exptl.method\", \"rcsb_entry_info\", \"rcsb_entry_info\"]\n",
    "# )\n",
    "#instance.parse_search(result)\n",
    "instance.download_structures_parallel(\n",
    "    [\"4HHB\", \"1A8I\", \"1A8J\", \"1A8K\", \"1A8L\", \"1A8M\"], \n",
    "    file_format=\"pdb\",\n",
    "    max_workers=2\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "ba91a187",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "with open(\"results.json\", \"w\") as f:\n",
    "    json.dump(result, f)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ed86b941",
   "metadata": {},
   "source": [
    "## Refseq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "9058d933",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import subprocess\n",
    "import json\n",
    "from Bio import Entrez, SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "581d5cb7",
   "metadata": {},
   "outputs": [],
   "source": [
    "Entrez.email = \"diego.fernandez@umag.cl\"\n",
    "handle = Entrez.einfo() # or esearch, efetch, ...\n",
    "record = Entrez.read(handle)\n",
    "handle.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "2f344da2",
   "metadata": {},
   "outputs": [],
   "source": [
    "handle = Entrez.efetch(db=\"protein\", id=\"XP_010804480.1\", retmode=\"xml\")\n",
    "record = Entrez.read(handle)\n",
    "handle.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "f2d846e8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'GBSeq_locus': 'XP_010804480', 'GBSeq_length': '259', 'GBSeq_moltype': 'AA', 'GBSeq_topology': 'linear', 'GBSeq_division': 'MAM', 'GBSeq_update-date': '26-JAN-2016', 'GBSeq_create-date': '30-DEC-2014', 'GBSeq_definition': 'PREDICTED: beta-casein isoform X1 [Bos taurus]', 'GBSeq_primary-accession': 'XP_010804480', 'GBSeq_accession-version': 'XP_010804480.1', 'GBSeq_other-seqids': ['ref|XP_010804480.1|', 'gi|741930202'], 'GBSeq_project': 'PRJNA33843', 'GBSeq_keywords': ['RefSeq'], 'GBSeq_source': 'Bos taurus (cattle)', 'GBSeq_organism': 'Bos taurus', 'GBSeq_taxonomy': 'Eukaryota; Metazoa; Chordata; Craniata; Vertebrata; Euteleostomi; Mammalia; Eutheria; Laurasiatheria; Cetartiodactyla; Ruminantia; Pecora; Bovidae; Bovinae; Bos', 'GBSeq_comment': \"MODEL REFSEQ: This record is predicted by automated computational analysis. This record is derived from a genomic sequence (NW_003103978.1) annotated using gene prediction method: Gnomon, supported by mRNA and EST evidence.~Also see:~ Documentation of NCBI's Annotation Process; [WARNING] On May 7, 2018 this sequence was replaced by XP_010804480.2.; ##Genome-Annotation-Data-START## ; Annotation Provider :: NCBI ; Annotation Status :: Full annotation ; Annotation Version :: Bos taurus Annotation Release 105 ; Annotation Pipeline :: NCBI eukaryotic genome annotation pipeline ; Annotation Software Version :: 6.5 ; Annotation Method :: Best-placed RefSeq; Gnomon ; Features Annotated :: Gene; mRNA; CDS; ncRNA ; ##Genome-Annotation-Data-END##; COMPLETENESS: full length.\", 'GBSeq_source-db': 'REFSEQ: accession XM_010806178.1', 'GBSeq_feature-table': [{'GBFeature_key': 'source', 'GBFeature_location': '1..259', 'GBFeature_intervals': [{'GBInterval_from': '1', 'GBInterval_to': '259', 'GBInterval_accession': 'XP_010804480.1'}], 'GBFeature_quals': [{'GBQualifier_name': 'organism', 'GBQualifier_value': 'Bos taurus'}, {'GBQualifier_name': 'db_xref', 'GBQualifier_value': 'taxon:9913'}, {'GBQualifier_name': 'chromosome', 'GBQualifier_value': '6'}, {'GBQualifier_name': 'breed', 'GBQualifier_value': 'Hereford'}]}, {'GBFeature_key': 'Protein', 'GBFeature_location': '1..259', 'GBFeature_intervals': [{'GBInterval_from': '1', 'GBInterval_to': '259', 'GBInterval_accession': 'XP_010804480.1'}], 'GBFeature_quals': [{'GBQualifier_name': 'product', 'GBQualifier_value': 'beta-casein isoform X1'}, {'GBQualifier_name': 'calculated_mol_wt', 'GBQualifier_value': '28981'}]}, {'GBFeature_key': 'CDS', 'GBFeature_location': '1..259', 'GBFeature_intervals': [{'GBInterval_from': '1', 'GBInterval_to': '259', 'GBInterval_accession': 'XP_010804480.1'}], 'GBFeature_quals': [{'GBQualifier_name': 'gene', 'GBQualifier_value': 'CSN2'}, {'GBQualifier_name': 'coded_by', 'GBQualifier_value': 'XM_010806178.1:124..903'}, {'GBQualifier_name': 'transl_table', 'GBQualifier_value': '1'}, {'GBQualifier_name': 'db_xref', 'GBQualifier_value': 'GeneID:281099'}, {'GBQualifier_name': 'db_xref', 'GBQualifier_value': 'BGD:BT11954'}]}], 'GBSeq_sequence': 'mplntiykqpqnqiiihsappsllvlyfgkkelramkvlilaclvalalareleelnvpgeiveslssseesitrinkkiekfqseeqqqtedelqdkihpfaqtqslvypfpgpipnslpqnippltqtpvvvppflqpevmgvskvkeamapkhkempfpkypvepftesqsltltdvenlhlplpllqswmhqphqplpptvmfppqsvlslsqskvlpvpqkavpypqrdmpiqafllyqepvlgpvrgpfpiiv', 'GBSeq_xrefs': [{'GBXref_dbname': 'BioProject', 'GBXref_id': 'PRJNA33843'}]}]"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "record"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "0eef5e2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "class RefSeqInterface():\n",
    "    def __init__(self):\n",
    "        pass\n",
    "\n",
    "    def get_refseq_id(self, id: str) -> pd.DataFrame:\n",
    "        handle = Entrez.efetch(db=\"protein\", id=id, retmode=\"xml\")\n",
    "        records = Entrez.read(handle)\n",
    "        handle.close()\n",
    "        return pd.DataFrame(records)\n",
    "\n",
    "    def get_refseq_ids(self, ids: list, max_workers: int = 5) -> pd.DataFrame:\n",
    "        \"\"\"\n",
    "        Get data in parallel for a list of RefSeq IDs.\n",
    "        Args:\n",
    "            ids (list): List of RefSeq IDs to fetch predictions for.\n",
    "            max_workers (int): Maximum number of parallel requests.\n",
    "        Returns:\n",
    "            pd.DataFrame: DataFrame containing parsed predictions.\n",
    "        \"\"\"\n",
    "        export_df = pd.DataFrame()\n",
    "        lock = threading.Lock()\n",
    "\n",
    "        with ThreadPoolExecutor(max_workers=max_workers) as executor:\n",
    "            futures = [executor.submit(self.get_refseq_id, uid) for uid in ids]\n",
    "            for future in as_completed(futures):\n",
    "                result = future.result()\n",
    "                if type(result) == pd.DataFrame:\n",
    "                    with lock:\n",
    "                        export_df = pd.concat([export_df, result], ignore_index=True)\n",
    "        return export_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ad306d9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GBSeq_locus</th>\n",
       "      <th>GBSeq_length</th>\n",
       "      <th>GBSeq_moltype</th>\n",
       "      <th>GBSeq_topology</th>\n",
       "      <th>GBSeq_division</th>\n",
       "      <th>GBSeq_update-date</th>\n",
       "      <th>GBSeq_create-date</th>\n",
       "      <th>GBSeq_definition</th>\n",
       "      <th>GBSeq_primary-accession</th>\n",
       "      <th>GBSeq_accession-version</th>\n",
       "      <th>...</th>\n",
       "      <th>GBSeq_project</th>\n",
       "      <th>GBSeq_keywords</th>\n",
       "      <th>GBSeq_source</th>\n",
       "      <th>GBSeq_organism</th>\n",
       "      <th>GBSeq_taxonomy</th>\n",
       "      <th>GBSeq_comment</th>\n",
       "      <th>GBSeq_source-db</th>\n",
       "      <th>GBSeq_feature-table</th>\n",
       "      <th>GBSeq_sequence</th>\n",
       "      <th>GBSeq_xrefs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>XP_010804480</td>\n",
       "      <td>259</td>\n",
       "      <td>AA</td>\n",
       "      <td>linear</td>\n",
       "      <td>MAM</td>\n",
       "      <td>26-JAN-2016</td>\n",
       "      <td>30-DEC-2014</td>\n",
       "      <td>PREDICTED: beta-casein isoform X1 [Bos taurus]</td>\n",
       "      <td>XP_010804480</td>\n",
       "      <td>XP_010804480.1</td>\n",
       "      <td>...</td>\n",
       "      <td>PRJNA33843</td>\n",
       "      <td>[RefSeq]</td>\n",
       "      <td>Bos taurus (cattle)</td>\n",
       "      <td>Bos taurus</td>\n",
       "      <td>Eukaryota; Metazoa; Chordata; Craniata; Verteb...</td>\n",
       "      <td>MODEL REFSEQ: This record is predicted by auto...</td>\n",
       "      <td>REFSEQ: accession XM_010806178.1</td>\n",
       "      <td>[{'GBFeature_key': 'source', 'GBFeature_locati...</td>\n",
       "      <td>mplntiykqpqnqiiihsappsllvlyfgkkelramkvlilaclva...</td>\n",
       "      <td>[{'GBXref_dbname': 'BioProject', 'GBXref_id': ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>XP_010804481</td>\n",
       "      <td>233</td>\n",
       "      <td>AA</td>\n",
       "      <td>linear</td>\n",
       "      <td>MAM</td>\n",
       "      <td>30-DEC-2014</td>\n",
       "      <td>30-DEC-2014</td>\n",
       "      <td>PREDICTED: proline-rich protein 27 isoform X1 ...</td>\n",
       "      <td>XP_010804481</td>\n",
       "      <td>XP_010804481.1</td>\n",
       "      <td>...</td>\n",
       "      <td>PRJNA33843</td>\n",
       "      <td>[RefSeq]</td>\n",
       "      <td>Bos taurus (cattle)</td>\n",
       "      <td>Bos taurus</td>\n",
       "      <td>Eukaryota; Metazoa; Chordata; Craniata; Verteb...</td>\n",
       "      <td>MODEL REFSEQ: This record is predicted by auto...</td>\n",
       "      <td>REFSEQ: accession XM_010806179.1</td>\n",
       "      <td>[{'GBFeature_key': 'source', 'GBFeature_locati...</td>\n",
       "      <td>mkfllwaclmyvsfardysdnigspypvnpsasisypvipsasipy...</td>\n",
       "      <td>[{'GBXref_dbname': 'BioProject', 'GBXref_id': ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>XP_010804482</td>\n",
       "      <td>218</td>\n",
       "      <td>AA</td>\n",
       "      <td>linear</td>\n",
       "      <td>MAM</td>\n",
       "      <td>30-DEC-2014</td>\n",
       "      <td>30-DEC-2014</td>\n",
       "      <td>PREDICTED: proline-rich protein 27 isoform X2 ...</td>\n",
       "      <td>XP_010804482</td>\n",
       "      <td>XP_010804482.1</td>\n",
       "      <td>...</td>\n",
       "      <td>PRJNA33843</td>\n",
       "      <td>[RefSeq]</td>\n",
       "      <td>Bos taurus (cattle)</td>\n",
       "      <td>Bos taurus</td>\n",
       "      <td>Eukaryota; Metazoa; Chordata; Craniata; Verteb...</td>\n",
       "      <td>MODEL REFSEQ: This record is predicted by auto...</td>\n",
       "      <td>REFSEQ: accession XM_010806180.1</td>\n",
       "      <td>[{'GBFeature_key': 'source', 'GBFeature_locati...</td>\n",
       "      <td>mfvgstnmkfllwaclmyvsfardysdnigspypvnpsasisypvi...</td>\n",
       "      <td>[{'GBXref_dbname': 'BioProject', 'GBXref_id': ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    GBSeq_locus GBSeq_length GBSeq_moltype GBSeq_topology GBSeq_division  \\\n",
       "0  XP_010804480          259            AA         linear            MAM   \n",
       "1  XP_010804481          233            AA         linear            MAM   \n",
       "2  XP_010804482          218            AA         linear            MAM   \n",
       "\n",
       "  GBSeq_update-date GBSeq_create-date  \\\n",
       "0       26-JAN-2016       30-DEC-2014   \n",
       "1       30-DEC-2014       30-DEC-2014   \n",
       "2       30-DEC-2014       30-DEC-2014   \n",
       "\n",
       "                                    GBSeq_definition GBSeq_primary-accession  \\\n",
       "0     PREDICTED: beta-casein isoform X1 [Bos taurus]            XP_010804480   \n",
       "1  PREDICTED: proline-rich protein 27 isoform X1 ...            XP_010804481   \n",
       "2  PREDICTED: proline-rich protein 27 isoform X2 ...            XP_010804482   \n",
       "\n",
       "  GBSeq_accession-version  ... GBSeq_project GBSeq_keywords  \\\n",
       "0          XP_010804480.1  ...    PRJNA33843       [RefSeq]   \n",
       "1          XP_010804481.1  ...    PRJNA33843       [RefSeq]   \n",
       "2          XP_010804482.1  ...    PRJNA33843       [RefSeq]   \n",
       "\n",
       "          GBSeq_source GBSeq_organism  \\\n",
       "0  Bos taurus (cattle)     Bos taurus   \n",
       "1  Bos taurus (cattle)     Bos taurus   \n",
       "2  Bos taurus (cattle)     Bos taurus   \n",
       "\n",
       "                                      GBSeq_taxonomy  \\\n",
       "0  Eukaryota; Metazoa; Chordata; Craniata; Verteb...   \n",
       "1  Eukaryota; Metazoa; Chordata; Craniata; Verteb...   \n",
       "2  Eukaryota; Metazoa; Chordata; Craniata; Verteb...   \n",
       "\n",
       "                                       GBSeq_comment  \\\n",
       "0  MODEL REFSEQ: This record is predicted by auto...   \n",
       "1  MODEL REFSEQ: This record is predicted by auto...   \n",
       "2  MODEL REFSEQ: This record is predicted by auto...   \n",
       "\n",
       "                    GBSeq_source-db  \\\n",
       "0  REFSEQ: accession XM_010806178.1   \n",
       "1  REFSEQ: accession XM_010806179.1   \n",
       "2  REFSEQ: accession XM_010806180.1   \n",
       "\n",
       "                                 GBSeq_feature-table  \\\n",
       "0  [{'GBFeature_key': 'source', 'GBFeature_locati...   \n",
       "1  [{'GBFeature_key': 'source', 'GBFeature_locati...   \n",
       "2  [{'GBFeature_key': 'source', 'GBFeature_locati...   \n",
       "\n",
       "                                      GBSeq_sequence  \\\n",
       "0  mplntiykqpqnqiiihsappsllvlyfgkkelramkvlilaclva...   \n",
       "1  mkfllwaclmyvsfardysdnigspypvnpsasisypvipsasipy...   \n",
       "2  mfvgstnmkfllwaclmyvsfardysdnigspypvnpsasisypvi...   \n",
       "\n",
       "                                         GBSeq_xrefs  \n",
       "0  [{'GBXref_dbname': 'BioProject', 'GBXref_id': ...  \n",
       "1  [{'GBXref_dbname': 'BioProject', 'GBXref_id': ...  \n",
       "2  [{'GBXref_dbname': 'BioProject', 'GBXref_id': ...  \n",
       "\n",
       "[3 rows x 21 columns]"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# TODO CSV o JSON para parseo\n",
    "instance = RefSeqInterface()\n",
    "#result = instance.get_refseq_id(\"XP_010804480.1\")\n",
    "df = instance.get_refseq_ids([\"XP_010804480.1\", \"XP_010804481.1\", \"XP_010804482.1\"], max_workers=2)\n",
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "64c69396",
   "metadata": {},
   "source": [
    "### Brenda"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "id": "c917ae92",
   "metadata": {},
   "outputs": [],
   "source": [
    "import string, os\n",
    "import hashlib\n",
    "from zeep import Client\n",
    "from zeep.helpers import serialize_object\n",
    "import pandas as pd\n",
    "from typing import List, Tuple\n",
    "import threading\n",
    "from concurrent.futures import ThreadPoolExecutor, as_completed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9caf8000",
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO Sleep random, ver lo de abajo\n",
    "# For aditional implementations see: https://www.brenda-enzymes.org/soap.php\n",
    "func_map = {\n",
    "    #'getReference': [\"ecNumber\", \"organism\"],\n",
    "    # 'getEcNumbersFromSequence': ,\n",
    "    # 'getOrganismsFromSequence': ,\n",
    "    'getKmValue': [\"ecNumber\", \"kmValue\", \"kmValueMaximum\", \"substrate\", \"commentary\", \"organism\", \"ligandStructureId\", \"literature\"],\n",
    "    # 'getGeneralInformation': ,\n",
    "    # 'getExpression': ,\n",
    "    'getIc50Value': [\"ecNumber\", \"ic50Value\", \"ic50ValueMaximum\", \"inhibitor\", \"commentary\", \"organism\", \"ligandStructureId\", \"literature\"],\n",
    "    'getKcatKmValue': [\"ecNumber\", \"kcatKmValue\", \"kcatKmValueMaximum\", \"substrate\", \"commentary\", \"organism\", \"ligandStructureId\", \"literature\"],\n",
    "    'getPhStability': [\"ecNumber\", \"phStability\", \"phStabilityMaximum\", \"commentary\", \"organism\", \"literature\"]\n",
    "    # 'getOxidationStability': ,\n",
    "    # 'getNaturalSubstratesProducts': ,\n",
    "    # 'getEngineering': ,\n",
    "    # 'getNaturalProduct': ,\n",
    "    # 'getMetalsIons': ,\n",
    "    # 'getActivatingCompound': ,\n",
    "    # 'getInhibitors': ,\n",
    "    # 'getCofactor': ,\n",
    "    # 'getGeneralStability': ,\n",
    "    # 'getNaturalSubstrate': ,\n",
    "    # 'getMolecularWeight': ,\n",
    "    # 'getCrystallization': ,\n",
    "    # 'getSubstratesProducts': ,\n",
    "    # 'getReactionType': ,\n",
    "    # 'getOrganismSynonyms': ,\n",
    "    # 'getEnzymeNames': ,\n",
    "    # 'getOrganicSolventStability': ,\n",
    "    # 'getApplication': ,\n",
    "    # 'getSynonyms': ,\n",
    "    # 'getTemperatureOptimum': ,\n",
    "    # 'getTemperatureStability': ,\n",
    "    # 'getPiValue': ,\n",
    "    # 'getTemperatureRange': ,\n",
    "    # 'getRecommendedName': ,\n",
    "    # 'getProduct': ,\n",
    "    # 'getCasRegistryNumber': ,\n",
    "    # 'getLocalization': ,\n",
    "    # 'getPosttranslationalModification': ,\n",
    "    # 'getSystematicName': ,\n",
    "    # 'getCloned': [\"ecNumber\", \"organism\"],\n",
    "    # 'getSpecificActivity': ,\n",
    "    # 'getSubunits': ,\n",
    "    # 'getLigands': ,\n",
    "    # 'getTurnoverNumber': ,\n",
    "    # 'getReaction': ,\n",
    "    # 'getSourceTissue': ,\n",
    "    # 'getSubstrate': ,\n",
    "    # 'getPhRange': ,\n",
    "    # 'getStorageStability': ,\n",
    "    # 'getPhOptimum': ,\n",
    "    # 'getDisease': ,\n",
    "    # 'getPurification': ,\n",
    "    # 'getRenatured': ,\n",
    "    # 'getKiValue': ,\n",
    "    # 'getPathway': ,\n",
    "    # 'getPdb': ,\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "id": "442623ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "class BrendaInstance():\n",
    "    def __init__(self, email: str, password: str):\n",
    "        wsdl = \"https://www.brenda-enzymes.org/soap/brenda_zeep.wsdl\"\n",
    "        self.email = email\n",
    "        self.password = hashlib.sha256(password.encode(\"utf-8\")).hexdigest()\n",
    "        self.client = Client(wsdl)\n",
    "        self.result = {}\n",
    "\n",
    "    def show_all_operations(self):\n",
    "        print(\"Available operations:\")\n",
    "        for service in self.client.wsdl.services.values():\n",
    "            for port in service.ports.values():\n",
    "                for operation_name in port.binding._operations.keys():\n",
    "                    print(f\"- {operation_name}\")\n",
    "    \n",
    "    def fetch_single_record(self, ec_number: str, organism: str) -> pd.DataFrame:\n",
    "        input_parameters = {\n",
    "            \"ecNumber\": ec_number,\n",
    "            \"organism\": organism\n",
    "        }\n",
    "        export_data = {}\n",
    "        \n",
    "        for operation_name in func_map.keys():\n",
    "            print(f\"Executing {operation_name}\")\n",
    "            try:\n",
    "                # Get the fields required for this function\n",
    "                field_names = func_map[operation_name]\n",
    "\n",
    "                # Build parameters in order\n",
    "                param_list = [f\"{key}*{input_parameters.get(key, '')}\" for key in field_names]\n",
    "\n",
    "                # Add credentials\n",
    "                parameters = [self.email, self.password] + param_list\n",
    "\n",
    "                func = getattr(self.client.service, operation_name)\n",
    "                result = serialize_object(func(*parameters))\n",
    "\n",
    "                # Convert to DataFrame\n",
    "                if isinstance(result, list):\n",
    "                    df = pd.DataFrame(result)\n",
    "                else:\n",
    "                    df = pd.DataFrame([result])\n",
    "                \n",
    "\n",
    "            \n",
    "            except Exception as e:\n",
    "                print(f\"Error executing {operation_name}: {e}\")\n",
    "                continue\n",
    "\n",
    "            # Add the DataFrame to the export data\n",
    "            df_name = operation_name.replace(\"get\", \"\")\n",
    "            export_data[df_name] = df\n",
    "        \n",
    "        return export_data\n",
    "    \n",
    "    def fetch_brenda_data_parallel(self, parameters: List[Tuple[str, str]], max_workers: int = 5) -> pd.DataFrame:\n",
    "        \"\"\"\n",
    "        Get data in parallel for a list of EC numbers and organisms.\n",
    "        Args:\n",
    "            parameters (List[str, str]): List of tuples containing EC numbers and organisms.\n",
    "        Returns:\n",
    "            pd.DataFrame: DataFrame containing parsed predictions.\n",
    "        \"\"\"\n",
    "        lock = threading.Lock()\n",
    "\n",
    "        with ThreadPoolExecutor(max_workers) as executor:\n",
    "            futures = [executor.submit(self.fetch_single_record, ec, org) for ec, org in parameters]\n",
    "            for future in as_completed(futures):\n",
    "                result = future.result()\n",
    "                if isinstance(result, dict):\n",
    "                    with lock:\n",
    "                        # Append to every correpsonding DataFrame in self.result\n",
    "                        for key, df in result.items():\n",
    "                            if key not in self.result:\n",
    "                                self.result[key] = df\n",
    "                            else:\n",
    "                                self.result[key] = pd.concat([self.result[key], df], ignore_index=True)\n",
    "    \n",
    "    def show_results(self):\n",
    "        \"\"\"\n",
    "        Show the results of the queries.\n",
    "        \"\"\"\n",
    "        if not self.result:\n",
    "            print(\"No results found.\")\n",
    "            return\n",
    "        for key, df in self.result.items():\n",
    "            print(f\"Results for {key}:\")\n",
    "            print(df.head())\n",
    "        \n",
    "\n",
    "    def save_results(self, directory: str):\n",
    "        \"\"\"\n",
    "        Save the results to the specified directory.\n",
    "        Args:\n",
    "            directory (str): Directory to save the results.\n",
    "        \"\"\"\n",
    "        if not os.path.exists(directory):\n",
    "            os.makedirs(directory)\n",
    "\n",
    "        for key, df in self.result.items():\n",
    "            if df.empty:\n",
    "                print(f\"No results for {key}.\")\n",
    "                continue\n",
    "            df.to_csv(os.path.join(directory, f\"{key}.csv\"), index=False)\n",
    "            print(f\"Saved {key} to {directory}/{key}.csv\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "id": "86193682",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Executing getIc50Value\n",
      "Executing getKcatKmValue\n",
      "Executing getPhStability\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'Ic50Value': Empty DataFrame\n",
       " Columns: []\n",
       " Index: [],\n",
       " 'KcatKmValue': Empty DataFrame\n",
       " Columns: []\n",
       " Index: [],\n",
       " 'PhStability':   phStabilityMaximum phStability literature commentary      organism ecNumber\n",
       " 0               10.6           7   [285567]     stable  Homo sapiens  1.1.1.1}"
      ]
     },
     "execution_count": 132,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "instance.fetch_single_record(\"1.1.1.1\", \"Homo sapiens\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "22d269e2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Executing getIc50Value\n",
      "Executing getKcatKmValue\n",
      "Executing getPhStability\n",
      "No results for Ic50Value.\n",
      "No results for KcatKmValue.\n",
      "Saved PhStability to results/brenda/PhStability.csv\n"
     ]
    }
   ],
   "source": [
    "# instance = BrendaInstance(\"diego.fernandez@umag.cl\", \"\")\n",
    "# instance.fetch_brenda_data_parallel([(\"1.1.1.1\", \"Homo sapiens\")])\n",
    "# instance.save_results(\"results/brenda\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6b8dec67",
   "metadata": {},
   "source": [
    "## Reactome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "34fcc3d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import requests\n",
    "import threading\n",
    "from concurrent.futures import ThreadPoolExecutor, as_completed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "e8df965f",
   "metadata": {},
   "outputs": [],
   "source": [
    "REACTOME_BASE_URL = \"https://reactome.org/ContentService/\"\n",
    "class ReactomeInstance():\n",
    "    def __init__(self):\n",
    "        pass\n",
    "\n",
    "    def get_pathway(self, id: str) -> dict:\n",
    "        \"\"\"\n",
    "        Download pathways from a given UniProt ID.\n",
    "        Args:\n",
    "            id (str): Reactome pathway ID.\n",
    "        Returns:\n",
    "            dict: Pathway data.\n",
    "        \"\"\"\n",
    "        url = f\"{REACTOME_BASE_URL}data/pathway/{id}/containedEvents\"\n",
    "        response = requests.get(url)\n",
    "        \n",
    "        if response.status_code == 200:\n",
    "            return response.json()\n",
    "        else:\n",
    "            print(f\"Failed to fetch pathways for {id}: {response.status_code}\")\n",
    "            return None\n",
    "    \n",
    "    def parse_pathway(self, raw_result: dict) -> pd.DataFrame:\n",
    "        \"\"\"\n",
    "        Parse the pathway data.\n",
    "        Args:\n",
    "            raw_result (dict): Raw pathway data.\n",
    "        Returns:\n",
    "            pd.DataFrame: Parsed pathway data.\n",
    "        \"\"\"\n",
    "        if not raw_result:\n",
    "            return pd.DataFrame()\n",
    "        \n",
    "        # Convert to DataFrame\n",
    "        df = pd.json_normalize(raw_result)\n",
    "        \n",
    "        # Rename columns\n",
    "        df.rename(columns={\"stId\": \"pathway_id\", \"displayName\": \"pathway_name\"}, inplace=True)\n",
    "        \n",
    "        return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e9b3601e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dbId</th>\n",
       "      <th>pathway_name</th>\n",
       "      <th>pathway_id</th>\n",
       "      <th>stIdVersion</th>\n",
       "      <th>isInDisease</th>\n",
       "      <th>isInferred</th>\n",
       "      <th>name</th>\n",
       "      <th>releaseDate</th>\n",
       "      <th>speciesName</th>\n",
       "      <th>category</th>\n",
       "      <th>className</th>\n",
       "      <th>schemaClass</th>\n",
       "      <th>hasDiagram</th>\n",
       "      <th>hasEHLD</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10790592</td>\n",
       "      <td>STING binds cyclic GMP-AMP</td>\n",
       "      <td>R-DME-3244643</td>\n",
       "      <td>R-DME-3244643.1</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[STING binds cyclic GMP-AMP]</td>\n",
       "      <td>2025-03-26</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>omitted</td>\n",
       "      <td>Reaction</td>\n",
       "      <td>BlackBoxEvent</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>10819344</td>\n",
       "      <td>DDX41 binds bacterial c-di-AMP, c-di-GMP</td>\n",
       "      <td>R-DME-9013869</td>\n",
       "      <td>R-DME-9013869.1</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[DDX41 binds bacterial c-di-AMP, c-di-GMP  ]</td>\n",
       "      <td>2025-03-26</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>binding</td>\n",
       "      <td>Reaction</td>\n",
       "      <td>Reaction</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>10790177</td>\n",
       "      <td>STING dimerization</td>\n",
       "      <td>R-DME-3134800</td>\n",
       "      <td>R-DME-3134800.1</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[STING dimerization]</td>\n",
       "      <td>2025-03-26</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>binding</td>\n",
       "      <td>Reaction</td>\n",
       "      <td>Reaction</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>10788129</td>\n",
       "      <td>STING binds c-di-GMP</td>\n",
       "      <td>R-DME-2396009</td>\n",
       "      <td>R-DME-2396009.1</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[STING binds c-di-GMP]</td>\n",
       "      <td>2025-03-26</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>omitted</td>\n",
       "      <td>Reaction</td>\n",
       "      <td>BlackBoxEvent</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>10829117</td>\n",
       "      <td>STAT6-mediated induction of chemokines</td>\n",
       "      <td>R-DME-3249367</td>\n",
       "      <td>R-DME-3249367.1</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[STAT6-mediated induction of chemokines]</td>\n",
       "      <td>2025-03-26</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Pathway</td>\n",
       "      <td>Pathway</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>10790821</td>\n",
       "      <td>STING recruits TBK1 and STAT6</td>\n",
       "      <td>R-DME-3249378</td>\n",
       "      <td>R-DME-3249378.1</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[STING recruits TBK1 and STAT6]</td>\n",
       "      <td>2025-03-26</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>binding</td>\n",
       "      <td>Reaction</td>\n",
       "      <td>Reaction</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>10790823</td>\n",
       "      <td>p-S407-STAT6 is phosphorylated at Tyr641</td>\n",
       "      <td>R-DME-3249379</td>\n",
       "      <td>R-DME-3249379.1</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[p-S407-STAT6 is phosphorylated at Tyr641]</td>\n",
       "      <td>2025-03-26</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>transition</td>\n",
       "      <td>Reaction</td>\n",
       "      <td>Reaction</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>10790813</td>\n",
       "      <td>TBK1 phosphorylates STAT6 at Ser407</td>\n",
       "      <td>R-DME-3249371</td>\n",
       "      <td>R-DME-3249371.1</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[TBK1 phosphorylates STAT6 at Ser407]</td>\n",
       "      <td>2025-03-26</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>transition</td>\n",
       "      <td>Reaction</td>\n",
       "      <td>Reaction</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>10790798</td>\n",
       "      <td>p-S407,Y641-STAT6 dimer migrates to the nucleus</td>\n",
       "      <td>R-DME-3249370</td>\n",
       "      <td>R-DME-3249370.1</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[p-S407,Y641-STAT6 dimer migrates to the nucleus]</td>\n",
       "      <td>2025-03-26</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>transition</td>\n",
       "      <td>Reaction</td>\n",
       "      <td>Reaction</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>10790815</td>\n",
       "      <td>p-S407,Y641-STAT6 is dimerized</td>\n",
       "      <td>R-DME-3249372</td>\n",
       "      <td>R-DME-3249372.1</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[p-S407,Y641-STAT6 is dimerized]</td>\n",
       "      <td>2025-03-26</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>binding</td>\n",
       "      <td>Reaction</td>\n",
       "      <td>Reaction</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>10790825</td>\n",
       "      <td>TBK1 is phosphorylated within STING:TBK1:STAT6...</td>\n",
       "      <td>R-DME-3249390</td>\n",
       "      <td>R-DME-3249390.1</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[TBK1 is phosphorylated within STING:TBK1:STAT...</td>\n",
       "      <td>2025-03-26</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>transition</td>\n",
       "      <td>Reaction</td>\n",
       "      <td>Reaction</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        dbId                                       pathway_name  \\\n",
       "0   10790592                         STING binds cyclic GMP-AMP   \n",
       "1   10819344         DDX41 binds bacterial c-di-AMP, c-di-GMP     \n",
       "2   10790177                                 STING dimerization   \n",
       "3   10788129                               STING binds c-di-GMP   \n",
       "4   10829117             STAT6-mediated induction of chemokines   \n",
       "5   10790821                      STING recruits TBK1 and STAT6   \n",
       "6   10790823           p-S407-STAT6 is phosphorylated at Tyr641   \n",
       "7   10790813                TBK1 phosphorylates STAT6 at Ser407   \n",
       "8   10790798    p-S407,Y641-STAT6 dimer migrates to the nucleus   \n",
       "9   10790815                     p-S407,Y641-STAT6 is dimerized   \n",
       "10  10790825  TBK1 is phosphorylated within STING:TBK1:STAT6...   \n",
       "\n",
       "       pathway_id      stIdVersion  isInDisease  isInferred  \\\n",
       "0   R-DME-3244643  R-DME-3244643.1        False        True   \n",
       "1   R-DME-9013869  R-DME-9013869.1        False        True   \n",
       "2   R-DME-3134800  R-DME-3134800.1        False        True   \n",
       "3   R-DME-2396009  R-DME-2396009.1        False        True   \n",
       "4   R-DME-3249367  R-DME-3249367.1        False        True   \n",
       "5   R-DME-3249378  R-DME-3249378.1        False        True   \n",
       "6   R-DME-3249379  R-DME-3249379.1        False        True   \n",
       "7   R-DME-3249371  R-DME-3249371.1        False        True   \n",
       "8   R-DME-3249370  R-DME-3249370.1        False        True   \n",
       "9   R-DME-3249372  R-DME-3249372.1        False        True   \n",
       "10  R-DME-3249390  R-DME-3249390.1        False        True   \n",
       "\n",
       "                                                 name releaseDate  \\\n",
       "0                        [STING binds cyclic GMP-AMP]  2025-03-26   \n",
       "1        [DDX41 binds bacterial c-di-AMP, c-di-GMP  ]  2025-03-26   \n",
       "2                                [STING dimerization]  2025-03-26   \n",
       "3                              [STING binds c-di-GMP]  2025-03-26   \n",
       "4            [STAT6-mediated induction of chemokines]  2025-03-26   \n",
       "5                     [STING recruits TBK1 and STAT6]  2025-03-26   \n",
       "6          [p-S407-STAT6 is phosphorylated at Tyr641]  2025-03-26   \n",
       "7               [TBK1 phosphorylates STAT6 at Ser407]  2025-03-26   \n",
       "8   [p-S407,Y641-STAT6 dimer migrates to the nucleus]  2025-03-26   \n",
       "9                    [p-S407,Y641-STAT6 is dimerized]  2025-03-26   \n",
       "10  [TBK1 is phosphorylated within STING:TBK1:STAT...  2025-03-26   \n",
       "\n",
       "                speciesName    category className    schemaClass hasDiagram  \\\n",
       "0   Drosophila melanogaster     omitted  Reaction  BlackBoxEvent        NaN   \n",
       "1   Drosophila melanogaster     binding  Reaction       Reaction        NaN   \n",
       "2   Drosophila melanogaster     binding  Reaction       Reaction        NaN   \n",
       "3   Drosophila melanogaster     omitted  Reaction  BlackBoxEvent        NaN   \n",
       "4   Drosophila melanogaster         NaN   Pathway        Pathway      False   \n",
       "5   Drosophila melanogaster     binding  Reaction       Reaction        NaN   \n",
       "6   Drosophila melanogaster  transition  Reaction       Reaction        NaN   \n",
       "7   Drosophila melanogaster  transition  Reaction       Reaction        NaN   \n",
       "8   Drosophila melanogaster  transition  Reaction       Reaction        NaN   \n",
       "9   Drosophila melanogaster     binding  Reaction       Reaction        NaN   \n",
       "10  Drosophila melanogaster  transition  Reaction       Reaction        NaN   \n",
       "\n",
       "   hasEHLD  \n",
       "0      NaN  \n",
       "1      NaN  \n",
       "2      NaN  \n",
       "3      NaN  \n",
       "4    False  \n",
       "5      NaN  \n",
       "6      NaN  \n",
       "7      NaN  \n",
       "8      NaN  \n",
       "9      NaN  \n",
       "10     NaN  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "instance = ReactomeInstance()\n",
    "respponse = instance.get_pathway(\"R-DME-1834941\")\n",
    "df = instance.parse_pathway(respponse)\n",
    "df"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
