{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "887ed803-c6dd-4da6-9845-3afb0c5f8392",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "from src.uniprot import UniprotInterface\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5bc2dc57",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e4834c7",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "def replace_char_at_index(s, i, new_char):\n",
    "    if i < 0 or i >= len(s):\n",
    "        raise IndexError(\"Index out of range.\")\n",
    "    return s[:i] + new_char + s[i+1:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b318b89f",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "ids = [\"Q75UA4\"]\n",
    "from_db = 'UniProtKB_AC-ID'\n",
    "to_db = 'UniProtKB'\n",
    "disease = \"CRC\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0a647744",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "downloader = UniprotInterface()\n",
    "\n",
    "job_id = downloader.submit_id_mapping(from_db=from_db, to_db=to_db, ids=ids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "eef323a8",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetched: 1 / 1\n"
     ]
    }
   ],
   "source": [
    "if downloader.check_id_mapping_results_ready(job_id):\n",
    "    link = downloader.get_id_mapping_results_link(job_id)\n",
    "    results = downloader.get_id_mapping_results_search(link)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "b9396a36",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'json' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mNameError\u001b[39m                                 Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[16]\u001b[39m\u001b[32m, line 2\u001b[39m\n\u001b[32m      1\u001b[39m \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28mopen\u001b[39m(\u001b[33m\"\u001b[39m\u001b[33mresults.json\u001b[39m\u001b[33m\"\u001b[39m, \u001b[33m\"\u001b[39m\u001b[33mw\u001b[39m\u001b[33m\"\u001b[39m) \u001b[38;5;28;01mas\u001b[39;00m f:\n\u001b[32m----> \u001b[39m\u001b[32m2\u001b[39m     \u001b[43mjson\u001b[49m.dump(results, f)\n",
      "\u001b[31mNameError\u001b[39m: name 'json' is not defined"
     ]
    }
   ],
   "source": [
    "with open(\"results.json\", \"w\") as f:\n",
    "    json.dump(results, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "243e35c2",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'MKFGKFVLLAASTALAVVGLGGPAAADSTPQAQPSIIGGSNATSGPWAARLFVNGRQNCTATIIAPQYILTAKHCVSSSGTYTFRIGSLDQTSGGTMATGSTITRYPGSADLAIVRLTTSVNATYSPLGSVGDVSVGQNVSVYGWGATSQCGSEINCQSRYLKVATVRVNSISCSDYTGGVAVCANRVNGITAGGDSGGPMFASGRQVGVASTSDRVNNTAYTNITRYRSWISQVAGV'"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results['results'][0]['to']['sequence']['value']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d06bc8c",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Q75UA4\n"
     ]
    }
   ],
   "source": [
    "for result in results['results']:\n",
    "    print(result['from'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f536ba57",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "export_data = []\n",
    "sequence = results['results'][0]['to']['sequence']['value']\n",
    "for feature in results['results'][0]['to']['features']:\n",
    "    row = []\n",
    "    if feature['type'] == 'Natural variant' and disease in feature['description']:     \n",
    "        row.append(feature['featureId'])\n",
    "        location_start = feature['location']['start']['value']\n",
    "        location_end = feature['location']['end']['value']\n",
    "        if location_start == location_end:\n",
    "            row.append(location_start)\n",
    "            original_sequence = feature['alternativeSequence']['originalSequence']\n",
    "            new_sequence = feature['alternativeSequence']['alternativeSequences'][0]\n",
    "            row.append(f\"{original_sequence}->{new_sequence}\")\n",
    "            row.append(replace_char_at_index(sequence, int(location_start)-1, new_sequence))\n",
    "        else:\n",
    "            row.append(f\"{location_start}-{location_end}\")\n",
    "            row.append(\"missing\")\n",
    "            row.append(sequence[:int(location_start)-1] + sequence[int(location_end)-1:])\n",
    "        export_data.append(row)\n",
    "export_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c43c304e",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "df = pd.DataFrame(export_data, columns=[\"variant id\", \"position\", \"change\", \"sequence\"])\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0b712ab6",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "df.to_csv(\"results.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41389a75",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[{'database': 'PubMed', 'id': '11133465'}, {'database': 'DOI', 'id': '10.1128/AEM.67.1.345-353.2001'}]\n",
      "[{'database': 'PubMed', 'id': '16237016'}, {'database': 'DOI', 'id': '10.1128/JB.187.21.7333-7340.2005'}]\n"
     ]
    }
   ],
   "source": [
    "result = results['results'][0]\n",
    "for reference in result['to']['references']:\n",
    "    print(reference['citation']['citationCrossReferences'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e9e1647-5992-4b1e-a28f-c1d73fd8af0f",
   "metadata": {},
   "outputs": [],
   "source": [
    "references_list = []\n",
    "result = results['results'][0]                 \n",
    "\n",
    "try:\n",
    "    for r in result['to']['references']:\n",
    "        tmp = {}\n",
    "        tmp[\"citacionCrossReferences\"] = r['citation']['citationCrossReferences']\n",
    "        tmp.update({\"title\": r['citation']['title']})\n",
    "        references_list.append(tmp)\n",
    "except KeyError:\n",
    "    pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "28767add-3dbd-45fd-9c7e-8a519ff59da0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'citacionCrossReferences': [{'database': 'PubMed', 'id': '11133465'},\n",
       "   {'database': 'DOI', 'id': '10.1128/AEM.67.1.345-353.2001'}],\n",
       "  'title': 'Purification and characterization of an extracellular poly(L-lactic acid) depolymerase from a soil isolate, Amycolatopsis sp. strain K104-1.'},\n",
       " {'citacionCrossReferences': [{'database': 'PubMed', 'id': '16237016'},\n",
       "   {'database': 'DOI', 'id': '10.1128/JB.187.21.7333-7340.2005'}],\n",
       "  'title': 'Gene cloning and molecular characterization of an extracellular poly(L-lactic acid) depolymerase from Amycolatopsis sp. strain K104-1.'}]"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "references_list"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1aa7cd1e",
   "metadata": {},
   "source": [
    "## Blast"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1102bd78",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, argparse\n",
    "import shutil\n",
    "import subprocess\n",
    "import tarfile\n",
    "from pathlib import Path\n",
    "from urllib.request import urlopen\n",
    "import re\n",
    "from typing import List\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "DB_DIR = os.path.join(\"scripts\", \"db\")\n",
    "BLAST_BASE_URL = \"https://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/LATEST/\"\n",
    "UNIPROT_BASE_URL = \"https://ftp.uniprot.org/pub/databases/uniprot/current_release\"\n",
    "BLAST_DIR = Path(\"blast_bin\")\n",
    "#https://ftp.uniprot.org/pub/databases/uniprot/uniref/uniref100/uniref100.fasta.gz\n",
    "#https://ftp.uniprot.org/pub/databases/uniprot/current_release/uniref/uniref100/uniref.xsd\n",
    "#https://ftp.uniprot.org/pub/databases/uniprot/uniref/uniref100/uniref100.xml.gz\n",
    "#https://ftp.uniprot.org/pub/databases/uniprot/uniref/uniref90/uniref90.xml.gz\n",
    "#https://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/complete/uniprot_sprot.xml.gz\n",
    "#https://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/complete/uniprot_trembl.fasta.gz\n",
    "\n",
    "databases = {\n",
    "    \"uniprotkb_reviewed\": \"knowledgebase/complete/uniprot_sprot\",\n",
    "    \"uniprotkb_unreviewed\": \"knowledgebase/complete/uniprot_trembl\",\n",
    "    \"uniref100\": \"uniref/niref100/uniref100\",\n",
    "    \"uniref90\": \"uniref/uniref90/uniref90\",\n",
    "    \"uniref50\": \"uniref/uniref50/uniref50\",\n",
    "}\n",
    "\n",
    "def download_uniprot_database(db_name: str, extension: str = \"xml\"):\n",
    "    \"\"\" Download a Uniprot database from the Uniprot FTP server.\n",
    "    Args:\n",
    "        db_name (str): Name of the database to download.\n",
    "        extension (str): File extension of the database. Default is \"xml\".\n",
    "    \"\"\"\n",
    "\n",
    "    if db_name not in databases:\n",
    "        raise ValueError(f\"Database {db_name} is not supported. Supported databases are: {', '.join(databases.keys())}.\")\n",
    "    \n",
    "    db_path = os.path.join(DB_DIR, f\"{db_name}.{extension}\")\n",
    "    \n",
    "    if not os.path.exists(db_path):\n",
    "        os.makedirs(DB_DIR, exist_ok=True)\n",
    "        url = f\"{UNIPROT_BASE_URL}/{databases[db_name]}.{extension}.gz\"\n",
    "        os.system(f\"wget {url} -O {db_path}.gz\")\n",
    "        print(f\"Unzipping {db_path}...\")\n",
    "        subprocess.run([\"gunzip\", db_path], check=True)\n",
    "    else:\n",
    "        print(f\"Database {db_name} already exists at {db_path}.\")\n",
    "\n",
    "def get_latest_version_url():\n",
    "    \"\"\"Retrieve the latest BLAST+ tarball URL from the NCBI FTP site.\"\"\"\n",
    "    with urlopen(BLAST_BASE_URL) as response:\n",
    "        html = response.read().decode(\"utf-8\")\n",
    "    # Look for something like: ncbi-blast-2.16.0+-x64-linux.tar.gz\n",
    "    match = re.search(r'ncbi-blast-(\\d+\\.\\d+\\.\\d+\\+)-x64-linux\\.tar\\.gz', html)\n",
    "    if match:\n",
    "        version = match.group(1)\n",
    "        tar_name = f\"ncbi-blast-{version}-x64-linux.tar.gz\"\n",
    "        return version, BLAST_BASE_URL + tar_name\n",
    "    else:\n",
    "        raise RuntimeError(\"Could not find the latest BLAST version from NCBI.\")\n",
    "\n",
    "def is_blast_installed():\n",
    "    \"\"\"Check if 'blastp' is available in the system PATH.\"\"\"\n",
    "    try:\n",
    "        subprocess.run([\"blastp\", \"-version\"], check=True, stdout=subprocess.DEVNULL)\n",
    "        return True\n",
    "    except (subprocess.CalledProcessError, FileNotFoundError):\n",
    "        return False\n",
    "\n",
    "\n",
    "def download_and_extract_blast(version: str, url: str):\n",
    "    \"\"\"Download and extract the BLAST+ tarball.\"\"\"\n",
    "    tarball_name = url.split(\"/\")[-1]\n",
    "    if not Path(tarball_name).exists():\n",
    "        print(f\"Downloading BLAST+ {version}...\")\n",
    "        subprocess.run([\"wget\", url], check=True)\n",
    "\n",
    "    print(\"Extracting BLAST+...\")\n",
    "    with tarfile.open(tarball_name, \"r:gz\") as tar:\n",
    "        tar.extractall(BLAST_DIR)\n",
    "    print(f\"BLAST extracted to: {BLAST_DIR.resolve()}\")\n",
    "\n",
    "\n",
    "def get_local_blastp_path(version: str):\n",
    "    \"\"\"Return the path to local blastp binary.\"\"\"\n",
    "    return BLAST_DIR / f\"ncbi-blast-{version}\" / \"bin\" / \"blastp\"\n",
    "\n",
    "\n",
    "def check_blast():\n",
    "    \"\"\"Ensure BLAST is installed. Return path to `blastp` binary.\"\"\"\n",
    "    if is_blast_installed():\n",
    "        print(\"System-wide BLAST is installed.\")\n",
    "        return shutil.which(\"blastp\")\n",
    "    else:\n",
    "        version, url = get_latest_version_url()\n",
    "        local_blastp = get_local_blastp_path(version)\n",
    "        if not local_blastp.exists():\n",
    "            print(f\"BLAST {version} not found locally. Installing...\")\n",
    "            BLAST_DIR.mkdir(exist_ok=True)\n",
    "            download_and_extract_blast(version, url)\n",
    "        else:\n",
    "            print(f\"Using already downloaded BLAST {version}.\")\n",
    "        return str(local_blastp)\n",
    "\n",
    "def make_blast_database(db_name: str, db_type: str = \"prot\", extension: str = \"xml\"):\n",
    "    \"\"\"Create a BLAST database from the Uniprot database.\"\"\"\n",
    "    db_path = os.path.join(DB_DIR, f\"{db_name}.{extension}\")\n",
    "    if not os.path.exists(db_path):\n",
    "        raise FileNotFoundError(f\"Database {db_name} not found at {db_path}. Please download it first.\")\n",
    "    \n",
    "    # Check if the database is already created\n",
    "    blast_db_path = os.path.join(DB_DIR, db_name)\n",
    "    extensions = [\".pdb\", \".phr\", \".pin\", \".psq\", \".pot\", \".psq\", \".ptf\", \".pto\"]\n",
    "    makedb = False\n",
    "    # For all extensions check if exists if there is one failing makedb again\n",
    "    for ext in extensions:\n",
    "        if not os.path.exists(blast_db_path + \"/db\" + ext):\n",
    "            makedb = True\n",
    "            break\n",
    "    if makedb:\n",
    "        print(f\"Creating BLAST database for {db_name}...\")\n",
    "        blast_db_cmd = [\n",
    "            \"makeblastdb\",\n",
    "            \"-in\", db_path,\n",
    "            \"-dbtype\", db_type,\n",
    "            \"-out\", os.path.join(DB_DIR, db_name) + \"/db\",\n",
    "        ]\n",
    "    \n",
    "        subprocess.run(blast_db_cmd, check=True)\n",
    "        print(f\"BLAST database created at: {os.path.join(DB_DIR, databases[db_name])}\")\n",
    "    else:\n",
    "        print(f\"BLAST database already exists at {blast_db_path}. No need to create it again.\")\n",
    "\n",
    "def run_blast(sequences: List[str], db_name: str, blast_type: str = \"blastp\", evalue: float = 0.001):\n",
    "    \"\"\"Run BLAST search.\"\"\"\n",
    "    blast_db_path = os.path.join(DB_DIR, db_name)\n",
    "    if not os.path.exists(blast_db_path):\n",
    "        raise FileNotFoundError(f\"Database {db_name} not found at {blast_db_path}. Please download it first.\")\n",
    "\n",
    "    # Make tmp directory if it does not exist\n",
    "    os.makedirs(\"tmp\", exist_ok=True)\n",
    "\n",
    "    # Write sequences to a temporary file\n",
    "    with open(\"tmp/sequences.fasta\", \"w\") as f:\n",
    "        for i, seq in enumerate(sequences):\n",
    "            f.write(f\">{i}\\n{seq}\\n\")\n",
    "    \n",
    "    blast_cmd = [\n",
    "        blast_type,\n",
    "        \"-query\", \"tmp/sequences.fasta\",\n",
    "        \"-db\", blast_db_path + \"/db\",\n",
    "        \"-outfmt\", \"6\",\n",
    "        \"-evalue\", str(evalue),\n",
    "    ]\n",
    "    \n",
    "    print(f\"Running BLAST search...\")\n",
    "    with open(\"tmp/blast_results.txt\", \"w\") as f:\n",
    "        subprocess.run(blast_cmd, stdout=f, check=True)\n",
    "    print(f\"BLAST results saved to tmp/blast_results.txt\")\n",
    "    # Clean up temporary file\n",
    "    os.remove(\"tmp/sequences.fasta\")\n",
    "\n",
    "def parse_blast_results(file_path: str, identity_threshold: float = 90.0):\n",
    "    \"\"\"Parse BLAST results from a file.\"\"\"\n",
    "    with open(file_path, \"r\") as f:\n",
    "        results = f.readlines()\n",
    "    \n",
    "    parsed_results = []\n",
    "    for line in results:\n",
    "        fields = line.strip().split(\"\\t\")\n",
    "        identity = float(fields[2])\n",
    "        if identity >= identity_threshold:\n",
    "            parsed_results.append({\n",
    "                \"query\": fields[0],\n",
    "                \"subject\": fields[1],\n",
    "                \"identity\": fields[2],\n",
    "                \"alignment_length\": fields[3],\n",
    "                \"evalue\": fields[4],\n",
    "                \"bit_score\": fields[5],\n",
    "            })\n",
    "    \n",
    "    return parsed_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "de905467",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Database uniprotkb_reviewed already exists at scripts/db/uniprotkb_reviewed.fasta.\n",
      "System-wide BLAST is installed.\n",
      "Using blastp at: /home/diego/micromamba/envs/bioseqdownloader/bin/blastp\n",
      "BLAST database already exists at scripts/db/uniprotkb_reviewed. No need to create it again.\n",
      "Running BLAST search...\n",
      "BLAST results saved to tmp/blast_results.txt\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequences</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MARPLLGKTSSVRRRLESLSACSIFFFLRKFCQKMASLVFLNSPVY...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MSFKVYDPIAELIATQFPTSNPDLQIINNDVLVVSPHKITLPMGPQ...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NYTETAQAIARSWRAGSHDRLKARGEAVAVTVHRLVAVPRGRDTPR...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           sequences  id\n",
       "0  MARPLLGKTSSVRRRLESLSACSIFFFLRKFCQKMASLVFLNSPVY...   0\n",
       "1  MSFKVYDPIAELIATQFPTSNPDLQIINNDVLVVSPHKITLPMGPQ...   1\n",
       "2  NYTETAQAIARSWRAGSHDRLKARGEAVAVTVHRLVAVPRGRDTPR...   2"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv(\"data/test.csv\")\n",
    "sequences = df[\"sequences\"].dropna().unique().tolist()\n",
    "    \n",
    "download_uniprot_database(\"uniprotkb_reviewed\", \"fasta\")\n",
    "    \n",
    "blastp_path = check_blast()\n",
    "print(f\"Using blastp at: {blastp_path}\")\n",
    "\n",
    "make_blast_database(\"uniprotkb_reviewed\", extension=\"fasta\")\n",
    "\n",
    "run_blast(sequences, \"uniprotkb_reviewed\", blast_type=\"blastp\", evalue=0.0001)\n",
    "\n",
    "results = parse_blast_results(\"tmp/blast_results.txt\")\n",
    "\n",
    "# Convert to DataFrame\n",
    "sequences_df = pd.DataFrame(sequences, columns=[\"sequences\"])\n",
    "sequences_df[\"id\"] = sequences_df.index\n",
    "\n",
    "sequences_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4009c945",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_blast = pd.DataFrame(results)\n",
    "\n",
    "df_blast = df_blast.rename(columns={\"query\": \"id\", \"subject\": \"subject_id\"})\n",
    "df_blast[\"id\"] = df_blast[\"id\"].astype(int)\n",
    "df_blast = df_blast.merge(sequences_df, on=\"id\", how=\"left\")\n",
    "df_blast = df_blast.drop(columns=[\"id\"])\n",
    "df_blast = df_blast.rename(columns={\"sequences\": \"sequence\"})\n",
    "\n",
    "# Separate subject into source, accession, entry_name\n",
    "df_blast[\"source\"] = df_blast[\"subject_id\"].apply(lambda x: x.split(\"|\")[0])\n",
    "df_blast[\"accession\"] = df_blast[\"subject_id\"].apply(lambda x: x.split(\"|\")[1])\n",
    "df_blast[\"entry_name\"] = df_blast[\"subject_id\"].apply(lambda x: x.split(\"|\")[2])\n",
    "df_blast = df_blast.drop(columns=[\"subject_id\"])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a32f8e17",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identity</th>\n",
       "      <th>alignment_length</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bit_score</th>\n",
       "      <th>sequence</th>\n",
       "      <th>source</th>\n",
       "      <th>accession</th>\n",
       "      <th>entry_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>100.000</td>\n",
       "      <td>438</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>MARPLLGKTSSVRRRLESLSACSIFFFLRKFCQKMASLVFLNSPVY...</td>\n",
       "      <td>sp</td>\n",
       "      <td>Q6GZX2</td>\n",
       "      <td>003R_FRG3G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>100.000</td>\n",
       "      <td>180</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>MSFKVYDPIAELIATQFPTSNPDLQIINNDVLVVSPHKITLPMGPQ...</td>\n",
       "      <td>sp</td>\n",
       "      <td>Q197F2</td>\n",
       "      <td>008L_IIV3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>100.000</td>\n",
       "      <td>50</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>MSFKVYDPIAELIATQFPTSNPDLQIINNDVLVVSPHKITLPMGPQ...</td>\n",
       "      <td>sp</td>\n",
       "      <td>Q6GZW6</td>\n",
       "      <td>009L_FRG3G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>100.000</td>\n",
       "      <td>345</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NYTETAQAIARSWRAGSHDRLKARGEAVAVTVHRLVAVPRGRDTPR...</td>\n",
       "      <td>sp</td>\n",
       "      <td>Q6GZW6</td>\n",
       "      <td>009L_FRG3G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  identity alignment_length evalue bit_score  \\\n",
       "0  100.000              438      0         0   \n",
       "1  100.000              180      0         0   \n",
       "2  100.000               50      0         0   \n",
       "3  100.000              345      0         0   \n",
       "\n",
       "                                            sequence source accession  \\\n",
       "0  MARPLLGKTSSVRRRLESLSACSIFFFLRKFCQKMASLVFLNSPVY...     sp    Q6GZX2   \n",
       "1  MSFKVYDPIAELIATQFPTSNPDLQIINNDVLVVSPHKITLPMGPQ...     sp    Q197F2   \n",
       "2  MSFKVYDPIAELIATQFPTSNPDLQIINNDVLVVSPHKITLPMGPQ...     sp    Q6GZW6   \n",
       "3  NYTETAQAIARSWRAGSHDRLKARGEAVAVTVHRLVAVPRGRDTPR...     sp    Q6GZW6   \n",
       "\n",
       "   entry_name  \n",
       "0  003R_FRG3G  \n",
       "1   008L_IIV3  \n",
       "2  009L_FRG3G  \n",
       "3  009L_FRG3G  "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_blast"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d944e22",
   "metadata": {},
   "source": [
    "## GO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5899c06f",
   "metadata": {},
   "outputs": [],
   "source": [
    "from src.description_go import *\n",
    "import os, ast\n",
    "import pandas as pd\n",
    "from tqdm import tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6b45cd2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[DESCRIPTION_GO] Getting Gen Ontology\n",
      "Docker version 28.0.0, build f9ced58158\n",
      "[DESCRIPTION_GO] Metastudent found.\n",
      "[DESCRIPTION_GO] Go terms found in input data.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>entry_type</th>\n",
       "      <th>protein_name</th>\n",
       "      <th>ec_numbers</th>\n",
       "      <th>organism</th>\n",
       "      <th>taxon_id</th>\n",
       "      <th>sequence</th>\n",
       "      <th>length</th>\n",
       "      <th>go_terms</th>\n",
       "      <th>pfam_ids</th>\n",
       "      <th>references</th>\n",
       "      <th>features</th>\n",
       "      <th>keywords</th>\n",
       "      <th>source_db</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [uniprot_id, entry_type, protein_name, ec_numbers, organism, taxon_id, sequence, length, go_terms, pfam_ids, references, features, keywords, source_db]\n",
       "Index: []"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DOCKER_IMAGE_NAME = \"metastudent\"\n",
    "DOCKER_CONTAINER_NAME = \"metastudent_container\"\n",
    "HOST_INPUT_FILE = os.path.abspath(\"tmp/sequences.fasta\")\n",
    "HOST_OUTPUT_DIR = os.path.abspath(\"tmp/\")\n",
    "CONTAINER_INPUT_FILE = \"/app/input.fasta\"\n",
    "CONTAINER_OUTPUT_FILE = \"/app/output.result\"\n",
    "\n",
    "\n",
    "print(\"[DESCRIPTION_GO] Getting Gen Ontology\")\n",
    "tqdm.pandas()\n",
    "\n",
    "if not check_dependencies(DOCKER_IMAGE_NAME):\n",
    "    print(\"[DESCRIPTION_GO] Metastudent not found. Installing...\")\n",
    "    install_dependencies(DOCKER_IMAGE_NAME)\n",
    "else:\n",
    "    print(\"[DESCRIPTION_GO] Metastudent found.\")\n",
    "\n",
    "input_df = pd.read_csv(\"results/umami_uniprot.csv\")\n",
    "obsolete_df = pd.read_csv(\"scripts/resources/amiGO_data.csv\", sep=\"\\t\", names=[\"id_go\", \"description\", \"is_obsolete\"])\n",
    "input_df['go_terms'] = input_df['go_terms'].apply(lambda x: ast.literal_eval(x) if isinstance(x, str) else x)\n",
    "\n",
    "parsed_df = pd.DataFrame()\n",
    "if os.path.isfile(f\"{HOST_OUTPUT_DIR}/output.BPO.txt\") and \\\n",
    "        os.path.isfile(f\"{HOST_OUTPUT_DIR}/output.CCO.txt\") and \\\n",
    "        os.path.isfile(f\"{HOST_OUTPUT_DIR}/output.MFO.txt\"):\n",
    "    print(\"[DESCRIPTION_GO] Metastudent results found.\")\n",
    "    parsed_df = parse_outputs(\"uniprot_id\")\n",
    "\n",
    "# Filter input_df with go_terms ~= null\n",
    "input_df_with_go_terms = input_df[input_df[\"go_terms\"].apply(lambda x: isinstance(x, list) and len(x) > 0)]\n",
    "input_df = input_df[input_df[\"go_terms\"].apply(lambda x: isinstance(x, list) and len(x) == 0)]\n",
    "\n",
    "if not input_df_with_go_terms.empty:\n",
    "    print(\"[DESCRIPTION_GO] Go terms found in input data.\")\n",
    "    input_df_with_go_terms = input_df_with_go_terms[[\"uniprot_id\", \"go_terms\"]]\n",
    "    input_df_with_go_terms = input_df_with_go_terms.explode(\"go_terms\")\n",
    "    parsed_df = pd.concat(\n",
    "        [\n",
    "            parsed_df,\n",
    "            pd.merge(\n",
    "                input_df_with_go_terms, \n",
    "                obsolete_df, \n",
    "                left_on=\"go_terms\", \n",
    "                right_on=\"id_go\", \n",
    "                how=\"left\"\n",
    "            )\n",
    "            .drop(columns=[\"go_terms\"])\n",
    "            .rename(columns={\"id_go\": \"go\"})  \n",
    "        ]\n",
    "    )\n",
    "\n",
    "input_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6a2a9d8e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[DESCRIPTION_GO] 0 sequences have not been processed.\n"
     ]
    }
   ],
   "source": [
    "if not parsed_df.empty:\n",
    "    # Check if all sequences have been processed\n",
    "    parsed_ids = parsed_df[\"uniprot_id\"].unique()\n",
    "    input_ids = input_df[\"uniprot_id\"].unique()\n",
    "    if len(parsed_ids) == len(input_ids):\n",
    "        print(\"[DESCRIPTION_GO] All sequences have been processed.\")\n",
    "        input_df = pd.DataFrame()\n",
    "    else:\n",
    "        input_df = input_df[~input_df[\"uniprot_id\"].isin(parsed_ids)]\n",
    "        print(f\"[DESCRIPTION_GO] {len(input_df)} sequences have not been processed.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d815e6cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.makedirs(HOST_OUTPUT_DIR, exist_ok=True)\n",
    "if not input_df.empty:\n",
    "    print(\"[DESCRIPTION_GO] Running in batches of 50...\")\n",
    "    for i in tqdm(range(0, len(input_df), 50)):\n",
    "        run_in_batches(input_df[i:i+50], HOST_OUTPUT_DIR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7cffac13",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File '/home/diego/Documents/PythonProjects/BioSeqDownloader/tmp/output.BPO.txt' not found.\n"
     ]
    },
    {
     "ename": "NameError",
     "evalue": "name 'exit' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mFileNotFoundError\u001b[39m                         Traceback (most recent call last)",
      "\u001b[36mFile \u001b[39m\u001b[32m~/Documents/PythonProjects/BioSeqDownloader/scripts/description_go.py:27\u001b[39m, in \u001b[36mparse_outputs\u001b[39m\u001b[34m(id_column)\u001b[39m\n\u001b[32m     26\u001b[39m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[32m---> \u001b[39m\u001b[32m27\u001b[39m     df_go = \u001b[43mpd\u001b[49m\u001b[43m.\u001b[49m\u001b[43mread_csv\u001b[49m\u001b[43m(\u001b[49m\n\u001b[32m     28\u001b[39m \u001b[43m            \u001b[49m\u001b[43mfile\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\n\u001b[32m     29\u001b[39m \u001b[43m            \u001b[49m\u001b[43mheader\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\n\u001b[32m     30\u001b[39m \u001b[43m            \u001b[49m\u001b[43msep\u001b[49m\u001b[43m=\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[38;5;130;43;01m\\t\u001b[39;49;00m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\n\u001b[32m     31\u001b[39m \u001b[43m            \u001b[49m\u001b[43mnames\u001b[49m\u001b[43m=\u001b[49m\u001b[43m[\u001b[49m\u001b[43mid_column\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mgo\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mprobability\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mterm\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m     32\u001b[39m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m     33\u001b[39m     df_go[\u001b[33m'\u001b[39m\u001b[33mgo_type\u001b[39m\u001b[33m'\u001b[39m] = file_type\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/micromamba/envs/bioseqdownloader/lib/python3.13/site-packages/pandas/io/parsers/readers.py:1026\u001b[39m, in \u001b[36mread_csv\u001b[39m\u001b[34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, date_format, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, encoding_errors, dialect, on_bad_lines, delim_whitespace, low_memory, memory_map, float_precision, storage_options, dtype_backend)\u001b[39m\n\u001b[32m   1024\u001b[39m kwds.update(kwds_defaults)\n\u001b[32m-> \u001b[39m\u001b[32m1026\u001b[39m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_read\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilepath_or_buffer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/micromamba/envs/bioseqdownloader/lib/python3.13/site-packages/pandas/io/parsers/readers.py:620\u001b[39m, in \u001b[36m_read\u001b[39m\u001b[34m(filepath_or_buffer, kwds)\u001b[39m\n\u001b[32m    619\u001b[39m \u001b[38;5;66;03m# Create the parser.\u001b[39;00m\n\u001b[32m--> \u001b[39m\u001b[32m620\u001b[39m parser = \u001b[43mTextFileReader\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilepath_or_buffer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m*\u001b[49m\u001b[43m*\u001b[49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    622\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m chunksize \u001b[38;5;129;01mor\u001b[39;00m iterator:\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/micromamba/envs/bioseqdownloader/lib/python3.13/site-packages/pandas/io/parsers/readers.py:1620\u001b[39m, in \u001b[36mTextFileReader.__init__\u001b[39m\u001b[34m(self, f, engine, **kwds)\u001b[39m\n\u001b[32m   1619\u001b[39m \u001b[38;5;28mself\u001b[39m.handles: IOHandles | \u001b[38;5;28;01mNone\u001b[39;00m = \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[32m-> \u001b[39m\u001b[32m1620\u001b[39m \u001b[38;5;28mself\u001b[39m._engine = \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43m_make_engine\u001b[49m\u001b[43m(\u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43mengine\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/micromamba/envs/bioseqdownloader/lib/python3.13/site-packages/pandas/io/parsers/readers.py:1880\u001b[39m, in \u001b[36mTextFileReader._make_engine\u001b[39m\u001b[34m(self, f, engine)\u001b[39m\n\u001b[32m   1879\u001b[39m         mode += \u001b[33m\"\u001b[39m\u001b[33mb\u001b[39m\u001b[33m\"\u001b[39m\n\u001b[32m-> \u001b[39m\u001b[32m1880\u001b[39m \u001b[38;5;28mself\u001b[39m.handles = \u001b[43mget_handle\u001b[49m\u001b[43m(\u001b[49m\n\u001b[32m   1881\u001b[39m \u001b[43m    \u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1882\u001b[39m \u001b[43m    \u001b[49m\u001b[43mmode\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1883\u001b[39m \u001b[43m    \u001b[49m\u001b[43mencoding\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43moptions\u001b[49m\u001b[43m.\u001b[49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mencoding\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1884\u001b[39m \u001b[43m    \u001b[49m\u001b[43mcompression\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43moptions\u001b[49m\u001b[43m.\u001b[49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mcompression\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1885\u001b[39m \u001b[43m    \u001b[49m\u001b[43mmemory_map\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43moptions\u001b[49m\u001b[43m.\u001b[49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mmemory_map\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1886\u001b[39m \u001b[43m    \u001b[49m\u001b[43mis_text\u001b[49m\u001b[43m=\u001b[49m\u001b[43mis_text\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1887\u001b[39m \u001b[43m    \u001b[49m\u001b[43merrors\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43moptions\u001b[49m\u001b[43m.\u001b[49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mencoding_errors\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mstrict\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1888\u001b[39m \u001b[43m    \u001b[49m\u001b[43mstorage_options\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43moptions\u001b[49m\u001b[43m.\u001b[49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mstorage_options\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1889\u001b[39m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m   1890\u001b[39m \u001b[38;5;28;01massert\u001b[39;00m \u001b[38;5;28mself\u001b[39m.handles \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/micromamba/envs/bioseqdownloader/lib/python3.13/site-packages/pandas/io/common.py:873\u001b[39m, in \u001b[36mget_handle\u001b[39m\u001b[34m(path_or_buf, mode, encoding, compression, memory_map, is_text, errors, storage_options)\u001b[39m\n\u001b[32m    871\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m ioargs.encoding \u001b[38;5;129;01mand\u001b[39;00m \u001b[33m\"\u001b[39m\u001b[33mb\u001b[39m\u001b[33m\"\u001b[39m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m ioargs.mode:\n\u001b[32m    872\u001b[39m     \u001b[38;5;66;03m# Encoding\u001b[39;00m\n\u001b[32m--> \u001b[39m\u001b[32m873\u001b[39m     handle = \u001b[38;5;28;43mopen\u001b[39;49m\u001b[43m(\u001b[49m\n\u001b[32m    874\u001b[39m \u001b[43m        \u001b[49m\u001b[43mhandle\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    875\u001b[39m \u001b[43m        \u001b[49m\u001b[43mioargs\u001b[49m\u001b[43m.\u001b[49m\u001b[43mmode\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    876\u001b[39m \u001b[43m        \u001b[49m\u001b[43mencoding\u001b[49m\u001b[43m=\u001b[49m\u001b[43mioargs\u001b[49m\u001b[43m.\u001b[49m\u001b[43mencoding\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    877\u001b[39m \u001b[43m        \u001b[49m\u001b[43merrors\u001b[49m\u001b[43m=\u001b[49m\u001b[43merrors\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    878\u001b[39m \u001b[43m        \u001b[49m\u001b[43mnewline\u001b[49m\u001b[43m=\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[32m    879\u001b[39m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    880\u001b[39m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[32m    881\u001b[39m     \u001b[38;5;66;03m# Binary mode\u001b[39;00m\n",
      "\u001b[31mFileNotFoundError\u001b[39m: [Errno 2] No such file or directory: '/home/diego/Documents/PythonProjects/BioSeqDownloader/tmp/output.BPO.txt'",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[31mNameError\u001b[39m                                 Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[23]\u001b[39m\u001b[32m, line 4\u001b[39m\n\u001b[32m      1\u001b[39m test = pd.concat(\n\u001b[32m      2\u001b[39m     [\n\u001b[32m      3\u001b[39m         parsed_df,\n\u001b[32m----> \u001b[39m\u001b[32m4\u001b[39m         \u001b[43mparse_outputs\u001b[49m\u001b[43m(\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43muniprot_id\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[32m      5\u001b[39m     ]\n\u001b[32m      6\u001b[39m )\n\u001b[32m      8\u001b[39m test = test.sort_values(by=\u001b[33m\"\u001b[39m\u001b[33muniprot_id\u001b[39m\u001b[33m\"\u001b[39m)\n\u001b[32m      9\u001b[39m test = test.merge(obsolete_df, left_on=\u001b[33m\"\u001b[39m\u001b[33mgo\u001b[39m\u001b[33m\"\u001b[39m, right_on=\u001b[33m\"\u001b[39m\u001b[33mid_go\u001b[39m\u001b[33m\"\u001b[39m, how=\u001b[33m\"\u001b[39m\u001b[33mleft\u001b[39m\u001b[33m\"\u001b[39m)\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/Documents/PythonProjects/BioSeqDownloader/scripts/description_go.py:40\u001b[39m, in \u001b[36mparse_outputs\u001b[39m\u001b[34m(id_column)\u001b[39m\n\u001b[32m     38\u001b[39m     \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mFileNotFoundError\u001b[39;00m:\n\u001b[32m     39\u001b[39m         \u001b[38;5;28mprint\u001b[39m(\u001b[33mf\u001b[39m\u001b[33m\"\u001b[39m\u001b[33mFile \u001b[39m\u001b[33m'\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mfile\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m'\u001b[39m\u001b[33m not found.\u001b[39m\u001b[33m\"\u001b[39m)\n\u001b[32m---> \u001b[39m\u001b[32m40\u001b[39m         \u001b[43mexit\u001b[49m(\u001b[32m1\u001b[39m)\n\u001b[32m     42\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m results:\n\u001b[32m     43\u001b[39m     \u001b[38;5;66;03m# df = pd.DataFrame({\u001b[39;00m\n\u001b[32m     44\u001b[39m     \u001b[38;5;66;03m#     \"id_seq\": [\"None\"],\u001b[39;00m\n\u001b[32m   (...)\u001b[39m\u001b[32m     48\u001b[39m     \u001b[38;5;66;03m#     \"go\": [\"No results found\"]\u001b[39;00m\n\u001b[32m     49\u001b[39m     \u001b[38;5;66;03m# })\u001b[39;00m\n\u001b[32m     50\u001b[39m     df = pd.DataFrame(columns=[id_column, \u001b[33m\"\u001b[39m\u001b[33mgo\u001b[39m\u001b[33m\"\u001b[39m, \u001b[33m\"\u001b[39m\u001b[33mprobability\u001b[39m\u001b[33m\"\u001b[39m, \u001b[33m\"\u001b[39m\u001b[33mterm\u001b[39m\u001b[33m\"\u001b[39m, \u001b[33m\"\u001b[39m\u001b[33mgo_type\u001b[39m\u001b[33m\"\u001b[39m])\n",
      "\u001b[31mNameError\u001b[39m: name 'exit' is not defined"
     ]
    }
   ],
   "source": [
    "test = pd.concat(\n",
    "    [\n",
    "        parsed_df,\n",
    "        parse_outputs(\"uniprot_id\")\n",
    "    ]\n",
    ")\n",
    "    \n",
    "test = test.sort_values(by=\"uniprot_id\")\n",
    "test = test.merge(obsolete_df, left_on=\"go\", right_on=\"id_go\", how=\"left\")\n",
    "test = test.drop(columns=[\"id_go\"])\n",
    "\n",
    "test"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "504746e9",
   "metadata": {},
   "source": [
    "## Uniprot query"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4a4b02aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "os.chdir(\"src\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec84c845",
   "metadata": {},
   "outputs": [],
   "source": [
    "query=\"organism_name:homo sapiens (human) AND length:[15 TO 30] AND reviewed:true\"\n",
    "fields=\"accession,protein_name,sequence,ec,lineage,organism_name,xref_pfam,xref_alphafolddb,xref_pdb,go_id\"\n",
    "sort=\"accession asc\"\n",
    "download=True\n",
    "format=\"json\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68f385ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "query=\"P05067\"\n",
    "fields=\"accession,xref_interpro\"\n",
    "sort=\"accession asc\"\n",
    "download=True\n",
    "format=\"json\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6f64e04b",
   "metadata": {},
   "outputs": [],
   "source": [
    "from src.uniprot import UniprotInterface\n",
    "\n",
    "uniprot = UniprotInterface()\n",
    "response = uniprot.submit_stream(\n",
    "    query=query,\n",
    "    fields=fields,\n",
    "    sort=sort,\n",
    "    include_isoform=True,\n",
    "    download=download,\n",
    "    format=format\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7caf7c38",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'results': [{'entryType': 'UniProtKB reviewed (Swiss-Prot)',\n",
       "   'primaryAccession': 'P05067',\n",
       "   'uniProtKBCrossReferences': [{'database': 'InterPro',\n",
       "     'id': 'IPR036669',\n",
       "     'properties': [{'key': 'EntryName', 'value': 'Amyloid_Cu-bd_sf'}]},\n",
       "    {'database': 'InterPro',\n",
       "     'id': 'IPR008155',\n",
       "     'properties': [{'key': 'EntryName', 'value': 'Amyloid_glyco'}]},\n",
       "    {'database': 'InterPro',\n",
       "     'id': 'IPR013803',\n",
       "     'properties': [{'key': 'EntryName', 'value': 'Amyloid_glyco_Abeta'}]},\n",
       "    {'database': 'InterPro',\n",
       "     'id': 'IPR037071',\n",
       "     'properties': [{'key': 'EntryName', 'value': 'Amyloid_glyco_Abeta_sf'}]},\n",
       "    {'database': 'InterPro',\n",
       "     'id': 'IPR011178',\n",
       "     'properties': [{'key': 'EntryName', 'value': 'Amyloid_glyco_Cu-bd'}]},\n",
       "    {'database': 'InterPro',\n",
       "     'id': 'IPR024329',\n",
       "     'properties': [{'key': 'EntryName', 'value': 'Amyloid_glyco_E2_domain'}]},\n",
       "    {'database': 'InterPro',\n",
       "     'id': 'IPR008154',\n",
       "     'properties': [{'key': 'EntryName', 'value': 'Amyloid_glyco_extra'}]},\n",
       "    {'database': 'InterPro',\n",
       "     'id': 'IPR015849',\n",
       "     'properties': [{'key': 'EntryName',\n",
       "       'value': 'Amyloid_glyco_heparin-bd'}]},\n",
       "    {'database': 'InterPro',\n",
       "     'id': 'IPR036454',\n",
       "     'properties': [{'key': 'EntryName',\n",
       "       'value': 'Amyloid_glyco_heparin-bd_sf'}]},\n",
       "    {'database': 'InterPro',\n",
       "     'id': 'IPR019745',\n",
       "     'properties': [{'key': 'EntryName',\n",
       "       'value': 'Amyloid_glyco_intracell_CS'}]},\n",
       "    {'database': 'InterPro',\n",
       "     'id': 'IPR019543',\n",
       "     'properties': [{'key': 'EntryName', 'value': 'APP_amyloid_C'}]},\n",
       "    {'database': 'InterPro',\n",
       "     'id': 'IPR019744',\n",
       "     'properties': [{'key': 'EntryName', 'value': 'APP_CUBD_CS'}]},\n",
       "    {'database': 'InterPro',\n",
       "     'id': 'IPR036176',\n",
       "     'properties': [{'key': 'EntryName', 'value': 'E2_sf'}]},\n",
       "    {'database': 'InterPro',\n",
       "     'id': 'IPR002223',\n",
       "     'properties': [{'key': 'EntryName', 'value': 'Kunitz_BPTI'}]},\n",
       "    {'database': 'InterPro',\n",
       "     'id': 'IPR036880',\n",
       "     'properties': [{'key': 'EntryName', 'value': 'Kunitz_BPTI_sf'}]},\n",
       "    {'database': 'InterPro',\n",
       "     'id': 'IPR011993',\n",
       "     'properties': [{'key': 'EntryName', 'value': 'PH-like_dom_sf'}]},\n",
       "    {'database': 'InterPro',\n",
       "     'id': 'IPR020901',\n",
       "     'properties': [{'key': 'EntryName', 'value': 'Prtase_inh_Kunz-CS'}]}],\n",
       "   'extraAttributes': {'uniParcId': 'UPI000002DB1C'}},\n",
       "  {'entryType': 'UniProtKB reviewed (Swiss-Prot)',\n",
       "   'primaryAccession': 'P05067-10',\n",
       "   'uniProtKBCrossReferences': [],\n",
       "   'extraAttributes': {'uniParcId': 'UPI00001BE539'}},\n",
       "  {'entryType': 'UniProtKB reviewed (Swiss-Prot)',\n",
       "   'primaryAccession': 'P05067-11',\n",
       "   'uniProtKBCrossReferences': [],\n",
       "   'extraAttributes': {'uniParcId': 'UPI00017A7091'}},\n",
       "  {'entryType': 'UniProtKB reviewed (Swiss-Prot)',\n",
       "   'primaryAccession': 'P05067-2',\n",
       "   'uniProtKBCrossReferences': [],\n",
       "   'extraAttributes': {'uniParcId': 'UPI000002A2F0'}},\n",
       "  {'entryType': 'UniProtKB reviewed (Swiss-Prot)',\n",
       "   'primaryAccession': 'P05067-3',\n",
       "   'uniProtKBCrossReferences': [],\n",
       "   'extraAttributes': {'uniParcId': 'UPI000002A2F1'}},\n",
       "  {'entryType': 'UniProtKB reviewed (Swiss-Prot)',\n",
       "   'primaryAccession': 'P05067-4',\n",
       "   'uniProtKBCrossReferences': [],\n",
       "   'extraAttributes': {'uniParcId': 'UPI000002A2F2'}},\n",
       "  {'entryType': 'UniProtKB reviewed (Swiss-Prot)',\n",
       "   'primaryAccession': 'P05067-5',\n",
       "   'uniProtKBCrossReferences': [],\n",
       "   'extraAttributes': {'uniParcId': 'UPI000002A2F3'}},\n",
       "  {'entryType': 'UniProtKB reviewed (Swiss-Prot)',\n",
       "   'primaryAccession': 'P05067-6',\n",
       "   'uniProtKBCrossReferences': [],\n",
       "   'extraAttributes': {'uniParcId': 'UPI000002A2F4'}},\n",
       "  {'entryType': 'UniProtKB reviewed (Swiss-Prot)',\n",
       "   'primaryAccession': 'P05067-7',\n",
       "   'uniProtKBCrossReferences': [],\n",
       "   'extraAttributes': {'uniParcId': 'UPI000002A2F5'}},\n",
       "  {'entryType': 'UniProtKB reviewed (Swiss-Prot)',\n",
       "   'primaryAccession': 'P05067-8',\n",
       "   'uniProtKBCrossReferences': [],\n",
       "   'extraAttributes': {'uniParcId': 'UPI000002A2F6'}},\n",
       "  {'entryType': 'UniProtKB reviewed (Swiss-Prot)',\n",
       "   'primaryAccession': 'P05067-9',\n",
       "   'uniProtKBCrossReferences': [],\n",
       "   'extraAttributes': {'uniParcId': 'UPI000002A2F7'}}]}"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "response.json()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "02c0f301",
   "metadata": {},
   "outputs": [],
   "source": [
    "parsed = uniprot.parse_stream_response(\n",
    "    query=query,\n",
    "    response=response\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8e8151b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query</th>\n",
       "      <th>accession</th>\n",
       "      <th>alphafold_ids</th>\n",
       "      <th>biogrid_ids</th>\n",
       "      <th>brenda_ids</th>\n",
       "      <th>go_terms</th>\n",
       "      <th>interpro_ids</th>\n",
       "      <th>kegg_ids</th>\n",
       "      <th>pdb_ids</th>\n",
       "      <th>pfam_ids</th>\n",
       "      <th>reactome_ids</th>\n",
       "      <th>refseq_ids</th>\n",
       "      <th>string_ids</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P05067</td>\n",
       "      <td>P05067</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[IPR036669, IPR008155, IPR013803, IPR037071, I...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>P05067</td>\n",
       "      <td>P05067-10</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>P05067</td>\n",
       "      <td>P05067-11</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>P05067</td>\n",
       "      <td>P05067-2</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>P05067</td>\n",
       "      <td>P05067-3</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>P05067</td>\n",
       "      <td>P05067-4</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>P05067</td>\n",
       "      <td>P05067-5</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>P05067</td>\n",
       "      <td>P05067-6</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>P05067</td>\n",
       "      <td>P05067-7</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>P05067</td>\n",
       "      <td>P05067-8</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>P05067</td>\n",
       "      <td>P05067-9</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     query  accession alphafold_ids biogrid_ids brenda_ids go_terms  \\\n",
       "0   P05067     P05067            []          []         []       []   \n",
       "1   P05067  P05067-10          None        None       None     None   \n",
       "2   P05067  P05067-11          None        None       None     None   \n",
       "3   P05067   P05067-2          None        None       None     None   \n",
       "4   P05067   P05067-3          None        None       None     None   \n",
       "5   P05067   P05067-4          None        None       None     None   \n",
       "6   P05067   P05067-5          None        None       None     None   \n",
       "7   P05067   P05067-6          None        None       None     None   \n",
       "8   P05067   P05067-7          None        None       None     None   \n",
       "9   P05067   P05067-8          None        None       None     None   \n",
       "10  P05067   P05067-9          None        None       None     None   \n",
       "\n",
       "                                         interpro_ids kegg_ids pdb_ids  \\\n",
       "0   [IPR036669, IPR008155, IPR013803, IPR037071, I...       []      []   \n",
       "1                                                None     None    None   \n",
       "2                                                None     None    None   \n",
       "3                                                None     None    None   \n",
       "4                                                None     None    None   \n",
       "5                                                None     None    None   \n",
       "6                                                None     None    None   \n",
       "7                                                None     None    None   \n",
       "8                                                None     None    None   \n",
       "9                                                None     None    None   \n",
       "10                                               None     None    None   \n",
       "\n",
       "   pfam_ids reactome_ids refseq_ids string_ids  \n",
       "0        []           []         []         []  \n",
       "1      None         None       None       None  \n",
       "2      None         None       None       None  \n",
       "3      None         None       None       None  \n",
       "4      None         None       None       None  \n",
       "5      None         None       None       None  \n",
       "6      None         None       None       None  \n",
       "7      None         None       None       None  \n",
       "8      None         None       None       None  \n",
       "9      None         None       None       None  \n",
       "10     None         None       None       None  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "parsed"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "301959e5",
   "metadata": {},
   "source": [
    "## Activity search"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "06c29588",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59f9589e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['uniprot_celiac-toxic.csv',\n",
       " 'uniprot_embryotoxic.csv',\n",
       " 'uniprot_ace-inhibitor.csv',\n",
       " 'uniprot_anuran-defense.csv',\n",
       " 'uniprot_campde-inhibitor.csv',\n",
       " 'uniprot_anti-neurotensive.csv',\n",
       " 'uniprot_antitrypanosomic.csv',\n",
       " 'uniprot_anticancer.csv',\n",
       " 'uniprot_anorectic.csv',\n",
       " 'uniprot_chemotactic.csv',\n",
       " 'uniprot_targeting-GP.csv',\n",
       " 'uniprot_Blood-Brain-Barrier.csv',\n",
       " 'uniprot_antitumor.csv',\n",
       " 'uniprot_cytotoxic.csv',\n",
       " 'uniprot_activating-ubiquitin-mediated-proteolysis.csv',\n",
       " 'uniprot_protein-kinase-c-inhibitor.csv',\n",
       " 'uniprot_antihiv.csv',\n",
       " 'uniprot_antidiabetic.csv',\n",
       " 'uniprot_calpain-2-inhibitor.csv',\n",
       " 'uniprot_antileishmania.csv',\n",
       " 'uniprot_inhibitor.csv',\n",
       " 'uniprot_antimicrobial.csv',\n",
       " 'uniprot_antituberculosis.csv',\n",
       " 'uniprot_antiviral.csv',\n",
       " 'uniprot_wound-healing.csv',\n",
       " 'uniprot_targeting-GN.csv',\n",
       " 'uniprot_hmg-coa-reductase-inhibitor.csv',\n",
       " 'uniprot_opioid-agonist.csv',\n",
       " 'uniprot_toxicology.csv',\n",
       " 'uniprot_antibacterial.csv',\n",
       " 'uniprot_antibiofilm.csv',\n",
       " 'uniprot_hypocholesterolemic.csv',\n",
       " 'uniprot_bacterial-permease-ligand.csv',\n",
       " 'uniprot_protease-inhibitor.csv',\n",
       " 'uniprot_antitoxin.csv',\n",
       " 'uniprot_polystyrene-surface-binding.csv',\n",
       " 'uniprot_stimulating.csv',\n",
       " 'uniprot_immunomodulating.csv',\n",
       " 'uniprot_opioid.csv',\n",
       " 'uniprot_dipeptidyl-peptidase-iii-inhibitor.csv',\n",
       " 'uniprot_tumor-homing.csv',\n",
       " 'uniprot_heparin-binding.csv',\n",
       " 'uniprot_anticoronavirus.csv',\n",
       " 'uniprot_targeting-mammals.csv',\n",
       " 'uniprot_antiplasmodial.csv',\n",
       " 'uniprot_uterotonic.csv',\n",
       " 'uniprot_cell-cell-comunication.csv',\n",
       " 'uniprot_binding.csv',\n",
       " 'uniprot_peptide-hormones.csv',\n",
       " 'uniprot_drug-delivery.csv',\n",
       " 'uniprot_anti-MRSA.csv',\n",
       " 'uniprot_antihypertensive.csv',\n",
       " 'uniprot_renin-inhibitor.csv',\n",
       " 'uniprot_neuroactive.csv',\n",
       " 'uniprot_surface-binding.csv',\n",
       " 'uniprot_antiantifreeze.csv',\n",
       " 'uniprot_insecticidal.csv',\n",
       " 'uniprot_antiprotozoal.csv',\n",
       " 'uniprot_alpha-amylase-inhibitor.csv',\n",
       " 'uniprot_antiparasitic.csv',\n",
       " 'uniprot_alpha-glucosidase-inhibitor.csv',\n",
       " 'uniprot_neuropeptide.csv',\n",
       " 'uniprot_trypsin-inhibitory.csv',\n",
       " 'uniprot_orphan-receptor-gpr14-agonist.csv',\n",
       " 'uniprot_antithrombotic.csv',\n",
       " 'uniprot_antiamnestic.csv',\n",
       " 'uniprot_antifungal.csv',\n",
       " 'uniprot_antiangiogenic.csv',\n",
       " 'uniprot_antioxidative.csv',\n",
       " 'uniprot_natriuretic.csv',\n",
       " 'uniprot_cell-penetrating.csv',\n",
       " 'uniprot_quorum-sensing.csv',\n",
       " 'uniprot_dipeptidyl-peptidaseiv.csv',\n",
       " 'uniprot_regulating.csv',\n",
       " 'uniprot_antioncogenic.csv',\n",
       " 'uniprot_contracting.csv',\n",
       " 'uniprot_immunostimulating.csv',\n",
       " 'uniprot_membrane-active-peptide.csv',\n",
       " 'uniprot_hypolipidemic.csv',\n",
       " 'uniprot_antiinflammatory.csv',\n",
       " 'uniprot_vasoconstrictor.csv',\n",
       " 'uniprot_opioid-antagonist.csv',\n",
       " 'uniprot_targeting-GV.csv',\n",
       " 'uniprot_hypotensive.csv',\n",
       " 'uniprot_edema-inducer.csv',\n",
       " 'uniprot_antiendotoxin.csv',\n",
       " 'uniprot_hemolytic.csv',\n",
       " 'uniprot_camkii-inhibitor.csv',\n",
       " 'uniprot_cytokine-growth-factor.csv']"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "uniprot_search_files = os.listdir(\"uniprot_search\")\n",
    "uniprot_search_files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ecb7ce2c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>activity</th>\n",
       "      <th>sequence_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Binding</td>\n",
       "      <td>411281</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Inhibitor</td>\n",
       "      <td>18950</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Surface binding</td>\n",
       "      <td>15663</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Regulating</td>\n",
       "      <td>11361</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Antimicrobial</td>\n",
       "      <td>5945</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>Dipeptidyl peptidaseiv</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>Antiamnestic</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>Edema inducer</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>Antiendotoxin</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>Total</td>\n",
       "      <td>502774</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>90 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  activity  sequence_count\n",
       "0                  Binding          411281\n",
       "1                Inhibitor           18950\n",
       "2          Surface binding           15663\n",
       "3               Regulating           11361\n",
       "4            Antimicrobial            5945\n",
       "..                     ...             ...\n",
       "85  Dipeptidyl peptidaseiv               0\n",
       "86            Antiamnestic               0\n",
       "87           Edema inducer               0\n",
       "88           Antiendotoxin               0\n",
       "89                   Total          502774\n",
       "\n",
       "[90 rows x 2 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create an empty list to store the results\n",
    "activity_data = []\n",
    "\n",
    "# Iterate over each file in uniprot_search_files\n",
    "for file in uniprot_search_files:\n",
    "    # Load the CSV file into a DataFrame\n",
    "    df = pd.read_csv(f\"uniprot_search/{file}\")\n",
    "    \n",
    "    # Extract the activity name from the file name\n",
    "    activity = file.replace(\"uniprot_\", \"\").replace(\".csv\", \"\").replace(\"-\", \" \").capitalize()\n",
    "    \n",
    "    # Count the number of sequences in the DataFrame\n",
    "    sequence_count = len(df)\n",
    "    \n",
    "    # Append the activity and sequence count to the list\n",
    "    activity_data.append({\"activity\": activity, \"sequence_count\": sequence_count})\n",
    "\n",
    "# Create a DataFrame from the activity data\n",
    "activity_df = pd.DataFrame(activity_data)\n",
    "\n",
    "# Sort the DataFrame by sequence count in descending order\n",
    "activity_df = activity_df.sort_values(by=\"sequence_count\", ascending=False)\n",
    "\n",
    "# Calculate the total number of sequences\n",
    "total_sequences = activity_df[\"sequence_count\"].sum()\n",
    "\n",
    "# Add a row for the total sequences\n",
    "activity_df = pd.concat([activity_df, pd.DataFrame([{\"activity\": \"Total\", \"sequence_count\": total_sequences}])], ignore_index=True)\n",
    "\n",
    "# Display the resulting DataFrame\n",
    "activity_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4abc5ec2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>activity</th>\n",
       "      <th>sequence_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Antimicrobial</td>\n",
       "      <td>5945</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Antiviral</td>\n",
       "      <td>3835</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Antibacterial</td>\n",
       "      <td>2366</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Antifungal</td>\n",
       "      <td>1209</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Antitoxin</td>\n",
       "      <td>832</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Antitumor</td>\n",
       "      <td>479</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Anticancer</td>\n",
       "      <td>393</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>Antiparasitic</td>\n",
       "      <td>140</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>Antiangiogenic</td>\n",
       "      <td>86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>Antioxidative</td>\n",
       "      <td>55</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>Antithrombotic</td>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>Antidiabetic</td>\n",
       "      <td>42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>Antiprotozoal</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>Anti mrsa</td>\n",
       "      <td>31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>Antiinflammatory</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>Antihypertensive</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>Antituberculosis</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>Antibiofilm</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>Antiplasmodial</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>Antioncogenic</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>Antileishmania</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>Antihiv</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>Antitrypanosomic</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>Anti neurotensive</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>Antiantifreeze</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>Anticoronavirus</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>Antiamnestic</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>Antiendotoxin</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             activity  sequence_count\n",
       "4       Antimicrobial            5945\n",
       "6           Antiviral            3835\n",
       "10      Antibacterial            2366\n",
       "17         Antifungal            1209\n",
       "19          Antitoxin             832\n",
       "24          Antitumor             479\n",
       "25         Anticancer             393\n",
       "32      Antiparasitic             140\n",
       "37     Antiangiogenic              86\n",
       "41      Antioxidative              55\n",
       "42     Antithrombotic              50\n",
       "45       Antidiabetic              42\n",
       "48      Antiprotozoal              33\n",
       "49          Anti mrsa              31\n",
       "52   Antiinflammatory              27\n",
       "56   Antihypertensive              21\n",
       "60   Antituberculosis              15\n",
       "66        Antibiofilm               9\n",
       "67     Antiplasmodial               8\n",
       "71      Antioncogenic               2\n",
       "73     Antileishmania               0\n",
       "74            Antihiv               0\n",
       "79   Antitrypanosomic               0\n",
       "80  Anti neurotensive               0\n",
       "81     Antiantifreeze               0\n",
       "83    Anticoronavirus               0\n",
       "86       Antiamnestic               0\n",
       "88      Antiendotoxin               0"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "anti_activities_df = activity_df[activity_df['activity'].str.startswith('Anti', na=False)]\n",
    "anti_activities_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3ae4e212",
   "metadata": {},
   "source": [
    "## Alphafold"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "ea2eb7a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "from src.alphafold import AlphafoldInterface"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "6bbf0e69",
   "metadata": {},
   "outputs": [],
   "source": [
    "instance = AlphafoldInterface(\n",
    "    structures=['pdb'],\n",
    "    output_dir=\"results\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "c527d645",
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "Unknown method 'None'",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mValueError\u001b[39m                                Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[8]\u001b[39m\u001b[32m, line 1\u001b[39m\n\u001b[32m----> \u001b[39m\u001b[32m1\u001b[39m \u001b[38;5;28mprint\u001b[39m(\u001b[43minstance\u001b[49m\u001b[43m.\u001b[49m\u001b[43mquery_usage\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m)\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/Documents/PythonProjects/BioSeqDownloader/src/alphafold.py:249\u001b[39m, in \u001b[36mAlphafoldInterface.query_usage\u001b[39m\u001b[34m(self)\u001b[39m\n\u001b[32m    229\u001b[39m \u001b[38;5;250m\u001b[39m\u001b[33;03m\"\"\"\u001b[39;00m\n\u001b[32m    230\u001b[39m \u001b[33;03mGet usage information for the Alphafold API.\u001b[39;00m\n\u001b[32m    231\u001b[39m \u001b[33;03mReturns:\u001b[39;00m\n\u001b[32m    232\u001b[39m \u001b[33;03m    str: Usage information.\u001b[39;00m\n\u001b[32m    233\u001b[39m \u001b[33;03m\"\"\"\u001b[39;00m\n\u001b[32m    234\u001b[39m usage = \u001b[33m\"\"\"\u001b[39m\u001b[33mUsage: To fetch predictions, use the UniProt ID as the query.\u001b[39m\n\u001b[32m    235\u001b[39m \u001b[33mExample: \u001b[39m\n\u001b[32m    236\u001b[39m \u001b[33m    - fetch_single(\u001b[39m\u001b[33m\"\u001b[39m\u001b[33mP02666\u001b[39m\u001b[33m\"\u001b[39m\u001b[33m, parse=True)\u001b[39m\n\u001b[32m   (...)\u001b[39m\u001b[32m    247\u001b[39m \u001b[33m    - bcif: Binary Crystallographic Information File format\u001b[39m\n\u001b[32m    248\u001b[39m \u001b[33m\u001b[39m\u001b[33m\"\"\"\u001b[39m\n\u001b[32m--> \u001b[39m\u001b[32m249\u001b[39m dummy = \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43mget_dummy\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    251\u001b[39m usage += \u001b[33m\"\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[33mExample fields in the response:\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[33m\"\u001b[39m\n\u001b[32m    252\u001b[39m \u001b[38;5;28;01mfor\u001b[39;00m key \u001b[38;5;129;01min\u001b[39;00m dummy.keys():\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/Documents/PythonProjects/BioSeqDownloader/src/alphafold.py:223\u001b[39m, in \u001b[36mAlphafoldInterface.get_dummy\u001b[39m\u001b[34m(self, **kwargs)\u001b[39m\n\u001b[32m    215\u001b[39m \u001b[38;5;250m\u001b[39m\u001b[33;03m\"\"\"\u001b[39;00m\n\u001b[32m    216\u001b[39m \u001b[33;03mGet a dummy response.\u001b[39;00m\n\u001b[32m    217\u001b[39m \u001b[33;03mUseful for knowing the structure of the data returned by the API.\u001b[39;00m\n\u001b[32m    218\u001b[39m \u001b[33;03mReturns:\u001b[39;00m\n\u001b[32m    219\u001b[39m \u001b[33;03m    Dict: Dummy response with example fields.\u001b[39;00m\n\u001b[32m    220\u001b[39m \u001b[33;03m\"\"\"\u001b[39;00m\n\u001b[32m    221\u001b[39m parse = kwargs.get(\u001b[33m\"\u001b[39m\u001b[33mparse\u001b[39m\u001b[33m\"\u001b[39m, \u001b[38;5;28;01mFalse\u001b[39;00m)\n\u001b[32m--> \u001b[39m\u001b[32m223\u001b[39m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43msuper\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[43m.\u001b[49m\u001b[43mget_dummy\u001b[49m\u001b[43m(\u001b[49m\n\u001b[32m    224\u001b[39m \u001b[43m    \u001b[49m\u001b[43mquery\u001b[49m\u001b[43m=\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mP02666\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[32m    225\u001b[39m \u001b[43m    \u001b[49m\u001b[43mparse\u001b[49m\u001b[43m=\u001b[49m\u001b[43mparse\u001b[49m\n\u001b[32m    226\u001b[39m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/Documents/PythonProjects/BioSeqDownloader/src/base.py:842\u001b[39m, in \u001b[36mBaseAPIInterface.get_dummy\u001b[39m\u001b[34m(self, *args, **kwargs)\u001b[39m\n\u001b[32m    839\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m query:\n\u001b[32m    840\u001b[39m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\u001b[33m\"\u001b[39m\u001b[33mQuery must be provided to get dummy data.\u001b[39m\u001b[33m\"\u001b[39m)\n\u001b[32m--> \u001b[39m\u001b[32m842\u001b[39m response = \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43mfetch_single\u001b[49m\u001b[43m(\u001b[49m\u001b[43mquery\u001b[49m\u001b[43m=\u001b[49m\u001b[43mquery\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mparse\u001b[49m\u001b[43m=\u001b[49m\u001b[43mparse\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m*\u001b[49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m*\u001b[49m\u001b[43m*\u001b[49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    843\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(response, \u001b[38;5;28mlist\u001b[39m):\n\u001b[32m    844\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m get_feature_keys(response[\u001b[32m0\u001b[39m] \u001b[38;5;28;01mif\u001b[39;00m response \u001b[38;5;28;01melse\u001b[39;00m {})\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/Documents/PythonProjects/BioSeqDownloader/src/alphafold.py:62\u001b[39m, in \u001b[36mAlphafoldInterface.fetch_single\u001b[39m\u001b[34m(self, query, parse, *args, **kwargs)\u001b[39m\n\u001b[32m     58\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(query, \u001b[38;5;28mstr\u001b[39m):\n\u001b[32m     59\u001b[39m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\u001b[33m\"\u001b[39m\u001b[33mQuery must be a string representing a UniProt ID.\u001b[39m\u001b[33m\"\u001b[39m)\n\u001b[32m---> \u001b[39m\u001b[32m62\u001b[39m result = \u001b[38;5;28;43msuper\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[43m.\u001b[49m\u001b[43mfetch_single\u001b[49m\u001b[43m(\u001b[49m\u001b[43mquery\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mparse\u001b[49m\u001b[43m=\u001b[49m\u001b[43mparse\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m*\u001b[49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m*\u001b[49m\u001b[43m*\u001b[49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m     64\u001b[39m new_result = {}\n\u001b[32m     65\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m.structures:\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/Documents/PythonProjects/BioSeqDownloader/src/base.py:683\u001b[39m, in \u001b[36mBaseAPIInterface.fetch_single\u001b[39m\u001b[34m(self, query, parse, *args, **kwargs)\u001b[39m\n\u001b[32m    680\u001b[39m option       = kwargs.get(\u001b[33m\"\u001b[39m\u001b[33moption\u001b[39m\u001b[33m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m)\n\u001b[32m    682\u001b[39m \u001b[38;5;66;03m# Get method specification\u001b[39;00m\n\u001b[32m--> \u001b[39m\u001b[32m683\u001b[39m spec = \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43m_get_method_spec\u001b[49m\u001b[43m(\u001b[49m\u001b[43m*\u001b[49m\u001b[43m*\u001b[49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    684\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m spec \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[32m    685\u001b[39m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\u001b[33mf\u001b[39m\u001b[33m\"\u001b[39m\u001b[33mMethod \u001b[39m\u001b[33m'\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mmethod\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m'\u001b[39m\u001b[33m is not supported. Available: \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mlist\u001b[39m(\u001b[38;5;28mself\u001b[39m.METHODS)\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m\"\u001b[39m)\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/Documents/PythonProjects/BioSeqDownloader/src/base.py:312\u001b[39m, in \u001b[36mBaseAPIInterface._get_method_spec\u001b[39m\u001b[34m(self, **kwargs)\u001b[39m\n\u001b[32m    310\u001b[39m method = kwargs.get(\u001b[33m\"\u001b[39m\u001b[33mmethod\u001b[39m\u001b[33m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m)\n\u001b[32m    311\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m method \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m.METHODS:\n\u001b[32m--> \u001b[39m\u001b[32m312\u001b[39m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\u001b[33mf\u001b[39m\u001b[33m\"\u001b[39m\u001b[33mUnknown method \u001b[39m\u001b[33m'\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mmethod\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m'\u001b[39m\u001b[33m\"\u001b[39m)\n\u001b[32m    313\u001b[39m option = kwargs.get(\u001b[33m\"\u001b[39m\u001b[33moption\u001b[39m\u001b[33m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m)\n\u001b[32m    315\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m option:\n",
      "\u001b[31mValueError\u001b[39m: Unknown method 'None'"
     ]
    }
   ],
   "source": [
    "print(instance.query_usage())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "137a29dc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cache_key: P02666_prediction\n",
      "Prepared request: https://alphafold.ebi.ac.uk/api/prediction/P02666\n",
      "Error fetching prediction for {'qualifier': 'P02666'}: HTTPSConnectionPool(host='alphafold.ebi.ac.uk', port=443): Max retries exceeded with url: /api/prediction/P02666 (Caused by ResponseError('too many 500 error responses'))\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>entry</th>\n",
       "      <th>gene</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>is_reviewed</th>\n",
       "      <th>is_reference</th>\n",
       "      <th>pdbUrl</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  entry  gene tax_id organism is_reviewed is_reference pdbUrl\n",
       "0  None  None   None     None        None         None   None"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "instance.fetch_single(\n",
    "    query=\"P02666\",\n",
    "    method=\"prediction\",\n",
    "    parse=True,\n",
    "    to_dataframe=True,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b5dbf5ff",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cache_key: P02666\n",
      "Cache_key: Q9TSI0\n",
      "Cache_key: P33048\n",
      "Cache_key: P11839\n",
      "Cache_key: O15552\n",
      "Cache_key: P76011\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>entryId</th>\n",
       "      <th>gene</th>\n",
       "      <th>sequenceChecksum</th>\n",
       "      <th>sequenceVersionDate</th>\n",
       "      <th>uniprotAccession</th>\n",
       "      <th>uniprotId</th>\n",
       "      <th>uniprotDescription</th>\n",
       "      <th>taxId</th>\n",
       "      <th>organismScientificName</th>\n",
       "      <th>uniprotStart</th>\n",
       "      <th>...</th>\n",
       "      <th>bcifUrl</th>\n",
       "      <th>cifUrl</th>\n",
       "      <th>pdbUrl</th>\n",
       "      <th>paeImageUrl</th>\n",
       "      <th>paeDocUrl</th>\n",
       "      <th>amAnnotationsUrl</th>\n",
       "      <th>amAnnotationsHg19Url</th>\n",
       "      <th>amAnnotationsHg38Url</th>\n",
       "      <th>isReviewed</th>\n",
       "      <th>isReferenceProteome</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AF-P02666-F1</td>\n",
       "      <td>CSN2</td>\n",
       "      <td>F0BBDD8148A238AE</td>\n",
       "      <td>1989-07-01</td>\n",
       "      <td>P02666</td>\n",
       "      <td>CASB_BOVIN</td>\n",
       "      <td>Beta-casein</td>\n",
       "      <td>9913</td>\n",
       "      <td>Bos taurus</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>https://alphafold.ebi.ac.uk/files/AF-P02666-F1...</td>\n",
       "      <td>https://alphafold.ebi.ac.uk/files/AF-P02666-F1...</td>\n",
       "      <td>https://alphafold.ebi.ac.uk/files/AF-P02666-F1...</td>\n",
       "      <td>https://alphafold.ebi.ac.uk/files/AF-P02666-F1...</td>\n",
       "      <td>https://alphafold.ebi.ac.uk/files/AF-P02666-F1...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AF-Q9TSI0-F1</td>\n",
       "      <td>CSN2</td>\n",
       "      <td>14FD3687DD17C5A9</td>\n",
       "      <td>2000-05-01</td>\n",
       "      <td>Q9TSI0</td>\n",
       "      <td>CASB_BUBBU</td>\n",
       "      <td>Beta-casein</td>\n",
       "      <td>89462</td>\n",
       "      <td>Bubalus bubalis</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>https://alphafold.ebi.ac.uk/files/AF-Q9TSI0-F1...</td>\n",
       "      <td>https://alphafold.ebi.ac.uk/files/AF-Q9TSI0-F1...</td>\n",
       "      <td>https://alphafold.ebi.ac.uk/files/AF-Q9TSI0-F1...</td>\n",
       "      <td>https://alphafold.ebi.ac.uk/files/AF-Q9TSI0-F1...</td>\n",
       "      <td>https://alphafold.ebi.ac.uk/files/AF-Q9TSI0-F1...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AF-P33048-F1</td>\n",
       "      <td>CSN2</td>\n",
       "      <td>96AE17746A01CD05</td>\n",
       "      <td>1993-10-01</td>\n",
       "      <td>P33048</td>\n",
       "      <td>CASB_CAPHI</td>\n",
       "      <td>Beta-casein</td>\n",
       "      <td>9925</td>\n",
       "      <td>Capra hircus</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>https://alphafold.ebi.ac.uk/files/AF-P33048-F1...</td>\n",
       "      <td>https://alphafold.ebi.ac.uk/files/AF-P33048-F1...</td>\n",
       "      <td>https://alphafold.ebi.ac.uk/files/AF-P33048-F1...</td>\n",
       "      <td>https://alphafold.ebi.ac.uk/files/AF-P33048-F1...</td>\n",
       "      <td>https://alphafold.ebi.ac.uk/files/AF-P33048-F1...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AF-P11839-F1</td>\n",
       "      <td>CSN2</td>\n",
       "      <td>061B4424DCB49EB1</td>\n",
       "      <td>1995-11-01</td>\n",
       "      <td>P11839</td>\n",
       "      <td>CASB_SHEEP</td>\n",
       "      <td>Beta-casein</td>\n",
       "      <td>9940</td>\n",
       "      <td>Ovis aries</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>https://alphafold.ebi.ac.uk/files/AF-P11839-F1...</td>\n",
       "      <td>https://alphafold.ebi.ac.uk/files/AF-P11839-F1...</td>\n",
       "      <td>https://alphafold.ebi.ac.uk/files/AF-P11839-F1...</td>\n",
       "      <td>https://alphafold.ebi.ac.uk/files/AF-P11839-F1...</td>\n",
       "      <td>https://alphafold.ebi.ac.uk/files/AF-P11839-F1...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AF-O15552-F1</td>\n",
       "      <td>FFAR2</td>\n",
       "      <td>F4A8AC6AFBDF1E90</td>\n",
       "      <td>1998-01-01</td>\n",
       "      <td>O15552</td>\n",
       "      <td>FFAR2_HUMAN</td>\n",
       "      <td>Free fatty acid receptor 2</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>https://alphafold.ebi.ac.uk/files/AF-O15552-F1...</td>\n",
       "      <td>https://alphafold.ebi.ac.uk/files/AF-O15552-F1...</td>\n",
       "      <td>https://alphafold.ebi.ac.uk/files/AF-O15552-F1...</td>\n",
       "      <td>https://alphafold.ebi.ac.uk/files/AF-O15552-F1...</td>\n",
       "      <td>https://alphafold.ebi.ac.uk/files/AF-O15552-F1...</td>\n",
       "      <td>https://alphafold.ebi.ac.uk/files/AF-O15552-F1...</td>\n",
       "      <td>https://alphafold.ebi.ac.uk/files/AF-O15552-F1...</td>\n",
       "      <td>https://alphafold.ebi.ac.uk/files/AF-O15552-F1...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>AF-P76011-F1</td>\n",
       "      <td>ymgE</td>\n",
       "      <td>78ED929220264E62</td>\n",
       "      <td>2001-07-11</td>\n",
       "      <td>P76011</td>\n",
       "      <td>YMGE_ECOLI</td>\n",
       "      <td>UPF0410 protein YmgE</td>\n",
       "      <td>83333</td>\n",
       "      <td>Escherichia coli (strain K12)</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>https://alphafold.ebi.ac.uk/files/AF-P76011-F1...</td>\n",
       "      <td>https://alphafold.ebi.ac.uk/files/AF-P76011-F1...</td>\n",
       "      <td>https://alphafold.ebi.ac.uk/files/AF-P76011-F1...</td>\n",
       "      <td>https://alphafold.ebi.ac.uk/files/AF-P76011-F1...</td>\n",
       "      <td>https://alphafold.ebi.ac.uk/files/AF-P76011-F1...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        entryId   gene  sequenceChecksum sequenceVersionDate uniprotAccession  \\\n",
       "0  AF-P02666-F1   CSN2  F0BBDD8148A238AE          1989-07-01           P02666   \n",
       "1  AF-Q9TSI0-F1   CSN2  14FD3687DD17C5A9          2000-05-01           Q9TSI0   \n",
       "2  AF-P33048-F1   CSN2  96AE17746A01CD05          1993-10-01           P33048   \n",
       "3  AF-P11839-F1   CSN2  061B4424DCB49EB1          1995-11-01           P11839   \n",
       "4  AF-O15552-F1  FFAR2  F4A8AC6AFBDF1E90          1998-01-01           O15552   \n",
       "5  AF-P76011-F1   ymgE  78ED929220264E62          2001-07-11           P76011   \n",
       "\n",
       "     uniprotId          uniprotDescription  taxId  \\\n",
       "0   CASB_BOVIN                 Beta-casein   9913   \n",
       "1   CASB_BUBBU                 Beta-casein  89462   \n",
       "2   CASB_CAPHI                 Beta-casein   9925   \n",
       "3   CASB_SHEEP                 Beta-casein   9940   \n",
       "4  FFAR2_HUMAN  Free fatty acid receptor 2   9606   \n",
       "5   YMGE_ECOLI        UPF0410 protein YmgE  83333   \n",
       "\n",
       "          organismScientificName  uniprotStart  ...  \\\n",
       "0                     Bos taurus             1  ...   \n",
       "1                Bubalus bubalis             1  ...   \n",
       "2                   Capra hircus             1  ...   \n",
       "3                     Ovis aries             1  ...   \n",
       "4                   Homo sapiens             1  ...   \n",
       "5  Escherichia coli (strain K12)             1  ...   \n",
       "\n",
       "                                             bcifUrl  \\\n",
       "0  https://alphafold.ebi.ac.uk/files/AF-P02666-F1...   \n",
       "1  https://alphafold.ebi.ac.uk/files/AF-Q9TSI0-F1...   \n",
       "2  https://alphafold.ebi.ac.uk/files/AF-P33048-F1...   \n",
       "3  https://alphafold.ebi.ac.uk/files/AF-P11839-F1...   \n",
       "4  https://alphafold.ebi.ac.uk/files/AF-O15552-F1...   \n",
       "5  https://alphafold.ebi.ac.uk/files/AF-P76011-F1...   \n",
       "\n",
       "                                              cifUrl  \\\n",
       "0  https://alphafold.ebi.ac.uk/files/AF-P02666-F1...   \n",
       "1  https://alphafold.ebi.ac.uk/files/AF-Q9TSI0-F1...   \n",
       "2  https://alphafold.ebi.ac.uk/files/AF-P33048-F1...   \n",
       "3  https://alphafold.ebi.ac.uk/files/AF-P11839-F1...   \n",
       "4  https://alphafold.ebi.ac.uk/files/AF-O15552-F1...   \n",
       "5  https://alphafold.ebi.ac.uk/files/AF-P76011-F1...   \n",
       "\n",
       "                                              pdbUrl  \\\n",
       "0  https://alphafold.ebi.ac.uk/files/AF-P02666-F1...   \n",
       "1  https://alphafold.ebi.ac.uk/files/AF-Q9TSI0-F1...   \n",
       "2  https://alphafold.ebi.ac.uk/files/AF-P33048-F1...   \n",
       "3  https://alphafold.ebi.ac.uk/files/AF-P11839-F1...   \n",
       "4  https://alphafold.ebi.ac.uk/files/AF-O15552-F1...   \n",
       "5  https://alphafold.ebi.ac.uk/files/AF-P76011-F1...   \n",
       "\n",
       "                                         paeImageUrl  \\\n",
       "0  https://alphafold.ebi.ac.uk/files/AF-P02666-F1...   \n",
       "1  https://alphafold.ebi.ac.uk/files/AF-Q9TSI0-F1...   \n",
       "2  https://alphafold.ebi.ac.uk/files/AF-P33048-F1...   \n",
       "3  https://alphafold.ebi.ac.uk/files/AF-P11839-F1...   \n",
       "4  https://alphafold.ebi.ac.uk/files/AF-O15552-F1...   \n",
       "5  https://alphafold.ebi.ac.uk/files/AF-P76011-F1...   \n",
       "\n",
       "                                           paeDocUrl  \\\n",
       "0  https://alphafold.ebi.ac.uk/files/AF-P02666-F1...   \n",
       "1  https://alphafold.ebi.ac.uk/files/AF-Q9TSI0-F1...   \n",
       "2  https://alphafold.ebi.ac.uk/files/AF-P33048-F1...   \n",
       "3  https://alphafold.ebi.ac.uk/files/AF-P11839-F1...   \n",
       "4  https://alphafold.ebi.ac.uk/files/AF-O15552-F1...   \n",
       "5  https://alphafold.ebi.ac.uk/files/AF-P76011-F1...   \n",
       "\n",
       "                                    amAnnotationsUrl  \\\n",
       "0                                               None   \n",
       "1                                               None   \n",
       "2                                               None   \n",
       "3                                               None   \n",
       "4  https://alphafold.ebi.ac.uk/files/AF-O15552-F1...   \n",
       "5                                               None   \n",
       "\n",
       "                                amAnnotationsHg19Url  \\\n",
       "0                                               None   \n",
       "1                                               None   \n",
       "2                                               None   \n",
       "3                                               None   \n",
       "4  https://alphafold.ebi.ac.uk/files/AF-O15552-F1...   \n",
       "5                                               None   \n",
       "\n",
       "                                amAnnotationsHg38Url isReviewed  \\\n",
       "0                                               None       True   \n",
       "1                                               None       True   \n",
       "2                                               None       True   \n",
       "3                                               None       True   \n",
       "4  https://alphafold.ebi.ac.uk/files/AF-O15552-F1...       True   \n",
       "5                                               None       True   \n",
       "\n",
       "  isReferenceProteome  \n",
       "0                True  \n",
       "1               False  \n",
       "2                True  \n",
       "3                True  \n",
       "4                True  \n",
       "5                True  \n",
       "\n",
       "[6 rows x 25 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "instance.fetch_batch(\n",
    "    queries=[\"P02666\", \"Q9TSI0\", \"P33048\", \"P11839\", \"O15552\", \"P76011\"],\n",
    "    parse=False,\n",
    "    to_dataframe=True,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d63b7d30",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'result' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mNameError\u001b[39m                                 Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[18]\u001b[39m\u001b[32m, line 2\u001b[39m\n\u001b[32m      1\u001b[39m instance.save(\n\u001b[32m----> \u001b[39m\u001b[32m2\u001b[39m     data = \u001b[43mresult\u001b[49m,\n\u001b[32m      3\u001b[39m     filename = \u001b[33m\"\u001b[39m\u001b[33malphafold2_results\u001b[39m\u001b[33m\"\u001b[39m,\n\u001b[32m      4\u001b[39m     extension= \u001b[33m\"\u001b[39m\u001b[33mcsv\u001b[39m\u001b[33m\"\u001b[39m\n\u001b[32m      5\u001b[39m )\n",
      "\u001b[31mNameError\u001b[39m: name 'result' is not defined"
     ]
    }
   ],
   "source": [
    "instance.save(\n",
    "    data = result,\n",
    "    filename = \"alphafold2_results\",\n",
    "    extension= \"csv\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "96e5bb46",
   "metadata": {},
   "source": [
    "## BioDBNet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "a9a230ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "from src.biodbnet import BioDBNetInterface"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "8ab5ee97",
   "metadata": {},
   "outputs": [],
   "source": [
    "instance = BioDBNetInterface()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "dfc37b40",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cache_key: APP_genesymbol_genesymbol_9606_db2db\n",
      "Cache_key: APP_affyid_genesymbol_9606_db2db\n",
      "Cache_key: APP_go-biologicalprocess_genesymbol_9606_db2db\n",
      "Cache_key: SLC9A9_genesymbol_genesymbol_9606_db2db\n",
      "Cache_key: SLC9A9_affyid_genesymbol_9606_db2db\n",
      "Cache_key: SLC9A9_go-biologicalprocess_genesymbol_9606_db2db\n",
      "Prepared request: https://biodbnet.abcc.ncifcrf.gov/webServices/rest.php/biodbnetRestApi.json?inputValues=APP%2CSLC9A9&outputs=genesymbol%2Caffyid%2Cgo-biologicalprocess&input=genesymbol&taxonId=9606&method=db2db\n",
      "Cache_key: APP_genesymbol_genesymbol_9606_db2db\n",
      "Cache_key: APP_affyid_genesymbol_9606_db2db\n",
      "Cache_key: APP_go-biologicalprocess_genesymbol_9606_db2db\n",
      "Cache_key: SLC9A9_genesymbol_genesymbol_9606_db2db\n",
      "Cache_key: SLC9A9_affyid_genesymbol_9606_db2db\n",
      "Cache_key: SLC9A9_go-biologicalprocess_genesymbol_9606_db2db\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>affy_id</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>go_biological_process</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>[11738008_s_at [Chip: PrimeView], 11746866_a_a...</td>\n",
       "      <td>APP</td>\n",
       "      <td>[GO:0006357 [Name: regulation of transcription...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>[11738008_s_at [Chip: PrimeView], 11746866_a_a...</td>\n",
       "      <td>APP</td>\n",
       "      <td>[GO:0006357 [Name: regulation of transcription...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>[11738008_s_at [Chip: PrimeView], 11746866_a_a...</td>\n",
       "      <td>APP</td>\n",
       "      <td>[GO:0006357 [Name: regulation of transcription...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>[227791_at [Chip: U133_Plus_2], 43140_at [Chip...</td>\n",
       "      <td>SLC9A9</td>\n",
       "      <td>[GO:0042742 [Name: defense response to bacteri...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>[227791_at [Chip: U133_Plus_2], 43140_at [Chip...</td>\n",
       "      <td>SLC9A9</td>\n",
       "      <td>[GO:0042742 [Name: defense response to bacteri...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>[227791_at [Chip: U133_Plus_2], 43140_at [Chip...</td>\n",
       "      <td>SLC9A9</td>\n",
       "      <td>[GO:0042742 [Name: defense response to bacteri...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>[11738008_s_at [Chip: PrimeView], 11746866_a_a...</td>\n",
       "      <td>APP</td>\n",
       "      <td>[GO:0006357 [Name: regulation of transcription...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>[227791_at [Chip: U133_Plus_2], 43140_at [Chip...</td>\n",
       "      <td>SLC9A9</td>\n",
       "      <td>[GO:0042742 [Name: defense response to bacteri...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             affy_id gene_symbol  \\\n",
       "0  [11738008_s_at [Chip: PrimeView], 11746866_a_a...         APP   \n",
       "1  [11738008_s_at [Chip: PrimeView], 11746866_a_a...         APP   \n",
       "2  [11738008_s_at [Chip: PrimeView], 11746866_a_a...         APP   \n",
       "3  [227791_at [Chip: U133_Plus_2], 43140_at [Chip...      SLC9A9   \n",
       "4  [227791_at [Chip: U133_Plus_2], 43140_at [Chip...      SLC9A9   \n",
       "5  [227791_at [Chip: U133_Plus_2], 43140_at [Chip...      SLC9A9   \n",
       "6  [11738008_s_at [Chip: PrimeView], 11746866_a_a...         APP   \n",
       "7  [227791_at [Chip: U133_Plus_2], 43140_at [Chip...      SLC9A9   \n",
       "\n",
       "                               go_biological_process  \n",
       "0  [GO:0006357 [Name: regulation of transcription...  \n",
       "1  [GO:0006357 [Name: regulation of transcription...  \n",
       "2  [GO:0006357 [Name: regulation of transcription...  \n",
       "3  [GO:0042742 [Name: defense response to bacteri...  \n",
       "4  [GO:0042742 [Name: defense response to bacteri...  \n",
       "5  [GO:0042742 [Name: defense response to bacteri...  \n",
       "6  [GO:0006357 [Name: regulation of transcription...  \n",
       "7  [GO:0042742 [Name: defense response to bacteri...  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "instance.fetch_single(\n",
    "    query={\n",
    "        \"input\": \"genesymbol\",\n",
    "        \"inputValues\": [\"APP\", \"SLC9A9\"],\n",
    "        \"outputs\": [\"genesymbol\", \"affyid\", \"go-biologicalprocess\"],\n",
    "        \"taxonId\": \"9606\"\n",
    "    },\n",
    "    method=\"db2db\",\n",
    "    parse=True,\n",
    "    to_dataframe=True\n",
    ").dropna(axis=1, how='all')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "31cdb9a1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cache_key: 1_9606_getpathways\n",
      "Prepared request: https://biodbnet.abcc.ncifcrf.gov/webServices/rest.php/biodbnetRestApi.json?pathways=1&taxonId=9606&method=getpathways\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>title</th>\n",
       "      <th>description</th>\n",
       "      <th>source_database</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NS1 Mediated Effects on Host Pathways</td>\n",
       "      <td>NS1 Mediated Effects on Host Pathways</td>\n",
       "      <td>NS1 Mediated Effects on Host Pathways</td>\n",
       "      <td>REACTOME</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2-LTR circle formation</td>\n",
       "      <td>2-LTR circle formation</td>\n",
       "      <td>2-LTR circle formation</td>\n",
       "      <td>REACTOME</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3' -UTR-mediated translational regulation</td>\n",
       "      <td>3' -UTR-mediated translational regulation</td>\n",
       "      <td>3' -UTR-mediated translational regulation</td>\n",
       "      <td>REACTOME</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5-Phosphoribose 1-diphosphate biosynthesis</td>\n",
       "      <td>PRPP biosynthesis</td>\n",
       "      <td>PRPP biosynthesis</td>\n",
       "      <td>REACTOME</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ABH2 mediated Reversal of Alkylation Damage</td>\n",
       "      <td>ABH2 mediated Reversal of Alkylation Damage</td>\n",
       "      <td>ABH2 mediated Reversal of Alkylation Damage</td>\n",
       "      <td>REACTOME</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1751</th>\n",
       "      <td>hsa05020</td>\n",
       "      <td>Prion diseases</td>\n",
       "      <td>Prion diseases</td>\n",
       "      <td>KEGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1752</th>\n",
       "      <td>hsa03320</td>\n",
       "      <td>PPAR signaling pathway</td>\n",
       "      <td>PPAR signaling pathway</td>\n",
       "      <td>KEGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1753</th>\n",
       "      <td>hsa00620</td>\n",
       "      <td>Pyruvate metabolism</td>\n",
       "      <td>Pyruvate metabolism</td>\n",
       "      <td>KEGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1754</th>\n",
       "      <td>hsa04070</td>\n",
       "      <td>Phosphatidylinositol signaling system</td>\n",
       "      <td>Phosphatidylinositol signaling system</td>\n",
       "      <td>KEGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1755</th>\n",
       "      <td>hsa00532</td>\n",
       "      <td>Glycosaminoglycan biosynthesis - chondroitin s...</td>\n",
       "      <td>Glycosaminoglycan biosynthesis - chondroitin s...</td>\n",
       "      <td>KEGG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1756 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             name  \\\n",
       "0           NS1 Mediated Effects on Host Pathways   \n",
       "1                          2-LTR circle formation   \n",
       "2       3' -UTR-mediated translational regulation   \n",
       "3      5-Phosphoribose 1-diphosphate biosynthesis   \n",
       "4     ABH2 mediated Reversal of Alkylation Damage   \n",
       "...                                           ...   \n",
       "1751                                     hsa05020   \n",
       "1752                                     hsa03320   \n",
       "1753                                     hsa00620   \n",
       "1754                                     hsa04070   \n",
       "1755                                     hsa00532   \n",
       "\n",
       "                                                  title  \\\n",
       "0                 NS1 Mediated Effects on Host Pathways   \n",
       "1                                2-LTR circle formation   \n",
       "2             3' -UTR-mediated translational regulation   \n",
       "3                                     PRPP biosynthesis   \n",
       "4           ABH2 mediated Reversal of Alkylation Damage   \n",
       "...                                                 ...   \n",
       "1751                                     Prion diseases   \n",
       "1752                             PPAR signaling pathway   \n",
       "1753                                Pyruvate metabolism   \n",
       "1754              Phosphatidylinositol signaling system   \n",
       "1755  Glycosaminoglycan biosynthesis - chondroitin s...   \n",
       "\n",
       "                                            description source_database  \n",
       "0                 NS1 Mediated Effects on Host Pathways        REACTOME  \n",
       "1                                2-LTR circle formation        REACTOME  \n",
       "2             3' -UTR-mediated translational regulation        REACTOME  \n",
       "3                                     PRPP biosynthesis        REACTOME  \n",
       "4           ABH2 mediated Reversal of Alkylation Damage        REACTOME  \n",
       "...                                                 ...             ...  \n",
       "1751                                     Prion diseases            KEGG  \n",
       "1752                             PPAR signaling pathway            KEGG  \n",
       "1753                                Pyruvate metabolism            KEGG  \n",
       "1754              Phosphatidylinositol signaling system            KEGG  \n",
       "1755  Glycosaminoglycan biosynthesis - chondroitin s...            KEGG  \n",
       "\n",
       "[1756 rows x 4 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "instance.fetch_single(\n",
    "    query={\n",
    "        \"pathways\": \"1\",\n",
    "        \"taxonId\": \"9606\"\n",
    "    },\n",
    "    method=\"getpathways\",\n",
    "    parse=True,\n",
    "    to_dataframe=True\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6968b1a7",
   "metadata": {},
   "source": [
    "## BioGRID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "8e5ecb97",
   "metadata": {},
   "outputs": [],
   "source": [
    "from src.biogrid import BioGRIDInterface\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "c7ab9d9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "from dotenv import load_dotenv\n",
    "load_dotenv()\n",
    "biogrid_api_key = os.getenv(\"biogrid_api_key\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "ab7b94a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "instance = BioGRIDInterface(\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "2cd9cd56",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cache_key: P53_interactions\n",
      "Cache_key: CDK2_interactions\n",
      "Cache_key: BRCA1_interactions\n",
      "Prepared request: https://webservice.thebiogrid.org/interactions?accessKey=87aee10241ab2fcc6e497352e5024380&start=0&max=10000&interSpeciesExclude=False&selfInteractionsExclude=False&includeEvidence=False&geneList=P53%7CCDK2%7CBRCA1&searchBiogridIds=False&taxId=All&searchIds=False&format=json\n",
      "Cache_key: P53_interactions\n",
      "Cache_key: CDK2_interactions\n",
      "Cache_key: BRCA1_interactions\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>interaction_b</th>\n",
       "      <th>synonyms_a</th>\n",
       "      <th>synonyms_b</th>\n",
       "      <th>organism_a</th>\n",
       "      <th>organism_b</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WWOX</td>\n",
       "      <td>BCC7|LFS1|P53|TRP53</td>\n",
       "      <td>D16S432E|FOR|FRA16D|HHCMA56|PRO0128|SCAR12|SDR...</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CDC14A</td>\n",
       "      <td>BCC7|LFS1|P53|TRP53</td>\n",
       "      <td>cdc14|hCDC14</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>betaTub60D</td>\n",
       "      <td>Dmel\\CG15450</td>\n",
       "      <td>143391_i_at|3t|B3t|BETA 60D|CG3401|D.m.BETA-60...</td>\n",
       "      <td>7227</td>\n",
       "      <td>7227</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>tamo</td>\n",
       "      <td>1323/07|1422/04|CG17117|Dm-HTH|Dmel\\CG17117|Me...</td>\n",
       "      <td>60B|CG4057|Dmel\\CG4057</td>\n",
       "      <td>7227</td>\n",
       "      <td>7227</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>exd</td>\n",
       "      <td>1323/07|1422/04|CG17117|Dm-HTH|Dmel\\CG17117|Me...</td>\n",
       "      <td>CG8933|DExd|Dm-EXD|Dmel\\CG8933|Dpbx|Pbx1|anon-...</td>\n",
       "      <td>7227</td>\n",
       "      <td>7227</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10035</th>\n",
       "      <td>PRR5</td>\n",
       "      <td>BRCAI|BRCC1|BROVCA1|FANCS|IRIS|PNCA4|PPP1R53|P...</td>\n",
       "      <td>FLJ20185k|PP610|PROTOR-1|PROTOR1</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10036</th>\n",
       "      <td>RICTOR</td>\n",
       "      <td>BRCAI|BRCC1|BROVCA1|FANCS|IRIS|PNCA4|PPP1R53|P...</td>\n",
       "      <td>AVO3|PIA|hAVO3</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10037</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>JC310|MIP1|SIN1|SIN1b|SIN1g</td>\n",
       "      <td>BRCAI|BRCC1|BROVCA1|FANCS|IRIS|PNCA4|PPP1R53|P...</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10038</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>FLJ20185k|PP610|PROTOR-1|PROTOR1</td>\n",
       "      <td>BRCAI|BRCC1|BROVCA1|FANCS|IRIS|PNCA4|PPP1R53|P...</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10039</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>AVO3|PIA|hAVO3</td>\n",
       "      <td>BRCAI|BRCC1|BROVCA1|FANCS|IRIS|PNCA4|PPP1R53|P...</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10040 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      interaction_b                                         synonyms_a  \\\n",
       "0              WWOX                                BCC7|LFS1|P53|TRP53   \n",
       "1            CDC14A                                BCC7|LFS1|P53|TRP53   \n",
       "2        betaTub60D                                       Dmel\\CG15450   \n",
       "3              tamo  1323/07|1422/04|CG17117|Dm-HTH|Dmel\\CG17117|Me...   \n",
       "4               exd  1323/07|1422/04|CG17117|Dm-HTH|Dmel\\CG17117|Me...   \n",
       "...             ...                                                ...   \n",
       "10035          PRR5  BRCAI|BRCC1|BROVCA1|FANCS|IRIS|PNCA4|PPP1R53|P...   \n",
       "10036        RICTOR  BRCAI|BRCC1|BROVCA1|FANCS|IRIS|PNCA4|PPP1R53|P...   \n",
       "10037         BRCA1                        JC310|MIP1|SIN1|SIN1b|SIN1g   \n",
       "10038         BRCA1                   FLJ20185k|PP610|PROTOR-1|PROTOR1   \n",
       "10039         BRCA1                                     AVO3|PIA|hAVO3   \n",
       "\n",
       "                                              synonyms_b  organism_a  \\\n",
       "0      D16S432E|FOR|FRA16D|HHCMA56|PRO0128|SCAR12|SDR...        9606   \n",
       "1                                           cdc14|hCDC14        9606   \n",
       "2      143391_i_at|3t|B3t|BETA 60D|CG3401|D.m.BETA-60...        7227   \n",
       "3                                 60B|CG4057|Dmel\\CG4057        7227   \n",
       "4      CG8933|DExd|Dm-EXD|Dmel\\CG8933|Dpbx|Pbx1|anon-...        7227   \n",
       "...                                                  ...         ...   \n",
       "10035                   FLJ20185k|PP610|PROTOR-1|PROTOR1        9606   \n",
       "10036                                     AVO3|PIA|hAVO3        9606   \n",
       "10037  BRCAI|BRCC1|BROVCA1|FANCS|IRIS|PNCA4|PPP1R53|P...        9606   \n",
       "10038  BRCAI|BRCC1|BROVCA1|FANCS|IRIS|PNCA4|PPP1R53|P...        9606   \n",
       "10039  BRCAI|BRCC1|BROVCA1|FANCS|IRIS|PNCA4|PPP1R53|P...        9606   \n",
       "\n",
       "       organism_b  \n",
       "0            9606  \n",
       "1            9606  \n",
       "2            7227  \n",
       "3            7227  \n",
       "4            7227  \n",
       "...           ...  \n",
       "10035        9606  \n",
       "10036        9606  \n",
       "10037        9606  \n",
       "10038        9606  \n",
       "10039        9606  \n",
       "\n",
       "[10040 rows x 5 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "instance.fetch_single(\n",
    "    query={\n",
    "        \"accessKey\": biogrid_api_key,\n",
    "        \"geneList\": [\"P53\", \"CDK2\", \"BRCA1\"],\n",
    "    },\n",
    "    method=\"interactions\",\n",
    "    parse=True,\n",
    "    to_dataframe=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "8bb0b540",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Usage: To fetch interactions, use the BioGRID API with the following parameters.\n",
      "        Example:\n",
      "            - fetch_single(method=\"interactions\", query={})\n",
      "        Available methods: interactions\n",
      "\n",
      "\n",
      "\n",
      "Example Query:\n",
      "\n",
      "        {\n",
      "            \"accessKey\": \"YOUR_ACCESS_KEY\",\n",
      "            \"geneList\": [\"P53\"],\n",
      "            \"max\": 10,\n",
      "            \"format\": \"json\"\n",
      "        }\n",
      "        \n",
      "\n",
      "Response Format:\n",
      "\n",
      "        {\n",
      "            \"BIOGRID_INTERACTION_ID\": \"int\",\n",
      "            \"ENTREZ_GENE_A\": \"str\",\n",
      "            \"ENTREZ_GENE_B\": \"str\",\n",
      "            \"BIOGRID_ID_A\": \"int\",\n",
      "            \"BIOGRID_ID_B\": \"int\",\n",
      "            \"SYSTEMATIC_NAME_A\": \"str\",\n",
      "            \"SYSTEMATIC_NAME_B\": \"str\",\n",
      "            \"OFFICIAL_SYMBOL_A\": \"str\",\n",
      "            \"OFFICIAL_SYMBOL_B\": \"str\",\n",
      "            \"SYNONYMS_A\": \"str\",\n",
      "            \"SYNONYMS_B\": \"str\",\n",
      "            \"EXPERIMENTAL_SYSTEM\": \"str\",\n",
      "            \"EXPERIMENTAL_SYSTEM_TYPE\": \"str\",\n",
      "            \"PUBMED_AUTHOR\": \"str\",\n",
      "            \"PUBMED_ID\": \"int\",\n",
      "            \"ORGANISM_A\": \"int\",\n",
      "            \"ORGANISM_B\": \"int\",\n",
      "            \"THROUGHPUT\": \"str\",\n",
      "            \"QUANTITATION\": \"str\",\n",
      "            \"MODIFICATION\": \"str\",\n",
      "            \"ONTOLOGY_TERMS\": {},\n",
      "            \"QUALIFICATIONS\": \"\",\n",
      "            \"TAGS\": \"\",\n",
      "            \"SOURCEDB\": \"\"\n",
      "        }\n"
     ]
    }
   ],
   "source": [
    "print(instance.query_usage())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "ac901370",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cache_key: P53_interactions\n",
      "Cache_key: CDK2_interactions\n",
      "Cache_key: BRCA1_interactions\n",
      "Cache_key: BRCA2_interactions\n",
      "Cache_key: ATM_interactions\n",
      "Cache_key: CHEK2_interactions\n",
      "Cache_key: BRCA2_interactions\n",
      "Cache_key: ATM_interactions\n",
      "Cache_key: CHEK2_interactions\n",
      "Prepared request: https://webservice.thebiogrid.org/interactions?accessKey=87aee10241ab2fcc6e497352e5024380&start=0&max=10000&interSpeciesExclude=False&selfInteractionsExclude=False&includeEvidence=False&geneList=BRCA2%7CATM%7CCHEK2&searchBiogridIds=False&taxId=All&searchIds=False&format=json\n",
      "Cache_key: BRCA2_interactions\n",
      "Cache_key: ATM_interactions\n",
      "Cache_key: CHEK2_interactions\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>interaction_b</th>\n",
       "      <th>synonyms_a</th>\n",
       "      <th>synonyms_b</th>\n",
       "      <th>organism_a</th>\n",
       "      <th>organism_b</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WWOX</td>\n",
       "      <td>BCC7|LFS1|P53|TRP53</td>\n",
       "      <td>D16S432E|FOR|FRA16D|HHCMA56|PRO0128|SCAR12|SDR...</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CDC14A</td>\n",
       "      <td>BCC7|LFS1|P53|TRP53</td>\n",
       "      <td>cdc14|hCDC14</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>betaTub60D</td>\n",
       "      <td>Dmel\\CG15450</td>\n",
       "      <td>143391_i_at|3t|B3t|BETA 60D|CG3401|D.m.BETA-60...</td>\n",
       "      <td>7227</td>\n",
       "      <td>7227</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>tamo</td>\n",
       "      <td>1323/07|1422/04|CG17117|Dm-HTH|Dmel\\CG17117|Me...</td>\n",
       "      <td>60B|CG4057|Dmel\\CG4057</td>\n",
       "      <td>7227</td>\n",
       "      <td>7227</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>exd</td>\n",
       "      <td>1323/07|1422/04|CG17117|Dm-HTH|Dmel\\CG17117|Me...</td>\n",
       "      <td>CG8933|DExd|Dm-EXD|Dmel\\CG8933|Dpbx|Pbx1|anon-...</td>\n",
       "      <td>7227</td>\n",
       "      <td>7227</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12049</th>\n",
       "      <td>CHEK2</td>\n",
       "      <td>MYL|PP8675|RNF71|TRIM19</td>\n",
       "      <td>CDS1|CHK2|HuCds1|LFS2|PP1425|RAD53|hCds1</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12050</th>\n",
       "      <td>CHEK2</td>\n",
       "      <td>CDS1|CHK2|HuCds1|LFS2|PP1425|RAD53|hCds1</td>\n",
       "      <td>CDS1|CHK2|HuCds1|LFS2|PP1425|RAD53|hCds1</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12051</th>\n",
       "      <td>CHEK2</td>\n",
       "      <td>CDS1|CHK2|HuCds1|LFS2|PP1425|RAD53|hCds1</td>\n",
       "      <td>CDS1|CHK2|HuCds1|LFS2|PP1425|RAD53|hCds1</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12052</th>\n",
       "      <td>CDC25C</td>\n",
       "      <td>CDS1|CHK2|HuCds1|LFS2|PP1425|RAD53|hCds1</td>\n",
       "      <td>CDC25|PPP1R60</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12053</th>\n",
       "      <td>CHEK2</td>\n",
       "      <td>ORF1ab polyprotein|1ab|SARS-CoV2 ORF1ab|SARS-C...</td>\n",
       "      <td>CDS1|CHK2|HuCds1|LFS2|PP1425|RAD53|hCds1</td>\n",
       "      <td>2697049</td>\n",
       "      <td>9606</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12054 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      interaction_b                                         synonyms_a  \\\n",
       "0              WWOX                                BCC7|LFS1|P53|TRP53   \n",
       "1            CDC14A                                BCC7|LFS1|P53|TRP53   \n",
       "2        betaTub60D                                       Dmel\\CG15450   \n",
       "3              tamo  1323/07|1422/04|CG17117|Dm-HTH|Dmel\\CG17117|Me...   \n",
       "4               exd  1323/07|1422/04|CG17117|Dm-HTH|Dmel\\CG17117|Me...   \n",
       "...             ...                                                ...   \n",
       "12049         CHEK2                            MYL|PP8675|RNF71|TRIM19   \n",
       "12050         CHEK2           CDS1|CHK2|HuCds1|LFS2|PP1425|RAD53|hCds1   \n",
       "12051         CHEK2           CDS1|CHK2|HuCds1|LFS2|PP1425|RAD53|hCds1   \n",
       "12052        CDC25C           CDS1|CHK2|HuCds1|LFS2|PP1425|RAD53|hCds1   \n",
       "12053         CHEK2  ORF1ab polyprotein|1ab|SARS-CoV2 ORF1ab|SARS-C...   \n",
       "\n",
       "                                              synonyms_b  organism_a  \\\n",
       "0      D16S432E|FOR|FRA16D|HHCMA56|PRO0128|SCAR12|SDR...        9606   \n",
       "1                                           cdc14|hCDC14        9606   \n",
       "2      143391_i_at|3t|B3t|BETA 60D|CG3401|D.m.BETA-60...        7227   \n",
       "3                                 60B|CG4057|Dmel\\CG4057        7227   \n",
       "4      CG8933|DExd|Dm-EXD|Dmel\\CG8933|Dpbx|Pbx1|anon-...        7227   \n",
       "...                                                  ...         ...   \n",
       "12049           CDS1|CHK2|HuCds1|LFS2|PP1425|RAD53|hCds1        9606   \n",
       "12050           CDS1|CHK2|HuCds1|LFS2|PP1425|RAD53|hCds1        9606   \n",
       "12051           CDS1|CHK2|HuCds1|LFS2|PP1425|RAD53|hCds1        9606   \n",
       "12052                                      CDC25|PPP1R60        9606   \n",
       "12053           CDS1|CHK2|HuCds1|LFS2|PP1425|RAD53|hCds1     2697049   \n",
       "\n",
       "       organism_b  \n",
       "0            9606  \n",
       "1            9606  \n",
       "2            7227  \n",
       "3            7227  \n",
       "4            7227  \n",
       "...           ...  \n",
       "12049        9606  \n",
       "12050        9606  \n",
       "12051        9606  \n",
       "12052        9606  \n",
       "12053        9606  \n",
       "\n",
       "[12054 rows x 5 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "instance.fetch_batch(\n",
    "    queries=[\n",
    "        {\n",
    "            \"accessKey\": biogrid_api_key,\n",
    "            \"geneList\": [\"P53\", \"CDK2\", \"BRCA1\"],\n",
    "        },\n",
    "        {\n",
    "            \"accessKey\": biogrid_api_key,\n",
    "            \"geneList\": [\"BRCA2\", \"ATM\", \"CHEK2\"],\n",
    "        }\n",
    "    ],\n",
    "    method=\"interactions\",\n",
    "    parse=True,\n",
    "    to_dataframe=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "19da9f81",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cache_key: cdc27_1_559292_interactions\n",
      "Cache_key: apc1_1_559292_interactions\n",
      "Cache_key: apc2_1_559292_interactions\n",
      "Prepared request: https://webservice.thebiogrid.org/interactions?accessKey=87aee10241ab2fcc6e497352e5024380&start=0&max=1&interSpeciesExclude=False&selfInteractionsExclude=False&includeEvidence=False&geneList=cdc27%7Capc1%7Capc2&searchBiogridIds=False&taxId=559292&searchIds=False&format=json\n",
      "Cache_key: cdc27_1_559292_interactions\n",
      "Cache_key: apc1_1_559292_interactions\n",
      "Cache_key: apc2_1_559292_interactions\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'interactions': {'BIOGRID_INTERACTION_ID': 'int',\n",
       "  'ENTREZ_GENE_A': 'str',\n",
       "  'ENTREZ_GENE_B': 'str',\n",
       "  'BIOGRID_ID_A': 'int',\n",
       "  'BIOGRID_ID_B': 'int',\n",
       "  'SYSTEMATIC_NAME_A': 'str',\n",
       "  'SYSTEMATIC_NAME_B': 'str',\n",
       "  'OFFICIAL_SYMBOL_A': 'str',\n",
       "  'OFFICIAL_SYMBOL_B': 'str',\n",
       "  'SYNONYMS_A': 'str',\n",
       "  'SYNONYMS_B': 'str',\n",
       "  'EXPERIMENTAL_SYSTEM': 'str',\n",
       "  'EXPERIMENTAL_SYSTEM_TYPE': 'str',\n",
       "  'PUBMED_AUTHOR': 'str',\n",
       "  'PUBMED_ID': 'int',\n",
       "  'ORGANISM_A': 'int',\n",
       "  'ORGANISM_B': 'int',\n",
       "  'THROUGHPUT': 'str',\n",
       "  'QUANTITATION': 'str',\n",
       "  'MODIFICATION': 'str',\n",
       "  'ONTOLOGY_TERMS': 'dict',\n",
       "  'QUALIFICATIONS': 'str',\n",
       "  'TAGS': 'str',\n",
       "  'SOURCEDB': 'str'}}"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "instance.get_dummy(biogrid_api_key, parse=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "fa1520af",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cache_key: cdc27_559292_interactions\n",
      "Cache_key: apc1_559292_interactions\n",
      "Cache_key: apc2_559292_interactions\n",
      "Prepared request: https://webservice.thebiogrid.org/interactions?accessKey=87aee10241ab2fcc6e497352e5024380&start=0&max=10000&interSpeciesExclude=False&selfInteractionsExclude=False&includeEvidence=False&geneList=cdc27%7Capc1%7Capc2&searchBiogridIds=False&taxId=559292&searchIds=False&format=json\n",
      "Cache_key: cdc27_559292_interactions\n",
      "Cache_key: apc1_559292_interactions\n",
      "Cache_key: apc2_559292_interactions\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>interaction_b</th>\n",
       "      <th>synonyms_a</th>\n",
       "      <th>synonyms_b</th>\n",
       "      <th>organism_a</th>\n",
       "      <th>organism_b</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>APC1</td>\n",
       "      <td>RSI1|TID2|anaphase promoting complex subunit 2...</td>\n",
       "      <td>anaphase promoting complex subunit 1|L000004053</td>\n",
       "      <td>559292</td>\n",
       "      <td>559292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CDC16</td>\n",
       "      <td>RSI1|TID2|anaphase promoting complex subunit 2...</td>\n",
       "      <td>anaphase promoting complex subunit CDC16|L0000...</td>\n",
       "      <td>559292</td>\n",
       "      <td>559292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CDC23</td>\n",
       "      <td>RSI1|TID2|anaphase promoting complex subunit 2...</td>\n",
       "      <td>anaphase promoting complex subunit CDC23|L0000...</td>\n",
       "      <td>559292</td>\n",
       "      <td>559292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CDC27</td>\n",
       "      <td>RSI1|TID2|anaphase promoting complex subunit 2...</td>\n",
       "      <td>APC3|SNB1|anaphase promoting complex subunit C...</td>\n",
       "      <td>559292</td>\n",
       "      <td>559292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>APC2</td>\n",
       "      <td>APC10|anaphase promoting complex subunit DOC1|...</td>\n",
       "      <td>RSI1|TID2|anaphase promoting complex subunit 2...</td>\n",
       "      <td>559292</td>\n",
       "      <td>559292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4423</th>\n",
       "      <td>APC1</td>\n",
       "      <td>CDK1|HSL5|SRM5|cyclin-dependent serine/threoni...</td>\n",
       "      <td>anaphase promoting complex subunit 1|L000004053</td>\n",
       "      <td>559292</td>\n",
       "      <td>559292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4424</th>\n",
       "      <td>APC2</td>\n",
       "      <td>CDK1|HSL5|SRM5|cyclin-dependent serine/threoni...</td>\n",
       "      <td>RSI1|TID2|anaphase promoting complex subunit 2...</td>\n",
       "      <td>559292</td>\n",
       "      <td>559292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4425</th>\n",
       "      <td>CDC27</td>\n",
       "      <td>CDK1|HSL5|SRM5|cyclin-dependent serine/threoni...</td>\n",
       "      <td>APC3|SNB1|anaphase promoting complex subunit C...</td>\n",
       "      <td>559292</td>\n",
       "      <td>559292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4426</th>\n",
       "      <td>APC1</td>\n",
       "      <td>MIND complex subunit DSN1|L000004645</td>\n",
       "      <td>anaphase promoting complex subunit 1|L000004053</td>\n",
       "      <td>559292</td>\n",
       "      <td>559292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4427</th>\n",
       "      <td>CDC27</td>\n",
       "      <td>MIND complex subunit DSN1|L000004645</td>\n",
       "      <td>APC3|SNB1|anaphase promoting complex subunit C...</td>\n",
       "      <td>559292</td>\n",
       "      <td>559292</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4428 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     interaction_b                                         synonyms_a  \\\n",
       "0             APC1  RSI1|TID2|anaphase promoting complex subunit 2...   \n",
       "1            CDC16  RSI1|TID2|anaphase promoting complex subunit 2...   \n",
       "2            CDC23  RSI1|TID2|anaphase promoting complex subunit 2...   \n",
       "3            CDC27  RSI1|TID2|anaphase promoting complex subunit 2...   \n",
       "4             APC2  APC10|anaphase promoting complex subunit DOC1|...   \n",
       "...            ...                                                ...   \n",
       "4423          APC1  CDK1|HSL5|SRM5|cyclin-dependent serine/threoni...   \n",
       "4424          APC2  CDK1|HSL5|SRM5|cyclin-dependent serine/threoni...   \n",
       "4425         CDC27  CDK1|HSL5|SRM5|cyclin-dependent serine/threoni...   \n",
       "4426          APC1               MIND complex subunit DSN1|L000004645   \n",
       "4427         CDC27               MIND complex subunit DSN1|L000004645   \n",
       "\n",
       "                                             synonyms_b  organism_a  \\\n",
       "0       anaphase promoting complex subunit 1|L000004053      559292   \n",
       "1     anaphase promoting complex subunit CDC16|L0000...      559292   \n",
       "2     anaphase promoting complex subunit CDC23|L0000...      559292   \n",
       "3     APC3|SNB1|anaphase promoting complex subunit C...      559292   \n",
       "4     RSI1|TID2|anaphase promoting complex subunit 2...      559292   \n",
       "...                                                 ...         ...   \n",
       "4423    anaphase promoting complex subunit 1|L000004053      559292   \n",
       "4424  RSI1|TID2|anaphase promoting complex subunit 2...      559292   \n",
       "4425  APC3|SNB1|anaphase promoting complex subunit C...      559292   \n",
       "4426    anaphase promoting complex subunit 1|L000004053      559292   \n",
       "4427  APC3|SNB1|anaphase promoting complex subunit C...      559292   \n",
       "\n",
       "      organism_b  \n",
       "0         559292  \n",
       "1         559292  \n",
       "2         559292  \n",
       "3         559292  \n",
       "4         559292  \n",
       "...          ...  \n",
       "4423      559292  \n",
       "4424      559292  \n",
       "4425      559292  \n",
       "4426      559292  \n",
       "4427      559292  \n",
       "\n",
       "[4428 rows x 5 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "instance.fetch_single(\n",
    "    query={\n",
    "        \"accessKey\": biogrid_api_key,\n",
    "        \"geneList\": [\"cdc27\", \"apc1\", \"apc2\"],\n",
    "        \"taxId\": \"559292\",\n",
    "    },\n",
    "    method=\"interactions\",\n",
    "    parse=True,\n",
    "    to_dataframe=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "0da7cebf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cache_key: cdc27_559292_interactions\n",
      "Cache_key: apc1_559292_interactions\n",
      "Cache_key: apc2_559292_interactions\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>interaction_b</th>\n",
       "      <th>synonyms_a</th>\n",
       "      <th>synonyms_b</th>\n",
       "      <th>organism_a</th>\n",
       "      <th>organism_b</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>APC1</td>\n",
       "      <td>RSI1|TID2|anaphase promoting complex subunit 2...</td>\n",
       "      <td>anaphase promoting complex subunit 1|L000004053</td>\n",
       "      <td>559292</td>\n",
       "      <td>559292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CDC16</td>\n",
       "      <td>RSI1|TID2|anaphase promoting complex subunit 2...</td>\n",
       "      <td>anaphase promoting complex subunit CDC16|L0000...</td>\n",
       "      <td>559292</td>\n",
       "      <td>559292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CDC23</td>\n",
       "      <td>RSI1|TID2|anaphase promoting complex subunit 2...</td>\n",
       "      <td>anaphase promoting complex subunit CDC23|L0000...</td>\n",
       "      <td>559292</td>\n",
       "      <td>559292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CDC27</td>\n",
       "      <td>RSI1|TID2|anaphase promoting complex subunit 2...</td>\n",
       "      <td>APC3|SNB1|anaphase promoting complex subunit C...</td>\n",
       "      <td>559292</td>\n",
       "      <td>559292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>APC2</td>\n",
       "      <td>APC10|anaphase promoting complex subunit DOC1|...</td>\n",
       "      <td>RSI1|TID2|anaphase promoting complex subunit 2...</td>\n",
       "      <td>559292</td>\n",
       "      <td>559292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4423</th>\n",
       "      <td>APC1</td>\n",
       "      <td>CDK1|HSL5|SRM5|cyclin-dependent serine/threoni...</td>\n",
       "      <td>anaphase promoting complex subunit 1|L000004053</td>\n",
       "      <td>559292</td>\n",
       "      <td>559292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4424</th>\n",
       "      <td>APC2</td>\n",
       "      <td>CDK1|HSL5|SRM5|cyclin-dependent serine/threoni...</td>\n",
       "      <td>RSI1|TID2|anaphase promoting complex subunit 2...</td>\n",
       "      <td>559292</td>\n",
       "      <td>559292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4425</th>\n",
       "      <td>CDC27</td>\n",
       "      <td>CDK1|HSL5|SRM5|cyclin-dependent serine/threoni...</td>\n",
       "      <td>APC3|SNB1|anaphase promoting complex subunit C...</td>\n",
       "      <td>559292</td>\n",
       "      <td>559292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4426</th>\n",
       "      <td>APC1</td>\n",
       "      <td>MIND complex subunit DSN1|L000004645</td>\n",
       "      <td>anaphase promoting complex subunit 1|L000004053</td>\n",
       "      <td>559292</td>\n",
       "      <td>559292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4427</th>\n",
       "      <td>CDC27</td>\n",
       "      <td>MIND complex subunit DSN1|L000004645</td>\n",
       "      <td>APC3|SNB1|anaphase promoting complex subunit C...</td>\n",
       "      <td>559292</td>\n",
       "      <td>559292</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4428 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     interaction_b                                         synonyms_a  \\\n",
       "0             APC1  RSI1|TID2|anaphase promoting complex subunit 2...   \n",
       "1            CDC16  RSI1|TID2|anaphase promoting complex subunit 2...   \n",
       "2            CDC23  RSI1|TID2|anaphase promoting complex subunit 2...   \n",
       "3            CDC27  RSI1|TID2|anaphase promoting complex subunit 2...   \n",
       "4             APC2  APC10|anaphase promoting complex subunit DOC1|...   \n",
       "...            ...                                                ...   \n",
       "4423          APC1  CDK1|HSL5|SRM5|cyclin-dependent serine/threoni...   \n",
       "4424          APC2  CDK1|HSL5|SRM5|cyclin-dependent serine/threoni...   \n",
       "4425         CDC27  CDK1|HSL5|SRM5|cyclin-dependent serine/threoni...   \n",
       "4426          APC1               MIND complex subunit DSN1|L000004645   \n",
       "4427         CDC27               MIND complex subunit DSN1|L000004645   \n",
       "\n",
       "                                             synonyms_b  organism_a  \\\n",
       "0       anaphase promoting complex subunit 1|L000004053      559292   \n",
       "1     anaphase promoting complex subunit CDC16|L0000...      559292   \n",
       "2     anaphase promoting complex subunit CDC23|L0000...      559292   \n",
       "3     APC3|SNB1|anaphase promoting complex subunit C...      559292   \n",
       "4     RSI1|TID2|anaphase promoting complex subunit 2...      559292   \n",
       "...                                                 ...         ...   \n",
       "4423    anaphase promoting complex subunit 1|L000004053      559292   \n",
       "4424  RSI1|TID2|anaphase promoting complex subunit 2...      559292   \n",
       "4425  APC3|SNB1|anaphase promoting complex subunit C...      559292   \n",
       "4426    anaphase promoting complex subunit 1|L000004053      559292   \n",
       "4427  APC3|SNB1|anaphase promoting complex subunit C...      559292   \n",
       "\n",
       "      organism_b  \n",
       "0         559292  \n",
       "1         559292  \n",
       "2         559292  \n",
       "3         559292  \n",
       "4         559292  \n",
       "...          ...  \n",
       "4423      559292  \n",
       "4424      559292  \n",
       "4425      559292  \n",
       "4426      559292  \n",
       "4427      559292  \n",
       "\n",
       "[4428 rows x 5 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "instance.fetch_single(\n",
    "    query={\n",
    "        \"accessKey\": biogrid_api_key,\n",
    "        \"geneList\": [\"cdc27\", \"apc1\", \"apc2\"],\n",
    "        \"taxId\": \"559292\",\n",
    "    },\n",
    "    method=\"interactions\",\n",
    "    parse=True,\n",
    "    to_dataframe=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "c0e77fd3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cache_key: MFS1_9606_interactions\n",
      "Prepared request: https://webservice.thebiogrid.org/interactions?accessKey=87aee10241ab2fcc6e497352e5024380&start=0&max=10000&interSpeciesExclude=False&selfInteractionsExclude=False&includeEvidence=False&geneList=MFS1&searchBiogridIds=False&taxId=9606&searchIds=False&format=json\n",
      "Cache_key: MFS1_9606_interactions\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>interaction_b</th>\n",
       "      <th>synonyms_a</th>\n",
       "      <th>synonyms_b</th>\n",
       "      <th>organism_a</th>\n",
       "      <th>organism_b</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MFAP2</td>\n",
       "      <td>ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...</td>\n",
       "      <td>MAGP|MAGP-1|MAGP1</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FBN1</td>\n",
       "      <td>-</td>\n",
       "      <td>ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FBN1</td>\n",
       "      <td>GLC1A|GPOA|JOAG|JOAG1|TIGR|myocilin</td>\n",
       "      <td>ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>FBN1</td>\n",
       "      <td>CSPG2|ERVR|GHAP|PG-M|WGN|WGN1</td>\n",
       "      <td>ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>FBN1</td>\n",
       "      <td>SVAS|WBS|WS</td>\n",
       "      <td>ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>FBN1</td>\n",
       "      <td>ADCAII|OPCA3|SCA7</td>\n",
       "      <td>ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>FBN1</td>\n",
       "      <td>hSPRY2</td>\n",
       "      <td>ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>FBN1</td>\n",
       "      <td>ZNF753|zfp-41</td>\n",
       "      <td>ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>FBN1</td>\n",
       "      <td>-</td>\n",
       "      <td>ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>FBN1</td>\n",
       "      <td>ZNF753|zfp-41</td>\n",
       "      <td>ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>FBN1</td>\n",
       "      <td>AI2A4</td>\n",
       "      <td>ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>FBN1</td>\n",
       "      <td>fSAP121</td>\n",
       "      <td>ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>FBN1</td>\n",
       "      <td>ORF1ab|ORF1ab-nsp9|SARS-CoV2 nsp9|SARS-CoV-2 n...</td>\n",
       "      <td>ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...</td>\n",
       "      <td>2697049</td>\n",
       "      <td>9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>HGS</td>\n",
       "      <td>ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...</td>\n",
       "      <td>HRS</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>SGTB</td>\n",
       "      <td>ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...</td>\n",
       "      <td>SGT2</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>GET4</td>\n",
       "      <td>ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...</td>\n",
       "      <td>C7orf20|CEE|TRC35</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>FBN1</td>\n",
       "      <td>ORF1ab|ORF1ab-nsp9|SARS-CoV2 nsp9|SARS-CoV-2 n...</td>\n",
       "      <td>ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...</td>\n",
       "      <td>2697049</td>\n",
       "      <td>9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>FBN1</td>\n",
       "      <td>CDW14</td>\n",
       "      <td>ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>FBN1</td>\n",
       "      <td>RIG-I|RIGI|RLR-1</td>\n",
       "      <td>ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>FBN1</td>\n",
       "      <td>FBG1|FBX2|Fbs1|NFB42|OCP1</td>\n",
       "      <td>ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>FBXO2</td>\n",
       "      <td>ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...</td>\n",
       "      <td>FBG1|FBX2|Fbs1|NFB42|OCP1</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>FBN1</td>\n",
       "      <td>MUL|POB1|TEF3</td>\n",
       "      <td>ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>FBN1</td>\n",
       "      <td>BSAS|HOX1|HOX1F</td>\n",
       "      <td>ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>FBN1</td>\n",
       "      <td>ZNF753|zfp-41</td>\n",
       "      <td>ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>FBN1</td>\n",
       "      <td>ARCI11|HAI|MT-SP1|MTSP1|PRSS14|SNC19|TADG15|TM...</td>\n",
       "      <td>ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>FBN1</td>\n",
       "      <td>-</td>\n",
       "      <td>ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>FBN1</td>\n",
       "      <td>AI2A4</td>\n",
       "      <td>ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>FBN1</td>\n",
       "      <td>FBG1|FBX2|Fbs1|NFB42|OCP1</td>\n",
       "      <td>ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>FBN1</td>\n",
       "      <td>HPRG|HRGP|THPH11</td>\n",
       "      <td>ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>FBN1</td>\n",
       "      <td>-</td>\n",
       "      <td>ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>FBN1</td>\n",
       "      <td>D6S115E|D6S2244E|H2-KE4|HKE4|KE4|RING5|ZIP7</td>\n",
       "      <td>ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>FBN1</td>\n",
       "      <td>-</td>\n",
       "      <td>ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>FBN1</td>\n",
       "      <td>bA350J20.1|p42</td>\n",
       "      <td>ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>FBN1</td>\n",
       "      <td>CUL-3|PHA2E</td>\n",
       "      <td>ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>FBN1</td>\n",
       "      <td>BPM-L|BPM90|RANBPM|RanBP7</td>\n",
       "      <td>ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>FBN1</td>\n",
       "      <td>TGN38|TGN46|TGN48|TGN51|TTGN2</td>\n",
       "      <td>ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>FBN1</td>\n",
       "      <td>MRX91</td>\n",
       "      <td>ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>FBN1</td>\n",
       "      <td>DYRK|DYRK1|HP86|MNB|MNBH|MRD7</td>\n",
       "      <td>ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>FBN1</td>\n",
       "      <td>CSC-21K|DDC8|KIAA1731NL</td>\n",
       "      <td>ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   interaction_b                                         synonyms_a  \\\n",
       "0          MFAP2  ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...   \n",
       "1           FBN1                                                  -   \n",
       "2           FBN1                GLC1A|GPOA|JOAG|JOAG1|TIGR|myocilin   \n",
       "3           FBN1                      CSPG2|ERVR|GHAP|PG-M|WGN|WGN1   \n",
       "4           FBN1                                        SVAS|WBS|WS   \n",
       "5           FBN1                                  ADCAII|OPCA3|SCA7   \n",
       "6           FBN1                                             hSPRY2   \n",
       "7           FBN1                                      ZNF753|zfp-41   \n",
       "8           FBN1                                                  -   \n",
       "9           FBN1                                      ZNF753|zfp-41   \n",
       "10          FBN1                                              AI2A4   \n",
       "11          FBN1                                            fSAP121   \n",
       "12          FBN1  ORF1ab|ORF1ab-nsp9|SARS-CoV2 nsp9|SARS-CoV-2 n...   \n",
       "13           HGS  ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...   \n",
       "14          SGTB  ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...   \n",
       "15          GET4  ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...   \n",
       "16          FBN1  ORF1ab|ORF1ab-nsp9|SARS-CoV2 nsp9|SARS-CoV-2 n...   \n",
       "17          FBN1                                              CDW14   \n",
       "18          FBN1                                   RIG-I|RIGI|RLR-1   \n",
       "19          FBN1                          FBG1|FBX2|Fbs1|NFB42|OCP1   \n",
       "20         FBXO2  ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...   \n",
       "21          FBN1                                      MUL|POB1|TEF3   \n",
       "22          FBN1                                    BSAS|HOX1|HOX1F   \n",
       "23          FBN1                                      ZNF753|zfp-41   \n",
       "24          FBN1  ARCI11|HAI|MT-SP1|MTSP1|PRSS14|SNC19|TADG15|TM...   \n",
       "25          FBN1                                                  -   \n",
       "26          FBN1                                              AI2A4   \n",
       "27          FBN1                          FBG1|FBX2|Fbs1|NFB42|OCP1   \n",
       "28          FBN1                                   HPRG|HRGP|THPH11   \n",
       "29          FBN1                                                  -   \n",
       "30          FBN1        D6S115E|D6S2244E|H2-KE4|HKE4|KE4|RING5|ZIP7   \n",
       "31          FBN1                                                  -   \n",
       "32          FBN1                                     bA350J20.1|p42   \n",
       "33          FBN1                                        CUL-3|PHA2E   \n",
       "34          FBN1                          BPM-L|BPM90|RANBPM|RanBP7   \n",
       "35          FBN1                      TGN38|TGN46|TGN48|TGN51|TTGN2   \n",
       "36          FBN1                                              MRX91   \n",
       "37          FBN1                      DYRK|DYRK1|HP86|MNB|MNBH|MRD7   \n",
       "38          FBN1                            CSC-21K|DDC8|KIAA1731NL   \n",
       "\n",
       "                                           synonyms_b  organism_a  organism_b  \n",
       "0                                   MAGP|MAGP-1|MAGP1        9606        9606  \n",
       "1   ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...        9606        9606  \n",
       "2   ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...        9606        9606  \n",
       "3   ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...        9606        9606  \n",
       "4   ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...        9606        9606  \n",
       "5   ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...        9606        9606  \n",
       "6   ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...        9606        9606  \n",
       "7   ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...        9606        9606  \n",
       "8   ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...        9606        9606  \n",
       "9   ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...        9606        9606  \n",
       "10  ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...        9606        9606  \n",
       "11  ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...        9606        9606  \n",
       "12  ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...     2697049        9606  \n",
       "13                                                HRS        9606        9606  \n",
       "14                                               SGT2        9606        9606  \n",
       "15                                  C7orf20|CEE|TRC35        9606        9606  \n",
       "16  ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...     2697049        9606  \n",
       "17  ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...        9606        9606  \n",
       "18  ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...        9606        9606  \n",
       "19  ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...        9606        9606  \n",
       "20                          FBG1|FBX2|Fbs1|NFB42|OCP1        9606        9606  \n",
       "21  ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...        9606        9606  \n",
       "22  ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...        9606        9606  \n",
       "23  ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...        9606        9606  \n",
       "24  ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...        9606        9606  \n",
       "25  ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...        9606        9606  \n",
       "26  ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...        9606        9606  \n",
       "27  ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...        9606        9606  \n",
       "28  ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...        9606        9606  \n",
       "29  ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...        9606        9606  \n",
       "30  ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...        9606        9606  \n",
       "31  ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...        9606        9606  \n",
       "32  ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...        9606        9606  \n",
       "33  ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...        9606        9606  \n",
       "34  ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...        9606        9606  \n",
       "35  ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...        9606        9606  \n",
       "36  ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...        9606        9606  \n",
       "37  ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...        9606        9606  \n",
       "38  ACMICD|ECTOL1|FBN|GPHYSD2|MASS|MFS1|OCTD|SGS|S...        9606        9606  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "instance.fetch_single(\n",
    "    query={\n",
    "        \"accessKey\": biogrid_api_key,\n",
    "        \"geneList\": ['MFS1'],\n",
    "        \"taxId\": \"9606\",\n",
    "    },\n",
    "    method=\"interactions\",\n",
    "    parse=True,\n",
    "    to_dataframe=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "2cf2850e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cache_key: 103_interactions\n",
      "Prepared request: https://webservice.thebiogrid.org/interactions?accessKey=87aee10241ab2fcc6e497352e5024380&id=103&start=0&max=10000&interSpeciesExclude=False&selfInteractionsExclude=False&includeEvidence=False&searchBiogridIds=False&taxId=All&searchIds=False&format=json\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>interaction_b</th>\n",
       "      <th>synonyms_a</th>\n",
       "      <th>synonyms_b</th>\n",
       "      <th>organism_a</th>\n",
       "      <th>organism_b</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>FLNC</td>\n",
       "      <td>JNKK|JNKK1|MAPKK4|MEK4|MKK4|PRKMK4|SAPKK-1|SAP...</td>\n",
       "      <td>ABP-280|ABP280A|ABPA|ABPL|FLN2|MFM5|MPD4</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACTN2</td>\n",
       "      <td>CMD1DD|CMH22|MYOP|RCM4</td>\n",
       "      <td>CMD1AA</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FNTA</td>\n",
       "      <td>ACTRI|ACVR1A|ACVRLK2|ALK2|FOP|SKR1|TSRI</td>\n",
       "      <td>FPTA|PGGT1A|PTAR2</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PML</td>\n",
       "      <td>DCML|IMD21|MONOMAC|NFE1B</td>\n",
       "      <td>MYL|PP8675|RNF71|TRIM19</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>STAT3</td>\n",
       "      <td>REPA2|RP-A p32|RP-A p34|RPA32</td>\n",
       "      <td>ADMIO|APRF|HIES</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9995</th>\n",
       "      <td>RpS17</td>\n",
       "      <td>CG13167|Dmel\\CG13167</td>\n",
       "      <td>BcDNA:RE44119|CG3922|D9|Dmel\\CG3922|M|M(3)67|M...</td>\n",
       "      <td>7227</td>\n",
       "      <td>7227</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9996</th>\n",
       "      <td>RpL15</td>\n",
       "      <td>CG13167|Dmel\\CG13167</td>\n",
       "      <td>80Fi|CG17420|CG40199|DmRpL15|Dmel\\CG17420|L15|...</td>\n",
       "      <td>7227</td>\n",
       "      <td>7227</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9997</th>\n",
       "      <td>CG15423</td>\n",
       "      <td>Dmel\\CG13168</td>\n",
       "      <td>Dmel\\CG15423</td>\n",
       "      <td>7227</td>\n",
       "      <td>7227</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9998</th>\n",
       "      <td>CG15549</td>\n",
       "      <td>Dmel\\CG13168</td>\n",
       "      <td>BcDNA:AT07554|Dmel\\CG15549</td>\n",
       "      <td>7227</td>\n",
       "      <td>7227</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9999</th>\n",
       "      <td>CG15578</td>\n",
       "      <td>Dmel\\CG13168</td>\n",
       "      <td>Dmel\\CG15578</td>\n",
       "      <td>7227</td>\n",
       "      <td>7227</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10000 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     interaction_b                                         synonyms_a  \\\n",
       "0             FLNC  JNKK|JNKK1|MAPKK4|MEK4|MKK4|PRKMK4|SAPKK-1|SAP...   \n",
       "1            ACTN2                             CMD1DD|CMH22|MYOP|RCM4   \n",
       "2             FNTA            ACTRI|ACVR1A|ACVRLK2|ALK2|FOP|SKR1|TSRI   \n",
       "3              PML                           DCML|IMD21|MONOMAC|NFE1B   \n",
       "4            STAT3                      REPA2|RP-A p32|RP-A p34|RPA32   \n",
       "...            ...                                                ...   \n",
       "9995         RpS17                               CG13167|Dmel\\CG13167   \n",
       "9996         RpL15                               CG13167|Dmel\\CG13167   \n",
       "9997       CG15423                                       Dmel\\CG13168   \n",
       "9998       CG15549                                       Dmel\\CG13168   \n",
       "9999       CG15578                                       Dmel\\CG13168   \n",
       "\n",
       "                                             synonyms_b  organism_a  \\\n",
       "0              ABP-280|ABP280A|ABPA|ABPL|FLN2|MFM5|MPD4        9606   \n",
       "1                                                CMD1AA        9606   \n",
       "2                                     FPTA|PGGT1A|PTAR2        9606   \n",
       "3                               MYL|PP8675|RNF71|TRIM19        9606   \n",
       "4                                       ADMIO|APRF|HIES        9606   \n",
       "...                                                 ...         ...   \n",
       "9995  BcDNA:RE44119|CG3922|D9|Dmel\\CG3922|M|M(3)67|M...        7227   \n",
       "9996  80Fi|CG17420|CG40199|DmRpL15|Dmel\\CG17420|L15|...        7227   \n",
       "9997                                       Dmel\\CG15423        7227   \n",
       "9998                         BcDNA:AT07554|Dmel\\CG15549        7227   \n",
       "9999                                       Dmel\\CG15578        7227   \n",
       "\n",
       "      organism_b  \n",
       "0           9606  \n",
       "1           9606  \n",
       "2           9606  \n",
       "3           9606  \n",
       "4           9606  \n",
       "...          ...  \n",
       "9995        7227  \n",
       "9996        7227  \n",
       "9997        7227  \n",
       "9998        7227  \n",
       "9999        7227  \n",
       "\n",
       "[10000 rows x 5 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "instance.fetch_single(\n",
    "    query={\n",
    "        \"id\" : \"103\",\n",
    "        \"accessKey\": biogrid_api_key,\n",
    "    },\n",
    "    method=\"interactions\",\n",
    "    parse=True,\n",
    "    to_dataframe=True\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "565e82a3",
   "metadata": {},
   "source": [
    "## Brenda"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "5b5684c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "from src.brenda import BrendaInstance\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "7f5f3a53",
   "metadata": {},
   "outputs": [],
   "source": [
    "from dotenv import load_dotenv\n",
    "load_dotenv()\n",
    "brenda_email = os.getenv(\"brenda_email\")\n",
    "brenda_password = os.getenv(\"brenda_password\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "b9707635",
   "metadata": {},
   "outputs": [],
   "source": [
    "instance = BrendaInstance(\n",
    "    email=brenda_email,\n",
    "    password=brenda_password,\n",
    "    #use_config=False,  # Set to True if you want to use a config file\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "0e9e87d1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cache_key: 1.1.1.10_Homo sapiens_getKmValue\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ec</th>\n",
       "      <th>organism</th>\n",
       "      <th>km</th>\n",
       "      <th>km_max</th>\n",
       "      <th>substrate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.1.1.10</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>0.001</td>\n",
       "      <td>None</td>\n",
       "      <td>NADP+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1.1.1.10</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>0.002</td>\n",
       "      <td>None</td>\n",
       "      <td>NADPH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1.1.1.10</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>0.002</td>\n",
       "      <td>None</td>\n",
       "      <td>NADPH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1.1.1.10</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>0.077</td>\n",
       "      <td>None</td>\n",
       "      <td>diacetyl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1.1.1.10</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>0.077</td>\n",
       "      <td>None</td>\n",
       "      <td>diacetyl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1.1.1.10</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>0.0945</td>\n",
       "      <td>None</td>\n",
       "      <td>diacetyl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1.1.1.10</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>0.17</td>\n",
       "      <td>None</td>\n",
       "      <td>NADH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1.1.1.10</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>0.21</td>\n",
       "      <td>None</td>\n",
       "      <td>L-xylulose</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1.1.1.10</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>0.22</td>\n",
       "      <td>None</td>\n",
       "      <td>diacetyl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1.1.1.10</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>0.85</td>\n",
       "      <td>None</td>\n",
       "      <td>NAD+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>1.1.1.10</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>1.2</td>\n",
       "      <td>None</td>\n",
       "      <td>DL-glyceraldehyde</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>1.1.1.10</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>1.5</td>\n",
       "      <td>None</td>\n",
       "      <td>acetoin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>1.1.1.10</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>1.6</td>\n",
       "      <td>None</td>\n",
       "      <td>dihydroxyacetone</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>1.1.1.10</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>1.9</td>\n",
       "      <td>None</td>\n",
       "      <td>D-threose</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>1.1.1.10</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>2.4</td>\n",
       "      <td>None</td>\n",
       "      <td>D-erythrose</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>1.1.1.10</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>2.5</td>\n",
       "      <td>None</td>\n",
       "      <td>L-erythrulose</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>1.1.1.10</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>4.2</td>\n",
       "      <td>None</td>\n",
       "      <td>L-threose</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>1.1.1.10</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>4.8</td>\n",
       "      <td>None</td>\n",
       "      <td>D-xylulose</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>1.1.1.10</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>5</td>\n",
       "      <td>None</td>\n",
       "      <td>D-ribulose</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>1.1.1.10</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>18</td>\n",
       "      <td>None</td>\n",
       "      <td>D-threitol</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>1.1.1.10</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>25</td>\n",
       "      <td>None</td>\n",
       "      <td>xylitol</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>1.1.1.10</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>26</td>\n",
       "      <td>None</td>\n",
       "      <td>diacetyl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>1.1.1.10</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>31</td>\n",
       "      <td>None</td>\n",
       "      <td>diacetyl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>1.1.1.10</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>95</td>\n",
       "      <td>None</td>\n",
       "      <td>NADPH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>1.1.1.10</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>97</td>\n",
       "      <td>None</td>\n",
       "      <td>NADPH</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          ec      organism      km km_max          substrate\n",
       "0   1.1.1.10  Homo sapiens   0.001   None              NADP+\n",
       "1   1.1.1.10  Homo sapiens   0.002   None              NADPH\n",
       "2   1.1.1.10  Homo sapiens   0.002   None              NADPH\n",
       "3   1.1.1.10  Homo sapiens   0.077   None           diacetyl\n",
       "4   1.1.1.10  Homo sapiens   0.077   None           diacetyl\n",
       "5   1.1.1.10  Homo sapiens  0.0945   None           diacetyl\n",
       "6   1.1.1.10  Homo sapiens    0.17   None               NADH\n",
       "7   1.1.1.10  Homo sapiens    0.21   None         L-xylulose\n",
       "8   1.1.1.10  Homo sapiens    0.22   None           diacetyl\n",
       "9   1.1.1.10  Homo sapiens    0.85   None               NAD+\n",
       "10  1.1.1.10  Homo sapiens     1.2   None  DL-glyceraldehyde\n",
       "11  1.1.1.10  Homo sapiens     1.5   None            acetoin\n",
       "12  1.1.1.10  Homo sapiens     1.6   None   dihydroxyacetone\n",
       "13  1.1.1.10  Homo sapiens     1.9   None          D-threose\n",
       "14  1.1.1.10  Homo sapiens     2.4   None        D-erythrose\n",
       "15  1.1.1.10  Homo sapiens     2.5   None      L-erythrulose\n",
       "16  1.1.1.10  Homo sapiens     4.2   None          L-threose\n",
       "17  1.1.1.10  Homo sapiens     4.8   None         D-xylulose\n",
       "18  1.1.1.10  Homo sapiens       5   None         D-ribulose\n",
       "19  1.1.1.10  Homo sapiens      18   None         D-threitol\n",
       "20  1.1.1.10  Homo sapiens      25   None            xylitol\n",
       "21  1.1.1.10  Homo sapiens      26   None           diacetyl\n",
       "22  1.1.1.10  Homo sapiens      31   None           diacetyl\n",
       "23  1.1.1.10  Homo sapiens      95   None              NADPH\n",
       "24  1.1.1.10  Homo sapiens      97   None              NADPH"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "instance.fetch_single(\n",
    "    query={\n",
    "            \"ecNumber\": \"1.1.1.10\",\n",
    "            \"organism\": \"Homo sapiens\",\n",
    "            #\"kcatKmValue\": \"0.5\"\n",
    "        }, \n",
    "    method=\"getKmValue\",\n",
    "    parse=True,\n",
    "    to_dataframe=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "3d4f521f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cache_key: 25-37_Homo sapiens_getTemperatureRange\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ec</th>\n",
       "      <th>organism</th>\n",
       "      <th>temperature_range</th>\n",
       "      <th>temperature_range_max</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.1.1.188</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>25</td>\n",
       "      <td>37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2.1.1.77</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>25</td>\n",
       "      <td>65</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2.4.1.38</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>25</td>\n",
       "      <td>45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2.4.1.90</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>25</td>\n",
       "      <td>45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2.7.1.21</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>25</td>\n",
       "      <td>37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>2.8.2.2</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>25</td>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>3.1.8.1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>25</td>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>3.1.11.1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>25</td>\n",
       "      <td>37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>3.4.21.108</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>25</td>\n",
       "      <td>55</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>3.5.1.1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>25</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>3.6.4.10</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>25</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>3.13.2.1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>25</td>\n",
       "      <td>49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>4.6.1.19</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>25</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>5.6.2.1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>25</td>\n",
       "      <td>37</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            ec      organism temperature_range temperature_range_max\n",
       "0    1.1.1.188  Homo sapiens                25                    37\n",
       "1     2.1.1.77  Homo sapiens                25                    65\n",
       "2     2.4.1.38  Homo sapiens                25                    45\n",
       "3     2.4.1.90  Homo sapiens                25                    45\n",
       "4     2.7.1.21  Homo sapiens                25                    37\n",
       "5      2.8.2.2  Homo sapiens                25                    50\n",
       "6      3.1.8.1  Homo sapiens                25                    50\n",
       "7     3.1.11.1  Homo sapiens                25                    37\n",
       "8   3.4.21.108  Homo sapiens                25                    55\n",
       "9      3.5.1.1  Homo sapiens                25                    60\n",
       "10    3.6.4.10  Homo sapiens                25                    70\n",
       "11    3.13.2.1  Homo sapiens                25                    49\n",
       "12    4.6.1.19  Homo sapiens                25                    30\n",
       "13     5.6.2.1  Homo sapiens                25                    37"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "instance.fetch_single(\n",
    "    query=\n",
    "        {\n",
    "            #\"ecNumber\": \"1.1.1.10\",\n",
    "            \"organism\": \"Homo sapiens\",\n",
    "            \"temperatureRange\": \"25-37\"\n",
    "        }, \n",
    "    method=\"getTemperatureRange\",\n",
    "    parse=True,\n",
    "    to_dataframe=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "2ebcdcd5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Usage: To fetch data from BRENDA, use the following parameters.\n",
      "        Example:\n",
      "            - fetch(query={}, methods=[\"getKmValue\", \"getIc50Value\"])\n",
      "        Available methods: getKmValue, getIc50Value, getKcatKmValue, getKiValue, getPhRange, getPhOptimum, getPhStability, getCofactor, getTemperatureOptimum, getTemperatureStability, getTemperatureRange\n",
      "\n",
      "For more information about each method, please refer to the BRENDA documentation.\n",
      "Or use `show_method({method_name})` to see the parameters required for each method.\n"
     ]
    }
   ],
   "source": [
    "print(instance.query_usage())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "6580cab2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cache_key: 1.1.1.1_Escherichia coli_getKmValue\n",
      "type(response)=[{'literature': [660650], 'substrate': 'ethanol', 'kmValue': '10.6', 'kmValueMaximum': None, 'commentary': None, 'organism': 'Escherichia coli', 'ecNumber': '1.1.1.1', 'ligandStructureId': 65}]\n",
      "Cache_key: 1.1.1.1_Escherichia coli_getIc50Value\n",
      "type(response)=[]\n",
      "Cache_key: 1.1.1.1_Escherichia coli_getKcatKmValue\n",
      "type(response)=[]\n",
      "Cache_key: 1.1.1.1_Escherichia coli_getKiValue\n",
      "type(response)=[{'kiValueMaximum': None, 'literature': [660650], 'kiValue': '18.26', 'commentary': None, 'inhibitor': '4-Methylpyrazole', 'organism': 'Escherichia coli', 'ecNumber': '1.1.1.1', 'ligandStructureId': 1804}]\n",
      "Cache_key: 1.1.1.1_Escherichia coli_getPhRange\n",
      "type(response)=[]\n",
      "Cache_key: 1.1.1.1_Escherichia coli_getPhOptimum\n",
      "type(response)=[]\n",
      "Cache_key: 1.1.1.1_Escherichia coli_getPhStability\n",
      "type(response)=[]\n",
      "Cache_key: 1.1.1.1_Escherichia coli_getCofactor\n",
      "type(response)=[{'literature': [660650], 'commentary': 'strongly preferred as cofactor', 'ecNumber': '1.1.1.1', 'organism': 'Escherichia coli', 'cofactor': 'NAD+', 'ligandStructureId': 4}, {'literature': [763497], 'commentary': None, 'ecNumber': '1.1.1.1', 'organism': 'Escherichia coli', 'cofactor': 'NAD+', 'ligandStructureId': 4}, {'literature': [763497], 'commentary': None, 'ecNumber': '1.1.1.1', 'organism': 'Escherichia coli', 'cofactor': 'NADH', 'ligandStructureId': 3}]\n",
      "Cache_key: 1.1.1.1_Escherichia coli_getTemperatureOptimum\n",
      "type(response)=[]\n",
      "Cache_key: 1.1.1.1_Escherichia coli_getTemperatureStability\n",
      "type(response)=[]\n",
      "Cache_key: 1.1.1.1_Escherichia coli_getTemperatureRange\n",
      "type(response)=[]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'getKmValue': {'literature': 'list(int)',\n",
       "  'substrate': 'str',\n",
       "  'kmValue': 'str',\n",
       "  'kmValueMaximum': 'NoneType',\n",
       "  'commentary': 'NoneType',\n",
       "  'organism': 'str',\n",
       "  'ecNumber': 'str',\n",
       "  'ligandStructureId': 'int'},\n",
       " 'getIc50Value': {},\n",
       " 'getKcatKmValue': {},\n",
       " 'getKiValue': {'kiValueMaximum': 'NoneType',\n",
       "  'literature': 'list(int)',\n",
       "  'kiValue': 'str',\n",
       "  'commentary': 'NoneType',\n",
       "  'inhibitor': 'str',\n",
       "  'organism': 'str',\n",
       "  'ecNumber': 'str',\n",
       "  'ligandStructureId': 'int'},\n",
       " 'getPhRange': {},\n",
       " 'getPhOptimum': {},\n",
       " 'getPhStability': {},\n",
       " 'getCofactor': {'literature': 'list(int)',\n",
       "  'commentary': 'str',\n",
       "  'ecNumber': 'str',\n",
       "  'organism': 'str',\n",
       "  'cofactor': 'str',\n",
       "  'ligandStructureId': 'int'},\n",
       " 'getTemperatureOptimum': {},\n",
       " 'getTemperatureStability': {},\n",
       " 'getTemperatureRange': {}}"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "instance.get_dummy()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3ddf5ea7",
   "metadata": {},
   "source": [
    "## Chembl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "8f530a89",
   "metadata": {},
   "outputs": [],
   "source": [
    "from src.chembl import ChEMBLInterface"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1a799566",
   "metadata": {},
   "outputs": [],
   "source": [
    "instance = ChEMBLInterface()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5d1d6a36",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cache_key: CHEMBL1824_activity\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>assay_id</th>\n",
       "      <th>type</th>\n",
       "      <th>units</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHEMBL650385</td>\n",
       "      <td>IC50</td>\n",
       "      <td>uM</td>\n",
       "      <td>2.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHEMBL880259</td>\n",
       "      <td>IC50</td>\n",
       "      <td>uM</td>\n",
       "      <td>2.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHEMBL1073381</td>\n",
       "      <td>IC50</td>\n",
       "      <td>uM</td>\n",
       "      <td>2.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHEMBL1932748</td>\n",
       "      <td>IC50</td>\n",
       "      <td>nM</td>\n",
       "      <td>2421.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHEMBL2320919</td>\n",
       "      <td>IC50</td>\n",
       "      <td>uM</td>\n",
       "      <td>2.38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CHEMBL3993804</td>\n",
       "      <td>IC50</td>\n",
       "      <td>nM</td>\n",
       "      <td>2400.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CHEMBL5444475</td>\n",
       "      <td>IC50</td>\n",
       "      <td>nM</td>\n",
       "      <td>2400.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        assay_id  type units   value\n",
       "0   CHEMBL650385  IC50    uM     2.4\n",
       "1   CHEMBL880259  IC50    uM     2.4\n",
       "2  CHEMBL1073381  IC50    uM     2.4\n",
       "3  CHEMBL1932748  IC50    nM  2421.0\n",
       "4  CHEMBL2320919  IC50    uM    2.38\n",
       "5  CHEMBL3993804  IC50    nM  2400.0\n",
       "6  CHEMBL5444475  IC50    nM  2400.0"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "instance.fetch_single(\n",
    "    query={\"target_chembl_id\": \"CHEMBL1824\", \"pchembl_value\": 5.62}, \n",
    "    method=\"activity\",\n",
    "    pages_to_fetch=1, \n",
    "    parse=True,\n",
    "    to_dataframe=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "0b163aae",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cache_key: CHEMBL1824_binding_site\n",
      "Fetching page: https://www.ebi.ac.uk/chembl/api/data/binding_site?target_chembl_id=CHEMBL1824&format=json for method binding_site with pages_to_fetch=1\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>site_components</th>\n",
       "      <th>site_id</th>\n",
       "      <th>site_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>[{'component_id': 4909, 'domain': {'domain_des...</td>\n",
       "      <td>2</td>\n",
       "      <td>UDP-glucuronosyltransferase 1-10, UDPGT domain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>[{'component_id': 5575, 'domain': {'domain_des...</td>\n",
       "      <td>3</td>\n",
       "      <td>Mitogen-activated protein kinase 8, Pkinase do...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>[{'component_id': 118, 'domain': {'domain_desc...</td>\n",
       "      <td>4</td>\n",
       "      <td>Inosine-5'-monophosphate dehydrogenase 1, IMPD...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>[{'component_id': 397, 'domain': {'domain_desc...</td>\n",
       "      <td>5</td>\n",
       "      <td>Dopamine D1 receptor, 7tm_1 domain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>[{'component_id': 218, 'domain': {'domain_desc...</td>\n",
       "      <td>6</td>\n",
       "      <td>GABA transporter 1, SNF domain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>[{'component_id': 1835, 'domain': {'domain_des...</td>\n",
       "      <td>7</td>\n",
       "      <td>LDL-associated phospholipase A2, PAF-AH_p_II d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>[{'component_id': 2632, 'domain': {'domain_des...</td>\n",
       "      <td>8</td>\n",
       "      <td>Purinergic receptor P2Y1, 7tm_1 domain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>[{'component_id': 873, 'domain': {'domain_desc...</td>\n",
       "      <td>9</td>\n",
       "      <td>3-phosphoinositide dependent protein kinase-1,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>[{'component_id': 970, 'domain': {'domain_desc...</td>\n",
       "      <td>10</td>\n",
       "      <td>Tyrosine-protein kinase CSK, Pkinase_Tyr domain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>[{'component_id': 3106, 'domain': {'domain_des...</td>\n",
       "      <td>11</td>\n",
       "      <td>Carbonic anhydrase VA, Carb_anhydrase domain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>[{'component_id': 2231, 'domain': {'domain_des...</td>\n",
       "      <td>12</td>\n",
       "      <td>Phosphodiesterase 3B, PDEase_I domain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>[{'component_id': 1388, 'domain': {'domain_des...</td>\n",
       "      <td>13</td>\n",
       "      <td>Glycine transporter 2, SNF domain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>[{'component_id': 2230, 'domain': {'domain_des...</td>\n",
       "      <td>14</td>\n",
       "      <td>11-beta-hydroxysteroid dehydrogenase 1, adh_sh...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>[{'component_id': 1377, 'domain': {'domain_des...</td>\n",
       "      <td>15</td>\n",
       "      <td>Monoamine oxidase B, Amino_oxidase domain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>[{'component_id': 1711, 'domain': {'domain_des...</td>\n",
       "      <td>16</td>\n",
       "      <td>Lactase-glycosylceramidase, Glyco_hydro_1 domain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>[{'component_id': 1188, 'domain': {'domain_des...</td>\n",
       "      <td>17</td>\n",
       "      <td>Mu opioid receptor, 7tm_1 domain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>[{'component_id': 2704, 'domain': {'domain_des...</td>\n",
       "      <td>18</td>\n",
       "      <td>Thymidylate kinase, Thymidylate_kin domain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>[{'component_id': 3285, 'domain': {'domain_des...</td>\n",
       "      <td>19</td>\n",
       "      <td>Excitatory amino acid transporter 2, SDF domain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>[{'component_id': 2600, 'domain': {'domain_des...</td>\n",
       "      <td>20</td>\n",
       "      <td>Group IID secretory phospholipase A2, Phosphol...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>[{'component_id': 2748, 'domain': {'domain_des...</td>\n",
       "      <td>21</td>\n",
       "      <td>Serotonin 2c (5-HT2c) receptor, 7tm_1 domain</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      site_components  site_id  \\\n",
       "0   [{'component_id': 4909, 'domain': {'domain_des...        2   \n",
       "1   [{'component_id': 5575, 'domain': {'domain_des...        3   \n",
       "2   [{'component_id': 118, 'domain': {'domain_desc...        4   \n",
       "3   [{'component_id': 397, 'domain': {'domain_desc...        5   \n",
       "4   [{'component_id': 218, 'domain': {'domain_desc...        6   \n",
       "5   [{'component_id': 1835, 'domain': {'domain_des...        7   \n",
       "6   [{'component_id': 2632, 'domain': {'domain_des...        8   \n",
       "7   [{'component_id': 873, 'domain': {'domain_desc...        9   \n",
       "8   [{'component_id': 970, 'domain': {'domain_desc...       10   \n",
       "9   [{'component_id': 3106, 'domain': {'domain_des...       11   \n",
       "10  [{'component_id': 2231, 'domain': {'domain_des...       12   \n",
       "11  [{'component_id': 1388, 'domain': {'domain_des...       13   \n",
       "12  [{'component_id': 2230, 'domain': {'domain_des...       14   \n",
       "13  [{'component_id': 1377, 'domain': {'domain_des...       15   \n",
       "14  [{'component_id': 1711, 'domain': {'domain_des...       16   \n",
       "15  [{'component_id': 1188, 'domain': {'domain_des...       17   \n",
       "16  [{'component_id': 2704, 'domain': {'domain_des...       18   \n",
       "17  [{'component_id': 3285, 'domain': {'domain_des...       19   \n",
       "18  [{'component_id': 2600, 'domain': {'domain_des...       20   \n",
       "19  [{'component_id': 2748, 'domain': {'domain_des...       21   \n",
       "\n",
       "                                            site_name  \n",
       "0      UDP-glucuronosyltransferase 1-10, UDPGT domain  \n",
       "1   Mitogen-activated protein kinase 8, Pkinase do...  \n",
       "2   Inosine-5'-monophosphate dehydrogenase 1, IMPD...  \n",
       "3                  Dopamine D1 receptor, 7tm_1 domain  \n",
       "4                      GABA transporter 1, SNF domain  \n",
       "5   LDL-associated phospholipase A2, PAF-AH_p_II d...  \n",
       "6              Purinergic receptor P2Y1, 7tm_1 domain  \n",
       "7   3-phosphoinositide dependent protein kinase-1,...  \n",
       "8     Tyrosine-protein kinase CSK, Pkinase_Tyr domain  \n",
       "9        Carbonic anhydrase VA, Carb_anhydrase domain  \n",
       "10              Phosphodiesterase 3B, PDEase_I domain  \n",
       "11                  Glycine transporter 2, SNF domain  \n",
       "12  11-beta-hydroxysteroid dehydrogenase 1, adh_sh...  \n",
       "13          Monoamine oxidase B, Amino_oxidase domain  \n",
       "14   Lactase-glycosylceramidase, Glyco_hydro_1 domain  \n",
       "15                   Mu opioid receptor, 7tm_1 domain  \n",
       "16         Thymidylate kinase, Thymidylate_kin domain  \n",
       "17    Excitatory amino acid transporter 2, SDF domain  \n",
       "18  Group IID secretory phospholipase A2, Phosphol...  \n",
       "19       Serotonin 2c (5-HT2c) receptor, 7tm_1 domain  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "instance.fetch_single(\n",
    "    query={\"target_chembl_id\": \"CHEMBL1824\"}, \n",
    "    method=\"binding_site\",\n",
    "    pages_to_fetch=1, \n",
    "    parse=True,\n",
    "    to_dataframe=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "9d2f646f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cache_key: CHEMBL1824_activity\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'action_type': 'NoneType',\n",
       " 'activity_comment': 'NoneType',\n",
       " 'activity_id': 'int',\n",
       " 'activity_properties': 'list',\n",
       " 'assay_chembl_id': 'str',\n",
       " 'assay_description': 'str',\n",
       " 'assay_type': 'str',\n",
       " 'assay_variant_accession': 'NoneType',\n",
       " 'assay_variant_mutation': 'NoneType',\n",
       " 'bao_endpoint': 'str',\n",
       " 'bao_format': 'str',\n",
       " 'bao_label': 'str',\n",
       " 'canonical_smiles': 'str',\n",
       " 'data_validity_comment': 'NoneType',\n",
       " 'data_validity_description': 'NoneType',\n",
       " 'document_chembl_id': 'str',\n",
       " 'document_journal': 'str',\n",
       " 'document_year': 'int',\n",
       " 'ligand_efficiency': 'NoneType',\n",
       " 'molecule_chembl_id': 'str',\n",
       " 'molecule_pref_name': 'NoneType',\n",
       " 'parent_molecule_chembl_id': 'str',\n",
       " 'pchembl_value': 'str',\n",
       " 'potential_duplicate': 'int',\n",
       " 'qudt_units': 'str',\n",
       " 'record_id': 'int',\n",
       " 'relation': 'str',\n",
       " 'src_id': 'int',\n",
       " 'standard_flag': 'int',\n",
       " 'standard_relation': 'str',\n",
       " 'standard_text_value': 'NoneType',\n",
       " 'standard_type': 'str',\n",
       " 'standard_units': 'str',\n",
       " 'standard_upper_value': 'NoneType',\n",
       " 'standard_value': 'str',\n",
       " 'target_chembl_id': 'str',\n",
       " 'target_organism': 'str',\n",
       " 'target_pref_name': 'str',\n",
       " 'target_tax_id': 'str',\n",
       " 'text_value': 'NoneType',\n",
       " 'toid': 'NoneType',\n",
       " 'type': 'str',\n",
       " 'units': 'str',\n",
       " 'uo_units': 'str',\n",
       " 'upper_value': 'NoneType',\n",
       " 'value': 'str'}"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "instance.get_dummy()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1a940eaa",
   "metadata": {},
   "source": [
    "## Chebi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "be87135f",
   "metadata": {},
   "outputs": [],
   "source": [
    "from src.chebi import ChEBIInterface"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f68cd545",
   "metadata": {},
   "outputs": [],
   "source": [
    "instance = ChEBIInterface()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ed4fd346",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cache_key: ['CHEBI:18357', 'CHEBI:29033']_compounds\n",
      "Prepared url: https://www.ebi.ac.uk/chebi/beta/api/public/compounds/?chebi_ids=CHEBI%3A18357%2CCHEBI%3A29033\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHEBI:18357</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHEBI:29033</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            id\n",
       "0  CHEBI:18357\n",
       "1  CHEBI:29033"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "instance.fetch_single(\n",
    "    query=[\"CHEBI:18357\",\"CHEBI:29033\"], \n",
    "    method=\"compounds\", \n",
    "    parse=True,\n",
    "    to_dataframe=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "8820419b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cache_key: {\"chebi_ids\": [\"CHEBI:18357\", \"CHEBI:29033\"]}_compounds\n",
      "Cache_key: ['CHEBI:18357', 'CHEBI:29033']_compounds\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHEBI:18357</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHEBI:29033</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            id\n",
       "0  CHEBI:18357\n",
       "1  CHEBI:29033"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "instance.fetch_batch(\n",
    "    queries=[\n",
    "        {\"chebi_ids\": [\"CHEBI:18357\", \"CHEBI:29033\"]}\n",
    "    ],\n",
    "    method=\"compounds\",\n",
    "    parse=True,\n",
    "    to_dataframe=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "7f14d4e1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cache_key: {\"term\": \"paracetamol\"}_es_search\n",
      "Cache_key: paracetamol_es_search\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>_index</th>\n",
       "      <th>_type</th>\n",
       "      <th>_id</th>\n",
       "      <th>_score</th>\n",
       "      <th>_source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chebi-prod-compounds-16042025</td>\n",
       "      <td>_doc</td>\n",
       "      <td>46195</td>\n",
       "      <td>66.144830</td>\n",
       "      <td>{'chebi_accession': 'CHEBI:46195', 'name': 'pa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chebi-prod-compounds-16042025</td>\n",
       "      <td>_doc</td>\n",
       "      <td>74529</td>\n",
       "      <td>58.117500</td>\n",
       "      <td>{'chebi_accession': 'CHEBI:74529', 'name': 'an...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chebi-prod-compounds-16042025</td>\n",
       "      <td>_doc</td>\n",
       "      <td>32635</td>\n",
       "      <td>32.586680</td>\n",
       "      <td>{'chebi_accession': 'CHEBI:32635', 'name': 'pa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chebi-prod-compounds-16042025</td>\n",
       "      <td>_doc</td>\n",
       "      <td>32636</td>\n",
       "      <td>31.397276</td>\n",
       "      <td>{'chebi_accession': 'CHEBI:32636', 'name': 'ac...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chebi-prod-compounds-16042025</td>\n",
       "      <td>_doc</td>\n",
       "      <td>139476</td>\n",
       "      <td>31.277231</td>\n",
       "      <td>{'chebi_accession': 'CHEBI:139476', 'name': '3...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chebi-prod-compounds-16042025</td>\n",
       "      <td>_doc</td>\n",
       "      <td>180460</td>\n",
       "      <td>29.612461</td>\n",
       "      <td>{'chebi_accession': 'CHEBI:180460', 'name': 'P...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chebi-prod-compounds-16042025</td>\n",
       "      <td>_doc</td>\n",
       "      <td>133066</td>\n",
       "      <td>26.956810</td>\n",
       "      <td>{'chebi_accession': 'CHEBI:133066', 'name': '&lt;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chebi-prod-compounds-16042025</td>\n",
       "      <td>_doc</td>\n",
       "      <td>133067</td>\n",
       "      <td>25.853018</td>\n",
       "      <td>{'chebi_accession': 'CHEBI:133067', 'name': '&lt;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chebi-prod-compounds-16042025</td>\n",
       "      <td>_doc</td>\n",
       "      <td>133435</td>\n",
       "      <td>20.886192</td>\n",
       "      <td>{'chebi_accession': 'CHEBI:133435', 'name': '&lt;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chebi-prod-compounds-16042025</td>\n",
       "      <td>_doc</td>\n",
       "      <td>133438</td>\n",
       "      <td>20.886192</td>\n",
       "      <td>{'chebi_accession': 'CHEBI:133438', 'name': '&lt;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chebi-prod-compounds-16042025</td>\n",
       "      <td>_doc</td>\n",
       "      <td>82942</td>\n",
       "      <td>1.077858</td>\n",
       "      <td>{'chebi_accession': 'CHEBI:82942', 'name': '4-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chebi-prod-compounds-16042025</td>\n",
       "      <td>_doc</td>\n",
       "      <td>82944</td>\n",
       "      <td>0.949729</td>\n",
       "      <td>{'chebi_accession': 'CHEBI:82944', 'name': '2-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chebi-prod-compounds-16042025</td>\n",
       "      <td>_doc</td>\n",
       "      <td>139354</td>\n",
       "      <td>0.861907</td>\n",
       "      <td>{'chebi_accession': 'CHEBI:139354', 'name': 'm...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chebi-prod-compounds-16042025</td>\n",
       "      <td>_doc</td>\n",
       "      <td>76987</td>\n",
       "      <td>0.727384</td>\n",
       "      <td>{'chebi_accession': 'CHEBI:76987', 'name': 'me...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chebi-prod-compounds-16042025</td>\n",
       "      <td>_doc</td>\n",
       "      <td>143107</td>\n",
       "      <td>0.666686</td>\n",
       "      <td>{'chebi_accession': 'CHEBI:143107', 'name': '2...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           _index _type     _id     _score  \\\n",
       "0   chebi-prod-compounds-16042025  _doc   46195  66.144830   \n",
       "1   chebi-prod-compounds-16042025  _doc   74529  58.117500   \n",
       "2   chebi-prod-compounds-16042025  _doc   32635  32.586680   \n",
       "3   chebi-prod-compounds-16042025  _doc   32636  31.397276   \n",
       "4   chebi-prod-compounds-16042025  _doc  139476  31.277231   \n",
       "5   chebi-prod-compounds-16042025  _doc  180460  29.612461   \n",
       "6   chebi-prod-compounds-16042025  _doc  133066  26.956810   \n",
       "7   chebi-prod-compounds-16042025  _doc  133067  25.853018   \n",
       "8   chebi-prod-compounds-16042025  _doc  133435  20.886192   \n",
       "9   chebi-prod-compounds-16042025  _doc  133438  20.886192   \n",
       "10  chebi-prod-compounds-16042025  _doc   82942   1.077858   \n",
       "11  chebi-prod-compounds-16042025  _doc   82944   0.949729   \n",
       "12  chebi-prod-compounds-16042025  _doc  139354   0.861907   \n",
       "13  chebi-prod-compounds-16042025  _doc   76987   0.727384   \n",
       "14  chebi-prod-compounds-16042025  _doc  143107   0.666686   \n",
       "\n",
       "                                              _source  \n",
       "0   {'chebi_accession': 'CHEBI:46195', 'name': 'pa...  \n",
       "1   {'chebi_accession': 'CHEBI:74529', 'name': 'an...  \n",
       "2   {'chebi_accession': 'CHEBI:32635', 'name': 'pa...  \n",
       "3   {'chebi_accession': 'CHEBI:32636', 'name': 'ac...  \n",
       "4   {'chebi_accession': 'CHEBI:139476', 'name': '3...  \n",
       "5   {'chebi_accession': 'CHEBI:180460', 'name': 'P...  \n",
       "6   {'chebi_accession': 'CHEBI:133066', 'name': '<...  \n",
       "7   {'chebi_accession': 'CHEBI:133067', 'name': '<...  \n",
       "8   {'chebi_accession': 'CHEBI:133435', 'name': '<...  \n",
       "9   {'chebi_accession': 'CHEBI:133438', 'name': '<...  \n",
       "10  {'chebi_accession': 'CHEBI:82942', 'name': '4-...  \n",
       "11  {'chebi_accession': 'CHEBI:82944', 'name': '2-...  \n",
       "12  {'chebi_accession': 'CHEBI:139354', 'name': 'm...  \n",
       "13  {'chebi_accession': 'CHEBI:76987', 'name': 'me...  \n",
       "14  {'chebi_accession': 'CHEBI:143107', 'name': '2...  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "instance.fetch_batch(\n",
    "    queries=[\n",
    "        {\"term\": \"paracetamol\"}\n",
    "    ],\n",
    "    method=\"es_search\",\n",
    "    parse=True,\n",
    "    to_dataframe=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "f6329444",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cache_key: {\"chebi_id\": \"CHEBI:18357\"}_ontology-children\n",
      "Cache_key: CHEBI:18357_ontology-children\n",
      "Prepared url: https://www.ebi.ac.uk/chebi/beta/api/public/ontology/children/CHEBI%3A18357\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>chebi_accession</th>\n",
       "      <th>ontology_relations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>18357</td>\n",
       "      <td>CHEBI:18357</td>\n",
       "      <td>{'incoming_relations': [{'init_id': 167178, 'i...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      id chebi_accession                                 ontology_relations\n",
       "0  18357     CHEBI:18357  {'incoming_relations': [{'init_id': 167178, 'i..."
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "instance.fetch_batch(\n",
    "    queries=[\n",
    "        {\"chebi_id\": \"CHEBI:18357\"}\n",
    "    ],\n",
    "    method=\"ontology-children\",\n",
    "    parse=True,\n",
    "    to_dataframe=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "0f5eacdf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cache_key: {\"chebi_id\": \"CHEBI:18357\"}_ontology-parents\n",
      "Cache_key: CHEBI:18357_ontology-parents\n",
      "Prepared url: https://www.ebi.ac.uk/chebi/beta/api/public/ontology/parents/CHEBI%3A18357\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>chebi_accession</th>\n",
       "      <th>ontology_relations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>18357</td>\n",
       "      <td>CHEBI:18357</td>\n",
       "      <td>{'outgoing_relations': [{'init_id': 18357, 'in...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      id chebi_accession                                 ontology_relations\n",
       "0  18357     CHEBI:18357  {'outgoing_relations': [{'init_id': 18357, 'in..."
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "instance.fetch_batch(\n",
    "    queries=[\n",
    "        {\"chebi_id\": \"CHEBI:18357\"}\n",
    "    ],\n",
    "    method=\"ontology-parents\",\n",
    "    parse=True,\n",
    "    to_dataframe=True\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b9fd8a37",
   "metadata": {},
   "source": [
    "## Gen Ontology"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "e147770a",
   "metadata": {},
   "outputs": [],
   "source": [
    "from src.genontology import GenOntologyInterface"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "049dc5f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "instance = GenOntologyInterface(\n",
    "    #fields_to_extract = [\"goid\", \"label\"]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "aa7d2183",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cache_key: GO:0008150_ontology-term_default\n",
      "Prepared request: https://api.geneontology.org/api/ontology/term/GO%3A0008150\n",
      "Cache_key: GO:0008150_ontology-term_graph\n",
      "Prepared request: https://api.geneontology.org/api/ontology/term/GO%3A0008150/graph\n",
      "Cache_key: GO:0008150_go_default\n",
      "Prepared request: https://api.geneontology.org/api/go/GO%3A0008150\n",
      "Cache_key: GO:0008150_bioentity-function_default\n",
      "Prepared request: https://api.geneontology.org/api/bioentity/function/GO%3A0008150\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'ontology-term_default': {'goid': 'str',\n",
       "  'label': 'str',\n",
       "  'definition': 'str',\n",
       "  'comment': 'str',\n",
       "  'creation_date': 'str',\n",
       "  'synonyms': 'list(str)',\n",
       "  'relatedSynonyms': 'list(str)',\n",
       "  'alternativeIds': 'list(str)',\n",
       "  'xrefs': 'list(str)',\n",
       "  'subsets': 'list(str)'},\n",
       " 'ontology-term_graph': {'topology_graph_json.nodes.id': 'dict(str)',\n",
       "  'topology_graph_json.nodes.lbl': 'dict(str)',\n",
       "  'topology_graph_json.edges.sub': 'dict(str)',\n",
       "  'topology_graph_json.edges.obj': 'dict(str)',\n",
       "  'topology_graph_json.edges.pred': 'dict(str)',\n",
       "  'topology_graph_json.meta': 'dict(dict)',\n",
       "  'topology_graph_json': 'dict'},\n",
       " 'go_default': {'goid': 'str',\n",
       "  'label': 'str',\n",
       "  'definition': 'str',\n",
       "  'comment': 'str',\n",
       "  'creation_date': 'str',\n",
       "  'synonyms': 'list(str)',\n",
       "  'relatedSynonyms': 'list(str)',\n",
       "  'alternativeIds': 'list(str)',\n",
       "  'xrefs': 'list(str)',\n",
       "  'subsets': 'list(str)'},\n",
       " 'bioentity-function_default': {'bioentity_label': 'str',\n",
       "  'bioentity_name': 'str',\n",
       "  'date': 'str',\n",
       "  'assigned_by': 'str',\n",
       "  'taxon': 'str',\n",
       "  'taxon_label': 'str',\n",
       "  'evidence_type': 'str',\n",
       "  'evidence': 'str',\n",
       "  'reference': 'list(str)',\n",
       "  'synonym': 'list(str)',\n",
       "  'annotation_extension_class': 'list(str)',\n",
       "  'annotation_extension_class_label': 'list(str)'}}"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "instance.get_dummy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "8279b55f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cache_key: GO:0006915_bioentity-function_default\n",
      "Prepared request: https://api.geneontology.org/api/bioentity/function/GO%3A0006915\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>label</th>\n",
       "      <th>name</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>evidence_type</th>\n",
       "      <th>evidence</th>\n",
       "      <th>synonym</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Lhx4</td>\n",
       "      <td>LIM homeobox protein 4</td>\n",
       "      <td>NCBITaxon:10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>IMP</td>\n",
       "      <td>ECO:0000315</td>\n",
       "      <td>[A330062J17Rik, Gsh-4, Gsh4]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Cflar</td>\n",
       "      <td>CASP8 and FADD-like apoptosis regulator</td>\n",
       "      <td>NCBITaxon:10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>IMP</td>\n",
       "      <td>ECO:0000315</td>\n",
       "      <td>[2310024N18Rik, A430105C05Rik, Cash, Casper, F...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Aipl1</td>\n",
       "      <td>aryl hydrocarbon receptor-interacting protein-...</td>\n",
       "      <td>NCBITaxon:10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>IMP</td>\n",
       "      <td>ECO:0000315</td>\n",
       "      <td>A930007I01Rik</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Gli3</td>\n",
       "      <td>GLI-Kruppel family member GLI3</td>\n",
       "      <td>NCBITaxon:10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>IMP</td>\n",
       "      <td>ECO:0000315</td>\n",
       "      <td>[Bph, brachyphalangy]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Aldh1a3</td>\n",
       "      <td>aldehyde dehydrogenase family 1, subfamily A3</td>\n",
       "      <td>NCBITaxon:10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>IGI</td>\n",
       "      <td>ECO:0000316</td>\n",
       "      <td>[ALDH6, RALDH3, V1, retinaldehyde dehydrogenas...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>Nf1</td>\n",
       "      <td>neurofibromin 1</td>\n",
       "      <td>NCBITaxon:10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>IMP</td>\n",
       "      <td>ECO:0000315</td>\n",
       "      <td>[Dsk9, Mhdadsk9, Nf-1, neurofibromin]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>Pim2</td>\n",
       "      <td>proviral integration site 2</td>\n",
       "      <td>NCBITaxon:10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>IMP</td>\n",
       "      <td>ECO:0000315</td>\n",
       "      <td>[DXCch3, Pim-2]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>Palb2</td>\n",
       "      <td>partner and localizer of BRCA2</td>\n",
       "      <td>NCBITaxon:10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>IMP</td>\n",
       "      <td>ECO:0000315</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>Syce3</td>\n",
       "      <td>synaptonemal complex central element protein 3</td>\n",
       "      <td>NCBITaxon:10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>IDA</td>\n",
       "      <td>ECO:0000314</td>\n",
       "      <td>1700007E06Rik</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>Syce3</td>\n",
       "      <td>synaptonemal complex central element protein 3</td>\n",
       "      <td>NCBITaxon:10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>IDA</td>\n",
       "      <td>ECO:0000314</td>\n",
       "      <td>1700007E06Rik</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      label                                               name  \\\n",
       "0      Lhx4                             LIM homeobox protein 4   \n",
       "1     Cflar            CASP8 and FADD-like apoptosis regulator   \n",
       "2     Aipl1  aryl hydrocarbon receptor-interacting protein-...   \n",
       "3      Gli3                     GLI-Kruppel family member GLI3   \n",
       "4   Aldh1a3      aldehyde dehydrogenase family 1, subfamily A3   \n",
       "..      ...                                                ...   \n",
       "95      Nf1                                    neurofibromin 1   \n",
       "96     Pim2                        proviral integration site 2   \n",
       "97    Palb2                     partner and localizer of BRCA2   \n",
       "98    Syce3     synaptonemal complex central element protein 3   \n",
       "99    Syce3     synaptonemal complex central element protein 3   \n",
       "\n",
       "             tax_id      organism evidence_type     evidence  \\\n",
       "0   NCBITaxon:10090  Mus musculus           IMP  ECO:0000315   \n",
       "1   NCBITaxon:10090  Mus musculus           IMP  ECO:0000315   \n",
       "2   NCBITaxon:10090  Mus musculus           IMP  ECO:0000315   \n",
       "3   NCBITaxon:10090  Mus musculus           IMP  ECO:0000315   \n",
       "4   NCBITaxon:10090  Mus musculus           IGI  ECO:0000316   \n",
       "..              ...           ...           ...          ...   \n",
       "95  NCBITaxon:10090  Mus musculus           IMP  ECO:0000315   \n",
       "96  NCBITaxon:10090  Mus musculus           IMP  ECO:0000315   \n",
       "97  NCBITaxon:10090  Mus musculus           IMP  ECO:0000315   \n",
       "98  NCBITaxon:10090  Mus musculus           IDA  ECO:0000314   \n",
       "99  NCBITaxon:10090  Mus musculus           IDA  ECO:0000314   \n",
       "\n",
       "                                              synonym  \n",
       "0                        [A330062J17Rik, Gsh-4, Gsh4]  \n",
       "1   [2310024N18Rik, A430105C05Rik, Cash, Casper, F...  \n",
       "2                                       A930007I01Rik  \n",
       "3                               [Bph, brachyphalangy]  \n",
       "4   [ALDH6, RALDH3, V1, retinaldehyde dehydrogenas...  \n",
       "..                                                ...  \n",
       "95              [Dsk9, Mhdadsk9, Nf-1, neurofibromin]  \n",
       "96                                    [DXCch3, Pim-2]  \n",
       "97                                               None  \n",
       "98                                      1700007E06Rik  \n",
       "99                                      1700007E06Rik  \n",
       "\n",
       "[100 rows x 7 columns]"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "instance.fetch_single(\n",
    "    method=\"bioentity-function\",\n",
    "    query=\"GO:0006915\",\n",
    "    parse=True,\n",
    "    to_dataframe=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "e6695e6f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cache_key: GO:0008150_bioentity-function_default\n",
      "Cache_key: GO:0006915_bioentity-function_default\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>label</th>\n",
       "      <th>name</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>evidence_type</th>\n",
       "      <th>evidence</th>\n",
       "      <th>synonym</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Rapgef4os3</td>\n",
       "      <td>Rap guanine nucleotide exchange factor (GEF) 4...</td>\n",
       "      <td>NCBITaxon:10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>ND</td>\n",
       "      <td>ECO:0000307</td>\n",
       "      <td>4732447D17Rik</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Nxpe3</td>\n",
       "      <td>neurexophilin and PC-esterase domain family, m...</td>\n",
       "      <td>NCBITaxon:10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>ND</td>\n",
       "      <td>ECO:0000307</td>\n",
       "      <td>[Fam55c, LOC208684, LOC385658]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Lsm14b</td>\n",
       "      <td>LSM family member 14B</td>\n",
       "      <td>NCBITaxon:10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>ND</td>\n",
       "      <td>ECO:0000307</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BC039771</td>\n",
       "      <td>cDNA sequence BC039771</td>\n",
       "      <td>NCBITaxon:10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>ND</td>\n",
       "      <td>ECO:0000307</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Gm14661</td>\n",
       "      <td>predicted gene 14661</td>\n",
       "      <td>NCBITaxon:10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>ND</td>\n",
       "      <td>ECO:0000307</td>\n",
       "      <td>ENSMUSG00000067835</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>195</th>\n",
       "      <td>Nf1</td>\n",
       "      <td>neurofibromin 1</td>\n",
       "      <td>NCBITaxon:10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>IMP</td>\n",
       "      <td>ECO:0000315</td>\n",
       "      <td>[Dsk9, Mhdadsk9, Nf-1, neurofibromin]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196</th>\n",
       "      <td>Pim2</td>\n",
       "      <td>proviral integration site 2</td>\n",
       "      <td>NCBITaxon:10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>IMP</td>\n",
       "      <td>ECO:0000315</td>\n",
       "      <td>[DXCch3, Pim-2]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>197</th>\n",
       "      <td>Palb2</td>\n",
       "      <td>partner and localizer of BRCA2</td>\n",
       "      <td>NCBITaxon:10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>IMP</td>\n",
       "      <td>ECO:0000315</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>198</th>\n",
       "      <td>Syce3</td>\n",
       "      <td>synaptonemal complex central element protein 3</td>\n",
       "      <td>NCBITaxon:10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>IDA</td>\n",
       "      <td>ECO:0000314</td>\n",
       "      <td>1700007E06Rik</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199</th>\n",
       "      <td>Syce3</td>\n",
       "      <td>synaptonemal complex central element protein 3</td>\n",
       "      <td>NCBITaxon:10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>IDA</td>\n",
       "      <td>ECO:0000314</td>\n",
       "      <td>1700007E06Rik</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>200 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          label                                               name  \\\n",
       "0    Rapgef4os3  Rap guanine nucleotide exchange factor (GEF) 4...   \n",
       "1         Nxpe3  neurexophilin and PC-esterase domain family, m...   \n",
       "2        Lsm14b                              LSM family member 14B   \n",
       "3      BC039771                             cDNA sequence BC039771   \n",
       "4       Gm14661                               predicted gene 14661   \n",
       "..          ...                                                ...   \n",
       "195         Nf1                                    neurofibromin 1   \n",
       "196        Pim2                        proviral integration site 2   \n",
       "197       Palb2                     partner and localizer of BRCA2   \n",
       "198       Syce3     synaptonemal complex central element protein 3   \n",
       "199       Syce3     synaptonemal complex central element protein 3   \n",
       "\n",
       "              tax_id      organism evidence_type     evidence  \\\n",
       "0    NCBITaxon:10090  Mus musculus            ND  ECO:0000307   \n",
       "1    NCBITaxon:10090  Mus musculus            ND  ECO:0000307   \n",
       "2    NCBITaxon:10090  Mus musculus            ND  ECO:0000307   \n",
       "3    NCBITaxon:10090  Mus musculus            ND  ECO:0000307   \n",
       "4    NCBITaxon:10090  Mus musculus            ND  ECO:0000307   \n",
       "..               ...           ...           ...          ...   \n",
       "195  NCBITaxon:10090  Mus musculus           IMP  ECO:0000315   \n",
       "196  NCBITaxon:10090  Mus musculus           IMP  ECO:0000315   \n",
       "197  NCBITaxon:10090  Mus musculus           IMP  ECO:0000315   \n",
       "198  NCBITaxon:10090  Mus musculus           IDA  ECO:0000314   \n",
       "199  NCBITaxon:10090  Mus musculus           IDA  ECO:0000314   \n",
       "\n",
       "                                   synonym  \n",
       "0                            4732447D17Rik  \n",
       "1           [Fam55c, LOC208684, LOC385658]  \n",
       "2                                     None  \n",
       "3                                     None  \n",
       "4                       ENSMUSG00000067835  \n",
       "..                                     ...  \n",
       "195  [Dsk9, Mhdadsk9, Nf-1, neurofibromin]  \n",
       "196                        [DXCch3, Pim-2]  \n",
       "197                                   None  \n",
       "198                          1700007E06Rik  \n",
       "199                          1700007E06Rik  \n",
       "\n",
       "[200 rows x 7 columns]"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "instance.fetch_batch(\n",
    "    method=\"bioentity-function\",\n",
    "    queries=[\"GO:0008150\", \"GO:0006915\"],\n",
    "    parse=True,\n",
    "    to_dataframe=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "2b8d35b5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cache_key: GO:0008150_ontology-term_default\n",
      "Cache_key: GO:0006915_ontology-term_default\n",
      "Cache_key: GO:0006915_ontology-term_default\n",
      "Prepared request: https://api.geneontology.org/api/ontology/term/GO%3A0006915\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>go_term</th>\n",
       "      <th>label</th>\n",
       "      <th>definition</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GO:0008150</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>A biological process is the execution of a gen...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GO:0006915</td>\n",
       "      <td>apoptotic process</td>\n",
       "      <td>A programmed cell death process which begins w...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      go_term               label  \\\n",
       "0  GO:0008150  biological_process   \n",
       "1  GO:0006915   apoptotic process   \n",
       "\n",
       "                                          definition  \n",
       "0  A biological process is the execution of a gen...  \n",
       "1  A programmed cell death process which begins w...  "
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "instance.fetch_batch(\n",
    "    method=\"ontology-term\",\n",
    "    queries=[\"GO:0008150\", \"GO:0006915\"],\n",
    "    parse=True,\n",
    "    to_dataframe=True\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e6a4a0ef",
   "metadata": {},
   "source": [
    "## Interpro"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "56f061c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "from src.interpro import InterproInstance\n",
    "from typing import List, Dict, Any, Union"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "9bc15d9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "instance = InterproInstance()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4dc364de",
   "metadata": {},
   "source": [
    "### Search Interpro Id IPR036669"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "ddd44ee4",
   "metadata": {},
   "outputs": [],
   "source": [
    "query = {\n",
    "    \"id\": \"IPR002223\",\n",
    "    \"db\" : \"InterPro\",\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cb1e9399",
   "metadata": {},
   "source": [
    "### Search uniProt id Q29537"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "524b97f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "query = {\n",
    "    \"db\" : \"InterPro\",\n",
    "    \"modifiers\" : {},\n",
    "    \"filters\" : [\n",
    "        {\n",
    "            \"type\": \"protein\",\n",
    "            \"db\": \"reviewed\",\n",
    "            \"value\": \"Q29537\"\n",
    "        }\n",
    "    ]\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "b61fb31c",
   "metadata": {},
   "outputs": [],
   "source": [
    "query = {\n",
    "    \"db\" : \"InterPro\",\n",
    "    \"modifiers\" : {},\n",
    "    \"filters\" : [\n",
    "        {\n",
    "            \"type\": \"protein\",\n",
    "            \"db\": \"reviewed\",\n",
    "            \"value\": \"P05067\"\n",
    "        }\n",
    "    ]\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "37ee27d3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cache_key: IPR002223_InterPro_entry\n",
      "Fetching data from InterPro API with URL: https://www.ebi.ac.uk:443/interpro/api/entry/InterPro/IPR002223/\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>length</th>\n",
       "      <th>organism_id</th>\n",
       "      <th>protein_source_db</th>\n",
       "      <th>in_alphafold</th>\n",
       "      <th>interpro_accession</th>\n",
       "      <th>name</th>\n",
       "      <th>type</th>\n",
       "      <th>source_database</th>\n",
       "      <th>integrated</th>\n",
       "      <th>...</th>\n",
       "      <th>prosite</th>\n",
       "      <th>smart</th>\n",
       "      <th>prints</th>\n",
       "      <th>panther</th>\n",
       "      <th>cathgene3d</th>\n",
       "      <th>ssf</th>\n",
       "      <th>go_identifier</th>\n",
       "      <th>go_name</th>\n",
       "      <th>go_category</th>\n",
       "      <th>locations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>IPR002223</td>\n",
       "      <td>{'name': 'Pancreatic trypsin inhibitor Kunitz ...</td>\n",
       "      <td>domain</td>\n",
       "      <td>interpro</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>{'SM00131': 'BPTI/Kunitz family of serine prot...</td>\n",
       "      <td>{'PR00759': 'BASICPTASE'}</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>GO:0004867</td>\n",
       "      <td>serine-type endopeptidase inhibitor activity</td>\n",
       "      <td>molecular_function</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  accession length organism_id protein_source_db in_alphafold  \\\n",
       "0      None   None        None              None         None   \n",
       "\n",
       "  interpro_accession                                               name  \\\n",
       "0          IPR002223  {'name': 'Pancreatic trypsin inhibitor Kunitz ...   \n",
       "\n",
       "     type source_database integrated  ... prosite  \\\n",
       "0  domain        interpro       None  ...    None   \n",
       "\n",
       "                                               smart  \\\n",
       "0  {'SM00131': 'BPTI/Kunitz family of serine prot...   \n",
       "\n",
       "                      prints panther cathgene3d   ssf go_identifier  \\\n",
       "0  {'PR00759': 'BASICPTASE'}    None       None  None    GO:0004867   \n",
       "\n",
       "                                        go_name         go_category locations  \n",
       "0  serine-type endopeptidase inhibitor activity  molecular_function      None  \n",
       "\n",
       "[1 rows x 21 columns]"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "instance.fetch_single(query=query, method=\"entry\", pages_to_fetch=1, parse=True, to_dataframe=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b60a1cc5",
   "metadata": {},
   "source": [
    "### Search in Batches"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "62156057",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cache_key: {\"db\": \"InterPro\", \"filters\": [{\"db\": \"reviewed\", \"type\": \"protein\", \"value\": \"Q29537\"}], \"modifiers\": {}}_entry\n",
      "Cache_key: {\"db\": \"InterPro\", \"filters\": [{\"db\": \"reviewed\", \"type\": \"protein\", \"value\": \"P05067\"}], \"modifiers\": {}}_entry\n",
      "Cache_key: InterPro_{}_[{'type': 'protein', 'db': 'reviewed', 'value': 'Q29537'}]_entry\n",
      "Fetching data from InterPro API with URL: https://www.ebi.ac.uk:443/interpro/api/entry/InterPro/protein/reviewed/Q29537/\n",
      "Cache_key: InterPro_{}_[{'type': 'protein', 'db': 'reviewed', 'value': 'P05067'}]_entry\n",
      "Fetching data from InterPro API with URL: https://www.ebi.ac.uk:443/interpro/api/entry/InterPro/protein/reviewed/P05067/\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>length</th>\n",
       "      <th>organism_id</th>\n",
       "      <th>protein_source_db</th>\n",
       "      <th>in_alphafold</th>\n",
       "      <th>interpro_accession</th>\n",
       "      <th>name</th>\n",
       "      <th>type</th>\n",
       "      <th>source_database</th>\n",
       "      <th>integrated</th>\n",
       "      <th>...</th>\n",
       "      <th>prosite</th>\n",
       "      <th>smart</th>\n",
       "      <th>prints</th>\n",
       "      <th>panther</th>\n",
       "      <th>cathgene3d</th>\n",
       "      <th>ssf</th>\n",
       "      <th>go_identifier</th>\n",
       "      <th>go_name</th>\n",
       "      <th>go_category</th>\n",
       "      <th>locations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>q29537</td>\n",
       "      <td>381</td>\n",
       "      <td>9615</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>True</td>\n",
       "      <td>IPR002117</td>\n",
       "      <td>p53 tumour suppressor family</td>\n",
       "      <td>family</td>\n",
       "      <td>interpro</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>{'PR00386': 'P53SUPPRESSR'}</td>\n",
       "      <td>{'PTHR11447': 'CELLULAR TUMOR ANTIGEN P53'}</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[GO:0003677, GO:0003700, GO:0006355, GO:000691...</td>\n",
       "      <td>[DNA binding, DNA-binding transcription factor...</td>\n",
       "      <td>[molecular_function, molecular_function, biolo...</td>\n",
       "      <td>{'fragments': [{'start': 3, 'end': 354, 'dc-st...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>q29537</td>\n",
       "      <td>381</td>\n",
       "      <td>9615</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>True</td>\n",
       "      <td>IPR008967</td>\n",
       "      <td>p53-like transcription factor, DNA-binding dom...</td>\n",
       "      <td>homologous_superfamily</td>\n",
       "      <td>interpro</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>{'SSF49417': 'p53-like transcription factors'}</td>\n",
       "      <td>[GO:0003700, GO:0006355]</td>\n",
       "      <td>[DNA-binding transcription factor activity, re...</td>\n",
       "      <td>[molecular_function, biological_process]</td>\n",
       "      <td>{'fragments': [{'start': 84, 'end': 275, 'dc-s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>q29537</td>\n",
       "      <td>381</td>\n",
       "      <td>9615</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>True</td>\n",
       "      <td>IPR010991</td>\n",
       "      <td>p53, tetramerisation domain</td>\n",
       "      <td>domain</td>\n",
       "      <td>interpro</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>GO:0051262</td>\n",
       "      <td>protein tetramerization</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>{'fragments': [{'start': 307, 'end': 345, 'dc-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>q29537</td>\n",
       "      <td>381</td>\n",
       "      <td>9615</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>True</td>\n",
       "      <td>IPR011615</td>\n",
       "      <td>p53, DNA-binding domain</td>\n",
       "      <td>domain</td>\n",
       "      <td>interpro</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>GO:0000976</td>\n",
       "      <td>transcription cis-regulatory region binding</td>\n",
       "      <td>molecular_function</td>\n",
       "      <td>{'fragments': [{'start': 88, 'end': 277, 'dc-s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>q29537</td>\n",
       "      <td>381</td>\n",
       "      <td>9615</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>True</td>\n",
       "      <td>IPR012346</td>\n",
       "      <td>p53/RUNT-type transcription factor, DNA-bindin...</td>\n",
       "      <td>homologous_superfamily</td>\n",
       "      <td>interpro</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>{'G3DSA:2.60.40.720': 'G3DSA:2.60.40.720'}</td>\n",
       "      <td>None</td>\n",
       "      <td>[GO:0003677, GO:0003700, GO:0006355, GO:0005634]</td>\n",
       "      <td>[DNA binding, DNA-binding transcription factor...</td>\n",
       "      <td>[molecular_function, molecular_function, biolo...</td>\n",
       "      <td>{'fragments': [{'start': 82, 'end': 282, 'dc-s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>q29537</td>\n",
       "      <td>381</td>\n",
       "      <td>9615</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>True</td>\n",
       "      <td>IPR013872</td>\n",
       "      <td>p53, transactivation domain</td>\n",
       "      <td>domain</td>\n",
       "      <td>interpro</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>GO:0005515</td>\n",
       "      <td>protein binding</td>\n",
       "      <td>molecular_function</td>\n",
       "      <td>{'fragments': [{'start': 6, 'end': 30, 'dc-sta...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>q29537</td>\n",
       "      <td>381</td>\n",
       "      <td>9615</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>True</td>\n",
       "      <td>IPR036674</td>\n",
       "      <td>p53-like tetramerisation domain superfamily</td>\n",
       "      <td>homologous_superfamily</td>\n",
       "      <td>interpro</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>{'G3DSA:4.10.170.10': 'p53-like tetramerisatio...</td>\n",
       "      <td>{'SSF47719': 'p53 tetramerization domain'}</td>\n",
       "      <td>GO:0051262</td>\n",
       "      <td>protein tetramerization</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>{'fragments': [{'start': 307, 'end': 348, 'dc-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>q29537</td>\n",
       "      <td>381</td>\n",
       "      <td>9615</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>True</td>\n",
       "      <td>IPR057064</td>\n",
       "      <td>p53, central conserved site</td>\n",
       "      <td>conserved_site</td>\n",
       "      <td>interpro</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>{'PS00348': 'p53 family signature'}</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>{'fragments': [{'start': 225, 'end': 237, 'dc-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>p05067</td>\n",
       "      <td>770</td>\n",
       "      <td>9606</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>True</td>\n",
       "      <td>IPR002223</td>\n",
       "      <td>Pancreatic trypsin inhibitor Kunitz domain</td>\n",
       "      <td>domain</td>\n",
       "      <td>interpro</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>{'SM00131': 'BPTI/Kunitz family of serine prot...</td>\n",
       "      <td>{'PR00759': 'BASICPTASE'}</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>GO:0004867</td>\n",
       "      <td>serine-type endopeptidase inhibitor activity</td>\n",
       "      <td>molecular_function</td>\n",
       "      <td>{'fragments': [{'start': 288, 'end': 342, 'dc-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>p05067</td>\n",
       "      <td>770</td>\n",
       "      <td>9606</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>True</td>\n",
       "      <td>IPR008154</td>\n",
       "      <td>Amyloidogenic glycoprotein, extracellular</td>\n",
       "      <td>domain</td>\n",
       "      <td>interpro</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>{'SM00006': 'amyloid A4'}</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>{'fragments': [{'start': 24, 'end': 189, 'dc-s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>p05067</td>\n",
       "      <td>770</td>\n",
       "      <td>9606</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>True</td>\n",
       "      <td>IPR008155</td>\n",
       "      <td>Amyloidogenic glycoprotein</td>\n",
       "      <td>family</td>\n",
       "      <td>interpro</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>{'PR00203': 'AMYLOIDA4'}</td>\n",
       "      <td>{'PTHR23103': 'ALZHEIMER'S DISEASE BETA-AMYLOI...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>GO:0016020</td>\n",
       "      <td>membrane</td>\n",
       "      <td>cellular_component</td>\n",
       "      <td>{'fragments': [{'start': 10, 'end': 770, 'dc-s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>p05067</td>\n",
       "      <td>770</td>\n",
       "      <td>9606</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>True</td>\n",
       "      <td>IPR011178</td>\n",
       "      <td>Amyloidogenic glycoprotein, copper-binding</td>\n",
       "      <td>domain</td>\n",
       "      <td>interpro</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>GO:0046914</td>\n",
       "      <td>transition metal ion binding</td>\n",
       "      <td>molecular_function</td>\n",
       "      <td>{'fragments': [{'start': 132, 'end': 188, 'dc-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>p05067</td>\n",
       "      <td>770</td>\n",
       "      <td>9606</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>True</td>\n",
       "      <td>IPR011993</td>\n",
       "      <td>PH-like domain superfamily</td>\n",
       "      <td>homologous_superfamily</td>\n",
       "      <td>interpro</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>{'G3DSA:2.30.29.30': 'Pleckstrin-homology doma...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>{'fragments': [{'start': 742, 'end': 770, 'dc-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>p05067</td>\n",
       "      <td>770</td>\n",
       "      <td>9606</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>True</td>\n",
       "      <td>IPR013803</td>\n",
       "      <td>Amyloidogenic glycoprotein, amyloid-beta peptide</td>\n",
       "      <td>domain</td>\n",
       "      <td>interpro</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>{'PR00204': 'BETAAMYLOID'}</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>GO:0016020</td>\n",
       "      <td>membrane</td>\n",
       "      <td>cellular_component</td>\n",
       "      <td>{'fragments': [{'start': 675, 'end': 713, 'dc-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>p05067</td>\n",
       "      <td>770</td>\n",
       "      <td>9606</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>True</td>\n",
       "      <td>IPR015849</td>\n",
       "      <td>Amyloidogenic glycoprotein, heparin-binding</td>\n",
       "      <td>domain</td>\n",
       "      <td>interpro</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>GO:0008201</td>\n",
       "      <td>heparin binding</td>\n",
       "      <td>molecular_function</td>\n",
       "      <td>{'fragments': [{'start': 31, 'end': 131, 'dc-s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>p05067</td>\n",
       "      <td>770</td>\n",
       "      <td>9606</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>True</td>\n",
       "      <td>IPR019543</td>\n",
       "      <td>Beta-amyloid precursor protein C-terminal</td>\n",
       "      <td>domain</td>\n",
       "      <td>interpro</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>{'fragments': [{'start': 716, 'end': 766, 'dc-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>p05067</td>\n",
       "      <td>770</td>\n",
       "      <td>9606</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>True</td>\n",
       "      <td>IPR019744</td>\n",
       "      <td>Amyloidogenic glycoprotein, copper-binding dom...</td>\n",
       "      <td>conserved_site</td>\n",
       "      <td>interpro</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>{'PS00319': 'Amyloid precursor protein (APP) c...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>{'fragments': [{'start': 181, 'end': 188, 'dc-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>p05067</td>\n",
       "      <td>770</td>\n",
       "      <td>9606</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>True</td>\n",
       "      <td>IPR019745</td>\n",
       "      <td>Amyloidogenic glycoprotein, intracellular doma...</td>\n",
       "      <td>conserved_site</td>\n",
       "      <td>interpro</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>{'PS00320': 'Amyloid precursor protein (APP) i...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>{'fragments': [{'start': 756, 'end': 763, 'dc-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>p05067</td>\n",
       "      <td>770</td>\n",
       "      <td>9606</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>True</td>\n",
       "      <td>IPR020901</td>\n",
       "      <td>Proteinase inhibitor I2, Kunitz, conserved site</td>\n",
       "      <td>conserved_site</td>\n",
       "      <td>interpro</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>{'PS00280': 'Pancreatic trypsin inhibitor (Kun...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>{'fragments': [{'start': 319, 'end': 337, 'dc-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>p05067</td>\n",
       "      <td>770</td>\n",
       "      <td>9606</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>True</td>\n",
       "      <td>IPR024329</td>\n",
       "      <td>Amyloidogenic glycoprotein, E2 domain</td>\n",
       "      <td>domain</td>\n",
       "      <td>interpro</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>{'fragments': [{'start': 367, 'end': 565, 'dc-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>p05067</td>\n",
       "      <td>770</td>\n",
       "      <td>9606</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>True</td>\n",
       "      <td>IPR036176</td>\n",
       "      <td>E2 domain superfamily</td>\n",
       "      <td>homologous_superfamily</td>\n",
       "      <td>interpro</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>{'G3DSA:1.20.120.770': 'Amyloid precursor prot...</td>\n",
       "      <td>{'SSF109843': 'CAPPD, an extracellular domain ...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>{'fragments': [{'start': 363, 'end': 580, 'dc-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>p05067</td>\n",
       "      <td>770</td>\n",
       "      <td>9606</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>True</td>\n",
       "      <td>IPR036454</td>\n",
       "      <td>Amyloidogenic glycoprotein, heparin-binding do...</td>\n",
       "      <td>homologous_superfamily</td>\n",
       "      <td>interpro</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>{'G3DSA:3.90.570.10': 'Amyloidogenic glycoprot...</td>\n",
       "      <td>{'SSF56491': 'A heparin-binding domain'}</td>\n",
       "      <td>GO:0008201</td>\n",
       "      <td>heparin binding</td>\n",
       "      <td>molecular_function</td>\n",
       "      <td>{'fragments': [{'start': 17, 'end': 123, 'dc-s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>p05067</td>\n",
       "      <td>770</td>\n",
       "      <td>9606</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>True</td>\n",
       "      <td>IPR036669</td>\n",
       "      <td>Amyloidogenic glycoprotein, copper-binding dom...</td>\n",
       "      <td>homologous_superfamily</td>\n",
       "      <td>interpro</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>{'G3DSA:3.30.1490.140': 'Amyloidogenic glycopr...</td>\n",
       "      <td>{'SSF89811': 'Amyloid beta a4 protein copper b...</td>\n",
       "      <td>GO:0046914</td>\n",
       "      <td>transition metal ion binding</td>\n",
       "      <td>molecular_function</td>\n",
       "      <td>{'fragments': [{'start': 124, 'end': 189, 'dc-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>p05067</td>\n",
       "      <td>770</td>\n",
       "      <td>9606</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>True</td>\n",
       "      <td>IPR036880</td>\n",
       "      <td>Pancreatic trypsin inhibitor Kunitz domain sup...</td>\n",
       "      <td>homologous_superfamily</td>\n",
       "      <td>interpro</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>{'G3DSA:4.10.410.10': 'Pancreatic trypsin inhi...</td>\n",
       "      <td>{'SSF57362': 'BPTI-like'}</td>\n",
       "      <td>GO:0004867</td>\n",
       "      <td>serine-type endopeptidase inhibitor activity</td>\n",
       "      <td>molecular_function</td>\n",
       "      <td>{'fragments': [{'start': 279, 'end': 352, 'dc-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>p05067</td>\n",
       "      <td>770</td>\n",
       "      <td>9606</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>True</td>\n",
       "      <td>IPR037071</td>\n",
       "      <td>Amyloidogenic glycoprotein, amyloid-beta pepti...</td>\n",
       "      <td>homologous_superfamily</td>\n",
       "      <td>interpro</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>{'G3DSA:4.10.230.10': 'Amyloidogenic glycoprot...</td>\n",
       "      <td>None</td>\n",
       "      <td>GO:0016020</td>\n",
       "      <td>membrane</td>\n",
       "      <td>cellular_component</td>\n",
       "      <td>{'fragments': [{'start': 672, 'end': 711, 'dc-...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>25 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   accession  length organism_id protein_source_db  in_alphafold  \\\n",
       "0     q29537     381        9615          reviewed          True   \n",
       "1     q29537     381        9615          reviewed          True   \n",
       "2     q29537     381        9615          reviewed          True   \n",
       "3     q29537     381        9615          reviewed          True   \n",
       "4     q29537     381        9615          reviewed          True   \n",
       "5     q29537     381        9615          reviewed          True   \n",
       "6     q29537     381        9615          reviewed          True   \n",
       "7     q29537     381        9615          reviewed          True   \n",
       "8     p05067     770        9606          reviewed          True   \n",
       "9     p05067     770        9606          reviewed          True   \n",
       "10    p05067     770        9606          reviewed          True   \n",
       "11    p05067     770        9606          reviewed          True   \n",
       "12    p05067     770        9606          reviewed          True   \n",
       "13    p05067     770        9606          reviewed          True   \n",
       "14    p05067     770        9606          reviewed          True   \n",
       "15    p05067     770        9606          reviewed          True   \n",
       "16    p05067     770        9606          reviewed          True   \n",
       "17    p05067     770        9606          reviewed          True   \n",
       "18    p05067     770        9606          reviewed          True   \n",
       "19    p05067     770        9606          reviewed          True   \n",
       "20    p05067     770        9606          reviewed          True   \n",
       "21    p05067     770        9606          reviewed          True   \n",
       "22    p05067     770        9606          reviewed          True   \n",
       "23    p05067     770        9606          reviewed          True   \n",
       "24    p05067     770        9606          reviewed          True   \n",
       "\n",
       "   interpro_accession                                               name  \\\n",
       "0           IPR002117                       p53 tumour suppressor family   \n",
       "1           IPR008967  p53-like transcription factor, DNA-binding dom...   \n",
       "2           IPR010991                        p53, tetramerisation domain   \n",
       "3           IPR011615                            p53, DNA-binding domain   \n",
       "4           IPR012346  p53/RUNT-type transcription factor, DNA-bindin...   \n",
       "5           IPR013872                        p53, transactivation domain   \n",
       "6           IPR036674        p53-like tetramerisation domain superfamily   \n",
       "7           IPR057064                        p53, central conserved site   \n",
       "8           IPR002223         Pancreatic trypsin inhibitor Kunitz domain   \n",
       "9           IPR008154          Amyloidogenic glycoprotein, extracellular   \n",
       "10          IPR008155                         Amyloidogenic glycoprotein   \n",
       "11          IPR011178         Amyloidogenic glycoprotein, copper-binding   \n",
       "12          IPR011993                         PH-like domain superfamily   \n",
       "13          IPR013803   Amyloidogenic glycoprotein, amyloid-beta peptide   \n",
       "14          IPR015849        Amyloidogenic glycoprotein, heparin-binding   \n",
       "15          IPR019543          Beta-amyloid precursor protein C-terminal   \n",
       "16          IPR019744  Amyloidogenic glycoprotein, copper-binding dom...   \n",
       "17          IPR019745  Amyloidogenic glycoprotein, intracellular doma...   \n",
       "18          IPR020901    Proteinase inhibitor I2, Kunitz, conserved site   \n",
       "19          IPR024329              Amyloidogenic glycoprotein, E2 domain   \n",
       "20          IPR036176                              E2 domain superfamily   \n",
       "21          IPR036454  Amyloidogenic glycoprotein, heparin-binding do...   \n",
       "22          IPR036669  Amyloidogenic glycoprotein, copper-binding dom...   \n",
       "23          IPR036880  Pancreatic trypsin inhibitor Kunitz domain sup...   \n",
       "24          IPR037071  Amyloidogenic glycoprotein, amyloid-beta pepti...   \n",
       "\n",
       "                      type source_database integrated  ...  \\\n",
       "0                   family        interpro       None  ...   \n",
       "1   homologous_superfamily        interpro       None  ...   \n",
       "2                   domain        interpro       None  ...   \n",
       "3                   domain        interpro       None  ...   \n",
       "4   homologous_superfamily        interpro       None  ...   \n",
       "5                   domain        interpro       None  ...   \n",
       "6   homologous_superfamily        interpro       None  ...   \n",
       "7           conserved_site        interpro       None  ...   \n",
       "8                   domain        interpro       None  ...   \n",
       "9                   domain        interpro       None  ...   \n",
       "10                  family        interpro       None  ...   \n",
       "11                  domain        interpro       None  ...   \n",
       "12  homologous_superfamily        interpro       None  ...   \n",
       "13                  domain        interpro       None  ...   \n",
       "14                  domain        interpro       None  ...   \n",
       "15                  domain        interpro       None  ...   \n",
       "16          conserved_site        interpro       None  ...   \n",
       "17          conserved_site        interpro       None  ...   \n",
       "18          conserved_site        interpro       None  ...   \n",
       "19                  domain        interpro       None  ...   \n",
       "20  homologous_superfamily        interpro       None  ...   \n",
       "21  homologous_superfamily        interpro       None  ...   \n",
       "22  homologous_superfamily        interpro       None  ...   \n",
       "23  homologous_superfamily        interpro       None  ...   \n",
       "24  homologous_superfamily        interpro       None  ...   \n",
       "\n",
       "                                              prosite  \\\n",
       "0                                                None   \n",
       "1                                                None   \n",
       "2                                                None   \n",
       "3                                                None   \n",
       "4                                                None   \n",
       "5                                                None   \n",
       "6                                                None   \n",
       "7                 {'PS00348': 'p53 family signature'}   \n",
       "8                                                None   \n",
       "9                                                None   \n",
       "10                                               None   \n",
       "11                                               None   \n",
       "12                                               None   \n",
       "13                                               None   \n",
       "14                                               None   \n",
       "15                                               None   \n",
       "16  {'PS00319': 'Amyloid precursor protein (APP) c...   \n",
       "17  {'PS00320': 'Amyloid precursor protein (APP) i...   \n",
       "18  {'PS00280': 'Pancreatic trypsin inhibitor (Kun...   \n",
       "19                                               None   \n",
       "20                                               None   \n",
       "21                                               None   \n",
       "22                                               None   \n",
       "23                                               None   \n",
       "24                                               None   \n",
       "\n",
       "                                                smart  \\\n",
       "0                                                None   \n",
       "1                                                None   \n",
       "2                                                None   \n",
       "3                                                None   \n",
       "4                                                None   \n",
       "5                                                None   \n",
       "6                                                None   \n",
       "7                                                None   \n",
       "8   {'SM00131': 'BPTI/Kunitz family of serine prot...   \n",
       "9                           {'SM00006': 'amyloid A4'}   \n",
       "10                                               None   \n",
       "11                                               None   \n",
       "12                                               None   \n",
       "13                                               None   \n",
       "14                                               None   \n",
       "15                                               None   \n",
       "16                                               None   \n",
       "17                                               None   \n",
       "18                                               None   \n",
       "19                                               None   \n",
       "20                                               None   \n",
       "21                                               None   \n",
       "22                                               None   \n",
       "23                                               None   \n",
       "24                                               None   \n",
       "\n",
       "                         prints  \\\n",
       "0   {'PR00386': 'P53SUPPRESSR'}   \n",
       "1                          None   \n",
       "2                          None   \n",
       "3                          None   \n",
       "4                          None   \n",
       "5                          None   \n",
       "6                          None   \n",
       "7                          None   \n",
       "8     {'PR00759': 'BASICPTASE'}   \n",
       "9                          None   \n",
       "10     {'PR00203': 'AMYLOIDA4'}   \n",
       "11                         None   \n",
       "12                         None   \n",
       "13   {'PR00204': 'BETAAMYLOID'}   \n",
       "14                         None   \n",
       "15                         None   \n",
       "16                         None   \n",
       "17                         None   \n",
       "18                         None   \n",
       "19                         None   \n",
       "20                         None   \n",
       "21                         None   \n",
       "22                         None   \n",
       "23                         None   \n",
       "24                         None   \n",
       "\n",
       "                                              panther  \\\n",
       "0         {'PTHR11447': 'CELLULAR TUMOR ANTIGEN P53'}   \n",
       "1                                                None   \n",
       "2                                                None   \n",
       "3                                                None   \n",
       "4                                                None   \n",
       "5                                                None   \n",
       "6                                                None   \n",
       "7                                                None   \n",
       "8                                                None   \n",
       "9                                                None   \n",
       "10  {'PTHR23103': 'ALZHEIMER'S DISEASE BETA-AMYLOI...   \n",
       "11                                               None   \n",
       "12                                               None   \n",
       "13                                               None   \n",
       "14                                               None   \n",
       "15                                               None   \n",
       "16                                               None   \n",
       "17                                               None   \n",
       "18                                               None   \n",
       "19                                               None   \n",
       "20                                               None   \n",
       "21                                               None   \n",
       "22                                               None   \n",
       "23                                               None   \n",
       "24                                               None   \n",
       "\n",
       "                                           cathgene3d  \\\n",
       "0                                                None   \n",
       "1                                                None   \n",
       "2                                                None   \n",
       "3                                                None   \n",
       "4          {'G3DSA:2.60.40.720': 'G3DSA:2.60.40.720'}   \n",
       "5                                                None   \n",
       "6   {'G3DSA:4.10.170.10': 'p53-like tetramerisatio...   \n",
       "7                                                None   \n",
       "8                                                None   \n",
       "9                                                None   \n",
       "10                                               None   \n",
       "11                                               None   \n",
       "12  {'G3DSA:2.30.29.30': 'Pleckstrin-homology doma...   \n",
       "13                                               None   \n",
       "14                                               None   \n",
       "15                                               None   \n",
       "16                                               None   \n",
       "17                                               None   \n",
       "18                                               None   \n",
       "19                                               None   \n",
       "20  {'G3DSA:1.20.120.770': 'Amyloid precursor prot...   \n",
       "21  {'G3DSA:3.90.570.10': 'Amyloidogenic glycoprot...   \n",
       "22  {'G3DSA:3.30.1490.140': 'Amyloidogenic glycopr...   \n",
       "23  {'G3DSA:4.10.410.10': 'Pancreatic trypsin inhi...   \n",
       "24  {'G3DSA:4.10.230.10': 'Amyloidogenic glycoprot...   \n",
       "\n",
       "                                                  ssf  \\\n",
       "0                                                None   \n",
       "1      {'SSF49417': 'p53-like transcription factors'}   \n",
       "2                                                None   \n",
       "3                                                None   \n",
       "4                                                None   \n",
       "5                                                None   \n",
       "6          {'SSF47719': 'p53 tetramerization domain'}   \n",
       "7                                                None   \n",
       "8                                                None   \n",
       "9                                                None   \n",
       "10                                               None   \n",
       "11                                               None   \n",
       "12                                               None   \n",
       "13                                               None   \n",
       "14                                               None   \n",
       "15                                               None   \n",
       "16                                               None   \n",
       "17                                               None   \n",
       "18                                               None   \n",
       "19                                               None   \n",
       "20  {'SSF109843': 'CAPPD, an extracellular domain ...   \n",
       "21           {'SSF56491': 'A heparin-binding domain'}   \n",
       "22  {'SSF89811': 'Amyloid beta a4 protein copper b...   \n",
       "23                          {'SSF57362': 'BPTI-like'}   \n",
       "24                                               None   \n",
       "\n",
       "                                        go_identifier  \\\n",
       "0   [GO:0003677, GO:0003700, GO:0006355, GO:000691...   \n",
       "1                            [GO:0003700, GO:0006355]   \n",
       "2                                          GO:0051262   \n",
       "3                                          GO:0000976   \n",
       "4    [GO:0003677, GO:0003700, GO:0006355, GO:0005634]   \n",
       "5                                          GO:0005515   \n",
       "6                                          GO:0051262   \n",
       "7                                                None   \n",
       "8                                          GO:0004867   \n",
       "9                                                None   \n",
       "10                                         GO:0016020   \n",
       "11                                         GO:0046914   \n",
       "12                                               None   \n",
       "13                                         GO:0016020   \n",
       "14                                         GO:0008201   \n",
       "15                                               None   \n",
       "16                                               None   \n",
       "17                                               None   \n",
       "18                                               None   \n",
       "19                                               None   \n",
       "20                                               None   \n",
       "21                                         GO:0008201   \n",
       "22                                         GO:0046914   \n",
       "23                                         GO:0004867   \n",
       "24                                         GO:0016020   \n",
       "\n",
       "                                              go_name  \\\n",
       "0   [DNA binding, DNA-binding transcription factor...   \n",
       "1   [DNA-binding transcription factor activity, re...   \n",
       "2                             protein tetramerization   \n",
       "3         transcription cis-regulatory region binding   \n",
       "4   [DNA binding, DNA-binding transcription factor...   \n",
       "5                                     protein binding   \n",
       "6                             protein tetramerization   \n",
       "7                                                None   \n",
       "8        serine-type endopeptidase inhibitor activity   \n",
       "9                                                None   \n",
       "10                                           membrane   \n",
       "11                       transition metal ion binding   \n",
       "12                                               None   \n",
       "13                                           membrane   \n",
       "14                                    heparin binding   \n",
       "15                                               None   \n",
       "16                                               None   \n",
       "17                                               None   \n",
       "18                                               None   \n",
       "19                                               None   \n",
       "20                                               None   \n",
       "21                                    heparin binding   \n",
       "22                       transition metal ion binding   \n",
       "23       serine-type endopeptidase inhibitor activity   \n",
       "24                                           membrane   \n",
       "\n",
       "                                          go_category  \\\n",
       "0   [molecular_function, molecular_function, biolo...   \n",
       "1            [molecular_function, biological_process]   \n",
       "2                                  biological_process   \n",
       "3                                  molecular_function   \n",
       "4   [molecular_function, molecular_function, biolo...   \n",
       "5                                  molecular_function   \n",
       "6                                  biological_process   \n",
       "7                                                None   \n",
       "8                                  molecular_function   \n",
       "9                                                None   \n",
       "10                                 cellular_component   \n",
       "11                                 molecular_function   \n",
       "12                                               None   \n",
       "13                                 cellular_component   \n",
       "14                                 molecular_function   \n",
       "15                                               None   \n",
       "16                                               None   \n",
       "17                                               None   \n",
       "18                                               None   \n",
       "19                                               None   \n",
       "20                                               None   \n",
       "21                                 molecular_function   \n",
       "22                                 molecular_function   \n",
       "23                                 molecular_function   \n",
       "24                                 cellular_component   \n",
       "\n",
       "                                            locations  \n",
       "0   {'fragments': [{'start': 3, 'end': 354, 'dc-st...  \n",
       "1   {'fragments': [{'start': 84, 'end': 275, 'dc-s...  \n",
       "2   {'fragments': [{'start': 307, 'end': 345, 'dc-...  \n",
       "3   {'fragments': [{'start': 88, 'end': 277, 'dc-s...  \n",
       "4   {'fragments': [{'start': 82, 'end': 282, 'dc-s...  \n",
       "5   {'fragments': [{'start': 6, 'end': 30, 'dc-sta...  \n",
       "6   {'fragments': [{'start': 307, 'end': 348, 'dc-...  \n",
       "7   {'fragments': [{'start': 225, 'end': 237, 'dc-...  \n",
       "8   {'fragments': [{'start': 288, 'end': 342, 'dc-...  \n",
       "9   {'fragments': [{'start': 24, 'end': 189, 'dc-s...  \n",
       "10  {'fragments': [{'start': 10, 'end': 770, 'dc-s...  \n",
       "11  {'fragments': [{'start': 132, 'end': 188, 'dc-...  \n",
       "12  {'fragments': [{'start': 742, 'end': 770, 'dc-...  \n",
       "13  {'fragments': [{'start': 675, 'end': 713, 'dc-...  \n",
       "14  {'fragments': [{'start': 31, 'end': 131, 'dc-s...  \n",
       "15  {'fragments': [{'start': 716, 'end': 766, 'dc-...  \n",
       "16  {'fragments': [{'start': 181, 'end': 188, 'dc-...  \n",
       "17  {'fragments': [{'start': 756, 'end': 763, 'dc-...  \n",
       "18  {'fragments': [{'start': 319, 'end': 337, 'dc-...  \n",
       "19  {'fragments': [{'start': 367, 'end': 565, 'dc-...  \n",
       "20  {'fragments': [{'start': 363, 'end': 580, 'dc-...  \n",
       "21  {'fragments': [{'start': 17, 'end': 123, 'dc-s...  \n",
       "22  {'fragments': [{'start': 124, 'end': 189, 'dc-...  \n",
       "23  {'fragments': [{'start': 279, 'end': 352, 'dc-...  \n",
       "24  {'fragments': [{'start': 672, 'end': 711, 'dc-...  \n",
       "\n",
       "[25 rows x 21 columns]"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "queries: List[Union[str, Dict[str, Any]]] = [\n",
    "    {\n",
    "        \"db\" : \"InterPro\",\n",
    "        \"modifiers\" : {},\n",
    "        \"filters\" : [\n",
    "            {\n",
    "                \"type\": \"protein\",\n",
    "                \"db\": \"reviewed\",\n",
    "                \"value\": \"Q29537\"\n",
    "            }\n",
    "        ]\n",
    "    },\n",
    "    {\n",
    "        \"db\": \"InterPro\",\n",
    "        \"modifiers\": {},\n",
    "        \"filters\": [\n",
    "            {\n",
    "                \"type\": \"protein\",\n",
    "                \"db\": \"reviewed\",\n",
    "                \"value\": \"P05067\"\n",
    "            }\n",
    "        ]\n",
    "    }\n",
    "]\n",
    "instance.fetch_batch(queries=queries, method=\"entry\", pages_to_fetch=1, parse=True, to_dataframe=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "57978983",
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'InterproInstance' object has no attribute 'fetch_to_dataframe'",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mAttributeError\u001b[39m                            Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[54]\u001b[39m\u001b[32m, line 1\u001b[39m\n\u001b[32m----> \u001b[39m\u001b[32m1\u001b[39m \u001b[43minstance\u001b[49m\u001b[43m.\u001b[49m\u001b[43mfetch_to_dataframe\u001b[49m(\n\u001b[32m      2\u001b[39m     [\n\u001b[32m      3\u001b[39m         {\n\u001b[32m      4\u001b[39m             \u001b[33m\"\u001b[39m\u001b[33mtype\u001b[39m\u001b[33m\"\u001b[39m: \u001b[33m\"\u001b[39m\u001b[33mentry\u001b[39m\u001b[33m\"\u001b[39m,\n\u001b[32m      5\u001b[39m             \u001b[33m\"\u001b[39m\u001b[33mdb\u001b[39m\u001b[33m\"\u001b[39m: \u001b[33m\"\u001b[39m\u001b[33mInterPro\u001b[39m\u001b[33m\"\u001b[39m,\n\u001b[32m      6\u001b[39m             \u001b[33m\"\u001b[39m\u001b[33mentry_integration\u001b[39m\u001b[33m\"\u001b[39m: \u001b[33m\"\u001b[39m\u001b[33m\"\u001b[39m,\n\u001b[32m      7\u001b[39m             \u001b[33m\"\u001b[39m\u001b[33mmodifiers\u001b[39m\u001b[33m\"\u001b[39m: {\n\u001b[32m      8\u001b[39m                 \u001b[33m\"\u001b[39m\u001b[33mgo_term\u001b[39m\u001b[33m\"\u001b[39m: \u001b[33m\"\u001b[39m\u001b[33mGO:0004867\u001b[39m\u001b[33m\"\u001b[39m\n\u001b[32m      9\u001b[39m             },\n\u001b[32m     10\u001b[39m             \u001b[33m\"\u001b[39m\u001b[33mfilter_type\u001b[39m\u001b[33m\"\u001b[39m: \u001b[33m\"\u001b[39m\u001b[33mprotein\u001b[39m\u001b[33m\"\u001b[39m,\n\u001b[32m     11\u001b[39m             \u001b[33m\"\u001b[39m\u001b[33mfilter_db\u001b[39m\u001b[33m\"\u001b[39m: \u001b[33m\"\u001b[39m\u001b[33mUniProt\u001b[39m\u001b[33m\"\u001b[39m,\n\u001b[32m     12\u001b[39m             \u001b[33m\"\u001b[39m\u001b[33mfilter_value\u001b[39m\u001b[33m\"\u001b[39m: \u001b[33m\"\u001b[39m\u001b[33mP05067\u001b[39m\u001b[33m\"\u001b[39m\n\u001b[32m     13\u001b[39m         }\n\u001b[32m     14\u001b[39m     ]\n\u001b[32m     15\u001b[39m )\n",
      "\u001b[31mAttributeError\u001b[39m: 'InterproInstance' object has no attribute 'fetch_to_dataframe'"
     ]
    }
   ],
   "source": [
    "instance.fetch_to_dataframe(\n",
    "    [\n",
    "        {\n",
    "            \"type\": \"entry\",\n",
    "            \"db\": \"InterPro\",\n",
    "            \"entry_integration\": \"\",\n",
    "            \"modifiers\": {\n",
    "                \"go_term\": \"GO:0004867\"\n",
    "            },\n",
    "            \"filter_type\": \"protein\",\n",
    "            \"filter_db\": \"UniProt\",\n",
    "            \"filter_value\": \"P05067\"\n",
    "        }\n",
    "    ]\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cf00f1af",
   "metadata": {},
   "source": [
    "## KEGG"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "38d0fe83",
   "metadata": {},
   "outputs": [],
   "source": [
    "from src.kegg import KEGGInterface"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "b3b4182c",
   "metadata": {},
   "outputs": [],
   "source": [
    "instance = KEGGInterface()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "63f4f4c4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cache_key: hsa:10458_get\n",
      "Cache_key: ece:Z5100_get\n",
      "Cache_key: hsa:10458_get\n",
      "Cache_key: ece:Z5100_get\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>entry</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10458             CDS       T01001</td>\n",
       "      <td>(RefSeq) BAR/IMD domain containing adaptor pro...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Z5100             CDS       T00044</td>\n",
       "      <td>(GenBank) espF</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                entry  \\\n",
       "0  10458             CDS       T01001   \n",
       "1  Z5100             CDS       T00044   \n",
       "\n",
       "                                                name  \n",
       "0  (RefSeq) BAR/IMD domain containing adaptor pro...  \n",
       "1                                     (GenBank) espF  "
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "instance.fetch_single(\n",
    "    query={\n",
    "        \"entries\": [\"hsa:10458\", \"ece:Z5100\"]\n",
    "    },\n",
    "    method=\"get\",\n",
    "    to_dataframe=True,\n",
    "    parse=True,\n",
    "    #fields_to_extract=[\"ENTRY\", \"NAME\", \"PATHWAY\"]\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e050909",
   "metadata": {},
   "source": [
    "## PathwayCommons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "b38e8b8a",
   "metadata": {},
   "outputs": [],
   "source": [
    "from src.pathwaycommons import PathwayCommonsInterface\n",
    "\n",
    "instance = PathwayCommonsInterface()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "2eb89c85",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cache_key: ['R-ALL-444824']_['interacts-with', 'used-to-produce']_fetch\n",
      "Prepared request: https://www.pathwaycommons.org/pc2/v2/fetch\n",
      "Fetching data with parameters: {'uri': ['R-ALL-444824'], 'format': 'jsonld', 'pattern': ['interacts-with', 'used-to-produce'], 'subpw': False}\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>db</th>\n",
       "      <th>cellular_location</th>\n",
       "      <th>chemical_formula</th>\n",
       "      <th>comment</th>\n",
       "      <th>component</th>\n",
       "      <th>component_stoichiometry</th>\n",
       "      <th>data_source</th>\n",
       "      <th>display_name</th>\n",
       "      <th>...</th>\n",
       "      <th>sequence_interval_begin</th>\n",
       "      <th>sequence_interval_end</th>\n",
       "      <th>sequence_position</th>\n",
       "      <th>standard_name</th>\n",
       "      <th>stoichiometric_coefficient</th>\n",
       "      <th>structure</th>\n",
       "      <th>structure_data</th>\n",
       "      <th>structure_format</th>\n",
       "      <th>term</th>\n",
       "      <th>xref</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHEBI:15551</td>\n",
       "      <td>None</td>\n",
       "      <td>chebi</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHEBI:15552</td>\n",
       "      <td>None</td>\n",
       "      <td>chebi</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHEBI:15555</td>\n",
       "      <td>None</td>\n",
       "      <td>chebi</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHEBI:16335</td>\n",
       "      <td>None</td>\n",
       "      <td>chebi</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHEBI:18397</td>\n",
       "      <td>None</td>\n",
       "      <td>chebi</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1822</th>\n",
       "      <td>Q9P1P5</td>\n",
       "      <td>None</td>\n",
       "      <td>uniprot</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1823</th>\n",
       "      <td>Q9P2W3</td>\n",
       "      <td>None</td>\n",
       "      <td>uniprot</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1824</th>\n",
       "      <td>Q9UBI6</td>\n",
       "      <td>None</td>\n",
       "      <td>uniprot</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1825</th>\n",
       "      <td>Q9UK08</td>\n",
       "      <td>None</td>\n",
       "      <td>uniprot</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1826</th>\n",
       "      <td>Q9Y5Y3</td>\n",
       "      <td>None</td>\n",
       "      <td>uniprot</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1827 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               id  name       db cellular_location chemical_formula comment  \\\n",
       "0     CHEBI:15551  None    chebi              None             None    None   \n",
       "1     CHEBI:15552  None    chebi              None             None    None   \n",
       "2     CHEBI:15555  None    chebi              None             None    None   \n",
       "3     CHEBI:16335  None    chebi              None             None    None   \n",
       "4     CHEBI:18397  None    chebi              None             None    None   \n",
       "...           ...   ...      ...               ...              ...     ...   \n",
       "1822       Q9P1P5  None  uniprot              None             None    None   \n",
       "1823       Q9P2W3  None  uniprot              None             None    None   \n",
       "1824       Q9UBI6  None  uniprot              None             None    None   \n",
       "1825       Q9UK08  None  uniprot              None             None    None   \n",
       "1826       Q9Y5Y3  None  uniprot              None             None    None   \n",
       "\n",
       "     component component_stoichiometry data_source display_name  ...  \\\n",
       "0         None                    None        None         None  ...   \n",
       "1         None                    None        None         None  ...   \n",
       "2         None                    None        None         None  ...   \n",
       "3         None                    None        None         None  ...   \n",
       "4         None                    None        None         None  ...   \n",
       "...        ...                     ...         ...          ...  ...   \n",
       "1822      None                    None        None         None  ...   \n",
       "1823      None                    None        None         None  ...   \n",
       "1824      None                    None        None         None  ...   \n",
       "1825      None                    None        None         None  ...   \n",
       "1826      None                    None        None         None  ...   \n",
       "\n",
       "     sequence_interval_begin sequence_interval_end sequence_position  \\\n",
       "0                       None                  None              None   \n",
       "1                       None                  None              None   \n",
       "2                       None                  None              None   \n",
       "3                       None                  None              None   \n",
       "4                       None                  None              None   \n",
       "...                      ...                   ...               ...   \n",
       "1822                    None                  None              None   \n",
       "1823                    None                  None              None   \n",
       "1824                    None                  None              None   \n",
       "1825                    None                  None              None   \n",
       "1826                    None                  None              None   \n",
       "\n",
       "     standard_name stoichiometric_coefficient structure structure_data  \\\n",
       "0             None                       None      None           None   \n",
       "1             None                       None      None           None   \n",
       "2             None                       None      None           None   \n",
       "3             None                       None      None           None   \n",
       "4             None                       None      None           None   \n",
       "...            ...                        ...       ...            ...   \n",
       "1822          None                       None      None           None   \n",
       "1823          None                       None      None           None   \n",
       "1824          None                       None      None           None   \n",
       "1825          None                       None      None           None   \n",
       "1826          None                       None      None           None   \n",
       "\n",
       "     structure_format  term  xref  \n",
       "0                None  None  None  \n",
       "1                None  None  None  \n",
       "2                None  None  None  \n",
       "3                None  None  None  \n",
       "4                None  None  None  \n",
       "...               ...   ...   ...  \n",
       "1822             None  None  None  \n",
       "1823             None  None  None  \n",
       "1824             None  None  None  \n",
       "1825             None  None  None  \n",
       "1826             None  None  None  \n",
       "\n",
       "[1827 rows x 34 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "instance.fetch_single(\n",
    "    query = {\n",
    "        \"uri\": [\"R-ALL-444824\"],\n",
    "        \"pattern\": [\"interacts-with\", \"used-to-produce\"]\n",
    "    },\n",
    "    method=\"fetch\",\n",
    "    parse=True,\n",
    "    to_dataframe=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f1cab3b4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cache_key: APP_['9606']_['reactome', 'uniprot']_top_pathways\n",
      "Prepared request: https://www.pathwaycommons.org/pc2/v2/top_pathways\n",
      "Fetching data with parameters: {'q': 'APP', 'organism': ['9606'], 'datasource': ['reactome', 'uniprot']}\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uri</th>\n",
       "      <th>biopax_class</th>\n",
       "      <th>name</th>\n",
       "      <th>data_source</th>\n",
       "      <th>organism</th>\n",
       "      <th>pathway</th>\n",
       "      <th>excerpt</th>\n",
       "      <th>num_participants</th>\n",
       "      <th>num_processes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>http://bioregistry.io/reactome:R-HSA-392499</td>\n",
       "      <td>Pathway</td>\n",
       "      <td>Metabolism of proteins</td>\n",
       "      <td>pc14:reactome</td>\n",
       "      <td>http://bioregistry.io/ncbitaxon:9606</td>\n",
       "      <td>[]</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>http://bioregistry.io/reactome:R-HSA-162582</td>\n",
       "      <td>Pathway</td>\n",
       "      <td>Signaling Pathways</td>\n",
       "      <td>pc14:reactome</td>\n",
       "      <td>http://bioregistry.io/ncbitaxon:9606</td>\n",
       "      <td>[]</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           uri biopax_class  \\\n",
       "0  http://bioregistry.io/reactome:R-HSA-392499      Pathway   \n",
       "1  http://bioregistry.io/reactome:R-HSA-162582      Pathway   \n",
       "\n",
       "                     name    data_source  \\\n",
       "0  Metabolism of proteins  pc14:reactome   \n",
       "1      Signaling Pathways  pc14:reactome   \n",
       "\n",
       "                               organism pathway excerpt  num_participants  \\\n",
       "0  http://bioregistry.io/ncbitaxon:9606      []    None                 0   \n",
       "1  http://bioregistry.io/ncbitaxon:9606      []    None                 0   \n",
       "\n",
       "   num_processes  \n",
       "0              9  \n",
       "1             17  "
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "instance.fetch_single(\n",
    "    query = {\n",
    "        \"q\": \"APP\",\n",
    "        \"organism\": [\n",
    "            \"9606\"\n",
    "        ],\n",
    "        \"datasource\": [\n",
    "            \"reactome\",\n",
    "            \"uniprot\"\n",
    "        ]\n",
    "    },\n",
    "    method=\"top_pathways\",\n",
    "    parse=True,\n",
    "    to_dataframe=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "714ab12a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cache_key: ['Q16602']_1_jsonld_['9606']_['reactome']_['interacts-with', 'used-to-produce']_True_undirected_neighborhood\n",
      "Prepared request: https://www.pathwaycommons.org/pc2/v2/neighborhood\n",
      "Fetching data with parameters: {'source': ['Q16602'], 'limit': 1, 'format': 'jsonld', 'organism': ['9606'], 'datasource': ['reactome'], 'pattern': ['interacts-with', 'used-to-produce'], 'subpw': True, 'direction': 'undirected'}\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>db</th>\n",
       "      <th>cellular_location</th>\n",
       "      <th>chemical_formula</th>\n",
       "      <th>comment</th>\n",
       "      <th>component</th>\n",
       "      <th>component_stoichiometry</th>\n",
       "      <th>control_type</th>\n",
       "      <th>controlled</th>\n",
       "      <th>...</th>\n",
       "      <th>standard_name</th>\n",
       "      <th>stoichiometric_coefficient</th>\n",
       "      <th>structure</th>\n",
       "      <th>structure_data</th>\n",
       "      <th>structure_format</th>\n",
       "      <th>term</th>\n",
       "      <th>title</th>\n",
       "      <th>url</th>\n",
       "      <th>xref</th>\n",
       "      <th>year</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHEBI:10100</td>\n",
       "      <td>None</td>\n",
       "      <td>chebi</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHEBI:15354</td>\n",
       "      <td>None</td>\n",
       "      <td>chebi</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6454</th>\n",
       "      <td>Q9UNW8</td>\n",
       "      <td>None</td>\n",
       "      <td>uniprot</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6455</th>\n",
       "      <td>Q9Y271</td>\n",
       "      <td>None</td>\n",
       "      <td>uniprot</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6456</th>\n",
       "      <td>Q9Y287</td>\n",
       "      <td>None</td>\n",
       "      <td>uniprot</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6457</th>\n",
       "      <td>Q9Y5X5</td>\n",
       "      <td>None</td>\n",
       "      <td>uniprot</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6458</th>\n",
       "      <td>Q9Y5Y3</td>\n",
       "      <td>None</td>\n",
       "      <td>uniprot</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6459 rows × 45 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               id  name       db cellular_location chemical_formula comment  \\\n",
       "0            None  None     None              None             None    None   \n",
       "1            None  None     None              None             None    None   \n",
       "2            None  None     None              None             None    None   \n",
       "3     CHEBI:10100  None    chebi              None             None    None   \n",
       "4     CHEBI:15354  None    chebi              None             None    None   \n",
       "...           ...   ...      ...               ...              ...     ...   \n",
       "6454       Q9UNW8  None  uniprot              None             None    None   \n",
       "6455       Q9Y271  None  uniprot              None             None    None   \n",
       "6456       Q9Y287  None  uniprot              None             None    None   \n",
       "6457       Q9Y5X5  None  uniprot              None             None    None   \n",
       "6458       Q9Y5Y3  None  uniprot              None             None    None   \n",
       "\n",
       "     component component_stoichiometry control_type controlled  ...  \\\n",
       "0         None                    None         None       None  ...   \n",
       "1         None                    None         None       None  ...   \n",
       "2         None                    None         None       None  ...   \n",
       "3         None                    None         None       None  ...   \n",
       "4         None                    None         None       None  ...   \n",
       "...        ...                     ...          ...        ...  ...   \n",
       "6454      None                    None         None       None  ...   \n",
       "6455      None                    None         None       None  ...   \n",
       "6456      None                    None         None       None  ...   \n",
       "6457      None                    None         None       None  ...   \n",
       "6458      None                    None         None       None  ...   \n",
       "\n",
       "     standard_name stoichiometric_coefficient structure structure_data  \\\n",
       "0             None                       None      None           None   \n",
       "1             None                       None      None           None   \n",
       "2             None                       None      None           None   \n",
       "3             None                       None      None           None   \n",
       "4             None                       None      None           None   \n",
       "...            ...                        ...       ...            ...   \n",
       "6454          None                       None      None           None   \n",
       "6455          None                       None      None           None   \n",
       "6456          None                       None      None           None   \n",
       "6457          None                       None      None           None   \n",
       "6458          None                       None      None           None   \n",
       "\n",
       "     structure_format  term title   url  xref  year  \n",
       "0                None  None  None  None  None  None  \n",
       "1                None  None  None  None  None  None  \n",
       "2                None  None  None  None  None  None  \n",
       "3                None  None  None  None  None  None  \n",
       "4                None  None  None  None  None  None  \n",
       "...               ...   ...   ...   ...   ...   ...  \n",
       "6454             None  None  None  None  None  None  \n",
       "6455             None  None  None  None  None  None  \n",
       "6456             None  None  None  None  None  None  \n",
       "6457             None  None  None  None  None  None  \n",
       "6458             None  None  None  None  None  None  \n",
       "\n",
       "[6459 rows x 45 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "instance.fetch_single(\n",
    "    query = {\n",
    "        \"source\": [\n",
    "            \"Q16602\"\n",
    "        ],\n",
    "        \"limit\": 1,\n",
    "        \"format\": \"jsonld\",\n",
    "        \"organism\": [\n",
    "            \"9606\"\n",
    "        ],\n",
    "        \"datasource\": [\n",
    "            \"reactome\"\n",
    "        ],\n",
    "        \"pattern\": [\n",
    "            \"interacts-with\",\n",
    "            \"used-to-produce\"\n",
    "        ],\n",
    "        \"subpw\": True,\n",
    "        \"direction\": \"undirected\"\n",
    "    },\n",
    "    method=\"neighborhood\",\n",
    "    parse=True,\n",
    "    to_dataframe=True\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "41ec1f4d",
   "metadata": {},
   "source": [
    "## PANTHER"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "2ce5522a",
   "metadata": {},
   "outputs": [],
   "source": [
    "from src.panther import PantherInterface"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "a25595e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "instance = PantherInterface()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "8d4180ce",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cache_key: PTHR10000_familyortholog\n",
      "Inputs for method 'familyortholog': {'family': 'PTHR10000'}\n",
      "Prepared request: https://pantherdb.org/services/oai/pantherdb/familyortholog?family=PTHR10000\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pid</th>\n",
       "      <th>sid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PTN000796677</td>\n",
       "      <td>PTN008492220</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PTN000796677</td>\n",
       "      <td>PTN008492221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PTN000796677</td>\n",
       "      <td>PTN000000088</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PTN000796677</td>\n",
       "      <td>PTN000796679</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PTN000796677</td>\n",
       "      <td>PTN001087724</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3652</th>\n",
       "      <td>PTN001597048</td>\n",
       "      <td>PTN007416238</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3653</th>\n",
       "      <td>PTN001597048</td>\n",
       "      <td>PTN001596985</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3654</th>\n",
       "      <td>PTN001597047</td>\n",
       "      <td>PTN007416238</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3655</th>\n",
       "      <td>PTN001597047</td>\n",
       "      <td>PTN001596985</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3656</th>\n",
       "      <td>PTN007416238</td>\n",
       "      <td>PTN001596985</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3657 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               pid           sid\n",
       "0     PTN000796677  PTN008492220\n",
       "1     PTN000796677  PTN008492221\n",
       "2     PTN000796677  PTN000000088\n",
       "3     PTN000796677  PTN000796679\n",
       "4     PTN000796677  PTN001087724\n",
       "...            ...           ...\n",
       "3652  PTN001597048  PTN007416238\n",
       "3653  PTN001597048  PTN001596985\n",
       "3654  PTN001597047  PTN007416238\n",
       "3655  PTN001597047  PTN001596985\n",
       "3656  PTN007416238  PTN001596985\n",
       "\n",
       "[3657 rows x 2 columns]"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "instance.fetch_single(\n",
    "    query={\n",
    "        \"family\": \"PTHR10000\"\n",
    "    },\n",
    "    method=\"familyortholog\",\n",
    "    parse=True,\n",
    "    to_dataframe=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "5672cc6e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cache_key: PTHR10000_familymsa\n",
      "Inputs for method 'familymsa': {'family': 'PTHR10000'}\n",
      "Prepared request: https://pantherdb.org/services/oai/pantherdb/familymsa?family=PTHR10000\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequence</th>\n",
       "      <th>persistent_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ml...............................................</td>\n",
       "      <td>PTN007416230</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>.................................................</td>\n",
       "      <td>PTN001597030</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>.................................................</td>\n",
       "      <td>PTN001597042</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>.................................................</td>\n",
       "      <td>PTN000000095</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ml.................xx..............xxnxs.........</td>\n",
       "      <td>PTN004118870</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>181</th>\n",
       "      <td>.................................................</td>\n",
       "      <td>PTN001597027</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>182</th>\n",
       "      <td>.................................................</td>\n",
       "      <td>PTN001087692</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>183</th>\n",
       "      <td>.................................................</td>\n",
       "      <td>PTN001087691</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>184</th>\n",
       "      <td>.................................................</td>\n",
       "      <td>PTN002558009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>185</th>\n",
       "      <td>.................................................</td>\n",
       "      <td>PTN002224257</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>186 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              sequence persistent_id\n",
       "0    ml...............................................  PTN007416230\n",
       "1    .................................................  PTN001597030\n",
       "2    .................................................  PTN001597042\n",
       "3    .................................................  PTN000000095\n",
       "4    ml.................xx..............xxnxs.........  PTN004118870\n",
       "..                                                 ...           ...\n",
       "181  .................................................  PTN001597027\n",
       "182  .................................................  PTN001087692\n",
       "183  .................................................  PTN001087691\n",
       "184  .................................................  PTN002558009\n",
       "185  .................................................  PTN002224257\n",
       "\n",
       "[186 rows x 2 columns]"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "instance.fetch_single(\n",
    "    query=\"PTHR10000\",\n",
    "    method=\"familymsa\",\n",
    "    parse=True,\n",
    "    to_dataframe=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "f15702e6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cache_key: BRCA1_9606_geneinfo\n",
      "Cache_key: CIROP_9606_geneinfo\n",
      "Inputs for method 'geneinfo': {'geneInputList': 'BRCA1,CIROP', 'organism': '9606'}\n",
      "Prepared request: https://pantherdb.org/services/oai/pantherdb/geneinfo?geneInputList=BRCA1%2CCIROP&organism=9606\n",
      "Cache_key: BRCA1_9606_geneinfo\n",
      "Cache_key: CIROP_9606_geneinfo\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PTHR13763</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PTHR10942</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          id\n",
       "0  PTHR13763\n",
       "1  PTHR10942"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "instance.fetch_single(\n",
    "    query={\n",
    "        \"geneInputList\": [\"BRCA1\", \"CIROP\"],\n",
    "        \"organism\": \"9606\"\n",
    "    },\n",
    "    method=\"geneinfo\",\n",
    "    parse=True,\n",
    "    to_dataframe=True,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "ea424379",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cache_key: BRCA1_9606_geneinfo\n",
      "Cache_key: CIROP_9606_geneinfo\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PTHR13763</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PTHR10942</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          id\n",
       "0  PTHR13763\n",
       "1  PTHR10942"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "instance.fetch_batch(\n",
    "    queries=[{\n",
    "        \"geneInputList\": [\"BRCA1\", \"CIROP\"],\n",
    "        \"organism\": \"9606\"\n",
    "    }],\n",
    "    method=\"geneinfo\",\n",
    "    parse=True,\n",
    "    to_dataframe=True,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9837e24a",
   "metadata": {},
   "source": [
    "## PDB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "ee51970f",
   "metadata": {},
   "outputs": [],
   "source": [
    "from src.proteindatabank import PDBInterface"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "a1025d14",
   "metadata": {},
   "outputs": [],
   "source": [
    "instance = PDBInterface(\n",
    "    download_structures=True,\n",
    "    return_data_list=[\"rcsb_id\", \"rcsb_comp_model_provenance\", \"rcsb_entry_info\"],\n",
    "    output_dir=\"results\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "32e10044",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Info: Structure for 4HHB already exists in pdb format.\n",
      "Cache_key: 4HHB_entry\n",
      "Prepared request: https://data.rcsb.org/rest/v1/core/entry/4HHB\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rcsb_id</th>\n",
       "      <th>model_provenance</th>\n",
       "      <th>branched_molecular_weight_minimum</th>\n",
       "      <th>resolution_combined</th>\n",
       "      <th>experimental_method</th>\n",
       "      <th>diffrn_resolution_high</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4HHB</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>1.74</td>\n",
       "      <td>X-ray</td>\n",
       "      <td>1.74</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  rcsb_id model_provenance branched_molecular_weight_minimum  \\\n",
       "0    4HHB             None                              None   \n",
       "\n",
       "   resolution_combined experimental_method  diffrn_resolution_high  \n",
       "0                 1.74               X-ray                    1.74  "
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "instance.fetch_single(query=\"4HHB\", method=\"entry\", parse=True, to_dataframe=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "991b6eed",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cache_key: 4HHB_entry\n",
      "Cache_key: 1A8I_entry\n",
      "Cache_key: 1A8J_entry\n",
      "Cache_key: 1A8K_entry\n",
      "Cache_key: 1A8L_entry\n",
      "Cache_key: 1A8M_entry\n",
      "Info: Structure for 1A8I already exists in pdb format.\n",
      "Cache_key: 1A8I_entry\n",
      "Prepared request: https://data.rcsb.org/rest/v1/core/entry/1A8I\n",
      "Info: Structure for 1A8J already exists in pdb format.\n",
      "Cache_key: 1A8J_entry\n",
      "Prepared request: https://data.rcsb.org/rest/v1/core/entry/1A8J\n",
      "Info: Structure for 1A8K already exists in pdb format.\n",
      "Cache_key: 1A8K_entry\n",
      "Prepared request: https://data.rcsb.org/rest/v1/core/entry/1A8K\n",
      "Info: Structure for 1A8L already exists in pdb format.\n",
      "Cache_key: 1A8L_entry\n",
      "Info: Structure for 1A8M already exists in pdb format.\n",
      "Cache_key: 1A8M_entry\n",
      "Prepared request: https://data.rcsb.org/rest/v1/core/entry/1A8L\n",
      "Prepared request: https://data.rcsb.org/rest/v1/core/entry/1A8M\n",
      "Info: Structure for 4HHB already exists in pdb format.\n",
      "Info: Structure for 1A8I already exists in pdb format.\n",
      "Info: Structure for 1A8J already exists in pdb format.\n",
      "Info: Structure for 1A8K already exists in pdb format.\n",
      "Info: Structure for 1A8L already exists in pdb format.\n",
      "Info: Structure for 1A8M already exists in pdb format.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rcsb_id</th>\n",
       "      <th>model_provenance</th>\n",
       "      <th>branched_molecular_weight_minimum</th>\n",
       "      <th>resolution_combined</th>\n",
       "      <th>experimental_method</th>\n",
       "      <th>diffrn_resolution_high</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4HHB</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>1.74</td>\n",
       "      <td>X-ray</td>\n",
       "      <td>1.74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1A8I</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>1.78</td>\n",
       "      <td>X-ray</td>\n",
       "      <td>1.78</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1A8J</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>2.70</td>\n",
       "      <td>X-ray</td>\n",
       "      <td>2.70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1A8K</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>2.00</td>\n",
       "      <td>X-ray</td>\n",
       "      <td>2.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1A8L</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>1.90</td>\n",
       "      <td>X-ray</td>\n",
       "      <td>1.90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1A8M</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>2.30</td>\n",
       "      <td>X-ray</td>\n",
       "      <td>2.30</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  rcsb_id model_provenance branched_molecular_weight_minimum  \\\n",
       "0    4HHB             None                              None   \n",
       "1    1A8I             None                              None   \n",
       "2    1A8J             None                              None   \n",
       "3    1A8K             None                              None   \n",
       "4    1A8L             None                              None   \n",
       "5    1A8M             None                              None   \n",
       "\n",
       "   resolution_combined experimental_method  diffrn_resolution_high  \n",
       "0                 1.74               X-ray                    1.74  \n",
       "1                 1.78               X-ray                    1.78  \n",
       "2                 2.70               X-ray                    2.70  \n",
       "3                 2.00               X-ray                    2.00  \n",
       "4                 1.90               X-ray                    1.90  \n",
       "5                 2.30               X-ray                    2.30  "
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "instance.fetch_batch(\n",
    "    queries=[\"4HHB\", \"1A8I\", \"1A8J\", \"1A8K\", \"1A8L\", \"1A8M\"],\n",
    "    method=\"entry\",\n",
    "    parse=True,\n",
    "    to_dataframe=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "50fcf4ea",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Info: Structure for 4HHB already exists in pdb format.\n",
      "Cache_key: 4HHB_entry\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'audit_author.name': 'str',\n",
       " 'audit_author.pdbx_ordinal': 'int',\n",
       " 'cell.angle_alpha': 'dict(float)',\n",
       " 'cell.angle_beta': 'dict(float)',\n",
       " 'cell.angle_gamma': 'dict(float)',\n",
       " 'cell.length_a': 'dict(float)',\n",
       " 'cell.length_b': 'dict(float)',\n",
       " 'cell.length_c': 'dict(float)',\n",
       " 'cell.zpdb': 'dict(int)',\n",
       " 'cell': 'dict',\n",
       " 'citation.country': 'str',\n",
       " 'citation.id': 'str',\n",
       " 'citation.journal_abbrev': 'str',\n",
       " 'citation.journal_id_astm': 'str',\n",
       " 'citation.journal_id_csd': 'str',\n",
       " 'citation.journal_id_issn': 'str',\n",
       " 'citation.journal_volume': 'str',\n",
       " 'citation.page_first': 'str',\n",
       " 'citation.page_last': 'str',\n",
       " 'citation.pdbx_database_id_doi': 'str',\n",
       " 'citation.pdbx_database_id_pub_med': 'int',\n",
       " 'citation.rcsb_authors': 'list(str)',\n",
       " 'citation.rcsb_is_primary': 'str',\n",
       " 'citation.rcsb_journal_abbrev': 'str',\n",
       " 'citation.title': 'str',\n",
       " 'citation.year': 'int',\n",
       " 'database2.database_code': 'str',\n",
       " 'database2.database_id': 'str',\n",
       " 'database2.pdbx_doi': 'str',\n",
       " 'database2.pdbx_database_accession': 'str',\n",
       " 'diffrn.crystal_id': 'str',\n",
       " 'diffrn.id': 'str',\n",
       " 'entry.id': 'dict(str)',\n",
       " 'entry': 'dict',\n",
       " 'exptl.method': 'str',\n",
       " 'exptl_crystal.density_matthews': 'float',\n",
       " 'exptl_crystal.density_percent_sol': 'float',\n",
       " 'exptl_crystal.id': 'str',\n",
       " 'pdbx_audit_revision_category.category': 'str',\n",
       " 'pdbx_audit_revision_category.data_content_type': 'str',\n",
       " 'pdbx_audit_revision_category.ordinal': 'int',\n",
       " 'pdbx_audit_revision_category.revision_ordinal': 'int',\n",
       " 'pdbx_audit_revision_details.data_content_type': 'str',\n",
       " 'pdbx_audit_revision_details.ordinal': 'int',\n",
       " 'pdbx_audit_revision_details.provider': 'str',\n",
       " 'pdbx_audit_revision_details.revision_ordinal': 'int',\n",
       " 'pdbx_audit_revision_details.type': 'str',\n",
       " 'pdbx_audit_revision_group.data_content_type': 'str',\n",
       " 'pdbx_audit_revision_group.group': 'str',\n",
       " 'pdbx_audit_revision_group.ordinal': 'int',\n",
       " 'pdbx_audit_revision_group.revision_ordinal': 'int',\n",
       " 'pdbx_audit_revision_history.data_content_type': 'str',\n",
       " 'pdbx_audit_revision_history.major_revision': 'int',\n",
       " 'pdbx_audit_revision_history.minor_revision': 'int',\n",
       " 'pdbx_audit_revision_history.ordinal': 'int',\n",
       " 'pdbx_audit_revision_history.revision_date': 'str',\n",
       " 'pdbx_audit_revision_item.data_content_type': 'str',\n",
       " 'pdbx_audit_revision_item.item': 'str',\n",
       " 'pdbx_audit_revision_item.ordinal': 'int',\n",
       " 'pdbx_audit_revision_item.revision_ordinal': 'int',\n",
       " 'pdbx_database_pdbobs_spr.date': 'str',\n",
       " 'pdbx_database_pdbobs_spr.id': 'str',\n",
       " 'pdbx_database_pdbobs_spr.pdb_id': 'str',\n",
       " 'pdbx_database_pdbobs_spr.replace_pdb_id': 'str',\n",
       " 'pdbx_database_related.content_type': 'str',\n",
       " 'pdbx_database_related.db_id': 'str',\n",
       " 'pdbx_database_related.db_name': 'str',\n",
       " 'pdbx_database_related.details': 'str',\n",
       " 'pdbx_database_status.pdb_format_compatible': 'dict(str)',\n",
       " 'pdbx_database_status.process_site': 'dict(str)',\n",
       " 'pdbx_database_status.recvd_initial_deposition_date': 'dict(str)',\n",
       " 'pdbx_database_status.status_code': 'dict(str)',\n",
       " 'pdbx_database_status': 'dict',\n",
       " 'pdbx_vrpt_summary.attempted_validation_steps': 'dict(str)',\n",
       " 'pdbx_vrpt_summary.ligands_for_buster_report': 'dict(str)',\n",
       " 'pdbx_vrpt_summary.report_creation_date': 'dict(str)',\n",
       " 'pdbx_vrpt_summary': 'dict',\n",
       " 'pdbx_vrpt_summary_geometry.angles_rmsz': 'float',\n",
       " 'pdbx_vrpt_summary_geometry.bonds_rmsz': 'float',\n",
       " 'pdbx_vrpt_summary_geometry.clashscore': 'float',\n",
       " 'pdbx_vrpt_summary_geometry.num_hreduce': 'int',\n",
       " 'pdbx_vrpt_summary_geometry.num_angles_rmsz': 'int',\n",
       " 'pdbx_vrpt_summary_geometry.num_bonds_rmsz': 'int',\n",
       " 'pdbx_vrpt_summary_geometry.percent_ramachandran_outliers': 'float',\n",
       " 'pdbx_vrpt_summary_geometry.percent_rotamer_outliers': 'float',\n",
       " 'rcsb_accession_info.deposit_date': 'dict(str)',\n",
       " 'rcsb_accession_info.has_released_experimental_data': 'dict(str)',\n",
       " 'rcsb_accession_info.initial_release_date': 'dict(str)',\n",
       " 'rcsb_accession_info.major_revision': 'dict(int)',\n",
       " 'rcsb_accession_info.minor_revision': 'dict(int)',\n",
       " 'rcsb_accession_info.revision_date': 'dict(str)',\n",
       " 'rcsb_accession_info.status_code': 'dict(str)',\n",
       " 'rcsb_accession_info': 'dict',\n",
       " 'rcsb_entry_container_identifiers.assembly_ids': 'dict(list(str))',\n",
       " 'rcsb_entry_container_identifiers.entity_ids': 'dict(list(str))',\n",
       " 'rcsb_entry_container_identifiers.entry_id': 'dict(str)',\n",
       " 'rcsb_entry_container_identifiers.model_ids': 'dict(list(int))',\n",
       " 'rcsb_entry_container_identifiers.non_polymer_entity_ids': 'dict(list(str))',\n",
       " 'rcsb_entry_container_identifiers.polymer_entity_ids': 'dict(list(str))',\n",
       " 'rcsb_entry_container_identifiers.pubmed_id': 'dict(int)',\n",
       " 'rcsb_entry_container_identifiers.rcsb_id': 'dict(str)',\n",
       " 'rcsb_entry_container_identifiers': 'dict',\n",
       " 'rcsb_entry_info.assembly_count': 'dict(int)',\n",
       " 'rcsb_entry_info.branched_entity_count': 'dict(int)',\n",
       " 'rcsb_entry_info.cis_peptide_count': 'dict(int)',\n",
       " 'rcsb_entry_info.deposited_atom_count': 'dict(int)',\n",
       " 'rcsb_entry_info.deposited_deuterated_water_count': 'dict(int)',\n",
       " 'rcsb_entry_info.deposited_hydrogen_atom_count': 'dict(int)',\n",
       " 'rcsb_entry_info.deposited_model_count': 'dict(int)',\n",
       " 'rcsb_entry_info.deposited_modeled_polymer_monomer_count': 'dict(int)',\n",
       " 'rcsb_entry_info.deposited_nonpolymer_entity_instance_count': 'dict(int)',\n",
       " 'rcsb_entry_info.deposited_polymer_entity_instance_count': 'dict(int)',\n",
       " 'rcsb_entry_info.deposited_polymer_monomer_count': 'dict(int)',\n",
       " 'rcsb_entry_info.deposited_solvent_atom_count': 'dict(int)',\n",
       " 'rcsb_entry_info.deposited_unmodeled_polymer_monomer_count': 'dict(int)',\n",
       " 'rcsb_entry_info.disulfide_bond_count': 'dict(int)',\n",
       " 'rcsb_entry_info.entity_count': 'dict(int)',\n",
       " 'rcsb_entry_info.experimental_method': 'dict(str)',\n",
       " 'rcsb_entry_info.experimental_method_count': 'dict(int)',\n",
       " 'rcsb_entry_info.inter_mol_covalent_bond_count': 'dict(int)',\n",
       " 'rcsb_entry_info.inter_mol_metalic_bond_count': 'dict(int)',\n",
       " 'rcsb_entry_info.molecular_weight': 'dict(float)',\n",
       " 'rcsb_entry_info.na_polymer_entity_types': 'dict(str)',\n",
       " 'rcsb_entry_info.nonpolymer_bound_components': 'dict(list(str))',\n",
       " 'rcsb_entry_info.nonpolymer_entity_count': 'dict(int)',\n",
       " 'rcsb_entry_info.nonpolymer_molecular_weight_maximum': 'dict(float)',\n",
       " 'rcsb_entry_info.nonpolymer_molecular_weight_minimum': 'dict(float)',\n",
       " 'rcsb_entry_info.polymer_composition': 'dict(str)',\n",
       " 'rcsb_entry_info.polymer_entity_count': 'dict(int)',\n",
       " 'rcsb_entry_info.polymer_entity_count_dna': 'dict(int)',\n",
       " 'rcsb_entry_info.polymer_entity_count_rna': 'dict(int)',\n",
       " 'rcsb_entry_info.polymer_entity_count_nucleic_acid': 'dict(int)',\n",
       " 'rcsb_entry_info.polymer_entity_count_nucleic_acid_hybrid': 'dict(int)',\n",
       " 'rcsb_entry_info.polymer_entity_count_protein': 'dict(int)',\n",
       " 'rcsb_entry_info.polymer_entity_taxonomy_count': 'dict(int)',\n",
       " 'rcsb_entry_info.polymer_molecular_weight_maximum': 'dict(float)',\n",
       " 'rcsb_entry_info.polymer_molecular_weight_minimum': 'dict(float)',\n",
       " 'rcsb_entry_info.polymer_monomer_count_maximum': 'dict(int)',\n",
       " 'rcsb_entry_info.polymer_monomer_count_minimum': 'dict(int)',\n",
       " 'rcsb_entry_info.resolution_combined': 'dict(list(float))',\n",
       " 'rcsb_entry_info.selected_polymer_entity_types': 'dict(str)',\n",
       " 'rcsb_entry_info.solvent_entity_count': 'dict(int)',\n",
       " 'rcsb_entry_info.structure_determination_methodology': 'dict(str)',\n",
       " 'rcsb_entry_info.structure_determination_methodology_priority': 'dict(int)',\n",
       " 'rcsb_entry_info.diffrn_resolution_high.provenance_source': 'dict(dict(str))',\n",
       " 'rcsb_entry_info.diffrn_resolution_high.value': 'dict(dict(float))',\n",
       " 'rcsb_entry_info.diffrn_resolution_high': 'dict(dict)',\n",
       " 'rcsb_entry_info': 'dict',\n",
       " 'rcsb_primary_citation.country': 'dict(str)',\n",
       " 'rcsb_primary_citation.id': 'dict(str)',\n",
       " 'rcsb_primary_citation.journal_abbrev': 'dict(str)',\n",
       " 'rcsb_primary_citation.journal_id_astm': 'dict(str)',\n",
       " 'rcsb_primary_citation.journal_id_csd': 'dict(str)',\n",
       " 'rcsb_primary_citation.journal_id_issn': 'dict(str)',\n",
       " 'rcsb_primary_citation.journal_volume': 'dict(str)',\n",
       " 'rcsb_primary_citation.page_first': 'dict(str)',\n",
       " 'rcsb_primary_citation.page_last': 'dict(str)',\n",
       " 'rcsb_primary_citation.pdbx_database_id_doi': 'dict(str)',\n",
       " 'rcsb_primary_citation.pdbx_database_id_pub_med': 'dict(int)',\n",
       " 'rcsb_primary_citation.rcsb_orcididentifiers': 'dict(list(str))',\n",
       " 'rcsb_primary_citation.rcsb_authors': 'dict(list(str))',\n",
       " 'rcsb_primary_citation.rcsb_journal_abbrev': 'dict(str)',\n",
       " 'rcsb_primary_citation.title': 'dict(str)',\n",
       " 'rcsb_primary_citation.year': 'dict(int)',\n",
       " 'rcsb_primary_citation': 'dict',\n",
       " 'refine.details': 'str',\n",
       " 'refine.ls_rfactor_rwork': 'float',\n",
       " 'refine.ls_dres_high': 'float',\n",
       " 'refine.pdbx_diffrn_id': 'list(str)',\n",
       " 'refine.pdbx_refine_id': 'str',\n",
       " 'refine_hist.cycle_id': 'str',\n",
       " 'refine_hist.d_res_high': 'float',\n",
       " 'refine_hist.number_atoms_solvent': 'int',\n",
       " 'refine_hist.number_atoms_total': 'int',\n",
       " 'refine_hist.pdbx_number_atoms_ligand': 'int',\n",
       " 'refine_hist.pdbx_number_atoms_nucleic_acid': 'int',\n",
       " 'refine_hist.pdbx_number_atoms_protein': 'int',\n",
       " 'refine_hist.pdbx_refine_id': 'str',\n",
       " 'struct.title': 'dict(str)',\n",
       " 'struct': 'dict',\n",
       " 'struct_keywords.pdbx_keywords': 'dict(str)',\n",
       " 'struct_keywords.text': 'dict(str)',\n",
       " 'struct_keywords': 'dict',\n",
       " 'symmetry.int_tables_number': 'dict(int)',\n",
       " 'symmetry.space_group_name_hm': 'dict(str)',\n",
       " 'symmetry': 'dict',\n",
       " 'rcsb_id': 'str'}"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "instance.get_dummy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "98fdbeb2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Info: Structure for 4HHB already exists in pdb format.\n",
      "Cache_key: 4HHB_entry\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'rcsb_id': '4HHB',\n",
       " 'model_provenance': None,\n",
       " 'branched_molecular_weight_minimum': None,\n",
       " 'resolution_combined': 1.74,\n",
       " 'experimental_method': 'X-ray',\n",
       " 'diffrn_resolution_high': 1.74}"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "instance.fetch_single(\n",
    "    query=\"4HHB\",\n",
    "    method=\"entry\",\n",
    "    parse=True\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8904e17d",
   "metadata": {},
   "source": [
    "## Pride"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "2e0e8165",
   "metadata": {},
   "outputs": [],
   "source": [
    "from src.pride import PrideInterface"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "09b7829a",
   "metadata": {},
   "outputs": [],
   "source": [
    "instance = PrideInterface()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "09a79764",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cache_key: cancer_search_projects\n",
      "Prepared request: https://www.ebi.ac.uk/pride/ws/archive/v3/search/projects?keyword=cancer&page=0&sortDirection=DESC&sortFields=submissionDate\n",
      "Fetching data with parameters: {'keyword': 'cancer', 'page': 0, 'sortDirection': 'DESC', 'sortFields': 'submissionDate'}\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>title</th>\n",
       "      <th>additional_attributes</th>\n",
       "      <th>project_description</th>\n",
       "      <th>sample_processing_protocol</th>\n",
       "      <th>data_processing_protocol</th>\n",
       "      <th>project_tags</th>\n",
       "      <th>keywords</th>\n",
       "      <th>doi</th>\n",
       "      <th>submission_type</th>\n",
       "      <th>...</th>\n",
       "      <th>experiment_types</th>\n",
       "      <th>quantification_methods</th>\n",
       "      <th>countries</th>\n",
       "      <th>sample_attributes</th>\n",
       "      <th>organisms</th>\n",
       "      <th>organism_parts</th>\n",
       "      <th>diseases</th>\n",
       "      <th>references</th>\n",
       "      <th>identified_ptm_strings</th>\n",
       "      <th>total_file_downloads</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PXD066683</td>\n",
       "      <td>New proteomic signature in circulating extrace...</td>\n",
       "      <td>None</td>\n",
       "      <td>The identification of noninvasive prognostic b...</td>\n",
       "      <td>Sample preparation Total protein lysates were ...</td>\n",
       "      <td>MS data processing. Proteins were identified w...</td>\n",
       "      <td>Human proteome project</td>\n",
       "      <td>[Tumor-draining vein, Proteomic, Extracellular...</td>\n",
       "      <td></td>\n",
       "      <td>PARTIAL</td>\n",
       "      <td>...</td>\n",
       "      <td>Bottom-up proteomics</td>\n",
       "      <td>Relative quantification</td>\n",
       "      <td>None</td>\n",
       "      <td>[blood cell, lung cancer, blood plasma, Homo s...</td>\n",
       "      <td>Homo sapiens (human)</td>\n",
       "      <td>None</td>\n",
       "      <td>Lung cancer</td>\n",
       "      <td>[]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PXD066623</td>\n",
       "      <td>Exon inclusion signatures enable accurate esti...</td>\n",
       "      <td>None</td>\n",
       "      <td>Splicing factors control exon inclusion in mes...</td>\n",
       "      <td>Cell culture BJ fibroblasts from the Danielsso...</td>\n",
       "      <td>Analysis of mass spectra Acquired spectra were...</td>\n",
       "      <td>[]</td>\n",
       "      <td>Cancer; alternative splicing; splicing factor;...</td>\n",
       "      <td></td>\n",
       "      <td>PARTIAL</td>\n",
       "      <td>...</td>\n",
       "      <td>Data-independent acquisition</td>\n",
       "      <td>label-free quantification</td>\n",
       "      <td>None</td>\n",
       "      <td>[fibroblast, cell culture, Homo sapiens (Human...</td>\n",
       "      <td>[Homo sapiens (human), Characteristics[organism]]</td>\n",
       "      <td>None</td>\n",
       "      <td>[Carcinoma, Characteristics[disease]]</td>\n",
       "      <td>Anglada-Girotto M, Moakley DF, Zhang C, Mirave...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PXD066297</td>\n",
       "      <td>Genome-scale CRISPR screens identify PTGES3 as...</td>\n",
       "      <td>None</td>\n",
       "      <td>The androgen receptor (AR) is a critical drive...</td>\n",
       "      <td>Approximately 10 million cells were lyesed usi...</td>\n",
       "      <td>The MS/MS raw data (.raw files) were processed...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[Ar, Crispr screens, Ptges3, Prostate cancer]</td>\n",
       "      <td></td>\n",
       "      <td>PARTIAL</td>\n",
       "      <td>...</td>\n",
       "      <td>Bottom-up proteomics</td>\n",
       "      <td>[]</td>\n",
       "      <td>None</td>\n",
       "      <td>[cell culture, Homo sapiens (Human)]</td>\n",
       "      <td>Homo sapiens (human)</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PXD066316</td>\n",
       "      <td>Design of combination therapeutics from protei...</td>\n",
       "      <td>None</td>\n",
       "      <td>High-grade serous ovarian cancer (HGSOC) remai...</td>\n",
       "      <td>Sample processing To harvest cells for mass sp...</td>\n",
       "      <td>The protein expression matrix contained protei...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[Cell culture, Drug pertubation, Hgsoc, Ovaria...</td>\n",
       "      <td></td>\n",
       "      <td>PARTIAL</td>\n",
       "      <td>...</td>\n",
       "      <td>[diaPASEF, Bottom-up proteomics]</td>\n",
       "      <td>[label-free quantification, Relative quantific...</td>\n",
       "      <td>None</td>\n",
       "      <td>[malignant neoplasm of ovary, cell culture, Ho...</td>\n",
       "      <td>Homo sapiens (human)</td>\n",
       "      <td>None</td>\n",
       "      <td>Malignant neoplasm of ovary</td>\n",
       "      <td>[]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PXD066069</td>\n",
       "      <td>Radiation induced extracellular matrix changes...</td>\n",
       "      <td>None</td>\n",
       "      <td>Muscle-invasive bladder cancer (MIBC) is a pre...</td>\n",
       "      <td>cells were grown for 7 days in normal tissue c...</td>\n",
       "      <td>MS analysis was performed following previously...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[Radiation, Extracellular matrix, In vitro, Bl...</td>\n",
       "      <td></td>\n",
       "      <td>PARTIAL</td>\n",
       "      <td>...</td>\n",
       "      <td>Gel-based experiment</td>\n",
       "      <td>Normalized Spectral Abundance Factor - NSAF</td>\n",
       "      <td>None</td>\n",
       "      <td>[cell culture, Homo sapiens (Human), permanent...</td>\n",
       "      <td>Homo sapiens (human)</td>\n",
       "      <td>None</td>\n",
       "      <td>Urinary bladder cancer</td>\n",
       "      <td>null--pubMed:0--doi: 10.3389/FONC.2025.1616943</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>PXD061698</td>\n",
       "      <td>Astral-DIA Proteomics Deciphers the Molecular ...</td>\n",
       "      <td>None</td>\n",
       "      <td>Male infertility is fundamentally rooted in de...</td>\n",
       "      <td>Leveraging the principles of DIA technology, w...</td>\n",
       "      <td>This study leverages state-of-the-art proteomi...</td>\n",
       "      <td>[]</td>\n",
       "      <td>Astral-dia; proteomics deciphers; molecular la...</td>\n",
       "      <td></td>\n",
       "      <td>PARTIAL</td>\n",
       "      <td>...</td>\n",
       "      <td>Data-independent acquisition</td>\n",
       "      <td>[]</td>\n",
       "      <td>None</td>\n",
       "      <td>[Homo sapiens (Human), spermatozoon]</td>\n",
       "      <td>Homo sapiens (human)</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>Perez-Riverol Y, Bandla C, Kundu DJ, Kamatchin...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>PXD061725</td>\n",
       "      <td>Astral-DIA Proteomics Deciphers the Molecular ...</td>\n",
       "      <td>None</td>\n",
       "      <td>Male infertility is fundamentally rooted in de...</td>\n",
       "      <td>To analyze the aberrant regulatory events in h...</td>\n",
       "      <td>Leveraging the principles of DIA technology, w...</td>\n",
       "      <td>[]</td>\n",
       "      <td>Astral-dia; proteomics; human spermatozoa</td>\n",
       "      <td></td>\n",
       "      <td>PARTIAL</td>\n",
       "      <td>...</td>\n",
       "      <td>Bottom-up proteomics</td>\n",
       "      <td>[]</td>\n",
       "      <td>None</td>\n",
       "      <td>[Homo sapiens (Human), spermatozoon]</td>\n",
       "      <td>Homo sapiens (human)</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>Perez-Riverol Y, Bandla C, Kundu DJ, Kamatchin...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>PXD061710</td>\n",
       "      <td>Overcoming preservation challenges to enable s...</td>\n",
       "      <td>None</td>\n",
       "      <td>Recent advances in single-cell proteomics (SCP...</td>\n",
       "      <td>Single cell seeding and processing was perform...</td>\n",
       "      <td>Spectronaut 18 was used for data analysis sear...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[Human, Scp, Mouse, Lc-ms/ms, Pancreas]</td>\n",
       "      <td></td>\n",
       "      <td>PARTIAL</td>\n",
       "      <td>...</td>\n",
       "      <td>[Data-independent acquisition, Bottom-up prote...</td>\n",
       "      <td>label-free quantification</td>\n",
       "      <td>None</td>\n",
       "      <td>[cell culture, Homo sapiens (Human), pancreas,...</td>\n",
       "      <td>[Homo sapiens (human), Mus musculus (mouse)]</td>\n",
       "      <td>None</td>\n",
       "      <td>Cancer</td>\n",
       "      <td>[]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>PXD061696</td>\n",
       "      <td>Recombinant Protein Spectral Library (rPSL) DI...</td>\n",
       "      <td>None</td>\n",
       "      <td>Data-independent acquisition mass spectrometry...</td>\n",
       "      <td>Frozen tissues were cryogenically homogenized ...</td>\n",
       "      <td>DIA data processing using DIA-NN: Raw DIA-MS d...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[Snap frozen tissues, Rpsl, Cancer- associated...</td>\n",
       "      <td></td>\n",
       "      <td>PARTIAL</td>\n",
       "      <td>...</td>\n",
       "      <td>[Data-dependent acquisition, Data-independent ...</td>\n",
       "      <td>Relative quantification</td>\n",
       "      <td>None</td>\n",
       "      <td>[colorectal cancer cell, Homo sapiens (Human),...</td>\n",
       "      <td>Homo sapiens (human)</td>\n",
       "      <td>None</td>\n",
       "      <td>[Colon cancer, Breast cancer]</td>\n",
       "      <td>Krishnamurthy S, Gunasegaran B, Paul-Heng M, M...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>PXD061655</td>\n",
       "      <td>SFX-01 is therapeutic against myeloproliferati...</td>\n",
       "      <td>None</td>\n",
       "      <td>• Activating mutations of Shp2 cause condition...</td>\n",
       "      <td>Bone marrow mononuclear cells were isolated fr...</td>\n",
       "      <td>For data analysis for phospho and total proteo...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[Shp2, Myeloid cell proliferation, Phosphoprot...</td>\n",
       "      <td>10.6019/PXD061655</td>\n",
       "      <td>COMPLETE</td>\n",
       "      <td>...</td>\n",
       "      <td>Bottom-up proteomics</td>\n",
       "      <td>Label free</td>\n",
       "      <td>None</td>\n",
       "      <td>[myeloid cell, juvenile myelomonocytic leukemi...</td>\n",
       "      <td>Mus musculus (mouse)</td>\n",
       "      <td>None</td>\n",
       "      <td>Juvenile myelomonocytic leukemia</td>\n",
       "      <td>[]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows × 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    accession                                              title  \\\n",
       "0   PXD066683  New proteomic signature in circulating extrace...   \n",
       "1   PXD066623  Exon inclusion signatures enable accurate esti...   \n",
       "2   PXD066297  Genome-scale CRISPR screens identify PTGES3 as...   \n",
       "3   PXD066316  Design of combination therapeutics from protei...   \n",
       "4   PXD066069  Radiation induced extracellular matrix changes...   \n",
       "..        ...                                                ...   \n",
       "95  PXD061698  Astral-DIA Proteomics Deciphers the Molecular ...   \n",
       "96  PXD061725  Astral-DIA Proteomics Deciphers the Molecular ...   \n",
       "97  PXD061710  Overcoming preservation challenges to enable s...   \n",
       "98  PXD061696  Recombinant Protein Spectral Library (rPSL) DI...   \n",
       "99  PXD061655  SFX-01 is therapeutic against myeloproliferati...   \n",
       "\n",
       "   additional_attributes                                project_description  \\\n",
       "0                   None  The identification of noninvasive prognostic b...   \n",
       "1                   None  Splicing factors control exon inclusion in mes...   \n",
       "2                   None  The androgen receptor (AR) is a critical drive...   \n",
       "3                   None  High-grade serous ovarian cancer (HGSOC) remai...   \n",
       "4                   None  Muscle-invasive bladder cancer (MIBC) is a pre...   \n",
       "..                   ...                                                ...   \n",
       "95                  None  Male infertility is fundamentally rooted in de...   \n",
       "96                  None  Male infertility is fundamentally rooted in de...   \n",
       "97                  None  Recent advances in single-cell proteomics (SCP...   \n",
       "98                  None  Data-independent acquisition mass spectrometry...   \n",
       "99                  None  • Activating mutations of Shp2 cause condition...   \n",
       "\n",
       "                           sample_processing_protocol  \\\n",
       "0   Sample preparation Total protein lysates were ...   \n",
       "1   Cell culture BJ fibroblasts from the Danielsso...   \n",
       "2   Approximately 10 million cells were lyesed usi...   \n",
       "3   Sample processing To harvest cells for mass sp...   \n",
       "4   cells were grown for 7 days in normal tissue c...   \n",
       "..                                                ...   \n",
       "95  Leveraging the principles of DIA technology, w...   \n",
       "96  To analyze the aberrant regulatory events in h...   \n",
       "97  Single cell seeding and processing was perform...   \n",
       "98  Frozen tissues were cryogenically homogenized ...   \n",
       "99  Bone marrow mononuclear cells were isolated fr...   \n",
       "\n",
       "                             data_processing_protocol            project_tags  \\\n",
       "0   MS data processing. Proteins were identified w...  Human proteome project   \n",
       "1   Analysis of mass spectra Acquired spectra were...                      []   \n",
       "2   The MS/MS raw data (.raw files) were processed...                      []   \n",
       "3   The protein expression matrix contained protei...                      []   \n",
       "4   MS analysis was performed following previously...                      []   \n",
       "..                                                ...                     ...   \n",
       "95  This study leverages state-of-the-art proteomi...                      []   \n",
       "96  Leveraging the principles of DIA technology, w...                      []   \n",
       "97  Spectronaut 18 was used for data analysis sear...                      []   \n",
       "98  DIA data processing using DIA-NN: Raw DIA-MS d...                      []   \n",
       "99  For data analysis for phospho and total proteo...                      []   \n",
       "\n",
       "                                             keywords                doi  \\\n",
       "0   [Tumor-draining vein, Proteomic, Extracellular...                      \n",
       "1   Cancer; alternative splicing; splicing factor;...                      \n",
       "2       [Ar, Crispr screens, Ptges3, Prostate cancer]                      \n",
       "3   [Cell culture, Drug pertubation, Hgsoc, Ovaria...                      \n",
       "4   [Radiation, Extracellular matrix, In vitro, Bl...                      \n",
       "..                                                ...                ...   \n",
       "95  Astral-dia; proteomics deciphers; molecular la...                      \n",
       "96          Astral-dia; proteomics; human spermatozoa                      \n",
       "97            [Human, Scp, Mouse, Lc-ms/ms, Pancreas]                      \n",
       "98  [Snap frozen tissues, Rpsl, Cancer- associated...                      \n",
       "99  [Shp2, Myeloid cell proliferation, Phosphoprot...  10.6019/PXD061655   \n",
       "\n",
       "   submission_type  ...                                   experiment_types  \\\n",
       "0          PARTIAL  ...                               Bottom-up proteomics   \n",
       "1          PARTIAL  ...                       Data-independent acquisition   \n",
       "2          PARTIAL  ...                               Bottom-up proteomics   \n",
       "3          PARTIAL  ...                   [diaPASEF, Bottom-up proteomics]   \n",
       "4          PARTIAL  ...                               Gel-based experiment   \n",
       "..             ...  ...                                                ...   \n",
       "95         PARTIAL  ...                       Data-independent acquisition   \n",
       "96         PARTIAL  ...                               Bottom-up proteomics   \n",
       "97         PARTIAL  ...  [Data-independent acquisition, Bottom-up prote...   \n",
       "98         PARTIAL  ...  [Data-dependent acquisition, Data-independent ...   \n",
       "99        COMPLETE  ...                               Bottom-up proteomics   \n",
       "\n",
       "                               quantification_methods countries  \\\n",
       "0                             Relative quantification      None   \n",
       "1                           label-free quantification      None   \n",
       "2                                                  []      None   \n",
       "3   [label-free quantification, Relative quantific...      None   \n",
       "4         Normalized Spectral Abundance Factor - NSAF      None   \n",
       "..                                                ...       ...   \n",
       "95                                                 []      None   \n",
       "96                                                 []      None   \n",
       "97                          label-free quantification      None   \n",
       "98                            Relative quantification      None   \n",
       "99                                         Label free      None   \n",
       "\n",
       "                                    sample_attributes  \\\n",
       "0   [blood cell, lung cancer, blood plasma, Homo s...   \n",
       "1   [fibroblast, cell culture, Homo sapiens (Human...   \n",
       "2                [cell culture, Homo sapiens (Human)]   \n",
       "3   [malignant neoplasm of ovary, cell culture, Ho...   \n",
       "4   [cell culture, Homo sapiens (Human), permanent...   \n",
       "..                                                ...   \n",
       "95               [Homo sapiens (Human), spermatozoon]   \n",
       "96               [Homo sapiens (Human), spermatozoon]   \n",
       "97  [cell culture, Homo sapiens (Human), pancreas,...   \n",
       "98  [colorectal cancer cell, Homo sapiens (Human),...   \n",
       "99  [myeloid cell, juvenile myelomonocytic leukemi...   \n",
       "\n",
       "                                            organisms organism_parts  \\\n",
       "0                                Homo sapiens (human)           None   \n",
       "1   [Homo sapiens (human), Characteristics[organism]]           None   \n",
       "2                                Homo sapiens (human)           None   \n",
       "3                                Homo sapiens (human)           None   \n",
       "4                                Homo sapiens (human)           None   \n",
       "..                                                ...            ...   \n",
       "95                               Homo sapiens (human)           None   \n",
       "96                               Homo sapiens (human)           None   \n",
       "97       [Homo sapiens (human), Mus musculus (mouse)]           None   \n",
       "98                               Homo sapiens (human)           None   \n",
       "99                               Mus musculus (mouse)           None   \n",
       "\n",
       "                                 diseases  \\\n",
       "0                             Lung cancer   \n",
       "1   [Carcinoma, Characteristics[disease]]   \n",
       "2                                      []   \n",
       "3             Malignant neoplasm of ovary   \n",
       "4                  Urinary bladder cancer   \n",
       "..                                    ...   \n",
       "95                                     []   \n",
       "96                                     []   \n",
       "97                                 Cancer   \n",
       "98          [Colon cancer, Breast cancer]   \n",
       "99       Juvenile myelomonocytic leukemia   \n",
       "\n",
       "                                           references identified_ptm_strings  \\\n",
       "0                                                  []                   None   \n",
       "1   Anglada-Girotto M, Moakley DF, Zhang C, Mirave...                   None   \n",
       "2                                                  []                   None   \n",
       "3                                                  []                   None   \n",
       "4      null--pubMed:0--doi: 10.3389/FONC.2025.1616943                   None   \n",
       "..                                                ...                    ...   \n",
       "95  Perez-Riverol Y, Bandla C, Kundu DJ, Kamatchin...                   None   \n",
       "96  Perez-Riverol Y, Bandla C, Kundu DJ, Kamatchin...                   None   \n",
       "97                                                 []                   None   \n",
       "98  Krishnamurthy S, Gunasegaran B, Paul-Heng M, M...                   None   \n",
       "99                                                 []                   None   \n",
       "\n",
       "   total_file_downloads  \n",
       "0                  None  \n",
       "1                  None  \n",
       "2                  None  \n",
       "3                  None  \n",
       "4                  None  \n",
       "..                  ...  \n",
       "95                 None  \n",
       "96                 None  \n",
       "97                 None  \n",
       "98                 None  \n",
       "99                 None  \n",
       "\n",
       "[100 rows x 27 columns]"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "instance.fetch_single(\n",
    "    query={\n",
    "        \"keyword\": \"cancer\",\n",
    "    },\n",
    "    method=\"search\",\n",
    "    option=\"projects\",\n",
    "    parse=True,\n",
    "    to_dataframe=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "fc0718a5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cache_key: PXD066763_projects_default\n",
      "Prepared request: https://www.ebi.ac.uk/pride/ws/archive/v3/projects/PXD066763\n",
      "Fetching data with parameters: {}\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>title</th>\n",
       "      <th>additional_attributes</th>\n",
       "      <th>project_description</th>\n",
       "      <th>sample_processing_protocol</th>\n",
       "      <th>data_processing_protocol</th>\n",
       "      <th>project_tags</th>\n",
       "      <th>keywords</th>\n",
       "      <th>doi</th>\n",
       "      <th>submission_type</th>\n",
       "      <th>...</th>\n",
       "      <th>experiment_types</th>\n",
       "      <th>quantification_methods</th>\n",
       "      <th>countries</th>\n",
       "      <th>sample_attributes</th>\n",
       "      <th>organisms</th>\n",
       "      <th>organism_parts</th>\n",
       "      <th>diseases</th>\n",
       "      <th>references</th>\n",
       "      <th>identified_ptm_strings</th>\n",
       "      <th>total_file_downloads</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PXD066763</td>\n",
       "      <td>Stereo-random Oligonucleotides Enable Efficien...</td>\n",
       "      <td>[]</td>\n",
       "      <td>We used our previously published isASO-ID prot...</td>\n",
       "      <td>Proximity biotinylation experiment. Cells were...</td>\n",
       "      <td>MS spectra were processed with MaxQuant softwa...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[Antisense oligonucleotide, Basu, Drug protein...</td>\n",
       "      <td></td>\n",
       "      <td>PARTIAL</td>\n",
       "      <td>...</td>\n",
       "      <td>{'@type': 'CvParam', 'cvLabel': 'PRIDE', 'acce...</td>\n",
       "      <td>{'@type': 'CvParam', 'cvLabel': 'PRIDE', 'acce...</td>\n",
       "      <td>Germany</td>\n",
       "      <td>[{'@type': 'Tuple', 'key': {'cvLabel': 'EFO', ...</td>\n",
       "      <td>{'@type': 'CvParam', 'cvLabel': 'NEWT', 'acces...</td>\n",
       "      <td>{'@type': 'CvParam', 'cvLabel': 'CL', 'accessi...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[{'@type': 'CvParam', 'cvLabel': 'MOD', 'acces...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   accession                                              title  \\\n",
       "0  PXD066763  Stereo-random Oligonucleotides Enable Efficien...   \n",
       "\n",
       "  additional_attributes                                project_description  \\\n",
       "0                    []  We used our previously published isASO-ID prot...   \n",
       "\n",
       "                          sample_processing_protocol  \\\n",
       "0  Proximity biotinylation experiment. Cells were...   \n",
       "\n",
       "                            data_processing_protocol project_tags  \\\n",
       "0  MS spectra were processed with MaxQuant softwa...           []   \n",
       "\n",
       "                                            keywords doi submission_type  ...  \\\n",
       "0  [Antisense oligonucleotide, Basu, Drug protein...             PARTIAL  ...   \n",
       "\n",
       "                                    experiment_types  \\\n",
       "0  {'@type': 'CvParam', 'cvLabel': 'PRIDE', 'acce...   \n",
       "\n",
       "                              quantification_methods countries  \\\n",
       "0  {'@type': 'CvParam', 'cvLabel': 'PRIDE', 'acce...   Germany   \n",
       "\n",
       "                                   sample_attributes  \\\n",
       "0  [{'@type': 'Tuple', 'key': {'cvLabel': 'EFO', ...   \n",
       "\n",
       "                                           organisms  \\\n",
       "0  {'@type': 'CvParam', 'cvLabel': 'NEWT', 'acces...   \n",
       "\n",
       "                                      organism_parts diseases references  \\\n",
       "0  {'@type': 'CvParam', 'cvLabel': 'CL', 'accessi...       []         []   \n",
       "\n",
       "                              identified_ptm_strings total_file_downloads  \n",
       "0  [{'@type': 'CvParam', 'cvLabel': 'MOD', 'acces...                    0  \n",
       "\n",
       "[1 rows x 27 columns]"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "instance.fetch_single(\n",
    "    query={\n",
    "        \"projectAccession\": \"PXD066763\",\n",
    "    },\n",
    "    method=\"projects\",\n",
    "    parse=True,\n",
    "    to_dataframe=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "3d96cce0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cache_key: PXD066763_projects_similarProjects\n",
      "Prepared request: https://www.ebi.ac.uk/pride/ws/archive/v3/projects/PXD066763/similarProjects?page=0&pageSize=10\n",
      "Fetching data with parameters: {'page': 0, 'pageSize': 10}\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>affiliations</th>\n",
       "      <th>avg_wonloads_per_file</th>\n",
       "      <th>data_processing_protocol</th>\n",
       "      <th>diseases</th>\n",
       "      <th>doi</th>\n",
       "      <th>download_count</th>\n",
       "      <th>experiment_types</th>\n",
       "      <th>instruments</th>\n",
       "      <th>keywords</th>\n",
       "      <th>...</th>\n",
       "      <th>sample_attributes</th>\n",
       "      <th>sample_processing_protocol</th>\n",
       "      <th>sdrf</th>\n",
       "      <th>softwares</th>\n",
       "      <th>submission_date</th>\n",
       "      <th>submission_type</th>\n",
       "      <th>submitters</th>\n",
       "      <th>title</th>\n",
       "      <th>updated_date</th>\n",
       "      <th>yearly_downloads</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PXD045992</td>\n",
       "      <td>Interfaculty Institute of Biochemistry, Univer...</td>\n",
       "      <td>3.333333</td>\n",
       "      <td>MS spectra were processed with MaxQuant softwa...</td>\n",
       "      <td>[]</td>\n",
       "      <td></td>\n",
       "      <td>300</td>\n",
       "      <td>Affinity purification coupled with mass spectr...</td>\n",
       "      <td>Q Exactive HF</td>\n",
       "      <td>[Live cell, Bioid, Proximity biotinylation, Be...</td>\n",
       "      <td>...</td>\n",
       "      <td>[Homo sapiens (Human), cervical cancer cell li...</td>\n",
       "      <td>Proximity biotinylation experiment. Hek293T Fl...</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>2023-10-09</td>\n",
       "      <td>PARTIAL</td>\n",
       "      <td>Mirita Franz-Wachtel</td>\n",
       "      <td>Profiling the interactome of oligonucleotide d...</td>\n",
       "      <td>2023-10-09</td>\n",
       "      <td>[{'year': '2024', 'count': 79}, {'year': '2025...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PXD013215</td>\n",
       "      <td>Interfaculty Institute of Biochemistry Univers...</td>\n",
       "      <td>37.461540</td>\n",
       "      <td>Spectra were processed with MaxQuant software ...</td>\n",
       "      <td>[]</td>\n",
       "      <td></td>\n",
       "      <td>487</td>\n",
       "      <td>[Affinity purification coupled with mass spect...</td>\n",
       "      <td>LTQ Orbitrap Elite</td>\n",
       "      <td>[Budding yeast yer156c interactome, Chromatin,...</td>\n",
       "      <td>...</td>\n",
       "      <td>Saccharomyces cerevisiae (Baker's yeast)</td>\n",
       "      <td>For SILAC APEX2 experiments the mitochondrial ...</td>\n",
       "      <td>None</td>\n",
       "      <td>MaxQuant</td>\n",
       "      <td>2019-03-25</td>\n",
       "      <td>PARTIAL</td>\n",
       "      <td>Nicolas Nalpas</td>\n",
       "      <td>APEX2-mediated proximity labeling of mitochond...</td>\n",
       "      <td>2019-03-25</td>\n",
       "      <td>[{'year': '2021', 'count': 84}, {'year': '2022...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PXD011594</td>\n",
       "      <td>Center for Plant Molecular Biology (ZMBP) Univ...</td>\n",
       "      <td>22.200000</td>\n",
       "      <td>Spectra were processed with MaxQuant software ...</td>\n",
       "      <td>[]</td>\n",
       "      <td></td>\n",
       "      <td>111</td>\n",
       "      <td>Affinity purification coupled with mass spectr...</td>\n",
       "      <td>LTQ Orbitrap XL</td>\n",
       "      <td>[Membrane traffic, Arf gtpases, Arabidopsis, A...</td>\n",
       "      <td>...</td>\n",
       "      <td>[Arabidopsis thaliana (Mouse-ear cress), seedl...</td>\n",
       "      <td>Protein extracts from transgenic seedlings exp...</td>\n",
       "      <td>None</td>\n",
       "      <td>MaxQuant</td>\n",
       "      <td>2018-11-07</td>\n",
       "      <td>PARTIAL</td>\n",
       "      <td>Nicolas Nalpas</td>\n",
       "      <td>Arabidopsis ARF-ARFGEF interaction analyzed by...</td>\n",
       "      <td>2018-11-07</td>\n",
       "      <td>[{'year': '2021', 'count': 48}, {'year': '2022...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PXD032366</td>\n",
       "      <td>Faculty of Biology &amp; Biotechnology Ruhr-Univer...</td>\n",
       "      <td>1.473684</td>\n",
       "      <td>MS spectra were processed with MaxQuant softwa...</td>\n",
       "      <td>[]</td>\n",
       "      <td></td>\n",
       "      <td>28</td>\n",
       "      <td>Affinity purification coupled with mass spectr...</td>\n",
       "      <td>Q Exactive</td>\n",
       "      <td>[Lc-msms, Bacterial effectors, Plants]</td>\n",
       "      <td>...</td>\n",
       "      <td>[Nicotiana benthamiana, leaf]</td>\n",
       "      <td>Proteins were purified by SDS-PAGE (short run)...</td>\n",
       "      <td></td>\n",
       "      <td>[Andromeda, MaxQuant]</td>\n",
       "      <td>2022-03-17</td>\n",
       "      <td>PARTIAL</td>\n",
       "      <td>Mirita Franz-Wachtel</td>\n",
       "      <td>Characterization of Xanthomonas effector prote...</td>\n",
       "      <td>2022-03-17</td>\n",
       "      <td>[{'year': '2023', 'count': 17}, {'year': '2024...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PXD010694</td>\n",
       "      <td>University of Tuebingen - Interfaculty Institu...</td>\n",
       "      <td>24.100000</td>\n",
       "      <td>Acquired MS spectra were processed with MaxQua...</td>\n",
       "      <td>[]</td>\n",
       "      <td></td>\n",
       "      <td>241</td>\n",
       "      <td>Affinity purification coupled with mass spectr...</td>\n",
       "      <td>LTQ Orbitrap Elite</td>\n",
       "      <td>[Rna localization, Bioid, Mouse, Rbps]</td>\n",
       "      <td>...</td>\n",
       "      <td>[embryonic fibroblast, Mus musculus (Mouse)]</td>\n",
       "      <td>For RNA-BioID, cells were incubated with 50 µM...</td>\n",
       "      <td>None</td>\n",
       "      <td>[Andromeda, MaxQuant]</td>\n",
       "      <td>2018-08-06</td>\n",
       "      <td>PARTIAL</td>\n",
       "      <td>Mirita Franz-Wachtel</td>\n",
       "      <td>β-actin RNA-BioID</td>\n",
       "      <td>2018-08-06</td>\n",
       "      <td>[{'year': '2021', 'count': 40}, {'year': '2022...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>PXD031610</td>\n",
       "      <td>University of Tuebingen, Interfaculty Institut...</td>\n",
       "      <td>14.250000</td>\n",
       "      <td>MS spectra were processed with MaxQuant softwa...</td>\n",
       "      <td>[]</td>\n",
       "      <td></td>\n",
       "      <td>57</td>\n",
       "      <td>Affinity purification coupled with mass spectr...</td>\n",
       "      <td>Q Exactive</td>\n",
       "      <td>[Mitochondria, Yeast, Signal-anchored proteins...</td>\n",
       "      <td>...</td>\n",
       "      <td>[cell culture, Saccharomyces cerevisiae (Baker...</td>\n",
       "      <td>Pull-down assays were performed with extract o...</td>\n",
       "      <td>None</td>\n",
       "      <td>[Andromeda, MaxQuant]</td>\n",
       "      <td>2022-02-11</td>\n",
       "      <td>PARTIAL</td>\n",
       "      <td>Mirita Franz-Wachtel</td>\n",
       "      <td>Searching for the interaction partners in yeas...</td>\n",
       "      <td>2022-02-11</td>\n",
       "      <td>[{'year': '2022', 'count': 12}, {'year': '2023...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>PXD045484</td>\n",
       "      <td>University of Halle-Wittenberg Institute for B...</td>\n",
       "      <td>10.615385</td>\n",
       "      <td>The MS data were processed with MaxQuant softw...</td>\n",
       "      <td>[]</td>\n",
       "      <td></td>\n",
       "      <td>138</td>\n",
       "      <td>Affinity purification coupled with mass spectr...</td>\n",
       "      <td>Q Exactive HF</td>\n",
       "      <td>[Comprehensive isolation by rna binding protei...</td>\n",
       "      <td>...</td>\n",
       "      <td>[plant cell, Arabidopsis thaliana (Mouse-ear c...</td>\n",
       "      <td>LC-MS/MS analyses of eluted samples were perfo...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>2023-09-19</td>\n",
       "      <td>PARTIAL</td>\n",
       "      <td>Irina Droste-Borel</td>\n",
       "      <td>Identification of U1 snRNA associated proteins</td>\n",
       "      <td>2023-09-19</td>\n",
       "      <td>[{'year': '2024', 'count': 116}, {'year': '202...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>PXD033585</td>\n",
       "      <td>Faculty of Biology &amp; Biotechnology Ruhr-Univer...</td>\n",
       "      <td>1.200000</td>\n",
       "      <td>MS spectra were processed with MaxQuant softwa...</td>\n",
       "      <td>[]</td>\n",
       "      <td></td>\n",
       "      <td>24</td>\n",
       "      <td>Affinity purification coupled with mass spectr...</td>\n",
       "      <td>Q Exactive</td>\n",
       "      <td>[Lc-msms, Plants]</td>\n",
       "      <td>...</td>\n",
       "      <td>[Nicotiana benthamiana, leaf]</td>\n",
       "      <td>: Proteins were purified by SDS-PAGE (short ru...</td>\n",
       "      <td>None</td>\n",
       "      <td>[Andromeda, MaxQuant]</td>\n",
       "      <td>2022-05-04</td>\n",
       "      <td>PARTIAL</td>\n",
       "      <td>Mirita Franz-Wachtel</td>\n",
       "      <td>Functional characterization of SH3P2 from Arab...</td>\n",
       "      <td>2022-05-04</td>\n",
       "      <td>[{'year': '2022', 'count': 5}, {'year': '2023'...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>PXD005597</td>\n",
       "      <td>Junior Research Group Infection Biology of Sal...</td>\n",
       "      <td>3.714286</td>\n",
       "      <td>The MS data were processed with MaxQuant softw...</td>\n",
       "      <td>[]</td>\n",
       "      <td></td>\n",
       "      <td>104</td>\n",
       "      <td>Affinity purification coupled with mass spectr...</td>\n",
       "      <td>LTQ Orbitrap</td>\n",
       "      <td>[T3ss, Flip, Flio, Export gate complex, Assembly]</td>\n",
       "      <td>...</td>\n",
       "      <td>Salmonella enterica subsp. enterica serovar Ty...</td>\n",
       "      <td>LC-MS/MS analyses were performed on an Easy na...</td>\n",
       "      <td>None</td>\n",
       "      <td>MaxQuant</td>\n",
       "      <td>2016-12-20</td>\n",
       "      <td>PARTIAL</td>\n",
       "      <td>Nicolas Nalpas</td>\n",
       "      <td>FliP subcomplex analysis -  A flagellum-specif...</td>\n",
       "      <td>2016-12-20</td>\n",
       "      <td>[{'year': '2021', 'count': 52}, {'year': '2022...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>PXD036942</td>\n",
       "      <td>University Tübingen, Center for Plant Molecula...</td>\n",
       "      <td>14.523809</td>\n",
       "      <td>Acquired MS spectra were processed with MaxQua...</td>\n",
       "      <td>[]</td>\n",
       "      <td></td>\n",
       "      <td>305</td>\n",
       "      <td>Affinity purification coupled with mass spectr...</td>\n",
       "      <td>Orbitrap Exploris 480</td>\n",
       "      <td>[Pti, Arabidopsis, Cell death, Bir3, Plant imm...</td>\n",
       "      <td>...</td>\n",
       "      <td>[Arabidopsis thaliana (Mouse-ear cress), leaf]</td>\n",
       "      <td>35S-BIR3-YFP expressing plants were used to im...</td>\n",
       "      <td>None</td>\n",
       "      <td>[Andromeda, MaxQuant]</td>\n",
       "      <td>2022-09-22</td>\n",
       "      <td>PARTIAL</td>\n",
       "      <td>Mirita Franz-Wachtel</td>\n",
       "      <td>Interactome of Arabidopsis BIR3 protein</td>\n",
       "      <td>2022-09-22</td>\n",
       "      <td>[{'year': '2022', 'count': 5}, {'year': '2023'...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 31 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   accession                                       affiliations  \\\n",
       "0  PXD045992  Interfaculty Institute of Biochemistry, Univer...   \n",
       "1  PXD013215  Interfaculty Institute of Biochemistry Univers...   \n",
       "2  PXD011594  Center for Plant Molecular Biology (ZMBP) Univ...   \n",
       "3  PXD032366  Faculty of Biology & Biotechnology Ruhr-Univer...   \n",
       "4  PXD010694  University of Tuebingen - Interfaculty Institu...   \n",
       "5  PXD031610  University of Tuebingen, Interfaculty Institut...   \n",
       "6  PXD045484  University of Halle-Wittenberg Institute for B...   \n",
       "7  PXD033585  Faculty of Biology & Biotechnology Ruhr-Univer...   \n",
       "8  PXD005597  Junior Research Group Infection Biology of Sal...   \n",
       "9  PXD036942  University Tübingen, Center for Plant Molecula...   \n",
       "\n",
       "   avg_wonloads_per_file                           data_processing_protocol  \\\n",
       "0               3.333333  MS spectra were processed with MaxQuant softwa...   \n",
       "1              37.461540  Spectra were processed with MaxQuant software ...   \n",
       "2              22.200000  Spectra were processed with MaxQuant software ...   \n",
       "3               1.473684  MS spectra were processed with MaxQuant softwa...   \n",
       "4              24.100000  Acquired MS spectra were processed with MaxQua...   \n",
       "5              14.250000  MS spectra were processed with MaxQuant softwa...   \n",
       "6              10.615385  The MS data were processed with MaxQuant softw...   \n",
       "7               1.200000  MS spectra were processed with MaxQuant softwa...   \n",
       "8               3.714286  The MS data were processed with MaxQuant softw...   \n",
       "9              14.523809  Acquired MS spectra were processed with MaxQua...   \n",
       "\n",
       "  diseases doi  download_count  \\\n",
       "0       []                 300   \n",
       "1       []                 487   \n",
       "2       []                 111   \n",
       "3       []                  28   \n",
       "4       []                 241   \n",
       "5       []                  57   \n",
       "6       []                 138   \n",
       "7       []                  24   \n",
       "8       []                 104   \n",
       "9       []                 305   \n",
       "\n",
       "                                    experiment_types            instruments  \\\n",
       "0  Affinity purification coupled with mass spectr...          Q Exactive HF   \n",
       "1  [Affinity purification coupled with mass spect...     LTQ Orbitrap Elite   \n",
       "2  Affinity purification coupled with mass spectr...        LTQ Orbitrap XL   \n",
       "3  Affinity purification coupled with mass spectr...             Q Exactive   \n",
       "4  Affinity purification coupled with mass spectr...     LTQ Orbitrap Elite   \n",
       "5  Affinity purification coupled with mass spectr...             Q Exactive   \n",
       "6  Affinity purification coupled with mass spectr...          Q Exactive HF   \n",
       "7  Affinity purification coupled with mass spectr...             Q Exactive   \n",
       "8  Affinity purification coupled with mass spectr...           LTQ Orbitrap   \n",
       "9  Affinity purification coupled with mass spectr...  Orbitrap Exploris 480   \n",
       "\n",
       "                                            keywords  ...  \\\n",
       "0  [Live cell, Bioid, Proximity biotinylation, Be...  ...   \n",
       "1  [Budding yeast yer156c interactome, Chromatin,...  ...   \n",
       "2  [Membrane traffic, Arf gtpases, Arabidopsis, A...  ...   \n",
       "3             [Lc-msms, Bacterial effectors, Plants]  ...   \n",
       "4             [Rna localization, Bioid, Mouse, Rbps]  ...   \n",
       "5  [Mitochondria, Yeast, Signal-anchored proteins...  ...   \n",
       "6  [Comprehensive isolation by rna binding protei...  ...   \n",
       "7                                  [Lc-msms, Plants]  ...   \n",
       "8  [T3ss, Flip, Flio, Export gate complex, Assembly]  ...   \n",
       "9  [Pti, Arabidopsis, Cell death, Bir3, Plant imm...  ...   \n",
       "\n",
       "                                   sample_attributes  \\\n",
       "0  [Homo sapiens (Human), cervical cancer cell li...   \n",
       "1           Saccharomyces cerevisiae (Baker's yeast)   \n",
       "2  [Arabidopsis thaliana (Mouse-ear cress), seedl...   \n",
       "3                      [Nicotiana benthamiana, leaf]   \n",
       "4       [embryonic fibroblast, Mus musculus (Mouse)]   \n",
       "5  [cell culture, Saccharomyces cerevisiae (Baker...   \n",
       "6  [plant cell, Arabidopsis thaliana (Mouse-ear c...   \n",
       "7                      [Nicotiana benthamiana, leaf]   \n",
       "8  Salmonella enterica subsp. enterica serovar Ty...   \n",
       "9     [Arabidopsis thaliana (Mouse-ear cress), leaf]   \n",
       "\n",
       "                          sample_processing_protocol  sdrf  \\\n",
       "0  Proximity biotinylation experiment. Hek293T Fl...         \n",
       "1  For SILAC APEX2 experiments the mitochondrial ...  None   \n",
       "2  Protein extracts from transgenic seedlings exp...  None   \n",
       "3  Proteins were purified by SDS-PAGE (short run)...         \n",
       "4  For RNA-BioID, cells were incubated with 50 µM...  None   \n",
       "5  Pull-down assays were performed with extract o...  None   \n",
       "6  LC-MS/MS analyses of eluted samples were perfo...  None   \n",
       "7  : Proteins were purified by SDS-PAGE (short ru...  None   \n",
       "8  LC-MS/MS analyses were performed on an Easy na...  None   \n",
       "9  35S-BIR3-YFP expressing plants were used to im...  None   \n",
       "\n",
       "               softwares  submission_date submission_type  \\\n",
       "0                   None       2023-10-09         PARTIAL   \n",
       "1               MaxQuant       2019-03-25         PARTIAL   \n",
       "2               MaxQuant       2018-11-07         PARTIAL   \n",
       "3  [Andromeda, MaxQuant]       2022-03-17         PARTIAL   \n",
       "4  [Andromeda, MaxQuant]       2018-08-06         PARTIAL   \n",
       "5  [Andromeda, MaxQuant]       2022-02-11         PARTIAL   \n",
       "6                   None       2023-09-19         PARTIAL   \n",
       "7  [Andromeda, MaxQuant]       2022-05-04         PARTIAL   \n",
       "8               MaxQuant       2016-12-20         PARTIAL   \n",
       "9  [Andromeda, MaxQuant]       2022-09-22         PARTIAL   \n",
       "\n",
       "             submitters                                              title  \\\n",
       "0  Mirita Franz-Wachtel  Profiling the interactome of oligonucleotide d...   \n",
       "1        Nicolas Nalpas  APEX2-mediated proximity labeling of mitochond...   \n",
       "2        Nicolas Nalpas  Arabidopsis ARF-ARFGEF interaction analyzed by...   \n",
       "3  Mirita Franz-Wachtel  Characterization of Xanthomonas effector prote...   \n",
       "4  Mirita Franz-Wachtel                                  β-actin RNA-BioID   \n",
       "5  Mirita Franz-Wachtel  Searching for the interaction partners in yeas...   \n",
       "6    Irina Droste-Borel     Identification of U1 snRNA associated proteins   \n",
       "7  Mirita Franz-Wachtel  Functional characterization of SH3P2 from Arab...   \n",
       "8        Nicolas Nalpas  FliP subcomplex analysis -  A flagellum-specif...   \n",
       "9  Mirita Franz-Wachtel            Interactome of Arabidopsis BIR3 protein   \n",
       "\n",
       "  updated_date                                   yearly_downloads  \n",
       "0   2023-10-09  [{'year': '2024', 'count': 79}, {'year': '2025...  \n",
       "1   2019-03-25  [{'year': '2021', 'count': 84}, {'year': '2022...  \n",
       "2   2018-11-07  [{'year': '2021', 'count': 48}, {'year': '2022...  \n",
       "3   2022-03-17  [{'year': '2023', 'count': 17}, {'year': '2024...  \n",
       "4   2018-08-06  [{'year': '2021', 'count': 40}, {'year': '2022...  \n",
       "5   2022-02-11  [{'year': '2022', 'count': 12}, {'year': '2023...  \n",
       "6   2023-09-19  [{'year': '2024', 'count': 116}, {'year': '202...  \n",
       "7   2022-05-04  [{'year': '2022', 'count': 5}, {'year': '2023'...  \n",
       "8   2016-12-20  [{'year': '2021', 'count': 52}, {'year': '2022...  \n",
       "9   2022-09-22  [{'year': '2022', 'count': 5}, {'year': '2023'...  \n",
       "\n",
       "[10 rows x 31 columns]"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "instance.fetch_single(\n",
    "    query={\n",
    "        \"accession\": \"PXD066763\",\n",
    "    },\n",
    "    method=\"projects\",\n",
    "    option=\"similarProjects\",\n",
    "    parse=True,\n",
    "    to_dataframe=True\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "25fa663c",
   "metadata": {},
   "source": [
    "## Pubchem"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "68906f76",
   "metadata": {},
   "outputs": [],
   "source": [
    "from src.pubchem import PubChemInterface"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "26b77076",
   "metadata": {},
   "outputs": [],
   "source": [
    "instance = PubChemInterface()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "f3bd8b0b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cache_key: APP_gene_summary\n",
      "Cache_key: ECD_gene_summary\n",
      "Prepared request: https://pubchem.ncbi.nlm.nih.gov/rest/pug/gene/genesymbol/APP,ECD/summary/json\n",
      "Cache_key: APP_gene_summary\n",
      "Cache_key: ECD_gene_summary\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>symbol</th>\n",
       "      <th>name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>description</th>\n",
       "      <th>synonym</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>351</td>\n",
       "      <td>APP</td>\n",
       "      <td>amyloid beta precursor protein</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens (human)</td>\n",
       "      <td>This gene encodes a cell surface receptor and ...</td>\n",
       "      <td>[AAA, ABETA, ABPP, AD1, APPI, CTFgamma, CVAP, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>11319</td>\n",
       "      <td>ECD</td>\n",
       "      <td>ecdysoneless cell cycle regulator</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens (human)</td>\n",
       "      <td>Enables histone acetyltransferase binding acti...</td>\n",
       "      <td>[GCR2, HSGT1, SGT1, protein ecdysoneless homol...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene_id symbol                               name  taxonomy_id  \\\n",
       "0      351    APP     amyloid beta precursor protein         9606   \n",
       "1    11319    ECD  ecdysoneless cell cycle regulator         9606   \n",
       "\n",
       "               taxonomy                                        description  \\\n",
       "0  Homo sapiens (human)  This gene encodes a cell surface receptor and ...   \n",
       "1  Homo sapiens (human)  Enables histone acetyltransferase binding acti...   \n",
       "\n",
       "                                             synonym  \n",
       "0  [AAA, ABETA, ABPP, AD1, APPI, CTFgamma, CVAP, ...  \n",
       "1  [GCR2, HSGT1, SGT1, protein ecdysoneless homol...  "
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "instance.fetch_single(\n",
    "    query={\n",
    "        \"genesymbol\": [\"APP\", \"ECD\"],\n",
    "    },\n",
    "    method=\"gene\",\n",
    "    option=\"summary\",\n",
    "    parse=True,\n",
    "    to_dataframe=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "8acc9ffd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cache_key: CC(=O)OC1=CC=CC=C1C(=O)O_['molecularformula', 'smiles', 'hbonddonorcount', 'hbondacceptorcount']_compound_default\n",
      "Prepared request: https://pubchem.ncbi.nlm.nih.gov/rest/pug/compound/smiles/CC(=O)OC1=CC=CC=C1C(=O)O/property/molecularformula,smiles,hbonddonorcount,hbondacceptorcount/json\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cid</th>\n",
       "      <th>molecular_formula</th>\n",
       "      <th>molecular_weight</th>\n",
       "      <th>smiles</th>\n",
       "      <th>connectivity_smiles</th>\n",
       "      <th>inchi</th>\n",
       "      <th>inchikey</th>\n",
       "      <th>iupac_name</th>\n",
       "      <th>title</th>\n",
       "      <th>xlogp</th>\n",
       "      <th>...</th>\n",
       "      <th>feature_acceptor_count_3d</th>\n",
       "      <th>feature_donor_count_3d</th>\n",
       "      <th>feature_anion_count_3d</th>\n",
       "      <th>feature_cation_count_3d</th>\n",
       "      <th>feature_ring_count_3d</th>\n",
       "      <th>feature_hydrophobe_count_3d</th>\n",
       "      <th>conformer_model_rmsd_3d</th>\n",
       "      <th>effective_rotor_count_3d</th>\n",
       "      <th>conformer_count_3d</th>\n",
       "      <th>fingerprint_2d</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2244</td>\n",
       "      <td>C9H8O4</td>\n",
       "      <td>None</td>\n",
       "      <td>CC(=O)OC1=CC=CC=C1C(=O)O</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 48 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    cid molecular_formula molecular_weight                    smiles  \\\n",
       "0  2244            C9H8O4             None  CC(=O)OC1=CC=CC=C1C(=O)O   \n",
       "\n",
       "  connectivity_smiles inchi inchikey iupac_name title xlogp  ...  \\\n",
       "0                None  None     None       None  None  None  ...   \n",
       "\n",
       "  feature_acceptor_count_3d feature_donor_count_3d feature_anion_count_3d  \\\n",
       "0                      None                   None                   None   \n",
       "\n",
       "  feature_cation_count_3d feature_ring_count_3d  feature_hydrophobe_count_3d  \\\n",
       "0                    None                  None                         None   \n",
       "\n",
       "   conformer_model_rmsd_3d effective_rotor_count_3d conformer_count_3d  \\\n",
       "0                     None                     None               None   \n",
       "\n",
       "  fingerprint_2d  \n",
       "0           None  \n",
       "\n",
       "[1 rows x 48 columns]"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# For more properties check the documentation: https://pubchem.ncbi.nlm.nih.gov/docs/pug-rest#section=Compound-Property-Tables\n",
    "instance.fetch_single(\n",
    "    query={\n",
    "        #\"cid\": [\"2244\", \"2245\"],\n",
    "        #\"name\": \"aspirin\",\n",
    "        \"smiles\": \"CC(=O)OC1=CC=CC=C1C(=O)O\",\n",
    "        \"property\": [\"molecularformula\", \"smiles\", \"hbonddonorcount\", \"hbondacceptorcount\"],\n",
    "    },\n",
    "    method=\"compound\",\n",
    "    #option=\"sids\",\n",
    "    parse=True,\n",
    "    to_dataframe=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "d39c7c21",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cache_key: CC(=O)OC1=CC=CC=C1C(=O)O_compound_cids\n",
      "Prepared request: https://pubchem.ncbi.nlm.nih.gov/rest/pug/compound/smiles/CC(=O)OC1=CC=CC=C1C(=O)O/cids/json\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2244</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    cid\n",
       "0  2244"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "instance.fetch_single(\n",
    "    query={\n",
    "        \"smiles\": \"CC(=O)OC1=CC=CC=C1C(=O)O\",\n",
    "    },\n",
    "    method=\"compound\",\n",
    "    option=\"cids\",\n",
    "    parse=True,\n",
    "    to_dataframe=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "95921c8b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cache_key: CC(=O)OC1=CC=CC=C1C(=O)O_compound_description\n",
      "Prepared request: https://pubchem.ncbi.nlm.nih.gov/rest/pug/compound/smiles/CC(=O)OC1=CC=CC=C1C(=O)O/description/json\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cid</th>\n",
       "      <th>title</th>\n",
       "      <th>description</th>\n",
       "      <th>description_source_name</th>\n",
       "      <th>description_url</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2244</td>\n",
       "      <td>Aspirin</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2244</td>\n",
       "      <td>None</td>\n",
       "      <td>Aspirin can cause developmental toxicity and f...</td>\n",
       "      <td>California Office of Environmental Health Haza...</td>\n",
       "      <td>https://oehha.ca.gov/proposition-65/chemicals/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2244</td>\n",
       "      <td>None</td>\n",
       "      <td>Acetylsalicylic acid is a member of the class ...</td>\n",
       "      <td>ChEBI</td>\n",
       "      <td>https://www.ebi.ac.uk/chebi/searchId.do?chebiI...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    cid    title                                        description  \\\n",
       "0  2244  Aspirin                                               None   \n",
       "1  2244     None  Aspirin can cause developmental toxicity and f...   \n",
       "2  2244     None  Acetylsalicylic acid is a member of the class ...   \n",
       "\n",
       "                             description_source_name  \\\n",
       "0                                               None   \n",
       "1  California Office of Environmental Health Haza...   \n",
       "2                                              ChEBI   \n",
       "\n",
       "                                     description_url  \n",
       "0                                               None  \n",
       "1  https://oehha.ca.gov/proposition-65/chemicals/...  \n",
       "2  https://www.ebi.ac.uk/chebi/searchId.do?chebiI...  "
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "instance.fetch_single(\n",
    "    query={\n",
    "        \"smiles\": \"CC(=O)OC1=CC=CC=C1C(=O)O\",\n",
    "    },\n",
    "    method=\"compound\",\n",
    "    option=\"description\",\n",
    "    parse=True,\n",
    "    to_dataframe=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "541e59d3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cache_key: P00533_protein_summary\n",
      "Prepared request: https://pubchem.ncbi.nlm.nih.gov/rest/pug/protein/accession/P00533/summary/json\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein_accession</th>\n",
       "      <th>name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>synonym</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P00533</td>\n",
       "      <td>Epidermal growth factor receptor</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens (human)</td>\n",
       "      <td>[EC 2.7.10.1, Proto-oncogene c-ErbB-1, Recepto...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  protein_accession                              name  taxonomy_id  \\\n",
       "0            P00533  Epidermal growth factor receptor         9606   \n",
       "\n",
       "               taxonomy                                            synonym  \n",
       "0  Homo sapiens (human)  [EC 2.7.10.1, Proto-oncogene c-ErbB-1, Recepto...  "
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "instance.fetch_single(\n",
    "    query={\n",
    "        \"accession\": \"P00533\",\n",
    "    },\n",
    "    method=\"protein\",\n",
    "    option=\"summary\",\n",
    "    parse=True,\n",
    "    to_dataframe=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "8b4df538",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cache_key: P00533_protein_aids\n",
      "Prepared request: https://pubchem.ncbi.nlm.nih.gov/rest/pug/protein/accession/P00533/aids/json\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein_accession</th>\n",
       "      <th>aid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P00533</td>\n",
       "      <td>[1433, 1726, 1727, 1729, 1731, 1742, 1982, 336...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  protein_accession                                                aid\n",
       "0            P00533  [1433, 1726, 1727, 1729, 1731, 1742, 1982, 336..."
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "instance.fetch_single(\n",
    "    query={\n",
    "        \"accession\": \"P00533\",\n",
    "    },\n",
    "    method=\"protein\",\n",
    "    option=\"aids\",\n",
    "    parse=True,\n",
    "    to_dataframe=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "12d1f789",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cache_key: P00533_protein_concise\n",
      "Prepared request: https://pubchem.ncbi.nlm.nih.gov/rest/pug/protein/accession/P00533/concise/json\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>aid</th>\n",
       "      <th>sid</th>\n",
       "      <th>cid</th>\n",
       "      <th>activity_outcome</th>\n",
       "      <th>target_accession</th>\n",
       "      <th>target_geneid</th>\n",
       "      <th>activity_name</th>\n",
       "      <th>activity_qualifier</th>\n",
       "      <th>activity_value_um</th>\n",
       "      <th>assay_name</th>\n",
       "      <th>assay_type</th>\n",
       "      <th>pubmed_id</th>\n",
       "      <th>rna_i</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3364</td>\n",
       "      <td>103262069</td>\n",
       "      <td>5329249</td>\n",
       "      <td>Active</td>\n",
       "      <td>None</td>\n",
       "      <td>1956</td>\n",
       "      <td>IC50</td>\n",
       "      <td>=</td>\n",
       "      <td>4.32</td>\n",
       "      <td>Inhibition of ligand-induced proliferation in ...</td>\n",
       "      <td>Confirmatory</td>\n",
       "      <td>12127526</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3364</td>\n",
       "      <td>103262170</td>\n",
       "      <td>10300451</td>\n",
       "      <td>Active</td>\n",
       "      <td>None</td>\n",
       "      <td>1956</td>\n",
       "      <td>IC50</td>\n",
       "      <td>=</td>\n",
       "      <td>0.62</td>\n",
       "      <td>Inhibition of ligand-induced proliferation in ...</td>\n",
       "      <td>Confirmatory</td>\n",
       "      <td>12127526</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3364</td>\n",
       "      <td>103262784</td>\n",
       "      <td>5329232</td>\n",
       "      <td>Active</td>\n",
       "      <td>None</td>\n",
       "      <td>1956</td>\n",
       "      <td>IC50</td>\n",
       "      <td>=</td>\n",
       "      <td>7.82</td>\n",
       "      <td>Inhibition of ligand-induced proliferation in ...</td>\n",
       "      <td>Confirmatory</td>\n",
       "      <td>12127526</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3364</td>\n",
       "      <td>103262821</td>\n",
       "      <td>5329244</td>\n",
       "      <td>Active</td>\n",
       "      <td>None</td>\n",
       "      <td>1956</td>\n",
       "      <td>IC50</td>\n",
       "      <td>=</td>\n",
       "      <td>1.29</td>\n",
       "      <td>Inhibition of ligand-induced proliferation in ...</td>\n",
       "      <td>Confirmatory</td>\n",
       "      <td>12127526</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3364</td>\n",
       "      <td>103262822</td>\n",
       "      <td>5329243</td>\n",
       "      <td>Active</td>\n",
       "      <td>None</td>\n",
       "      <td>1956</td>\n",
       "      <td>IC50</td>\n",
       "      <td>=</td>\n",
       "      <td>7.5</td>\n",
       "      <td>Inhibition of ligand-induced proliferation in ...</td>\n",
       "      <td>Confirmatory</td>\n",
       "      <td>12127526</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46762</th>\n",
       "      <td>2061069</td>\n",
       "      <td>513841252</td>\n",
       "      <td>154674422</td>\n",
       "      <td>Inactive</td>\n",
       "      <td>None</td>\n",
       "      <td>1956</td>\n",
       "      <td>IC50</td>\n",
       "      <td>&gt;</td>\n",
       "      <td>20</td>\n",
       "      <td>WT-EGFR Kinase Assay from US Patent US12227501...</td>\n",
       "      <td>Confirmatory</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46763</th>\n",
       "      <td>2061069</td>\n",
       "      <td>513841253</td>\n",
       "      <td>154674465</td>\n",
       "      <td>Inactive</td>\n",
       "      <td>None</td>\n",
       "      <td>1956</td>\n",
       "      <td>IC50</td>\n",
       "      <td>&gt;</td>\n",
       "      <td>20</td>\n",
       "      <td>WT-EGFR Kinase Assay from US Patent US12227501...</td>\n",
       "      <td>Confirmatory</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46764</th>\n",
       "      <td>2061069</td>\n",
       "      <td>513841254</td>\n",
       "      <td>154674465</td>\n",
       "      <td>Inactive</td>\n",
       "      <td>None</td>\n",
       "      <td>1956</td>\n",
       "      <td>IC50</td>\n",
       "      <td>&gt;</td>\n",
       "      <td>20</td>\n",
       "      <td>WT-EGFR Kinase Assay from US Patent US12227501...</td>\n",
       "      <td>Confirmatory</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46765</th>\n",
       "      <td>2061069</td>\n",
       "      <td>513841255</td>\n",
       "      <td>154674465</td>\n",
       "      <td>Inactive</td>\n",
       "      <td>None</td>\n",
       "      <td>1956</td>\n",
       "      <td>IC50</td>\n",
       "      <td>&gt;</td>\n",
       "      <td>20</td>\n",
       "      <td>WT-EGFR Kinase Assay from US Patent US12227501...</td>\n",
       "      <td>Confirmatory</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46766</th>\n",
       "      <td>2061069</td>\n",
       "      <td>513841256</td>\n",
       "      <td>154674285</td>\n",
       "      <td>Active</td>\n",
       "      <td>None</td>\n",
       "      <td>1956</td>\n",
       "      <td>IC50</td>\n",
       "      <td></td>\n",
       "      <td>0.419</td>\n",
       "      <td>WT-EGFR Kinase Assay from US Patent US12227501...</td>\n",
       "      <td>Confirmatory</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>46767 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           aid        sid        cid activity_outcome target_accession  \\\n",
       "0         3364  103262069    5329249           Active             None   \n",
       "1         3364  103262170   10300451           Active             None   \n",
       "2         3364  103262784    5329232           Active             None   \n",
       "3         3364  103262821    5329244           Active             None   \n",
       "4         3364  103262822    5329243           Active             None   \n",
       "...        ...        ...        ...              ...              ...   \n",
       "46762  2061069  513841252  154674422         Inactive             None   \n",
       "46763  2061069  513841253  154674465         Inactive             None   \n",
       "46764  2061069  513841254  154674465         Inactive             None   \n",
       "46765  2061069  513841255  154674465         Inactive             None   \n",
       "46766  2061069  513841256  154674285           Active             None   \n",
       "\n",
       "      target_geneid activity_name activity_qualifier activity_value_um  \\\n",
       "0              1956          IC50                  =              4.32   \n",
       "1              1956          IC50                  =              0.62   \n",
       "2              1956          IC50                  =              7.82   \n",
       "3              1956          IC50                  =              1.29   \n",
       "4              1956          IC50                  =               7.5   \n",
       "...             ...           ...                ...               ...   \n",
       "46762          1956          IC50                  >                20   \n",
       "46763          1956          IC50                  >                20   \n",
       "46764          1956          IC50                  >                20   \n",
       "46765          1956          IC50                  >                20   \n",
       "46766          1956          IC50                                0.419   \n",
       "\n",
       "                                              assay_name    assay_type  \\\n",
       "0      Inhibition of ligand-induced proliferation in ...  Confirmatory   \n",
       "1      Inhibition of ligand-induced proliferation in ...  Confirmatory   \n",
       "2      Inhibition of ligand-induced proliferation in ...  Confirmatory   \n",
       "3      Inhibition of ligand-induced proliferation in ...  Confirmatory   \n",
       "4      Inhibition of ligand-induced proliferation in ...  Confirmatory   \n",
       "...                                                  ...           ...   \n",
       "46762  WT-EGFR Kinase Assay from US Patent US12227501...  Confirmatory   \n",
       "46763  WT-EGFR Kinase Assay from US Patent US12227501...  Confirmatory   \n",
       "46764  WT-EGFR Kinase Assay from US Patent US12227501...  Confirmatory   \n",
       "46765  WT-EGFR Kinase Assay from US Patent US12227501...  Confirmatory   \n",
       "46766  WT-EGFR Kinase Assay from US Patent US12227501...  Confirmatory   \n",
       "\n",
       "      pubmed_id rna_i  \n",
       "0      12127526  None  \n",
       "1      12127526  None  \n",
       "2      12127526  None  \n",
       "3      12127526  None  \n",
       "4      12127526  None  \n",
       "...         ...   ...  \n",
       "46762            None  \n",
       "46763            None  \n",
       "46764            None  \n",
       "46765            None  \n",
       "46766            None  \n",
       "\n",
       "[46767 rows x 13 columns]"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "instance.fetch_single(\n",
    "    query={\n",
    "        \"accession\": \"P00533\",\n",
    "    },\n",
    "    method=\"protein\",\n",
    "    option=\"concise\",\n",
    "    parse=True,\n",
    "    to_dataframe=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "95601e01",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cache_key: P00533_protein_pwaccs\n",
      "Prepared request: https://pubchem.ncbi.nlm.nih.gov/rest/pug/protein/accession/P00533/pwaccs/json\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein_accession</th>\n",
       "      <th>pathway_accession</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P00533</td>\n",
       "      <td>[PathBank:SMP0000472, PathBank:SMP0000473, Pat...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  protein_accession                                  pathway_accession\n",
       "0            P00533  [PathBank:SMP0000472, PathBank:SMP0000473, Pat..."
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "instance.fetch_single(\n",
    "    query={\n",
    "        \"accession\": \"P00533\",\n",
    "    },\n",
    "    method=\"protein\",\n",
    "    option=\"pwaccs\",\n",
    "    parse=True,\n",
    "    to_dataframe=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "a905da8d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cache_key: APP_gene_summary\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[[{'gene_id': 351,\n",
       "   'symbol': 'APP',\n",
       "   'name': 'amyloid beta precursor protein',\n",
       "   'taxonomy_id': 9606,\n",
       "   'taxonomy': 'Homo sapiens (human)',\n",
       "   'description': 'This gene encodes a cell surface receptor and transmembrane precursor protein that is cleaved by secretases to form a number of peptides. Some of these peptides are secreted and can bind to the acetyltransferase complex APBB1/TIP60 to promote transcriptional activation, while others form the protein basis of the amyloid plaques found in the brains of patients with Alzheimer disease. In addition, two of the peptides are antimicrobial peptides, having been shown to have bacteriocidal and antifungal activities. Mutations in this gene have been implicated in autosomal dominant Alzheimer disease and cerebroarterial amyloidosis (cerebral amyloid angiopathy). Multiple transcript variants encoding several different isoforms have been found for this gene. [provided by RefSeq, Aug 2014]',\n",
       "   'synonym': ['AAA',\n",
       "    'ABETA',\n",
       "    'ABPP',\n",
       "    'AD1',\n",
       "    'APPI',\n",
       "    'CTFgamma',\n",
       "    'CVAP',\n",
       "    'PN-II',\n",
       "    'PN2',\n",
       "    'alpha-sAPP',\n",
       "    'preA4',\n",
       "    'amyloid-beta precursor protein',\n",
       "    'alzheimer disease amyloid A4 protein homolog',\n",
       "    'alzheimer disease amyloid protein',\n",
       "    'amyloid beta (A4) precursor protein',\n",
       "    'amyloid beta A4 protein',\n",
       "    'amyloid precursor protein',\n",
       "    'beta-amyloid peptide',\n",
       "    'beta-amyloid peptide(1-40)',\n",
       "    'beta-amyloid peptide(1-42)',\n",
       "    'beta-amyloid precursor protein',\n",
       "    'cerebral vascular amyloid peptide',\n",
       "    'peptidase nexin-II',\n",
       "    'protease nexin-II',\n",
       "    'testicular tissue protein Li 2']}]]"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "instance.fetch_single(\n",
    "    query={\n",
    "        \"genesymbol\": [\"APP\"]\n",
    "    },\n",
    "    method=\"gene\",\n",
    "    option=\"summary\",\n",
    "    parse=True,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "27fdd012",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cache_key: APP_gene_summary\n",
      "Cache_key: ECD_gene_summary\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>symbol</th>\n",
       "      <th>name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>description</th>\n",
       "      <th>synonym</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>351</td>\n",
       "      <td>APP</td>\n",
       "      <td>amyloid beta precursor protein</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens (human)</td>\n",
       "      <td>This gene encodes a cell surface receptor and ...</td>\n",
       "      <td>[AAA, ABETA, ABPP, AD1, APPI, CTFgamma, CVAP, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>11319</td>\n",
       "      <td>ECD</td>\n",
       "      <td>ecdysoneless cell cycle regulator</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens (human)</td>\n",
       "      <td>Enables histone acetyltransferase binding acti...</td>\n",
       "      <td>[GCR2, HSGT1, SGT1, protein ecdysoneless homol...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene_id symbol                               name  taxonomy_id  \\\n",
       "0      351    APP     amyloid beta precursor protein         9606   \n",
       "1    11319    ECD  ecdysoneless cell cycle regulator         9606   \n",
       "\n",
       "               taxonomy                                        description  \\\n",
       "0  Homo sapiens (human)  This gene encodes a cell surface receptor and ...   \n",
       "1  Homo sapiens (human)  Enables histone acetyltransferase binding acti...   \n",
       "\n",
       "                                             synonym  \n",
       "0  [AAA, ABETA, ABPP, AD1, APPI, CTFgamma, CVAP, ...  \n",
       "1  [GCR2, HSGT1, SGT1, protein ecdysoneless homol...  "
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "instance.fetch_single(\n",
    "    query={\n",
    "        \"genesymbol\": [\"APP\", \"ECD\"],\n",
    "    },\n",
    "    method=\"gene\",\n",
    "    option=\"summary\",\n",
    "    parse=True,\n",
    "    to_dataframe=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "26e0b584",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cache_key: APP_gene_concise\n",
      "Prepared request: https://pubchem.ncbi.nlm.nih.gov/rest/pug/gene/genesymbol/APP/concise/json\n",
      "Cache_key: APP_gene_concise\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>aid</th>\n",
       "      <th>sid</th>\n",
       "      <th>cid</th>\n",
       "      <th>activity_outcome</th>\n",
       "      <th>target_accession</th>\n",
       "      <th>target_geneid</th>\n",
       "      <th>activity_name</th>\n",
       "      <th>activity_qualifier</th>\n",
       "      <th>activity_value_um</th>\n",
       "      <th>assay_name</th>\n",
       "      <th>assay_type</th>\n",
       "      <th>pubmed_id</th>\n",
       "      <th>rna_i</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>374278</td>\n",
       "      <td>103655342</td>\n",
       "      <td>42603587</td>\n",
       "      <td>Active</td>\n",
       "      <td>P05067</td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Binding affinity to beta-amyloid plaques in su...</td>\n",
       "      <td>Other</td>\n",
       "      <td>18992967</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>374278</td>\n",
       "      <td>103655344</td>\n",
       "      <td>42603588</td>\n",
       "      <td>Active</td>\n",
       "      <td>P05067</td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Binding affinity to beta-amyloid plaques in su...</td>\n",
       "      <td>Other</td>\n",
       "      <td>18992967</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>374282</td>\n",
       "      <td>103655343</td>\n",
       "      <td>42603729</td>\n",
       "      <td>Active</td>\n",
       "      <td>P05067</td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Binding affinity to vascular beta-amyloid plaq...</td>\n",
       "      <td>Other</td>\n",
       "      <td>18992967</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>473568</td>\n",
       "      <td>103747372</td>\n",
       "      <td>11574638</td>\n",
       "      <td>Unspecified</td>\n",
       "      <td>P05067</td>\n",
       "      <td>None</td>\n",
       "      <td>EC50</td>\n",
       "      <td>=</td>\n",
       "      <td>16</td>\n",
       "      <td>Inhibition of human recombinant APP in CHO cel...</td>\n",
       "      <td>Confirmatory</td>\n",
       "      <td>20223661</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>473568</td>\n",
       "      <td>103747583</td>\n",
       "      <td>11596660</td>\n",
       "      <td>Active</td>\n",
       "      <td>P05067</td>\n",
       "      <td>None</td>\n",
       "      <td>EC50</td>\n",
       "      <td>=</td>\n",
       "      <td>4.8</td>\n",
       "      <td>Inhibition of human recombinant APP in CHO cel...</td>\n",
       "      <td>Confirmatory</td>\n",
       "      <td>20223661</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123</th>\n",
       "      <td>1831909</td>\n",
       "      <td>242634277</td>\n",
       "      <td>66665322</td>\n",
       "      <td>Active</td>\n",
       "      <td>P05067</td>\n",
       "      <td>None</td>\n",
       "      <td>IC50</td>\n",
       "      <td>=</td>\n",
       "      <td>0.003</td>\n",
       "      <td>Inhibition of wild type human APP expressed in...</td>\n",
       "      <td>Confirmatory</td>\n",
       "      <td>33844524</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>1515584</td>\n",
       "      <td>103218648</td>\n",
       "      <td>656629</td>\n",
       "      <td>Active</td>\n",
       "      <td>P05067</td>\n",
       "      <td>None</td>\n",
       "      <td>IC50</td>\n",
       "      <td>=</td>\n",
       "      <td>0.19953</td>\n",
       "      <td>Inhibition of APP (unknown origin)</td>\n",
       "      <td>Confirmatory</td>\n",
       "      <td>30655952</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>1515584</td>\n",
       "      <td>440111474</td>\n",
       "      <td>45102601</td>\n",
       "      <td>Unspecified</td>\n",
       "      <td>P05067</td>\n",
       "      <td>None</td>\n",
       "      <td>IC50</td>\n",
       "      <td>&gt;</td>\n",
       "      <td>10</td>\n",
       "      <td>Inhibition of APP (unknown origin)</td>\n",
       "      <td>Confirmatory</td>\n",
       "      <td>30655952</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>1831886</td>\n",
       "      <td>242634277</td>\n",
       "      <td>66665322</td>\n",
       "      <td>Active</td>\n",
       "      <td>P05067</td>\n",
       "      <td>None</td>\n",
       "      <td>IC50</td>\n",
       "      <td>=</td>\n",
       "      <td>0.004</td>\n",
       "      <td>Inhibition of wild type human APP expressed in...</td>\n",
       "      <td>Confirmatory</td>\n",
       "      <td>33844524</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>1831911</td>\n",
       "      <td>242634277</td>\n",
       "      <td>66665322</td>\n",
       "      <td>Active</td>\n",
       "      <td>P05067</td>\n",
       "      <td>None</td>\n",
       "      <td>IC50</td>\n",
       "      <td>=</td>\n",
       "      <td>0.003</td>\n",
       "      <td>Inhibition of wild type human APP expressed in...</td>\n",
       "      <td>Confirmatory</td>\n",
       "      <td>33844524</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>128 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         aid        sid       cid activity_outcome target_accession  \\\n",
       "0     374278  103655342  42603587           Active           P05067   \n",
       "1     374278  103655344  42603588           Active           P05067   \n",
       "2     374282  103655343  42603729           Active           P05067   \n",
       "3     473568  103747372  11574638      Unspecified           P05067   \n",
       "4     473568  103747583  11596660           Active           P05067   \n",
       "..       ...        ...       ...              ...              ...   \n",
       "123  1831909  242634277  66665322           Active           P05067   \n",
       "124  1515584  103218648    656629           Active           P05067   \n",
       "125  1515584  440111474  45102601      Unspecified           P05067   \n",
       "126  1831886  242634277  66665322           Active           P05067   \n",
       "127  1831911  242634277  66665322           Active           P05067   \n",
       "\n",
       "    target_geneid activity_name activity_qualifier activity_value_um  \\\n",
       "0            None                                                      \n",
       "1            None                                                      \n",
       "2            None                                                      \n",
       "3            None          EC50                  =                16   \n",
       "4            None          EC50                  =               4.8   \n",
       "..            ...           ...                ...               ...   \n",
       "123          None          IC50                  =             0.003   \n",
       "124          None          IC50                  =           0.19953   \n",
       "125          None          IC50                  >                10   \n",
       "126          None          IC50                  =             0.004   \n",
       "127          None          IC50                  =             0.003   \n",
       "\n",
       "                                            assay_name    assay_type  \\\n",
       "0    Binding affinity to beta-amyloid plaques in su...         Other   \n",
       "1    Binding affinity to beta-amyloid plaques in su...         Other   \n",
       "2    Binding affinity to vascular beta-amyloid plaq...         Other   \n",
       "3    Inhibition of human recombinant APP in CHO cel...  Confirmatory   \n",
       "4    Inhibition of human recombinant APP in CHO cel...  Confirmatory   \n",
       "..                                                 ...           ...   \n",
       "123  Inhibition of wild type human APP expressed in...  Confirmatory   \n",
       "124                 Inhibition of APP (unknown origin)  Confirmatory   \n",
       "125                 Inhibition of APP (unknown origin)  Confirmatory   \n",
       "126  Inhibition of wild type human APP expressed in...  Confirmatory   \n",
       "127  Inhibition of wild type human APP expressed in...  Confirmatory   \n",
       "\n",
       "    pubmed_id rna_i  \n",
       "0    18992967        \n",
       "1    18992967        \n",
       "2    18992967        \n",
       "3    20223661        \n",
       "4    20223661        \n",
       "..        ...   ...  \n",
       "123  33844524        \n",
       "124  30655952        \n",
       "125  30655952        \n",
       "126  33844524        \n",
       "127  33844524        \n",
       "\n",
       "[128 rows x 13 columns]"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "instance.fetch_single(\n",
    "    query={\n",
    "        \"genesymbol\": [\"APP\"],\n",
    "    },\n",
    "    method=\"gene\",\n",
    "    option=\"concise\",\n",
    "    parse=True,\n",
    "    to_dataframe=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "c5d56d97",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cache_key: APP_gene_pwaccs\n",
      "Prepared request: https://pubchem.ncbi.nlm.nih.gov/rest/pug/gene/genesymbol/APP/pwaccs/json\n",
      "Cache_key: APP_gene_pwaccs\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>pathway_accession</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>351</td>\n",
       "      <td>[COVID-19 Disease Map:718, PANTHER:P00003, Pat...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene_id                                  pathway_accession\n",
       "0      351  [COVID-19 Disease Map:718, PANTHER:P00003, Pat..."
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "instance.fetch_single(\n",
    "    query={\n",
    "        \"genesymbol\": [\"APP\"],\n",
    "    },\n",
    "    method=\"gene\",\n",
    "    option=\"pwaccs\",\n",
    "    parse=True,\n",
    "    to_dataframe=True\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6b8dec67",
   "metadata": {},
   "source": [
    "## Reactome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "ac49e8e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "from src.reactome import ReactomeInstance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "7eddbce6",
   "metadata": {},
   "outputs": [],
   "source": [
    "instance = ReactomeInstance()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "e9b3601e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cache_key: R-DME-1834941_data-discover\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>description</th>\n",
       "      <th>url</th>\n",
       "      <th>version</th>\n",
       "      <th>keywords</th>\n",
       "      <th>includedInDataCatalog</th>\n",
       "      <th>distribution</th>\n",
       "      <th>license</th>\n",
       "      <th>@context</th>\n",
       "      <th>@type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>STING mediated induction of host immune responses</td>\n",
       "      <td>This event has been computationally inferred f...</td>\n",
       "      <td>https://reactome.org/PathwayBrowser/#/R-DME-18...</td>\n",
       "      <td>93</td>\n",
       "      <td>[Pathway]</td>\n",
       "      <td>{'name': 'Reactome', 'url': 'https://reactome....</td>\n",
       "      <td>[{'contentUrl': 'https://reactome.org/ContentS...</td>\n",
       "      <td>https://creativecommons.org/licenses/by/4.0/</td>\n",
       "      <td>http://schema.org/</td>\n",
       "      <td>DataSet</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                name  \\\n",
       "0  STING mediated induction of host immune responses   \n",
       "\n",
       "                                         description  \\\n",
       "0  This event has been computationally inferred f...   \n",
       "\n",
       "                                                 url version   keywords  \\\n",
       "0  https://reactome.org/PathwayBrowser/#/R-DME-18...      93  [Pathway]   \n",
       "\n",
       "                               includedInDataCatalog  \\\n",
       "0  {'name': 'Reactome', 'url': 'https://reactome....   \n",
       "\n",
       "                                        distribution  \\\n",
       "0  [{'contentUrl': 'https://reactome.org/ContentS...   \n",
       "\n",
       "                                        license            @context    @type  \n",
       "0  https://creativecommons.org/licenses/by/4.0/  http://schema.org/  DataSet  "
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "instance.fetch_single(query=\"R-DME-1834941\", parse=True, method=\"data-discover\", to_dataframe=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bed9c7ca",
   "metadata": {},
   "source": [
    "## RHEA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "43249840",
   "metadata": {},
   "outputs": [],
   "source": [
    "from src.rhea import RheaInterface"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "08161c73",
   "metadata": {},
   "outputs": [],
   "source": [
    "instance = RheaInterface()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "049b4b42",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cache_key: uniprot:*_rhea\n",
      "Prepared request: https://www.rhea-db.org/rhea/?query=uniprot%3A%2A&columns=rhea-id%2Cequation%2Cchebi%2Cchebi-id%2Cec%2Cuniprot%2Cgo&format=json&limit=100\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>equation</th>\n",
       "      <th>status</th>\n",
       "      <th>htmlequation</th>\n",
       "      <th>comment</th>\n",
       "      <th>balanced</th>\n",
       "      <th>transport</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>21252</td>\n",
       "      <td>(S)-2-hydroxyglutarate + A = 2-oxoglutarate + AH2</td>\n",
       "      <td>approved</td>\n",
       "      <td>&lt;span class=\"participant\"&gt;&lt;span class=\"stoichi...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>21256</td>\n",
       "      <td>3-phosphoshikimate + phosphoenolpyruvate = 5-O...</td>\n",
       "      <td>approved</td>\n",
       "      <td>&lt;span class=\"participant\"&gt;&lt;span class=\"stoichi...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>21260</td>\n",
       "      <td>[thioredoxin]-disulfide + L-methionine + H2O =...</td>\n",
       "      <td>approved</td>\n",
       "      <td>&lt;span class=\"participant\"&gt;&lt;span class=\"stoichi...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>21264</td>\n",
       "      <td>glycolate + A = glyoxylate + AH2</td>\n",
       "      <td>approved</td>\n",
       "      <td>&lt;span class=\"participant\"&gt;&lt;span class=\"stoichi...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>21272</td>\n",
       "      <td>spermidine + glutathione + ATP = glutathionyls...</td>\n",
       "      <td>approved</td>\n",
       "      <td>&lt;span class=\"participant\"&gt;&lt;span class=\"stoichi...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>21800</td>\n",
       "      <td>all-trans-antheraxanthin + L-ascorbate = all-t...</td>\n",
       "      <td>approved</td>\n",
       "      <td>&lt;span class=\"participant\"&gt;&lt;span class=\"stoichi...</td>\n",
       "      <td>RHEA:21800 part of RHEA:32371 \\n\\nPublished in...</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>21804</td>\n",
       "      <td>myo-inositol + ATP = 1D-myo-inositol 3-phospha...</td>\n",
       "      <td>approved</td>\n",
       "      <td>&lt;span class=\"participant\"&gt;&lt;span class=\"stoichi...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>21812</td>\n",
       "      <td>cholesterol + reduced [NADPH--hemoprotein redu...</td>\n",
       "      <td>approved</td>\n",
       "      <td>&lt;span class=\"participant\"&gt;&lt;span class=\"stoichi...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>21816</td>\n",
       "      <td>a D-alpha-amino acid + O2 + H2O = a 2-oxocarbo...</td>\n",
       "      <td>approved</td>\n",
       "      <td>&lt;span class=\"participant\"&gt;&lt;span class=\"stoichi...</td>\n",
       "      <td>Multi-step reaction: RHEA:78799 + RHEA:78803</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>21824</td>\n",
       "      <td>L-aspartate + 2-oxoglutarate = oxaloacetate + ...</td>\n",
       "      <td>approved</td>\n",
       "      <td>&lt;span class=\"participant\"&gt;&lt;span class=\"stoichi...</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       id                                           equation    status  \\\n",
       "0   21252  (S)-2-hydroxyglutarate + A = 2-oxoglutarate + AH2  approved   \n",
       "1   21256  3-phosphoshikimate + phosphoenolpyruvate = 5-O...  approved   \n",
       "2   21260  [thioredoxin]-disulfide + L-methionine + H2O =...  approved   \n",
       "3   21264                   glycolate + A = glyoxylate + AH2  approved   \n",
       "4   21272  spermidine + glutathione + ATP = glutathionyls...  approved   \n",
       "..    ...                                                ...       ...   \n",
       "95  21800  all-trans-antheraxanthin + L-ascorbate = all-t...  approved   \n",
       "96  21804  myo-inositol + ATP = 1D-myo-inositol 3-phospha...  approved   \n",
       "97  21812  cholesterol + reduced [NADPH--hemoprotein redu...  approved   \n",
       "98  21816  a D-alpha-amino acid + O2 + H2O = a 2-oxocarbo...  approved   \n",
       "99  21824  L-aspartate + 2-oxoglutarate = oxaloacetate + ...  approved   \n",
       "\n",
       "                                         htmlequation  \\\n",
       "0   <span class=\"participant\"><span class=\"stoichi...   \n",
       "1   <span class=\"participant\"><span class=\"stoichi...   \n",
       "2   <span class=\"participant\"><span class=\"stoichi...   \n",
       "3   <span class=\"participant\"><span class=\"stoichi...   \n",
       "4   <span class=\"participant\"><span class=\"stoichi...   \n",
       "..                                                ...   \n",
       "95  <span class=\"participant\"><span class=\"stoichi...   \n",
       "96  <span class=\"participant\"><span class=\"stoichi...   \n",
       "97  <span class=\"participant\"><span class=\"stoichi...   \n",
       "98  <span class=\"participant\"><span class=\"stoichi...   \n",
       "99  <span class=\"participant\"><span class=\"stoichi...   \n",
       "\n",
       "                                              comment  balanced  transport  \n",
       "0                                                          True      False  \n",
       "1                                                          True      False  \n",
       "2                                                          True      False  \n",
       "3                                                          True      False  \n",
       "4                                                          True      False  \n",
       "..                                                ...       ...        ...  \n",
       "95  RHEA:21800 part of RHEA:32371 \\n\\nPublished in...      True      False  \n",
       "96                                                         True      False  \n",
       "97                                                         True      False  \n",
       "98       Multi-step reaction: RHEA:78799 + RHEA:78803      True      False  \n",
       "99                                                         True      False  \n",
       "\n",
       "[100 rows x 7 columns]"
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "instance.fetch_single(\n",
    "    query={\n",
    "        \"query\": \"uniprot:*\"\n",
    "    },\n",
    "    method=\"rhea\",\n",
    "    parse=True,\n",
    "    to_dataframe=True\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ed86b941",
   "metadata": {},
   "source": [
    "## Refseq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "700d5f92",
   "metadata": {},
   "outputs": [],
   "source": [
    "from src.refseq import RefSeqInterface"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "82f71c4c",
   "metadata": {},
   "outputs": [],
   "source": [
    "instance = RefSeqInterface()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "63f53aa2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cache_key: ['XP_010804480.1', 'XP_010804481.1', 'XP_010804482.1']_protein\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>locus</th>\n",
       "      <th>project_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>XP_010804480</td>\n",
       "      <td>PRJNA33843</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>XP_010804481</td>\n",
       "      <td>PRJNA33843</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>XP_010804482</td>\n",
       "      <td>PRJNA33843</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          locus  project_id\n",
       "0  XP_010804480  PRJNA33843\n",
       "1  XP_010804481  PRJNA33843\n",
       "2  XP_010804482  PRJNA33843"
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "instance.fetch_single(\n",
    "    query=[\"XP_010804480.1\", \"XP_010804481.1\", \"XP_010804482.1\"],\n",
    "    method=\"protein\",\n",
    "    parse=True,\n",
    "    to_dataframe=True,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "b4fc2896",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cache_key: XP_010804480.1_protein\n",
      "Cache_key: XP_010804481.1_protein\n",
      "Cache_key: XP_010804482.1_protein\n",
      "Cache_key: XP_010804480.1_protein\n",
      "Cache_key: XP_010804481.1_protein\n",
      "Cache_key: XP_010804482.1_protein\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>locus</th>\n",
       "      <th>project_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>XP_010804480</td>\n",
       "      <td>PRJNA33843</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>XP_010804481</td>\n",
       "      <td>PRJNA33843</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>XP_010804482</td>\n",
       "      <td>PRJNA33843</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          locus  project_id\n",
       "0  XP_010804480  PRJNA33843\n",
       "1  XP_010804481  PRJNA33843\n",
       "2  XP_010804482  PRJNA33843"
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "instance.fetch_batch(\n",
    "    queries=[\"XP_010804480.1\", \"XP_010804481.1\", \"XP_010804482.1\"],\n",
    "    method=\"protein\",\n",
    "    parse=True,\n",
    "    to_dataframe=True,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "57034340",
   "metadata": {},
   "source": [
    "## STRING"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "6ed9cccf",
   "metadata": {},
   "outputs": [],
   "source": [
    "from src.stringdb import StringInterface"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "dea4ee3d",
   "metadata": {},
   "outputs": [],
   "source": [
    "instance = StringInterface()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "073277ea",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cache_key: p53_get_string_ids\n",
      "Prepared request URL: https://string-db.org/api/json/get_string_ids?identifiers=p53&echo_query=0\n",
      "Cache_key: p53_get_string_ids\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query_index</th>\n",
       "      <th>query_item</th>\n",
       "      <th>str_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>p53</td>\n",
       "      <td>9606.ENSP00000269305</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   query_index query_item                str_id\n",
       "0            0        p53  9606.ENSP00000269305"
      ]
     },
     "execution_count": 104,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "instance.fetch_single(\n",
    "    query={\n",
    "        \"identifiers\": [\"p53\"],\n",
    "    },\n",
    "    method=\"get_string_ids\",\n",
    "    parse=True,\n",
    "    to_dataframe=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "0f0a55ca",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cache_key: p53_interaction_partners\n",
      "Cache_key: cdk2_interaction_partners\n",
      "Prepared request URL: https://string-db.org/api/json/interaction_partners?identifiers=p53%250dcdk2&species=9606&network_type=functional\n",
      "Cache_key: p53_interaction_partners\n",
      "Cache_key: cdk2_interaction_partners\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id_a</th>\n",
       "      <th>id_b</th>\n",
       "      <th>name_a</th>\n",
       "      <th>name_b</th>\n",
       "      <th>score</th>\n",
       "      <th>nscore</th>\n",
       "      <th>fscore</th>\n",
       "      <th>pscore</th>\n",
       "      <th>ascore</th>\n",
       "      <th>escore</th>\n",
       "      <th>dscore</th>\n",
       "      <th>tscore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9606.ENSP00000266970</td>\n",
       "      <td>9606.ENSP00000481380</td>\n",
       "      <td>CDK2</td>\n",
       "      <td>CCNA2</td>\n",
       "      <td>0.999</td>\n",
       "      <td>0</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.453</td>\n",
       "      <td>0.999</td>\n",
       "      <td>0.9</td>\n",
       "      <td>0.999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9606.ENSP00000266970</td>\n",
       "      <td>9606.ENSP00000413720</td>\n",
       "      <td>CDK2</td>\n",
       "      <td>CDKN1C</td>\n",
       "      <td>0.999</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.085</td>\n",
       "      <td>0.859</td>\n",
       "      <td>0.9</td>\n",
       "      <td>0.970</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>9606.ENSP00000266970</td>\n",
       "      <td>9606.ENSP00000228872</td>\n",
       "      <td>CDK2</td>\n",
       "      <td>CDKN1B</td>\n",
       "      <td>0.999</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.085</td>\n",
       "      <td>0.999</td>\n",
       "      <td>0.9</td>\n",
       "      <td>0.999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9606.ENSP00000266970</td>\n",
       "      <td>9606.ENSP00000429089</td>\n",
       "      <td>CDK2</td>\n",
       "      <td>CCNE2</td>\n",
       "      <td>0.999</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.200</td>\n",
       "      <td>0.995</td>\n",
       "      <td>0.9</td>\n",
       "      <td>0.996</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9606.ENSP00000266970</td>\n",
       "      <td>9606.ENSP00000255465</td>\n",
       "      <td>CDK2</td>\n",
       "      <td>CCNA1</td>\n",
       "      <td>0.999</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.292</td>\n",
       "      <td>0.942</td>\n",
       "      <td>0.9</td>\n",
       "      <td>0.999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5917</th>\n",
       "      <td>9606.ENSP00000269305</td>\n",
       "      <td>9606.ENSP00000310928</td>\n",
       "      <td>TP53</td>\n",
       "      <td>PPARD</td>\n",
       "      <td>0.400</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5918</th>\n",
       "      <td>9606.ENSP00000269305</td>\n",
       "      <td>9606.ENSP00000419945</td>\n",
       "      <td>TP53</td>\n",
       "      <td>ERVW-1</td>\n",
       "      <td>0.400</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5919</th>\n",
       "      <td>9606.ENSP00000269305</td>\n",
       "      <td>9606.ENSP00000462980</td>\n",
       "      <td>TP53</td>\n",
       "      <td>TAF4B</td>\n",
       "      <td>0.400</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.4</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5920</th>\n",
       "      <td>9606.ENSP00000269305</td>\n",
       "      <td>9606.ENSP00000431885</td>\n",
       "      <td>TP53</td>\n",
       "      <td>TYK2</td>\n",
       "      <td>0.400</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.083</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.372</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5921</th>\n",
       "      <td>9606.ENSP00000269305</td>\n",
       "      <td>9606.ENSP00000356247</td>\n",
       "      <td>TP53</td>\n",
       "      <td>LGR6</td>\n",
       "      <td>0.400</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.165</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.311</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5922 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      id_a                  id_b name_a  name_b  score  \\\n",
       "0     9606.ENSP00000266970  9606.ENSP00000481380   CDK2   CCNA2  0.999   \n",
       "1     9606.ENSP00000266970  9606.ENSP00000413720   CDK2  CDKN1C  0.999   \n",
       "2     9606.ENSP00000266970  9606.ENSP00000228872   CDK2  CDKN1B  0.999   \n",
       "3     9606.ENSP00000266970  9606.ENSP00000429089   CDK2   CCNE2  0.999   \n",
       "4     9606.ENSP00000266970  9606.ENSP00000255465   CDK2   CCNA1  0.999   \n",
       "...                    ...                   ...    ...     ...    ...   \n",
       "5917  9606.ENSP00000269305  9606.ENSP00000310928   TP53   PPARD  0.400   \n",
       "5918  9606.ENSP00000269305  9606.ENSP00000419945   TP53  ERVW-1  0.400   \n",
       "5919  9606.ENSP00000269305  9606.ENSP00000462980   TP53   TAF4B  0.400   \n",
       "5920  9606.ENSP00000269305  9606.ENSP00000431885   TP53    TYK2  0.400   \n",
       "5921  9606.ENSP00000269305  9606.ENSP00000356247   TP53    LGR6  0.400   \n",
       "\n",
       "      nscore  fscore  pscore  ascore  escore  dscore  tscore  \n",
       "0          0   0.003     0.0   0.453   0.999     0.9   0.999  \n",
       "1          0   0.000     0.0   0.085   0.859     0.9   0.970  \n",
       "2          0   0.000     0.0   0.085   0.999     0.9   0.999  \n",
       "3          0   0.000     0.0   0.200   0.995     0.9   0.996  \n",
       "4          0   0.000     0.0   0.292   0.942     0.9   0.999  \n",
       "...      ...     ...     ...     ...     ...     ...     ...  \n",
       "5917       0   0.000     0.0   0.000   0.000     0.0   0.400  \n",
       "5918       0   0.000     0.0   0.000   0.000     0.0   0.400  \n",
       "5919       0   0.000     0.0   0.000   0.000     0.4   0.000  \n",
       "5920       0   0.000     0.0   0.083   0.000     0.0   0.372  \n",
       "5921       0   0.000     0.0   0.000   0.165     0.0   0.311  \n",
       "\n",
       "[5922 rows x 12 columns]"
      ]
     },
     "execution_count": 105,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "instance.fetch_single(\n",
    "    method=\"interaction_partners\",\n",
    "    query={\n",
    "        \"identifiers\": [\"p53\", \"cdk2\"],\n",
    "        \"species\": 9606,\n",
    "    },\n",
    "    parse=True,\n",
    "    to_dataframe=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "9635fbc2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cache_key: p53_interaction_partners\n",
      "Cache_key: cdk2_interaction_partners\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id_a</th>\n",
       "      <th>id_b</th>\n",
       "      <th>name_a</th>\n",
       "      <th>name_b</th>\n",
       "      <th>score</th>\n",
       "      <th>nscore</th>\n",
       "      <th>fscore</th>\n",
       "      <th>pscore</th>\n",
       "      <th>ascore</th>\n",
       "      <th>escore</th>\n",
       "      <th>dscore</th>\n",
       "      <th>tscore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9606.ENSP00000266970</td>\n",
       "      <td>9606.ENSP00000481380</td>\n",
       "      <td>CDK2</td>\n",
       "      <td>CCNA2</td>\n",
       "      <td>0.999</td>\n",
       "      <td>0</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.453</td>\n",
       "      <td>0.999</td>\n",
       "      <td>0.9</td>\n",
       "      <td>0.999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9606.ENSP00000266970</td>\n",
       "      <td>9606.ENSP00000413720</td>\n",
       "      <td>CDK2</td>\n",
       "      <td>CDKN1C</td>\n",
       "      <td>0.999</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.085</td>\n",
       "      <td>0.859</td>\n",
       "      <td>0.9</td>\n",
       "      <td>0.970</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>9606.ENSP00000266970</td>\n",
       "      <td>9606.ENSP00000228872</td>\n",
       "      <td>CDK2</td>\n",
       "      <td>CDKN1B</td>\n",
       "      <td>0.999</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.085</td>\n",
       "      <td>0.999</td>\n",
       "      <td>0.9</td>\n",
       "      <td>0.999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9606.ENSP00000266970</td>\n",
       "      <td>9606.ENSP00000429089</td>\n",
       "      <td>CDK2</td>\n",
       "      <td>CCNE2</td>\n",
       "      <td>0.999</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.200</td>\n",
       "      <td>0.995</td>\n",
       "      <td>0.9</td>\n",
       "      <td>0.996</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9606.ENSP00000266970</td>\n",
       "      <td>9606.ENSP00000255465</td>\n",
       "      <td>CDK2</td>\n",
       "      <td>CCNA1</td>\n",
       "      <td>0.999</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.292</td>\n",
       "      <td>0.942</td>\n",
       "      <td>0.9</td>\n",
       "      <td>0.999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5917</th>\n",
       "      <td>9606.ENSP00000269305</td>\n",
       "      <td>9606.ENSP00000310928</td>\n",
       "      <td>TP53</td>\n",
       "      <td>PPARD</td>\n",
       "      <td>0.400</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5918</th>\n",
       "      <td>9606.ENSP00000269305</td>\n",
       "      <td>9606.ENSP00000419945</td>\n",
       "      <td>TP53</td>\n",
       "      <td>ERVW-1</td>\n",
       "      <td>0.400</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5919</th>\n",
       "      <td>9606.ENSP00000269305</td>\n",
       "      <td>9606.ENSP00000462980</td>\n",
       "      <td>TP53</td>\n",
       "      <td>TAF4B</td>\n",
       "      <td>0.400</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.4</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5920</th>\n",
       "      <td>9606.ENSP00000269305</td>\n",
       "      <td>9606.ENSP00000431885</td>\n",
       "      <td>TP53</td>\n",
       "      <td>TYK2</td>\n",
       "      <td>0.400</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.083</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.372</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5921</th>\n",
       "      <td>9606.ENSP00000269305</td>\n",
       "      <td>9606.ENSP00000356247</td>\n",
       "      <td>TP53</td>\n",
       "      <td>LGR6</td>\n",
       "      <td>0.400</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.165</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.311</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5922 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      id_a                  id_b name_a  name_b  score  \\\n",
       "0     9606.ENSP00000266970  9606.ENSP00000481380   CDK2   CCNA2  0.999   \n",
       "1     9606.ENSP00000266970  9606.ENSP00000413720   CDK2  CDKN1C  0.999   \n",
       "2     9606.ENSP00000266970  9606.ENSP00000228872   CDK2  CDKN1B  0.999   \n",
       "3     9606.ENSP00000266970  9606.ENSP00000429089   CDK2   CCNE2  0.999   \n",
       "4     9606.ENSP00000266970  9606.ENSP00000255465   CDK2   CCNA1  0.999   \n",
       "...                    ...                   ...    ...     ...    ...   \n",
       "5917  9606.ENSP00000269305  9606.ENSP00000310928   TP53   PPARD  0.400   \n",
       "5918  9606.ENSP00000269305  9606.ENSP00000419945   TP53  ERVW-1  0.400   \n",
       "5919  9606.ENSP00000269305  9606.ENSP00000462980   TP53   TAF4B  0.400   \n",
       "5920  9606.ENSP00000269305  9606.ENSP00000431885   TP53    TYK2  0.400   \n",
       "5921  9606.ENSP00000269305  9606.ENSP00000356247   TP53    LGR6  0.400   \n",
       "\n",
       "      nscore  fscore  pscore  ascore  escore  dscore  tscore  \n",
       "0          0   0.003     0.0   0.453   0.999     0.9   0.999  \n",
       "1          0   0.000     0.0   0.085   0.859     0.9   0.970  \n",
       "2          0   0.000     0.0   0.085   0.999     0.9   0.999  \n",
       "3          0   0.000     0.0   0.200   0.995     0.9   0.996  \n",
       "4          0   0.000     0.0   0.292   0.942     0.9   0.999  \n",
       "...      ...     ...     ...     ...     ...     ...     ...  \n",
       "5917       0   0.000     0.0   0.000   0.000     0.0   0.400  \n",
       "5918       0   0.000     0.0   0.000   0.000     0.0   0.400  \n",
       "5919       0   0.000     0.0   0.000   0.000     0.4   0.000  \n",
       "5920       0   0.000     0.0   0.083   0.000     0.0   0.372  \n",
       "5921       0   0.000     0.0   0.000   0.165     0.0   0.311  \n",
       "\n",
       "[5922 rows x 12 columns]"
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "instance.fetch_batch(\n",
    "    outfmt=\"json\",\n",
    "    method=\"interaction_partners\",\n",
    "    queries=[{\n",
    "        \"identifiers\": [\"p53\", \"cdk2\"],\n",
    "        \"species\": 9606,\n",
    "    }],\n",
    "    parse=True,\n",
    "    to_dataframe=True\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e13128ee",
   "metadata": {},
   "source": [
    "### Show graph test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "65e0ff5b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "✅ Grafo guardado en: mi_grafo.html\n"
     ]
    }
   ],
   "source": [
    "import json\n",
    "\n",
    "# Convierte a formato Cytoscape.js\n",
    "def convert_to_cytoscape_format(graph_json):\n",
    "    elements = []\n",
    "\n",
    "    for node in graph_json['nodes']:\n",
    "        elements.append({\n",
    "            \"data\": {\n",
    "                \"id\": node[\"id\"],\n",
    "                \"label\": node[\"lbl\"]\n",
    "            }\n",
    "        })\n",
    "\n",
    "    for edge in graph_json['edges']:\n",
    "        elements.append({\n",
    "            \"data\": {\n",
    "                \"source\": edge[\"sub\"],\n",
    "                \"target\": edge[\"obj\"],\n",
    "                \"label\": edge[\"pred\"]\n",
    "            }\n",
    "        })\n",
    "\n",
    "    return elements\n",
    "\n",
    "# Genera el HTML\n",
    "def create_cytoscape_html(graph_json):\n",
    "    elements = convert_to_cytoscape_format(graph_json)\n",
    "\n",
    "    html = f\"\"\"\n",
    "    <!DOCTYPE html>\n",
    "    <html>\n",
    "    <head>\n",
    "        <meta charset=\"utf-8\">\n",
    "        <title>Grafo GO Interactivo</title>\n",
    "        <script src=\"https://unpkg.com/cytoscape@3.19.0/dist/cytoscape.min.js\"></script>\n",
    "        <script src=\"https://unpkg.com/dagre@0.8.5/dist/dagre.min.js\"></script>\n",
    "        <script src=\"https://unpkg.com/cytoscape-dagre@2.3.2/cytoscape-dagre.js\"></script>\n",
    "        <style>\n",
    "            html, body {{\n",
    "                margin: 0;\n",
    "                padding: 0;\n",
    "                height: 100%;\n",
    "                width: 100%;\n",
    "                font-family: Arial, sans-serif;\n",
    "            }}\n",
    "            #cy {{\n",
    "                height: 100%;\n",
    "                width: 100%;\n",
    "                display: block;\n",
    "            }}\n",
    "        </style>\n",
    "    </head>\n",
    "    <body>\n",
    "        <div id=\"cy\"></div>\n",
    "        <script>\n",
    "            cytoscape.use(cytoscapeDagre);\n",
    "\n",
    "            var cy = cytoscape({{\n",
    "                container: document.getElementById('cy'),\n",
    "                elements: {json.dumps(elements)},\n",
    "                style: [\n",
    "                    {{\n",
    "                        selector: 'node',\n",
    "                        style: {{\n",
    "                            'shape': 'roundrectangle',\n",
    "                            'label': 'data(label)',\n",
    "                            'text-valign': 'center',\n",
    "                            'text-halign': 'center',\n",
    "                            'background-color': '#AED6F1',\n",
    "                            'color': '#1B2631',\n",
    "                            'font-size': '8px',\n",
    "                            'width': 'label',\n",
    "                            'height': 'label',\n",
    "                            'padding': '6px',\n",
    "                            'border-width': 1,\n",
    "                            'border-color': '#2980B9'\n",
    "                        }}\n",
    "                    }},\n",
    "                    {{\n",
    "                        selector: 'edge',\n",
    "                        style: {{\n",
    "                            'width': 2,\n",
    "                            'label': 'data(label)',\n",
    "                            'line-color': '#B2BABB',\n",
    "                            'target-arrow-color': '#B2BABB',\n",
    "                            'target-arrow-shape': 'triangle',\n",
    "                            'curve-style': 'bezier',\n",
    "                            'font-size': '7px',\n",
    "                            'color': '#5D6D7E',\n",
    "                            'text-background-opacity': 1,\n",
    "                            'text-background-color': '#fff',\n",
    "                            'text-background-shape': 'roundrectangle',\n",
    "                            'text-background-padding': 2\n",
    "                        }}\n",
    "                    }}\n",
    "                ],\n",
    "                layout: {{\n",
    "                    name: 'dagre',\n",
    "                    rankDir: 'TB',\n",
    "                    nodeSep: 30,\n",
    "                    edgeSep: 10,\n",
    "                    rankSep: 50\n",
    "                }},\n",
    "                zoomingEnabled: true,\n",
    "                userZoomingEnabled: true,\n",
    "                boxSelectionEnabled: false\n",
    "            }});\n",
    "        </script>\n",
    "    </body>\n",
    "    </html>\n",
    "    \"\"\"\n",
    "    return html\n",
    "\n",
    "# Guardar archivo HTML\n",
    "def save_graph_to_html(graph_json, output_file='cytoscape_graph.html'):\n",
    "    html = create_cytoscape_html(graph_json)\n",
    "    with open(output_file, 'w', encoding='utf-8') as f:\n",
    "        f.write(html)\n",
    "    print(f\"✅ Grafo guardado en: {output_file}\")\n",
    "\n",
    "# Ejecutar\n",
    "save_graph_to_html(instance.parse(response)[\"topology_graph_json\"], 'mi_grafo.html')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1c9b2b8a",
   "metadata": {},
   "source": [
    "# Uniprot + Databases union"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "55ee52a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import ast\n",
    "import os\n",
    "\n",
    "from dotenv import load_dotenv\n",
    "import pandas as pd \n",
    "from src.uniprot import UniprotInterface\n",
    "from src.alphafold import AlphafoldInterface\n",
    "from src.brenda import BrendaInstance\n",
    "from src.brenda import methods as brenda_methods\n",
    "from src.biogrid import BioGRIDInterface\n",
    "from src.interpro import InterproInstance\n",
    "from src.genontology import GenOntologyInterface\n",
    "from src.kegg import KEGGInterface\n",
    "from src.proteindatabank import PDBInterface\n",
    "from src.reactome import ReactomeInstance\n",
    "from src.refseq import RefSeqInterface\n",
    "from src.stringdb import StringInterface"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9e30a4e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "load_dotenv()\n",
    "\n",
    "# Retrieve the BioGRID API key\n",
    "biogrid_api_key = os.getenv(\"biogrid_api_key\")\n",
    "# Retrieve the Brenda email and password\n",
    "brenda_email = os.getenv(\"brenda_email\")\n",
    "brenda_password = os.getenv(\"brenda_password\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51fef622",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Argparse arguments\n",
    "output = \"results/output.csv\"\n",
    "#query = \"antimicrobial peptide\"\n",
    "query = \"organism_name:homo sapiens (human) AND length:[15 TO 30] AND reviewed:true\"\n",
    "fields = \"accession,protein_name,sequence,ec,lineage,organism_name,gene_primary\"\n",
    "out_db = \"brenda\"\n",
    "sort = \"accession asc\"\n",
    "fmt = \"json\"\n",
    "\n",
    "include_isoform = False\n",
    "download = False\n",
    "\n",
    "# Outside db argument\n",
    "outside_db = {\n",
    "    \"alphafold\": \"xref_alphafolddb\",\n",
    "    \"biogrid\": \"xref_biogrid\",\n",
    "    \"brenda\": \"xref_brenda\",\n",
    "    \"go\": \"go_id\",\n",
    "    \"interpro\": \"xref_interpro\",\n",
    "    \"kegg\": \"xref_kegg\",\n",
    "    \"pfam\": \"xref_pfam\",\n",
    "    \"pdb\": \"xref_pdb\",\n",
    "    \"reactome\": \"xref_reactome\",\n",
    "    \"refseq\": \"xref_refseq\",   \n",
    "    \"string\": \"xref_string\",\n",
    "}\n",
    "\n",
    "interfaces = {\n",
    "    \"alphafold\": AlphafoldInterface(\n",
    "        structures=['pdb'],\n",
    "        output_dir=\"results\"\n",
    "    ),\n",
    "    \"biogrid\": BioGRIDInterface(),\n",
    "    \"brenda\": BrendaInstance(\n",
    "        email=brenda_email,\n",
    "        password=brenda_password\n",
    "    ),\n",
    "    \"genontology\": GenOntologyInterface(),\n",
    "    \"interpro\": InterproInstance(),\n",
    "    \"kegg\": KEGGInterface(),\n",
    "    \"pdb\": PDBInterface(\n",
    "        download_structures=True,\n",
    "        return_data_list=[\"rcsb_id\", \"rcsb_comp_model_provenance\"],\n",
    "        output_dir=output.split(\"/\")[0]\n",
    "    ),\n",
    "    \"reactome\": ReactomeInstance(),\n",
    "    \"refseq\": RefSeqInterface(),\n",
    "    \"string\": StringInterface(),\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2877023d",
   "metadata": {},
   "source": [
    "### Download external db"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fcbad7ad",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query</th>\n",
       "      <th>accession</th>\n",
       "      <th>protein_name</th>\n",
       "      <th>organism_name</th>\n",
       "      <th>gene_primary</th>\n",
       "      <th>taxon_id</th>\n",
       "      <th>ineage</th>\n",
       "      <th>sequence</th>\n",
       "      <th>length</th>\n",
       "      <th>alphafold_ids</th>\n",
       "      <th>biogrid_ids</th>\n",
       "      <th>brenda_ids</th>\n",
       "      <th>go_terms</th>\n",
       "      <th>interpro_ids</th>\n",
       "      <th>kegg_ids</th>\n",
       "      <th>pdb_ids</th>\n",
       "      <th>pfam_ids</th>\n",
       "      <th>reactome_ids</th>\n",
       "      <th>refseq_ids</th>\n",
       "      <th>string_ids</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>A0A075B6S0</td>\n",
       "      <td>T cell receptor gamma joining 1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>['TRGJ1']</td>\n",
       "      <td>9606</td>\n",
       "      <td>['Eukaryota', 'Metazoa', 'Chordata', 'Craniata...</td>\n",
       "      <td>NYYKKLFGSGTTLVVT</td>\n",
       "      <td>16</td>\n",
       "      <td>['A0A075B6S0']</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>A0A075B6Y3</td>\n",
       "      <td>T cell receptor alpha joining 3</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>['TRAJ3']</td>\n",
       "      <td>9606</td>\n",
       "      <td>['Eukaryota', 'Metazoa', 'Chordata', 'Craniata...</td>\n",
       "      <td>GYSSASKIIFGSGTRLSIRP</td>\n",
       "      <td>20</td>\n",
       "      <td>['A0A075B6Y3']</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>A0A075B6Y9</td>\n",
       "      <td>T cell receptor alpha joining 42</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>['TRAJ42']</td>\n",
       "      <td>9606</td>\n",
       "      <td>['Eukaryota', 'Metazoa', 'Chordata', 'Craniata...</td>\n",
       "      <td>YGGSQGNLIFGKGTKLSVKP</td>\n",
       "      <td>20</td>\n",
       "      <td>['A0A075B6Y9']</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>A0A075B700</td>\n",
       "      <td>T cell receptor alpha joining 31</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>['TRAJ31']</td>\n",
       "      <td>9606</td>\n",
       "      <td>['Eukaryota', 'Metazoa', 'Chordata', 'Craniata...</td>\n",
       "      <td>NNNARLMFGDGTQLVVKP</td>\n",
       "      <td>18</td>\n",
       "      <td>['A0A075B700']</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>organism_name:homo sapiens (human) AND length:...</td>\n",
       "      <td>A0A075B706</td>\n",
       "      <td>T cell receptor delta joining 1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>['TRDJ1']</td>\n",
       "      <td>9606</td>\n",
       "      <td>['Eukaryota', 'Metazoa', 'Chordata', 'Craniata...</td>\n",
       "      <td>TDKLIFGKGTRVTVEP</td>\n",
       "      <td>16</td>\n",
       "      <td>['A0A075B706']</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               query   accession  \\\n",
       "0  organism_name:homo sapiens (human) AND length:...  A0A075B6S0   \n",
       "1  organism_name:homo sapiens (human) AND length:...  A0A075B6Y3   \n",
       "2  organism_name:homo sapiens (human) AND length:...  A0A075B6Y9   \n",
       "3  organism_name:homo sapiens (human) AND length:...  A0A075B700   \n",
       "4  organism_name:homo sapiens (human) AND length:...  A0A075B706   \n",
       "\n",
       "                       protein_name organism_name gene_primary  taxon_id  \\\n",
       "0   T cell receptor gamma joining 1  Homo sapiens    ['TRGJ1']      9606   \n",
       "1   T cell receptor alpha joining 3  Homo sapiens    ['TRAJ3']      9606   \n",
       "2  T cell receptor alpha joining 42  Homo sapiens   ['TRAJ42']      9606   \n",
       "3  T cell receptor alpha joining 31  Homo sapiens   ['TRAJ31']      9606   \n",
       "4   T cell receptor delta joining 1  Homo sapiens    ['TRDJ1']      9606   \n",
       "\n",
       "                                              ineage              sequence  \\\n",
       "0  ['Eukaryota', 'Metazoa', 'Chordata', 'Craniata...      NYYKKLFGSGTTLVVT   \n",
       "1  ['Eukaryota', 'Metazoa', 'Chordata', 'Craniata...  GYSSASKIIFGSGTRLSIRP   \n",
       "2  ['Eukaryota', 'Metazoa', 'Chordata', 'Craniata...  YGGSQGNLIFGKGTKLSVKP   \n",
       "3  ['Eukaryota', 'Metazoa', 'Chordata', 'Craniata...    NNNARLMFGDGTQLVVKP   \n",
       "4  ['Eukaryota', 'Metazoa', 'Chordata', 'Craniata...      TDKLIFGKGTRVTVEP   \n",
       "\n",
       "   length   alphafold_ids biogrid_ids brenda_ids go_terms interpro_ids  \\\n",
       "0      16  ['A0A075B6S0']          []         []       []           []   \n",
       "1      20  ['A0A075B6Y3']          []         []       []           []   \n",
       "2      20  ['A0A075B6Y9']          []         []       []           []   \n",
       "3      18  ['A0A075B700']          []         []       []           []   \n",
       "4      16  ['A0A075B706']          []         []       []           []   \n",
       "\n",
       "  kegg_ids pdb_ids pfam_ids reactome_ids refseq_ids string_ids  \n",
       "0       []      []       []           []         []         []  \n",
       "1       []      []       []           []         []         []  \n",
       "2       []      []       []           []         []         []  \n",
       "3       []      []       []           []         []         []  \n",
       "4       []      []       []           []         []         []  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "export_df = pd.read_csv(\"results/output.csv\")\n",
    "export_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cf359a1b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_ids(df, db_name):\n",
    "    raw_ids = df.get(f\"{db_name}_ids\", pd.Series(dtype=str)).dropna().unique().tolist()\n",
    "    ids = []\n",
    "    for entry in raw_ids:\n",
    "        parsed = ast.literal_eval(entry) if isinstance(entry, str) and entry.startswith(\"[\") else [entry]\n",
    "        ids.extend(parsed)\n",
    "\n",
    "    # Cleaning: remove \"[]\"\n",
    "    return [id for id in ids if id != \"[]\"]\n",
    "\n",
    "# Fetch data from alphafold given a ID from xref_alphafolddb\n",
    "# Those without alphafold_ids are skipped\n",
    "def fetch_alphafold(df):\n",
    "    instance = AlphafoldInterface(\n",
    "        structures=['pdb'],\n",
    "        output_dir=\"results\"\n",
    "    )\n",
    "\n",
    "    ids = parse_ids(df, \"alphafold\")\n",
    "    if not ids:\n",
    "        return []\n",
    "    # TODO: Remove Limit\n",
    "    print(f\"{len(ids)} IDs to fetch from alphafold\")\n",
    "    return instance.fetch_batch(queries=ids[:4], parse=True)\n",
    "\n",
    "# Fetch data from biogrid given a list of IDs from xref_biogrid or gene_primary and taxon_id fields\n",
    "def fetch_biogrid(df):\n",
    "    instance = BioGRIDInterface()\n",
    "    queries = []\n",
    "\n",
    "    # Part 1: by ID if they exist\n",
    "    ids = parse_ids(df, \"biogrid\")\n",
    "    if ids:\n",
    "        queries.extend([{\n",
    "            \"id\": id_,\n",
    "            \"accessKey\": biogrid_api_key\n",
    "        } for id_ in ids])\n",
    "\n",
    "    # Part 2: build queries by taxId and geneList\n",
    "    tmp_df = df[df[\"biogrid_ids\"].astype(str).isin([\"[]\", \"nan\", \"NaN\", \"\"])]\n",
    "    tmp_df = tmp_df.dropna(subset=[\"gene_primary\", \"taxon_id\"])\n",
    "    tmp_df = tmp_df[[\"gene_primary\", \"taxon_id\"]].drop_duplicates()\n",
    "\n",
    "    tmp_df[\"gene_primary\"] = tmp_df[\"gene_primary\"].apply(\n",
    "        lambda x: ast.literal_eval(x) if isinstance(x, str) and x.startswith(\"[\") else [x]\n",
    "    )\n",
    "\n",
    "    tmp_df = tmp_df.explode(\"gene_primary\")\n",
    "    grouped = tmp_df.groupby(\"taxon_id\")[\"gene_primary\"].agg(list).reset_index()\n",
    "\n",
    "    queries.extend([\n",
    "        {\n",
    "            \"accessKey\": biogrid_api_key,\n",
    "            \"geneList\": row[\"gene_primary\"],\n",
    "            \"taxId\": row[\"taxon_id\"],\n",
    "            \"format\": \"json\"\n",
    "        }\n",
    "        for _, row in grouped.iterrows()\n",
    "    ])\n",
    "\n",
    "    print(f\"{len(queries)} queries to fetch from biogrid\")\n",
    "\n",
    "    return instance.fetch_batch(\n",
    "        queries=queries,\n",
    "        method=\"interactions\",\n",
    "        parse=True\n",
    "    )\n",
    "\n",
    "# Those without brenda_ids are skipped\n",
    "def fetch_brenda(df, method: str):\n",
    "    \n",
    "    instance = BrendaInstance(\n",
    "        email=brenda_email,\n",
    "        password=brenda_password\n",
    "    )\n",
    "    queries = []\n",
    "    results = []\n",
    "\n",
    "    ids = df[~df[\"brenda_ids\"].astype(str).isin([\"[]\", \"nan\", \"NaN\", \"\"])]\n",
    "    ids.loc[:, 'brenda_ids'] = ids[\"brenda_ids\"].apply(\n",
    "        lambda x: ast.literal_eval(x) if isinstance(x, str) and x.startswith(\"[\") else [x]\n",
    "    )\n",
    "    ids = ids.explode(\"brenda_ids\").drop_duplicates(subset=[\"brenda_ids\"])\n",
    "    # TODO Remove Limit\n",
    "    ids = ids[:4]  # Limit to 4 for testing purposes\n",
    "\n",
    "    if not ids.empty:\n",
    "        queries.extend([{\n",
    "            \"ecNumber\": id[\"brenda_ids\"],\n",
    "            \"organism\": id[\"organism_name\"]\n",
    "        } for _, id in ids.iterrows() if isinstance(id[\"brenda_ids\"], str) and id[\"brenda_ids\"] != \"[]\"])\n",
    "\n",
    "    results.append(\n",
    "        {\n",
    "            \"method\": method,\n",
    "            \"data\": instance.fetch_batch(\n",
    "                queries=queries,\n",
    "                method=method,\n",
    "                parse=True\n",
    "                to_dataframe=True\n",
    "            )\n",
    "        })\n",
    "    \n",
    "    return results\n",
    "\n",
    "# Those without go_terms are skipped\n",
    "# TODO: Retorna una lista anidada y por eso no puedo hacerlo un dataframe ver que pasa\n",
    "def fetch_genontology(df):\n",
    "    queries = []\n",
    "    ids = df[\"go_terms\"].dropna().unique().tolist()\n",
    "    ids = [ast.literal_eval(id) if isinstance(id, str) and id.startswith(\"[\") else [id] for id in ids]\n",
    "    ids = [item for sublist in ids for item in sublist]  # Flatten\n",
    "    ids = list(set(ids))  # Remove duplicates\n",
    "\n",
    "    # TODO Remove Limit\n",
    "    ids = ids[:4]  # Limit to 4 for testing purposes\n",
    "    \n",
    "    # TODO It can be also added more methods to fetch\n",
    "    return interface.fetch_batch(\n",
    "        method=\"ontology-term\",\n",
    "        queries=ids,\n",
    "        option=None,\n",
    "        parse=True\n",
    "    )\n",
    "\n",
    "def fetch_interpro(df):\n",
    "    queries = []\n",
    "\n",
    "    ids = df[~df[\"interpro_ids\"].astype(str).isin([\"[]\", \"nan\", \"NaN\", \"\"])]\n",
    "    ids.loc[:, 'interpro_ids'] = ids[\"interpro_ids\"].apply(\n",
    "        lambda x: ast.literal_eval(x) if isinstance(x, str) and x.startswith(\"[\") else [x]\n",
    "    )\n",
    "    ids = ids.explode(\"interpro_ids\").drop_duplicates(subset=[\"interpro_ids\"])\n",
    "\n",
    "    if not ids.empty:\n",
    "        queries.extend([{\n",
    "            \"id\": id[\"interpro_ids\"],\n",
    "            \"db\": \"InterPro\",\n",
    "            \"filters\": {\n",
    "                \"type\": \"protein\",\n",
    "                \"db\": \"reviewed\",\n",
    "                \"value\": id[\"accession\"]\n",
    "            }\n",
    "        } for _, id in ids.iterrows() if isinstance(id[\"interpro_ids\"], str) and id[\"interpro_ids\"] != \"[]\"])\n",
    "\n",
    "\n",
    "    tmp_df = df[df[\"interpro_ids\"].astype(str).isin([\"[]\", \"nan\", \"NaN\", \"\"])]\n",
    "    tmp_df = df[[\"accession\", \"taxon_id\"]].drop_duplicates()\n",
    "    tmp_df = tmp_df.dropna(subset=[\"accession\", \"taxon_id\"])\n",
    "\n",
    "    if not tmp_df.empty:\n",
    "        queries.extend([\n",
    "            {\n",
    "                \"db\": \"InterPro\",\n",
    "                \"filters\": [\n",
    "                    {\n",
    "                        \"type\": \"protein\",\n",
    "                        \"db\": \"reviewed\",\n",
    "                        \"value\": row[\"accession\"]\n",
    "                    },\n",
    "                    {\n",
    "                        \"type\": \"taxonomy\",\n",
    "                        \"db\": \"uniprot\",\n",
    "                        \"value\": row[\"taxon_id\"]\n",
    "                    }\n",
    "                ]\n",
    "\n",
    "            }\n",
    "            for _, row in tmp_df.iterrows()\n",
    "        ])\n",
    "    \n",
    "    print(len(queries), \"queries to fetch from interpro\")\n",
    "\n",
    "    return interface.fetch_batch(\n",
    "        queries=queries,\n",
    "        method=\"entry\",\n",
    "        pages_to_fetch=1,\n",
    "        parse=True\n",
    "    )\n",
    "\n",
    "# Those without kegg_ids are skipped\n",
    "def fetch_kegg(df):\n",
    "    queries = []\n",
    "    ids = parse_ids(df, \"kegg\")\n",
    "\n",
    "    if not ids:\n",
    "        return []\n",
    "    \n",
    "    for id in ids:\n",
    "        queries.append({\n",
    "            \"entries\": [id]\n",
    "        })\n",
    "\n",
    "    return interface.fetch_batch(\n",
    "        method=\"get\",\n",
    "        queries=queries[:4],  # TODO Remove Limit\n",
    "        parse=True\n",
    "    )\n",
    "\n",
    "# TODO: PDB retorna una lista anidada y por eso no puedo hacerlo un dataframe ver que pasa\n",
    "def fetch_pdb(df):\n",
    "    queries = []\n",
    "    ids = parse_ids(df, \"pdb\")\n",
    "    \n",
    "    if not ids:\n",
    "        return []\n",
    "    \n",
    "    queries = ids[:4]  # TODO Remove Limit\n",
    "\n",
    "    return interface.fetch_batch(\n",
    "        queries=queries,\n",
    "        parse=True\n",
    "    )\n",
    "\n",
    "def fetch_reactome(df):\n",
    "    queries = []\n",
    "    ids = parse_ids(df, \"reactome\")\n",
    "\n",
    "    if not ids:\n",
    "        return []\n",
    "\n",
    "    return interface.fetch_batch(\n",
    "        queries=ids[:4],  # TODO Remove Limit\n",
    "        parse=True,\n",
    "        method=\"data/discover\",\n",
    "        option=\"\"\n",
    "    )\n",
    "\n",
    "def fetch_refseq(df):\n",
    "    queries = []\n",
    "    ids = parse_ids(df, \"refseq\")\n",
    "\n",
    "    if not ids:\n",
    "        return []\n",
    "\n",
    "    # TODO Remove Limit\n",
    "    queries = ids[:4]\n",
    "\n",
    "    return interface.fetch_batch(\n",
    "        queries=queries,\n",
    "        method=\"protein\",\n",
    "        parse=True\n",
    "    )\n",
    "\n",
    "# No string_ids this time, only gene_primary and taxon_id\n",
    "# TODO make sure string_ids are working\n",
    "def fetch_string(df):\n",
    "    queries = []\n",
    "    \n",
    "    # 1. Fetch by string_ids if they exist\n",
    "    tmp_df = df[~df[\"string_ids\"].astype(str).isin([\"[]\", \"nan\", \"NaN\", \"\"])]\n",
    "    if not tmp_df.empty:\n",
    "        tmp_df = tmp_df.dropna(subset=[\"string_ids\"])\n",
    "        tmp_df = tmp_df[[\"string_ids\", \"taxon_id\"]].drop_duplicates\n",
    "        tmp_df[\"string_ids\"] = tmp_df[\"string_ids\"].apply(\n",
    "            lambda x: ast.literal_eval(x) if isinstance(x, str) and x.startswith(\"[\") else [x]\n",
    "        )\n",
    "        tmp_df = tmp_df.explode(\"string_ids\")\n",
    "        grouped = tmp_df.groupby(\"taxon_id\")[\"string_ids\"].agg(list).reset_index()\n",
    "        queries.extend([\n",
    "            {\n",
    "                \"identifiers\": row[\"string_ids\"],\n",
    "                \"species\": row[\"taxon_id\"]\n",
    "            }\n",
    "            for _, row in grouped.iterrows()\n",
    "        ])\n",
    "    \n",
    "    # 2. Build queries by taxon_id and gene_primary\n",
    "    tmp_df = df[df[\"string_ids\"].astype(str).isin([\"[]\", \"nan\", \"NaN\", \"\"])]\n",
    "    tmp_df = tmp_df.dropna(subset=[\"gene_primary\", \"taxon_id\"])\n",
    "    tmp_df = tmp_df[[\"gene_primary\", \"taxon_id\"]].drop_duplicates()\n",
    "\n",
    "    tmp_df[\"gene_primary\"] = tmp_df[\"gene_primary\"].apply(\n",
    "        lambda x: ast.literal_eval(x) if isinstance(x, str) and x.startswith(\"[\") else [x]\n",
    "    )\n",
    "\n",
    "    tmp_df = tmp_df.explode(\"gene_primary\")\n",
    "    grouped = tmp_df.groupby(\"taxon_id\")[\"gene_primary\"].agg(list).reset_index()\n",
    "\n",
    "    queries.extend([\n",
    "        {\n",
    "            \"identifiers\": row[\"gene_primary\"],\n",
    "            \"species\": row[\"taxon_id\"]\n",
    "        }\n",
    "        for _, row in grouped.iterrows()\n",
    "    ])\n",
    "\n",
    "    return interface.fetch_batch(\n",
    "        outfmt=\"json\",\n",
    "        method=\"interaction_partners\",\n",
    "        queries=queries[:4],  # TODO Remove Limit\n",
    "        parse=True\n",
    "    )\n",
    "\n",
    "# TODO Add a parameter to convert to dataframe.\n",
    "# fetch_batch(..., parse=True, to_dataframe=True)\n",
    "\n",
    "def fetch_from_external(db_name, df):\n",
    "    match db_name:\n",
    "        case \"alphafold\":\n",
    "            return fetch_alphafold(df)\n",
    "        case \"biogrid\":\n",
    "            return fetch_biogrid(df)\n",
    "        case brenda if brenda.startswith(\"brenda_\"):\n",
    "            print(f\"aa\")\n",
    "            method = brenda.split(\"_\", 1)[1]\n",
    "            return fetch_brenda(df, method)\n",
    "        case \"genontology\":\n",
    "            return fetch_genontology(df)\n",
    "        case \"interpro\":\n",
    "            return fetch_interpro(df)\n",
    "        case \"kegg\":\n",
    "            return fetch_kegg(df)\n",
    "        case \"pdb\":\n",
    "            return fetch_pdb(df)\n",
    "        case \"reactome\":\n",
    "            return fetch_reactome(df)\n",
    "        case \"refseq\":\n",
    "            return fetch_refseq(df)\n",
    "        case \"string\":\n",
    "            return fetch_string(df)\n",
    "        case _:\n",
    "            raise ValueError(f\"Unsupported database: {db_name}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "824c9000",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetching data from brenda_getKmValue...\n",
      "aa\n",
      "Fetching data from brenda_getIc50Value...\n",
      "aa\n",
      "Fetching data from brenda_getKcatKmValue...\n",
      "aa\n",
      "Fetching data from brenda_getKiValue...\n",
      "aa\n",
      "Fetching data from brenda_getPhRange...\n",
      "aa\n",
      "Fetching data from brenda_getPhOptimum...\n",
      "aa\n",
      "Fetching data from brenda_getPhStability...\n",
      "aa\n",
      "Fetching data from brenda_getCofactor...\n",
      "aa\n",
      "Fetching data from brenda_getTemperatureOptimum...\n",
      "aa\n",
      "Fetching data from brenda_getTemperatureStability...\n",
      "aa\n",
      "Fetching data from brenda_getTemperatureRange...\n",
      "aa\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'brenda_getKmValue': [{'method': 'getKmValue', 'data': []}],\n",
       " 'brenda_getIc50Value': [{'method': 'getIc50Value', 'data': []}],\n",
       " 'brenda_getKcatKmValue': [{'method': 'getKcatKmValue', 'data': []}],\n",
       " 'brenda_getKiValue': [{'method': 'getKiValue', 'data': []}],\n",
       " 'brenda_getPhRange': [{'method': 'getPhRange', 'data': []}],\n",
       " 'brenda_getPhOptimum': [{'method': 'getPhOptimum', 'data': []}],\n",
       " 'brenda_getPhStability': [{'method': 'getPhStability', 'data': []}],\n",
       " 'brenda_getCofactor': [{'method': 'getCofactor', 'data': []}],\n",
       " 'brenda_getTemperatureOptimum': [{'method': 'getTemperatureOptimum',\n",
       "   'data': []}],\n",
       " 'brenda_getTemperatureStability': [{'method': 'getTemperatureStability',\n",
       "   'data': []}],\n",
       " 'brenda_getTemperatureRange': [{'method': 'getTemperatureRange', 'data': []}]}"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "export_dfs = {}\n",
    "databases = out_db.split(\",\")\n",
    "\n",
    "if \"biogrid\" in databases:\n",
    "    biogrid_api_key = os.getenv(\"biogrid_api_key\")\n",
    "    if not biogrid_api_key:\n",
    "        raise ValueError(\"Please set the 'biogrid_api_key' environment variable.\")\n",
    "    \n",
    "\n",
    "if \"brenda\" in databases:\n",
    "    databases.remove(\"brenda\")\n",
    "    # Add all brenda methods to the databases list\n",
    "    databases.extend([f\"brenda_{method}\" for method in brenda_methods.keys()])\n",
    "    if not brenda_email or not brenda_password:\n",
    "        raise ValueError(\"Please set the 'brenda_email' and 'brenda_password' environment variables.\")\n",
    "\n",
    "for db in databases:\n",
    "    print(f\"Fetching data from {db}...\")\n",
    "    test = fetch_from_external(db, export_df)\n",
    "    export_dfs[db] = test\n",
    "\n",
    "export_dfs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "724cf711",
   "metadata": {},
   "source": [
    "# Interface"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3622f012",
   "metadata": {},
   "outputs": [],
   "source": [
    "import ipywidgets as widgets\n",
    "from IPython.display import display, JSON, clear_output\n",
    "from src.interpro import InterproInstance, data_types, entry_integration_types, filter_types, db_types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab93d194",
   "metadata": {},
   "outputs": [],
   "source": [
    "filters = {\n",
    "    \"protein\": [(\"UniProtKB/Swiss-Prot\", \"reviewed\"),\n",
    "                (\"UniProtKB\", \"UniProt\"),\n",
    "                (\"UniProtKB/TrEMBL\", \"unreviewed\")],\n",
    "    \"structure\": [(\"PDB\", \"PDB\")],\n",
    "    \"taxonomy\": [(\"UniProtKB\", \"uniprot\")],\n",
    "    \"proteome\": [(\"UniProtKB\", \"uniprot\")],\n",
    "    \"set\": [(\"CDD\", \"cdd\"),\n",
    "            (\"Pfam\", \"pfam\"),\n",
    "            (\"PIRSF\", \"pirsf\")],\n",
    "}\n",
    "\n",
    "def interface_interpro():\n",
    "    output = widgets.Output()\n",
    "    dynamic_output = widgets.Output()\n",
    "    active_sections = {}\n",
    "\n",
    "    # ------------------------\n",
    "    # Sección de configuración base\n",
    "    # ------------------------\n",
    "\n",
    "    method_dropdown = widgets.Dropdown(\n",
    "        options=data_types,\n",
    "        value=\"entry\",\n",
    "        description=\"Choose a main data type:\"\n",
    "    )\n",
    "\n",
    "    db_dropdown = widgets.Dropdown(\n",
    "        options=db_types[\"entry\"],\n",
    "        value=\"InterPro\",\n",
    "        description=\"DB:\"\n",
    "    )\n",
    "\n",
    "    # Actualiza db_dropdown en base a method_dropdown\n",
    "    def update_db_options(change):\n",
    "        new_type = change['new']\n",
    "        db_dropdown.options = db_types.get(new_type, [])\n",
    "\n",
    "    method_dropdown.observe(update_db_options, names='value')\n",
    "\n",
    "    # ------------------------\n",
    "    # Sección de filtros dinámicos\n",
    "    # ------------------------\n",
    "\n",
    "    container = widgets.VBox()\n",
    "    button_box = widgets.HBox()\n",
    "    add_buttons = {}\n",
    "\n",
    "    # Crear sección individual de filtro\n",
    "    def create_filter_section(filter_type):\n",
    "        options = filters[filter_type]\n",
    "        labels = [label for label, _ in options]\n",
    "        label_to_value = {label: val for label, val in options}\n",
    "\n",
    "        db_dropdown_f = widgets.Dropdown(\n",
    "            options=labels,\n",
    "            description=f\"{filter_type} DB:\"\n",
    "        )\n",
    "\n",
    "        value_input = widgets.Text(\n",
    "            placeholder=f\"{filter_type} accession...\",\n",
    "            description=f\"{filter_type} accession:\"\n",
    "        )\n",
    "\n",
    "        def clear_value(b):\n",
    "            value_input.value = ''\n",
    "\n",
    "        clear_btn = widgets.Button(description=\"Clear\", button_style='info')\n",
    "        clear_btn.on_click(clear_value)\n",
    "\n",
    "        def remove_section(b):\n",
    "            container.children = [c for c in container.children if c != section]\n",
    "            add_buttons[filter_type].disabled = False\n",
    "            del active_sections[filter_type]\n",
    "\n",
    "        remove_btn = widgets.Button(description=\"Remove\", button_style='danger')\n",
    "        remove_btn.on_click(remove_section)\n",
    "\n",
    "        section = widgets.VBox([\n",
    "            widgets.HTML(f\"<b>{filter_type}</b>\"),\n",
    "            widgets.HBox([widgets.Label(\"Filter type:\"), widgets.Label(filter_type)]),\n",
    "            widgets.HBox([db_dropdown_f, clear_btn]),\n",
    "            value_input,\n",
    "            remove_btn\n",
    "        ])\n",
    "        active_sections[filter_type] = (db_dropdown_f, value_input, label_to_value)\n",
    "        return section\n",
    "\n",
    "    # Botones de filtros\n",
    "    for key in filters:\n",
    "        btn = widgets.Button(description=key.capitalize(), button_style='info', layout=widgets.Layout(width='auto'))\n",
    "        add_buttons[key] = btn\n",
    "\n",
    "        def make_handler(k):\n",
    "            def handler(b):\n",
    "                section = create_filter_section(k)\n",
    "                container.children += (section,)\n",
    "                add_buttons[k].disabled = True\n",
    "            return handler\n",
    "\n",
    "        btn.on_click(make_handler(key))\n",
    "\n",
    "    button_box.children = list(add_buttons.values())\n",
    "\n",
    "    # ------------------------\n",
    "    # Botón para generar query\n",
    "    # ------------------------\n",
    "\n",
    "    generate_button = widgets.Button(\n",
    "        description=\"Generar Query\",\n",
    "        button_style=\"success\",\n",
    "        icon=\"check\"\n",
    "    )\n",
    "\n",
    "    def on_generate_click(b):\n",
    "        with output:\n",
    "            clear_output()\n",
    "            query = {\n",
    "                \"db\": db_dropdown.value,\n",
    "                \"filters\": []\n",
    "            }\n",
    "            for ftype, (db_dd, val_input, label_map) in active_sections.items():\n",
    "                db_label = db_dd.value\n",
    "                db_value = label_map[db_label]\n",
    "                value = val_input.value.strip()\n",
    "                if value:\n",
    "                    query[\"filters\"].append({\n",
    "                        \"type\": ftype,\n",
    "                        \"db\": db_value,\n",
    "                        \"value\": value\n",
    "                    })\n",
    "            display(JSON(query, expanded=True))\n",
    "            # Aquí puedes llamar a la función de InterPro para enviar la consulta\n",
    "            interpro_instance = InterproInstance()\n",
    "            response = interpro_instance.fetch_single(\n",
    "                query=query,\n",
    "                method=method_dropdown.value,\n",
    "                pages_to_fetch=1,\n",
    "                parse=True\n",
    "            )\n",
    "            if response:\n",
    "                display(JSON(response, expanded=True))\n",
    "            else:\n",
    "                print(\"No response received from InterPro API.\")\n",
    "\n",
    "    generate_button.on_click(on_generate_click)\n",
    "\n",
    "    # ------------------------\n",
    "    # Render\n",
    "    # ------------------------\n",
    "\n",
    "    display(widgets.VBox([\n",
    "        widgets.HTML(\"<h3>InterPro Query Interface</h3>\"),\n",
    "        widgets.HTML(\"<b>Configuración principal:</b>\"),\n",
    "        method_dropdown,\n",
    "        db_dropdown,\n",
    "        widgets.HTML(\"<hr><b>Filtros dinámicos:</b>\"),\n",
    "        widgets.Label(\"Agregar un filtro:\"),\n",
    "        button_box,\n",
    "        container,\n",
    "        generate_button,\n",
    "        output\n",
    "    ]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "075aa410",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "9bff09a45cf8480e98198aa0d24320d4",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "VBox(children=(HTML(value='<h3>InterPro Query Interface</h3>'), HTML(value='<b>Configuración principal:</b>'),…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "interface_interpro()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ffe9f398",
   "metadata": {},
   "source": [
    "# Random Test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bb04dada",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import ast"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a494d073",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(67027, 21)"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv(\"test.csv\")\n",
    "df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6622ca8d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query</th>\n",
       "      <th>accession</th>\n",
       "      <th>protein_name</th>\n",
       "      <th>ec_numbers</th>\n",
       "      <th>organism_name</th>\n",
       "      <th>gene_primary</th>\n",
       "      <th>taxon_id</th>\n",
       "      <th>ineage</th>\n",
       "      <th>sequence</th>\n",
       "      <th>length</th>\n",
       "      <th>...</th>\n",
       "      <th>biogrid_ids</th>\n",
       "      <th>brenda_ids</th>\n",
       "      <th>go_terms</th>\n",
       "      <th>interpro_ids</th>\n",
       "      <th>kegg_ids</th>\n",
       "      <th>pdb_ids</th>\n",
       "      <th>pfam_ids</th>\n",
       "      <th>reactome_ids</th>\n",
       "      <th>refseq_ids</th>\n",
       "      <th>string_ids</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>19934</th>\n",
       "      <td>(go:DNA binding OR go:ATP binding OR keyword:A...</td>\n",
       "      <td>O55076</td>\n",
       "      <td>Cyclin-dependent kinase 2</td>\n",
       "      <td>['2.7.11.22']</td>\n",
       "      <td>Cricetulus griseus</td>\n",
       "      <td>CDK2</td>\n",
       "      <td>10029</td>\n",
       "      <td>['Eukaryota', 'Metazoa', 'Chordata', 'Craniata...</td>\n",
       "      <td>MENFQKVEKIGEGTYGVVYKAKNKLTGEVVALKKIRLDTETEGVPS...</td>\n",
       "      <td>298</td>\n",
       "      <td>...</td>\n",
       "      <td>[]</td>\n",
       "      <td>['2.7.11.22']</td>\n",
       "      <td>['GO:0015030', 'GO:0005813', 'GO:0000307', 'GO...</td>\n",
       "      <td>['IPR050108', 'IPR011009', 'IPR000719', 'IPR01...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>['PF00069']</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26084</th>\n",
       "      <td>(go:DNA binding OR go:ATP binding OR keyword:A...</td>\n",
       "      <td>P20248</td>\n",
       "      <td>Cyclin-A2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>CCNA2</td>\n",
       "      <td>9606</td>\n",
       "      <td>['Eukaryota', 'Metazoa', 'Chordata', 'Craniata...</td>\n",
       "      <td>MLGNSAPGPATREAGSALLALQQTALQEDQENINPEKAAPVQQPRT...</td>\n",
       "      <td>432</td>\n",
       "      <td>...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>['GO:0097122', 'GO:0097124', 'GO:0000307', 'GO...</td>\n",
       "      <td>['IPR039361', 'IPR032447', 'IPR013763', 'IPR03...</td>\n",
       "      <td>['hsa:890']</td>\n",
       "      <td>['1E9H', '1FIN', '1FVV', '1GY3', '1H1P', '1H1Q...</td>\n",
       "      <td>['PF02984', 'PF00134', 'PF16500']</td>\n",
       "      <td>['R-HSA-1362300', 'R-HSA-1538133', 'R-HSA-1701...</td>\n",
       "      <td>['NP_001228.2']</td>\n",
       "      <td>['9606.ENSP00000481380']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26805</th>\n",
       "      <td>(go:DNA binding OR go:ATP binding OR keyword:A...</td>\n",
       "      <td>P24941</td>\n",
       "      <td>Cyclin-dependent kinase 2</td>\n",
       "      <td>['2.7.11.22']</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>CDK2</td>\n",
       "      <td>9606</td>\n",
       "      <td>['Eukaryota', 'Metazoa', 'Chordata', 'Craniata...</td>\n",
       "      <td>MENFQKVEKIGEGTYGVVYKARNKLTGEVVALKKIRLDTETEGVPS...</td>\n",
       "      <td>298</td>\n",
       "      <td>...</td>\n",
       "      <td>[]</td>\n",
       "      <td>['2.7.11.22']</td>\n",
       "      <td>['GO:0015030', 'GO:0005813', 'GO:0000781', 'GO...</td>\n",
       "      <td>['IPR050108', 'IPR011009', 'IPR000719', 'IPR01...</td>\n",
       "      <td>['hsa:1017']</td>\n",
       "      <td>['1AQ1', '1B38', '1B39', '1BUH', '1CKP', '1DI8...</td>\n",
       "      <td>['PF00069']</td>\n",
       "      <td>['R-HSA-1538133', 'R-HSA-171319', 'R-HSA-17618...</td>\n",
       "      <td>['NP_001277159.1', 'NP_001789.2', 'NP_439892.2']</td>\n",
       "      <td>['9606.ENSP00000266970']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27540</th>\n",
       "      <td>(go:DNA binding OR go:ATP binding OR keyword:A...</td>\n",
       "      <td>P30274</td>\n",
       "      <td>Cyclin-A2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bos taurus</td>\n",
       "      <td>CCNA2</td>\n",
       "      <td>9913</td>\n",
       "      <td>['Eukaryota', 'Metazoa', 'Chordata', 'Craniata...</td>\n",
       "      <td>MLGSSAHGPAAREAGSAVTLQQTAFQEDQENVNPEKAAPAQQPRTR...</td>\n",
       "      <td>430</td>\n",
       "      <td>...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>['GO:0097124', 'GO:0000307', 'GO:0005737', 'GO...</td>\n",
       "      <td>['IPR039361', 'IPR032447', 'IPR013763', 'IPR03...</td>\n",
       "      <td>['bta:281667']</td>\n",
       "      <td>['1VIN', '2G9X', '3BHT', '3BHU', '3BHV', '3DDP...</td>\n",
       "      <td>['PF02984', 'PF00134', 'PF16500']</td>\n",
       "      <td>[]</td>\n",
       "      <td>['NP_001068591.1']</td>\n",
       "      <td>['9913.ENSBTAP00000006503']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28601</th>\n",
       "      <td>(go:DNA binding OR go:ATP binding OR keyword:A...</td>\n",
       "      <td>P37881</td>\n",
       "      <td>Cyclin-A2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Mesocricetus auratus</td>\n",
       "      <td>CCNA2</td>\n",
       "      <td>10036</td>\n",
       "      <td>['Eukaryota', 'Metazoa', 'Chordata', 'Craniata...</td>\n",
       "      <td>MPGSSRQSGREAGSALLSLQQEDQENVNPEKAAPDQRARAALKTGN...</td>\n",
       "      <td>421</td>\n",
       "      <td>...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>['GO:0097122', 'GO:0097124', 'GO:0000307', 'GO...</td>\n",
       "      <td>['IPR039361', 'IPR032447', 'IPR013763', 'IPR03...</td>\n",
       "      <td>['maua:101835930']</td>\n",
       "      <td>[]</td>\n",
       "      <td>['PF02984', 'PF00134', 'PF16500']</td>\n",
       "      <td>[]</td>\n",
       "      <td>['NP_001268563.1']</td>\n",
       "      <td>['10036.ENSMAUP00000020143']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30029</th>\n",
       "      <td>(go:DNA binding OR go:ATP binding OR keyword:A...</td>\n",
       "      <td>P48963</td>\n",
       "      <td>Cyclin-dependent kinase 2</td>\n",
       "      <td>['2.7.11.22']</td>\n",
       "      <td>Mesocricetus auratus</td>\n",
       "      <td>CDK2</td>\n",
       "      <td>10036</td>\n",
       "      <td>['Eukaryota', 'Metazoa', 'Chordata', 'Craniata...</td>\n",
       "      <td>MENFQKVEKIGEGTYGVVYKAKNKLTGEVVALKKIRLDTETEGVPS...</td>\n",
       "      <td>298</td>\n",
       "      <td>...</td>\n",
       "      <td>[]</td>\n",
       "      <td>['2.7.11.22']</td>\n",
       "      <td>['GO:0015030', 'GO:0005813', 'GO:0000307', 'GO...</td>\n",
       "      <td>['IPR050108', 'IPR011009', 'IPR000719', 'IPR01...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>['PF00069']</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>['10036.ENSMAUP00000021672']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46192</th>\n",
       "      <td>(go:DNA binding OR go:ATP binding OR keyword:A...</td>\n",
       "      <td>Q5E9Y0</td>\n",
       "      <td>Cyclin-dependent kinase 2</td>\n",
       "      <td>['2.7.11.22']</td>\n",
       "      <td>Bos taurus</td>\n",
       "      <td>CDK2</td>\n",
       "      <td>9913</td>\n",
       "      <td>['Eukaryota', 'Metazoa', 'Chordata', 'Craniata...</td>\n",
       "      <td>MENFQKVEKIGEGTYGVVYKAKNKLTGEVVALKKIRLDTETEGVPS...</td>\n",
       "      <td>298</td>\n",
       "      <td>...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>['GO:0015030', 'GO:0005813', 'GO:0000307', 'GO...</td>\n",
       "      <td>['IPR050108', 'IPR011009', 'IPR000719', 'IPR01...</td>\n",
       "      <td>['bta:519217']</td>\n",
       "      <td>[]</td>\n",
       "      <td>['PF00069']</td>\n",
       "      <td>['R-BTA-1538133', 'R-BTA-171319', 'R-BTA-17618...</td>\n",
       "      <td>['NP_001014934.1']</td>\n",
       "      <td>['9913.ENSBTAP00000005252']</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                   query accession  \\\n",
       "19934  (go:DNA binding OR go:ATP binding OR keyword:A...    O55076   \n",
       "26084  (go:DNA binding OR go:ATP binding OR keyword:A...    P20248   \n",
       "26805  (go:DNA binding OR go:ATP binding OR keyword:A...    P24941   \n",
       "27540  (go:DNA binding OR go:ATP binding OR keyword:A...    P30274   \n",
       "28601  (go:DNA binding OR go:ATP binding OR keyword:A...    P37881   \n",
       "30029  (go:DNA binding OR go:ATP binding OR keyword:A...    P48963   \n",
       "46192  (go:DNA binding OR go:ATP binding OR keyword:A...    Q5E9Y0   \n",
       "\n",
       "                    protein_name     ec_numbers         organism_name  \\\n",
       "19934  Cyclin-dependent kinase 2  ['2.7.11.22']    Cricetulus griseus   \n",
       "26084                  Cyclin-A2            NaN          Homo sapiens   \n",
       "26805  Cyclin-dependent kinase 2  ['2.7.11.22']          Homo sapiens   \n",
       "27540                  Cyclin-A2            NaN            Bos taurus   \n",
       "28601                  Cyclin-A2            NaN  Mesocricetus auratus   \n",
       "30029  Cyclin-dependent kinase 2  ['2.7.11.22']  Mesocricetus auratus   \n",
       "46192  Cyclin-dependent kinase 2  ['2.7.11.22']            Bos taurus   \n",
       "\n",
       "      gene_primary  taxon_id  \\\n",
       "19934         CDK2     10029   \n",
       "26084        CCNA2      9606   \n",
       "26805         CDK2      9606   \n",
       "27540        CCNA2      9913   \n",
       "28601        CCNA2     10036   \n",
       "30029         CDK2     10036   \n",
       "46192         CDK2      9913   \n",
       "\n",
       "                                                  ineage  \\\n",
       "19934  ['Eukaryota', 'Metazoa', 'Chordata', 'Craniata...   \n",
       "26084  ['Eukaryota', 'Metazoa', 'Chordata', 'Craniata...   \n",
       "26805  ['Eukaryota', 'Metazoa', 'Chordata', 'Craniata...   \n",
       "27540  ['Eukaryota', 'Metazoa', 'Chordata', 'Craniata...   \n",
       "28601  ['Eukaryota', 'Metazoa', 'Chordata', 'Craniata...   \n",
       "30029  ['Eukaryota', 'Metazoa', 'Chordata', 'Craniata...   \n",
       "46192  ['Eukaryota', 'Metazoa', 'Chordata', 'Craniata...   \n",
       "\n",
       "                                                sequence  length  ...  \\\n",
       "19934  MENFQKVEKIGEGTYGVVYKAKNKLTGEVVALKKIRLDTETEGVPS...     298  ...   \n",
       "26084  MLGNSAPGPATREAGSALLALQQTALQEDQENINPEKAAPVQQPRT...     432  ...   \n",
       "26805  MENFQKVEKIGEGTYGVVYKARNKLTGEVVALKKIRLDTETEGVPS...     298  ...   \n",
       "27540  MLGSSAHGPAAREAGSAVTLQQTAFQEDQENVNPEKAAPAQQPRTR...     430  ...   \n",
       "28601  MPGSSRQSGREAGSALLSLQQEDQENVNPEKAAPDQRARAALKTGN...     421  ...   \n",
       "30029  MENFQKVEKIGEGTYGVVYKAKNKLTGEVVALKKIRLDTETEGVPS...     298  ...   \n",
       "46192  MENFQKVEKIGEGTYGVVYKAKNKLTGEVVALKKIRLDTETEGVPS...     298  ...   \n",
       "\n",
       "      biogrid_ids     brenda_ids  \\\n",
       "19934          []  ['2.7.11.22']   \n",
       "26084          []             []   \n",
       "26805          []  ['2.7.11.22']   \n",
       "27540          []             []   \n",
       "28601          []             []   \n",
       "30029          []  ['2.7.11.22']   \n",
       "46192          []             []   \n",
       "\n",
       "                                                go_terms  \\\n",
       "19934  ['GO:0015030', 'GO:0005813', 'GO:0000307', 'GO...   \n",
       "26084  ['GO:0097122', 'GO:0097124', 'GO:0000307', 'GO...   \n",
       "26805  ['GO:0015030', 'GO:0005813', 'GO:0000781', 'GO...   \n",
       "27540  ['GO:0097124', 'GO:0000307', 'GO:0005737', 'GO...   \n",
       "28601  ['GO:0097122', 'GO:0097124', 'GO:0000307', 'GO...   \n",
       "30029  ['GO:0015030', 'GO:0005813', 'GO:0000307', 'GO...   \n",
       "46192  ['GO:0015030', 'GO:0005813', 'GO:0000307', 'GO...   \n",
       "\n",
       "                                            interpro_ids            kegg_ids  \\\n",
       "19934  ['IPR050108', 'IPR011009', 'IPR000719', 'IPR01...                  []   \n",
       "26084  ['IPR039361', 'IPR032447', 'IPR013763', 'IPR03...         ['hsa:890']   \n",
       "26805  ['IPR050108', 'IPR011009', 'IPR000719', 'IPR01...        ['hsa:1017']   \n",
       "27540  ['IPR039361', 'IPR032447', 'IPR013763', 'IPR03...      ['bta:281667']   \n",
       "28601  ['IPR039361', 'IPR032447', 'IPR013763', 'IPR03...  ['maua:101835930']   \n",
       "30029  ['IPR050108', 'IPR011009', 'IPR000719', 'IPR01...                  []   \n",
       "46192  ['IPR050108', 'IPR011009', 'IPR000719', 'IPR01...      ['bta:519217']   \n",
       "\n",
       "                                                 pdb_ids  \\\n",
       "19934                                                 []   \n",
       "26084  ['1E9H', '1FIN', '1FVV', '1GY3', '1H1P', '1H1Q...   \n",
       "26805  ['1AQ1', '1B38', '1B39', '1BUH', '1CKP', '1DI8...   \n",
       "27540  ['1VIN', '2G9X', '3BHT', '3BHU', '3BHV', '3DDP...   \n",
       "28601                                                 []   \n",
       "30029                                                 []   \n",
       "46192                                                 []   \n",
       "\n",
       "                                pfam_ids  \\\n",
       "19934                        ['PF00069']   \n",
       "26084  ['PF02984', 'PF00134', 'PF16500']   \n",
       "26805                        ['PF00069']   \n",
       "27540  ['PF02984', 'PF00134', 'PF16500']   \n",
       "28601  ['PF02984', 'PF00134', 'PF16500']   \n",
       "30029                        ['PF00069']   \n",
       "46192                        ['PF00069']   \n",
       "\n",
       "                                            reactome_ids  \\\n",
       "19934                                                 []   \n",
       "26084  ['R-HSA-1362300', 'R-HSA-1538133', 'R-HSA-1701...   \n",
       "26805  ['R-HSA-1538133', 'R-HSA-171319', 'R-HSA-17618...   \n",
       "27540                                                 []   \n",
       "28601                                                 []   \n",
       "30029                                                 []   \n",
       "46192  ['R-BTA-1538133', 'R-BTA-171319', 'R-BTA-17618...   \n",
       "\n",
       "                                             refseq_ids  \\\n",
       "19934                                                []   \n",
       "26084                                   ['NP_001228.2']   \n",
       "26805  ['NP_001277159.1', 'NP_001789.2', 'NP_439892.2']   \n",
       "27540                                ['NP_001068591.1']   \n",
       "28601                                ['NP_001268563.1']   \n",
       "30029                                                []   \n",
       "46192                                ['NP_001014934.1']   \n",
       "\n",
       "                         string_ids  \n",
       "19934                            []  \n",
       "26084      ['9606.ENSP00000481380']  \n",
       "26805      ['9606.ENSP00000266970']  \n",
       "27540   ['9913.ENSBTAP00000006503']  \n",
       "28601  ['10036.ENSMAUP00000020143']  \n",
       "30029  ['10036.ENSMAUP00000021672']  \n",
       "46192   ['9913.ENSBTAP00000005252']  \n",
       "\n",
       "[7 rows x 21 columns]"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Search CDK2 and CCNA2 in gene primary use is in or something like that\n",
    "df[\"gene_primary\"] = df[\"gene_primary\"].apply(\n",
    "    lambda x: ast.literal_eval(x) if isinstance(x, str) and x.startswith(\"[\") else [x]\n",
    ")\n",
    "df = df.explode(\"gene_primary\")\n",
    "df = df[df[\"gene_primary\"].isin([\"CDK2\", \"CCNA2\"])]\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8458301d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>taxon_id</th>\n",
       "      <th>gene_primary</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>720</th>\n",
       "      <td>9606</td>\n",
       "      <td>[TTF1, ATN1, SS18, MNDA, NHEJ1, EFEMP1, NCOA4,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>779</th>\n",
       "      <td>10090</td>\n",
       "      <td>[Htatip2, Ddb1, Has3, Tbx19, Nsd1, Sohlh2, Ddi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>359</th>\n",
       "      <td>3702</td>\n",
       "      <td>[RH9, NFYB1, HMGA, GATA3, ETC2, NSN1, PCNA2, A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>785</th>\n",
       "      <td>10116</td>\n",
       "      <td>[Htatip2, Ddb1, Sohlh2, Ddit4, Pagr1, Mcidas, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3153</th>\n",
       "      <td>559292</td>\n",
       "      <td>[PSF3, ADF1, RPO31, KIN28, TPK1, RPT1, YAP7, N...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>25</td>\n",
       "      <td>[ssb]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>61</td>\n",
       "      <td>[uvrA]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>84</td>\n",
       "      <td>[sgaR]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>141</td>\n",
       "      <td>[hbb]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>142</td>\n",
       "      <td>[hbb]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3520 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      taxon_id                                       gene_primary\n",
       "720       9606  [TTF1, ATN1, SS18, MNDA, NHEJ1, EFEMP1, NCOA4,...\n",
       "779      10090  [Htatip2, Ddb1, Has3, Tbx19, Nsd1, Sohlh2, Ddi...\n",
       "359       3702  [RH9, NFYB1, HMGA, GATA3, ETC2, NSN1, PCNA2, A...\n",
       "785      10116  [Htatip2, Ddb1, Sohlh2, Ddit4, Pagr1, Mcidas, ...\n",
       "3153    559292  [PSF3, ADF1, RPO31, KIN28, TPK1, RPT1, YAP7, N...\n",
       "...        ...                                                ...\n",
       "1           25                                              [ssb]\n",
       "3           61                                             [uvrA]\n",
       "5           84                                             [sgaR]\n",
       "7          141                                              [hbb]\n",
       "8          142                                              [hbb]\n",
       "\n",
       "[3520 rows x 2 columns]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Group and separate by taxon_id and get gene_primary\n",
    "grouped = df.groupby(\"taxon_id\").agg({\n",
    "    \"gene_primary\": lambda x: list(\n",
    "        set(\n",
    "            item\n",
    "            for sublist in x.dropna()\n",
    "            for item in (ast.literal_eval(sublist) if isinstance(sublist, str) and sublist.startswith(\"[\") else [sublist])\n",
    "        )\n",
    "    )\n",
    "})\n",
    "# Show those with the most gene_primary\n",
    "grouped = grouped[grouped[\"gene_primary\"].str.len() > 0].reset_index()\n",
    "grouped = grouped.sort_values(by=\"gene_primary\", key=lambda x: x.str.len(), ascending=False)\n",
    "grouped"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0eaae7cf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'TTF1%0dATN1%0dSS18%0dNHEJ1%0dMNDA%0dEFEMP1%0dNCOA4%0dDUX4L7%0dRHOG%0dLDB1%0dWNK1%0dGATA3%0dATXN7L3%0dZNF224%0dTGIF2LX%0dSMARCA4%0dTCF3%0dCCNE1%0dHLTF%0dZIC3%0dRALY%0dSMARCB1%0dTCF20%0dPLAGL2%0dTSPY1%0dZNF117%0dHOXC13%0dCC2D1A%0dTFEB%0dOTX1%0dLPIN1%0dNRG1%0dZNF638%0dIRX4%0dEHMT1%0dRUVBL2%0dUBTFL1%0dSFMBT1%0dMCM3%0dKAT7%0dHES5%0dZNF507%0dSOX21%0dMYO6%0dRPL6%0dKHDC3L%0dZNF721%0dRPRD1A%0dDNAJC1%0dGAS6%0dKAT2A%0dLRP6%0dCDC34%0dPRAP1%0dZFPM2%0dN4BP2%0dCRYM%0dHMGB4%0dTGFB1%0dARMCX3%0dHDAC9%0dHES4%0dFOXD3%0dSUV39H1%0dCASP8AP2%0dLITAF%0dUSPL1%0dTSC22D2%0dPALB2%0dPIF1%0dNOD2%0dPRDM5%0dSIM2%0dIFNB1%0dHES3%0dCXorf65%0dHOXC5%0dCASP2%0dESX1%0dMTERF3%0dHEMK1%0dENDOG%0dGTF2F2%0dZNF780A%0dSPIN2A%0dNPAS4%0dMAGED1%0dKAT2B%0dCHUK%0dPPP1R12A%0dTRIM34%0dNUDT15%0dHOPX%0dFGF9%0dCCT7%0dGEN1%0dTAF11L8%0dATR%0dPRNP%0dZNF280A'"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# From tax_id 9606 get the gene_primary list\n",
    "gene_primary_9606 = grouped[grouped[\"taxon_id\"] == 9606][\"gene_primary\"].values\n",
    "#convert to a flat list\n",
    "gene_primary_9606 = [item for sublist in gene_primary_9606 for item in sublist]\n",
    "gene_primary_9606 = list(set(gene_primary_9606))  # Remove duplicates\n",
    "\"%0d\".join(gene_primary_9606[:100])\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
